{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.append('../../')\n",
    "\n",
    "import mudata\n",
    "\n",
    "from src.evaluation import compute_categorical_association\n",
    "from src.evaluation import compute_explained_variance_ratio\n",
    "from src.evaluation import compute_geneset_enrichment\n",
    "from src.evaluation import compute_trait_enrichment\n",
    "from src.evaluation import compute_motif_enrichment\n",
    "\n",
    "from plotly.subplots import make_subplots\n",
    "\n",
    "import plotly.graph_objects as go\n",
    "import plotly.express as px\n",
    "\n",
    "import ipywidgets as w\n",
    "from IPython.display import display\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from scipy import stats\n",
    "\n",
    "from tqdm.auto import tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/oak/stanford/groups/akundaje/regu/miniforge3/envs/evaluation/lib/python3.9/site-packages/anndata/_core/anndata.py:522: FutureWarning: The dtype argument is deprecated and will be removed in late 2024.\n",
      "  warnings.warn(\n",
      "/oak/stanford/groups/akundaje/regu/miniforge3/envs/evaluation/lib/python3.9/site-packages/anndata/_core/anndata.py:522: FutureWarning: The dtype argument is deprecated and will be removed in late 2024.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<pre>MuData object with n_obs × n_vars = 4000 × 17532\n",
       "  2 modalities\n",
       "    rna:\t4000 x 17472\n",
       "      obs:\t&#x27;barcodes&#x27;, &#x27;n_genes&#x27;, &#x27;n_counts&#x27;, &#x27;sample&#x27;, &#x27;batch&#x27;\n",
       "      var:\t&#x27;n_cells&#x27;\n",
       "    cNMF:\t4000 x 60\n",
       "      obs:\t&#x27;sample&#x27;, &#x27;batch&#x27;\n",
       "      uns:\t&#x27;var_names&#x27;\n",
       "      varm:\t&#x27;loadings&#x27;, &#x27;loadings_zscore&#x27;</pre>"
      ],
      "text/plain": [
       "MuData object with n_obs × n_vars = 4000 × 17532\n",
       "  2 modalities\n",
       "    rna:\t4000 x 17472\n",
       "      obs:\t'barcodes', 'n_genes', 'n_counts', 'sample', 'batch'\n",
       "      var:\t'n_cells'\n",
       "    cNMF:\t4000 x 60\n",
       "      obs:\t'sample', 'batch'\n",
       "      uns:\t'var_names'\n",
       "      varm:\t'loadings', 'loadings_zscore'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Import test data\n",
    "mdata = mudata.read('../../../../../data/TeloHAEC_Perturb-seq_2kG/2kG.library_K60_kangh.h5mu')\n",
    "\n",
    "sel_idx = []\n",
    "for batch in mdata['cNMF'].obs['batch'].unique():\n",
    "    for samp in mdata['cNMF'].obs['sample'].unique():\n",
    "        mdata_obs_ = mdata['cNMF'].obs.loc[(mdata['cNMF'].obs['batch']==batch) & \\\n",
    "                                           (mdata['cNMF'].obs['sample']==samp)]\n",
    "\n",
    "        sel_idx.extend(mdata_obs_.iloc[:200].index.tolist())\n",
    "\n",
    "mdata = mdata[sel_idx].copy()\n",
    "mdata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "8941700d66694d99a7d7727e766b3a3c",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Computing explained variance:   0%|          | 0/60 [00:00<?, ?programs/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Explained variance ratio plot\n",
    "explained_ratios = compute_explained_variance_ratio(mdata, prog_key='cNMF', data_key='rna', inplace=False)\n",
    "\n",
    "# Since original data wasn't available we use this as a stand in\n",
    "from sklearn.decomposition import PCA\n",
    "\n",
    "pca = PCA(n_components=60)\n",
    "\n",
    "explained_ratios['explained_variance_ratio_X'] = pca.fit(mdata['cNMF'].X).explained_variance_ratio_\n",
    "\n",
    "# Categorical association plot\n",
    "association_df, posthoc_df = compute_categorical_association(mdata, prog_key='cNMF',\n",
    "                                                             pseudobulk_key='sample',\n",
    "                                                             categorical_key='batch',\n",
    "                                                             inplace=False, \n",
    "                                                             n_jobs=-1)\n",
    "\n",
    "# Geneset enrichment \n",
    "gsea_df = compute_geneset_enrichment(mdata, prog_key='cNMF', library='Reactome_2013', inplace=False, n_jobs=-1)\n",
    "\n",
    "# Trait enrichment \n",
    "gwas_df = compute_trait_enrichment(mdata, '../../smk/resources/OpenTargets_L2G_Filtered.csv.gz', prog_key='cNMF', inplace=False)\n",
    "\n",
    "# Motif enrichment\n",
    "motif_match_df, motif_count_df, motif_enrichment_df = \\\n",
    "compute_motif_enrichment(mdata, prog_key='cNMF', data_key='rna', motif_file='../tests/test_data/motifs.meme',\n",
    "                         seq_file='../../../../../data/hg38/hg38.fa', coords_file='../tests/test_data/p2g_links.txt',\n",
    "                         n_jobs=1, inplace=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'make_subplots' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[1], line 15\u001b[0m\n\u001b[1;32m     10\u001b[0m     counts_df \u001b[39m=\u001b[39m pd\u001b[39m.\u001b[39mDataFrame(counts_df\u001b[39m.\u001b[39mreset_index()\u001b[39m.\u001b[39mvalues, \n\u001b[1;32m     11\u001b[0m                              columns\u001b[39m=\u001b[39m[categorical_var,\n\u001b[1;32m     12\u001b[0m                                       count_var])\n\u001b[1;32m     13\u001b[0m     \u001b[39mreturn\u001b[39;00m counts_df\n\u001b[0;32m---> 15\u001b[0m fig \u001b[39m=\u001b[39m make_subplots(\n\u001b[1;32m     16\u001b[0m     rows\u001b[39m=\u001b[39m\u001b[39m2\u001b[39m, cols\u001b[39m=\u001b[39m\u001b[39m2\u001b[39m,\n\u001b[1;32m     17\u001b[0m     specs\u001b[39m=\u001b[39m[\n\u001b[1;32m     18\u001b[0m            [{}, {}],\n\u001b[1;32m     19\u001b[0m            [{}, {}],\n\u001b[1;32m     20\u001b[0m           ],\n\u001b[1;32m     21\u001b[0m     print_grid\u001b[39m=\u001b[39m\u001b[39mTrue\u001b[39;00m,\n\u001b[1;32m     22\u001b[0m     subplot_titles\u001b[39m=\u001b[39m(\u001b[39m'\u001b[39m\u001b[39mCummulative component wise R2 scores\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mCummulative enriched gene-sets\u001b[39m\u001b[39m'\u001b[39m,\n\u001b[1;32m     23\u001b[0m                      \u001b[39m'\u001b[39m\u001b[39mCummulative enriched GWAS traits\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mCummulative enriched Promoter Motifs\u001b[39m\u001b[39m'\u001b[39m),\n\u001b[1;32m     24\u001b[0m     vertical_spacing \u001b[39m=\u001b[39m \u001b[39m0.05\u001b[39m, horizontal_spacing \u001b[39m=\u001b[39m \u001b[39m0.1\u001b[39m)\n\u001b[1;32m     26\u001b[0m \u001b[39m# Explained variance ratios\u001b[39;00m\n\u001b[1;32m     27\u001b[0m explained_ratios \u001b[39m=\u001b[39m explained_ratios\u001b[39m.\u001b[39msort_values(explained_ratios\u001b[39m.\u001b[39mcolumns[\u001b[39m0\u001b[39m], ascending\u001b[39m=\u001b[39m\u001b[39mFalse\u001b[39;00m)\n",
      "\u001b[0;31mNameError\u001b[0m: name 'make_subplots' is not defined"
     ]
    }
   ],
   "source": [
    "# Assemble dashboard - page 1\n",
    "\n",
    "# Plot unique terms per program\n",
    "def count_unique(categorical_var, count_var, dataframe):\n",
    "\n",
    "    counts_df = dataframe.value_counts([categorical_var, count_var])\n",
    "    counts_df = counts_df.groupby(categorical_var).sum()\n",
    "    counts_df = counts_df.sort_values(ascending=False).cumsum()\n",
    "\n",
    "    counts_df = pd.DataFrame(counts_df.reset_index().values, \n",
    "                             columns=[categorical_var,\n",
    "                                      count_var])\n",
    "    return counts_df\n",
    "\n",
    "fig = make_subplots(\n",
    "    rows=2, cols=2,\n",
    "    specs=[\n",
    "           [{}, {}],\n",
    "           [{}, {}],\n",
    "          ],\n",
    "    print_grid=True,\n",
    "    subplot_titles=('Cummulative component wise R2 scores', 'Cummulative enriched gene-sets',\n",
    "                     'Cummulative enriched GWAS traits', 'Cummulative enriched Promoter Motifs'),\n",
    "    vertical_spacing = 0.05, horizontal_spacing = 0.1)\n",
    "\n",
    "# Explained variance ratios\n",
    "explained_ratios = explained_ratios.sort_values(explained_ratios.columns[0], ascending=False)\n",
    "\n",
    "# Plot unique GSEA Terms per program\n",
    "gsea_unique_df = count_unique('program_name', 'Term', gsea_df.loc[gsea_df['FDR q-val']<=0.05])\n",
    "\n",
    "# Plot unique GWAS Terms per program\n",
    "gwas_unique_df = count_unique('program_name', 'Term', gwas_df.loc[gwas_df['Adjusted P-value']<=0.05])\n",
    "\n",
    "# Plot unique Motif Terms per program\n",
    "motif_enrichment_df['fdr'] = stats.false_discovery_control(motif_enrichment_df.pval.astype(float))\n",
    "\n",
    "# Generate example data for dashapp\n",
    "\n",
    "a = explained_ratios.reset_index()\n",
    "\n",
    "a['prog_name'] = a['index']\n",
    "a['Cummulative component wise R2 scores'] = a['explained_variance_ratio_X'].cumsum()\n",
    "a = a.drop(['index', 'explained_variance_ratio_X'], axis=1)\n",
    "\n",
    "gwas_unique_df['Cummulative enriched GWAS traits'] = gwas_unique_df['Term']\n",
    "gwas_unique_df['prog_name'] = gwas_unique_df['program_name']\n",
    "a = a.merge(gwas_unique_df.loc[:, ['prog_name', 'Cummulative enriched GWAS traits']], left_on='prog_name', right_on='prog_name', how='left')\n",
    "\n",
    "gsea_unique_df['Cummulative enriched gene-sets'] = gsea_unique_df['Term']\n",
    "gsea_unique_df['prog_name'] = gsea_unique_df['program_name']\n",
    "a = a.merge(gsea_unique_df.loc[:, ['prog_name', 'Cummulative enriched gene-sets']], left_on='prog_name', right_on='prog_name', how='left')\n",
    "\n",
    "motif_unique_df['Cummulative enriched Promoter Motifs'] = motif_unique_df['motif']\n",
    "motif_unique_df['prog_name'] = motif_unique_df['index']\n",
    "a = a.merge(motif_unique_df.loc[:, ['prog_name', 'Cummulative enriched Promoter Motifs']], left_on='prog_name', right_on='prog_name', how='left')\n",
    "\n",
    "a['Cummulative enriched GWAS traits'] = a['Cummulative enriched GWAS traits'].fillna(a['Cummulative enriched GWAS traits'].max())\n",
    "a['Cummulative enriched gene-sets'] = a['Cummulative enriched gene-sets'].fillna(a['Cummulative enriched gene-sets'].max())\n",
    "a['Cummulative enriched Promoter Motifs'] = a['Cummulative enriched Promoter Motifs'].fillna(a['Cummulative enriched Promoter Motifs'].max())\n",
    "\n",
    "a.to_csv('dashapp/example_data/fit_metrics.txt', sep='\\t', index=False)\n",
    "\n",
    "plots = {}\n",
    "for i, col in enumerate([col for col in a.columns if 'prog_name' not in col]):\n",
    "\n",
    "    plots[col] = px.scatter(x=a.sort_values(col)['prog_name'], \n",
    "                            y=a.sort_values(col)[col],)\n",
    "    plots[col].update_layout(xaxis_title='Components', yaxis_title=col)\n",
    "\n",
    "    row_num = int((i)/2) + 1\n",
    "    col_num = i - (row_num-1)*2 + 1\n",
    "    fig.add_trace(plots[col]['data'][0], row=row_num, col=col_num)\n",
    "    fig.update_xaxes(showticklabels=False, row=row_num, col=col_num)\n",
    "    fig.update_yaxes(ticksuffix = \"  \", row=row_num, col=col_num)\n",
    "\n",
    "fig.update_traces(hovertemplate=\"Program Name: %{x} <br> Value: %{y}\", marker_color='silver')\n",
    "fig.update_layout(height=700, width=1000, \n",
    "                  plot_bgcolor='whitesmoke',\n",
    "                  title_text=\"GEP Dashboard - v0.1 - Goodness of fit measures\")\n",
    "fig.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'stats' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[2], line 4\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[39m# Page 2 - program assessment\u001b[39;00m\n\u001b[1;32m      2\u001b[0m \n\u001b[1;32m      3\u001b[0m \u001b[39m# Batch assocation\u001b[39;00m\n\u001b[0;32m----> 4\u001b[0m association_df[\u001b[39m'\u001b[39m\u001b[39mbatch_sample_kruskall_wallis_fdr\u001b[39m\u001b[39m'\u001b[39m] \u001b[39m=\u001b[39m stats\u001b[39m.\u001b[39mfalse_discovery_control(association_df\u001b[39m.\u001b[39mbatch_sample_kruskall_wallis_pval)\n\u001b[1;32m      5\u001b[0m association_df[\u001b[39m'\u001b[39m\u001b[39mbatch_sample_kruskall_wallis_neg_log_fdr\u001b[39m\u001b[39m'\u001b[39m] \u001b[39m=\u001b[39m association_df[\u001b[39m'\u001b[39m\u001b[39mbatch_sample_kruskall_wallis_fdr\u001b[39m\u001b[39m'\u001b[39m]\u001b[39m.\u001b[39mapply(\u001b[39mlambda\u001b[39;00m x: \u001b[39m-\u001b[39mnp\u001b[39m.\u001b[39mlog(x))\n\u001b[1;32m      7\u001b[0m association_df[\u001b[39m'\u001b[39m\u001b[39mbatch_sample_kruskall_wallis_log_stat\u001b[39m\u001b[39m'\u001b[39m] \u001b[39m=\u001b[39m association_df[\u001b[39m'\u001b[39m\u001b[39mbatch_sample_kruskall_wallis_stat\u001b[39m\u001b[39m'\u001b[39m]\u001b[39m.\u001b[39mapply(\u001b[39mlambda\u001b[39;00m x: np\u001b[39m.\u001b[39mlog(x))\n",
      "\u001b[0;31mNameError\u001b[0m: name 'stats' is not defined"
     ]
    }
   ],
   "source": [
    "# Page 2 - program assessment\n",
    "\n",
    "# Batch assocation\n",
    "association_df['batch_sample_kruskall_wallis_fdr'] = stats.false_discovery_control(association_df.batch_sample_kruskall_wallis_pval)\n",
    "association_df['batch_sample_kruskall_wallis_neg_log_fdr'] = association_df['batch_sample_kruskall_wallis_fdr'].apply(lambda x: -np.log(x))\n",
    "\n",
    "association_df['batch_sample_kruskall_wallis_log_stat'] = association_df['batch_sample_kruskall_wallis_stat'].apply(lambda x: np.log(x))\n",
    "\n",
    "# Program scores across batch\n",
    "prog_df = pd.DataFrame(mdata['cNMF'].X, index=mdata['cNMF'].obs.index)\n",
    "prog_df['sample'] = mdata['cNMF'].obs['sample']\n",
    "prog_df['batch'] = mdata['cNMF'].obs['batch']\n",
    "\n",
    "# Loadings\n",
    "loadings = pd.DataFrame(mdata['cNMF'].varm['loadings'], \n",
    "                        index=mdata['cNMF'].var_names,\n",
    "                        columns=mdata['cNMF'].uns['var_names']).T    \n",
    "loadings = loadings/abs(loadings).max(axis=0)\n",
    "\n",
    "# Make figure\n",
    "fig = make_subplots(rows=2, cols=2,\n",
    "                    specs=[\n",
    "                           [{}, {}],\n",
    "                           [{'colspan':2}, None],\n",
    "                          ],\n",
    "                    print_grid=True,\n",
    "                    subplot_titles=('Enrichment w.r.t. batch', \n",
    "                                    'Program distribution across batch', \n",
    "                                    'Program-gene loadings (normalised)'),\n",
    "                    vertical_spacing = 0.1, \n",
    "                    horizontal_spacing = 0.1)\n",
    "\n",
    "for k, r in enumerate([col for col in prog_df.columns if col not in ['sample', 'batch']]):\n",
    "       # Plot volcano\n",
    "       volcano = go.Scatter(x=association_df.batch_sample_kruskall_wallis_log_stat,\n",
    "                            y=association_df.batch_sample_kruskall_wallis_neg_log_fdr,\n",
    "                            customdata=association_df.index.values,\n",
    "                            hovertemplate=\" Program Name: %{customdata}\",\n",
    "                            showlegend=False, mode='markers', \n",
    "                            visible=True if k == 0 else False,\n",
    "                            marker_color='silver',\n",
    "                            )\n",
    "       fig.add_trace(volcano, row=1, col=1)\n",
    "       fig.update_xaxes(showticklabels=False, row=1, col=1)\n",
    "       fig.update_yaxes(title='Neg. log adjusted pval', ticksuffix = \"  \", row=1, col=1)\n",
    "\n",
    "# Plot box\n",
    "for k, r in enumerate([col for col in prog_df.columns if col not in ['sample', 'batch']]):\n",
    "\n",
    "    fig.add_trace(\n",
    "                  go.Box(x=prog_df.batch, \n",
    "                         y=prog_df[r],\n",
    "                         visible=True if k == 0 else False,\n",
    "                         customdata=prog_df.index.values,\n",
    "                         hovertemplate=\"Cell barcode: %{customdata}\",\n",
    "                         marker_color='silver',\n",
    "                        ), \n",
    "                   row=1, col=2)\n",
    "# fig.update_xaxes(showticklabels=False, row=2, col=1)\n",
    "fig.update_yaxes(ticksuffix = \"  \", row=1, col=2)\n",
    "\n",
    "# Plot loadings\n",
    "for k, r in enumerate(loadings.columns):\n",
    "\n",
    "    dfp = loadings.sort_values(r, ascending=False)[:100]\n",
    "    fig.add_trace(\n",
    "                  go.Bar(x=dfp.index, \n",
    "                         y=dfp[r],\n",
    "                         name='', \n",
    "                         orientation='v',\n",
    "                         hovertemplate=\"Gene Name: %{x} <br> Loading: %{y}\",\n",
    "                         visible=True if k == 0 else False,\n",
    "                         marker_color='silver',\n",
    "                        ), \n",
    "                   row=2, col=1)\n",
    "fig.update_xaxes(showticklabels=False, row=2, col=1)\n",
    "fig.update_yaxes(tickvals=np.arange(0,1.25,0.25), ticksuffix = \"  \", row=2, col=1)\n",
    "        \n",
    "# Define buttons for dropdown\n",
    "col_opts = list(loadings.columns)\n",
    "buttons_opts = []\n",
    "for i, opt in enumerate(col_opts):\n",
    "    args = [False] * len(col_opts)\n",
    "    args[i] = True\n",
    "    buttons_opts.append(\n",
    "        dict(\n",
    "             method='update',\n",
    "             label=opt,\n",
    "             args=[{\n",
    "                    'visible': args, #this is the key line!\n",
    "                    'title': opt,\n",
    "                    'showlegend': False\n",
    "                   }]\n",
    "            ))\n",
    "    \n",
    "fig.update_layout(height=700, width=1000, \n",
    "                  plot_bgcolor='whitesmoke',\n",
    "\n",
    "                  title_text=\"GEP Dashboard - v0.1 - Investigate GEPs\",\n",
    "                  updatemenus = [go.layout.Updatemenu(\n",
    "                                 active=0,\n",
    "                                 buttons=buttons_opts,\n",
    "                                 x=-0.1,\n",
    "                                 xanchor='left',\n",
    "                                 y=1,\n",
    "                                 yanchor='bottom')]\n",
    "                 )\n",
    "fig.show()\n",
    "\n",
    "\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "evaluation",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
