{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 354,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "sys.path.append('/cellar/users/aklie/opt/gene_program_evaluation')\n",
    "sys.path.append('/cellar/users/aklie/opt/gene_program_evaluation/src/plotting/dashapp/example_app')\n",
    "import mudata\n",
    "\n",
    "from data_processing import load_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 355,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Insert geneset enrichment into mudata\n",
    "def insert_enrichment(mdata, df, library=\"GSEA\", prog_key=\"prog\",\n",
    "                      geneset_index=\"Term\", program_index=\"program_name\",\n",
    "                      varmap_name_prefix=\"gsea_varmap\"):\n",
    "    \n",
    "    # Create a mudata key to column name mapping dictionary\n",
    "    mudata_keys_dict = {}\n",
    "    for col in df.columns:\n",
    "        if col not in [geneset_index, program_index]:\n",
    "            key = f\"{col}_{library}\"\n",
    "            key = key.replace(' ', '_').replace('%', 'percent')\n",
    "            mudata_keys_dict[key] = col\n",
    "\n",
    "    print(mudata_keys_dict)\n",
    "\n",
    "    # Insert the values from the dataframe into the array for each key\n",
    "    for key, colname in mudata_keys_dict.items():\n",
    "        \n",
    "        # Create an empty dataframe with the right dimensions\n",
    "        all_progs_df = pd.DataFrame(index=df[geneset_index].unique(), \n",
    "                                    columns=mdata[prog_key].var.index)\n",
    "        \n",
    "        # Pivot the dataframe for gene sets and programs\n",
    "        pivot_df = df[[geneset_index, program_index, colname]].pivot(index=geneset_index, \n",
    "                                                                     columns=program_index, \n",
    "                                                                     values=colname)\n",
    "        \n",
    "        # Update the empty dataframe with new values\n",
    "        all_progs_df[pivot_df.columns] = pivot_df\n",
    "        \n",
    "        # Convert dataframe to a numpy array\n",
    "        if all_progs_df.dtypes[0] == 'float64':\n",
    "            all_progs_array = all_progs_df.T.to_numpy(dtype='float64')\n",
    "        else:\n",
    "            all_progs_array = all_progs_df.T.to_numpy()\n",
    "        \n",
    "        # Add the array into the MuData object\n",
    "        mdata[prog_key].varm[key] = all_progs_array\n",
    "        \n",
    "    # Add the varmap to the mudata object\n",
    "    varmap_name = f\"{varmap_name_prefix}_{library}\"\n",
    "    mdata[prog_key].uns[varmap_name] = all_progs_df.index\n",
    "\n",
    "\n",
    "# Now make a function that inverses the above, and takes the values from the mudata object and puts them into a dataframe\n",
    "def extract_enrichment(\n",
    "    mdata, library=\"GSEA\", prog_key=\"prog\", geneset_index=\"Term\", program_index=\"program_name\",\n",
    "    varmap_name_prefix=\"gsea_varmap\"\n",
    "):\n",
    "    # Create a mudata key to column name mapping dictionary\n",
    "    mudata_keys_dict = {}\n",
    "    for key in mdata[prog_key].varm.keys():\n",
    "        if library in key:\n",
    "            colname = key.replace(f\"_{library}\", \"\")\n",
    "            mudata_keys_dict[colname] = key\n",
    "    print(mudata_keys_dict)\n",
    "    \n",
    "    # Unpivot the programs and genesets\n",
    "    programs = mdata[prog_key].var.index.tolist()\n",
    "    genesets = mdata[prog_key].uns[f\"{varmap_name_prefix}_{library}\"].tolist()\n",
    "    \n",
    "    # Create an empty dataframe\n",
    "    df = pd.DataFrame(index=genesets, columns=programs)\n",
    "\n",
    "    # Melting the dataframe\n",
    "    df = df.melt(value_vars=programs, var_name=program_index, value_name=\"value\", ignore_index=False).reset_index()\n",
    "    df.columns = [geneset_index, program_index, \"value\"]\n",
    "    df = df.drop(columns=\"value\")\n",
    "    \n",
    "    # For each key in the dictionary, extract the values from the mudata object and put them into the dataframe\n",
    "    for colname, key in mudata_keys_dict.items():\n",
    "        # Extract the values from the mudata object\n",
    "        print(colname, key)\n",
    "        all_progs_array = mdata[prog_key].varm[key].flatten()\n",
    "        \n",
    "        # Add the values to the dataframe\n",
    "        df[colname] = all_progs_array\n",
    "\n",
    "    # Drop any rows with NaN values\n",
    "    df = df.dropna()\n",
    "    \n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 356,
   "metadata": {},
   "outputs": [],
   "source": [
    "path_excel = \"/cellar/users/aklie/opt/gene_program_evaluation/src/plotting/dashapp/example_data/cNMF_evaluation_output.xlsx\"\n",
    "path_mdata = \"/cellar/users/aklie/opt/gene_program_evaluation/src/plotting/dashapp/example_data/cNMF_evaluation_dashapp_data.h5mu\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 357,
   "metadata": {},
   "outputs": [],
   "source": [
    "path_out = \"/cellar/users/aklie/opt/gene_program_evaluation/src/plotting/dashapp/example_data/example_data.h5mu\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 358,
   "metadata": {},
   "outputs": [],
   "source": [
    "explained_variance, enrichment_gsea, enrichment_motif, enrichment_gwas = load_data(path_excel)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 359,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/cellar/users/aklie/opt/miniconda3/envs/test_celloracle/lib/python3.10/site-packages/anndata/_core/anndata.py:522: FutureWarning: The dtype argument is deprecated and will be removed in late 2024.\n",
      "  warnings.warn(\n",
      "/cellar/users/aklie/opt/miniconda3/envs/test_celloracle/lib/python3.10/site-packages/anndata/_core/anndata.py:522: FutureWarning: The dtype argument is deprecated and will be removed in late 2024.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<pre>MuData object with n_obs × n_vars = 214449 × 17532\n",
       "  2 modalities\n",
       "    rna:\t214449 x 17472\n",
       "      obs:\t&#x27;barcodes&#x27;, &#x27;n_genes&#x27;, &#x27;n_counts&#x27;, &#x27;sample&#x27;, &#x27;batch&#x27;\n",
       "      var:\t&#x27;n_cells&#x27;\n",
       "    cNMF:\t214449 x 60\n",
       "      obs:\t&#x27;sample&#x27;, &#x27;batch&#x27;\n",
       "      uns:\t&#x27;var_names&#x27;\n",
       "      varm:\t&#x27;loadings&#x27;, &#x27;loadings_zscore&#x27;</pre>"
      ],
      "text/plain": [
       "MuData object with n_obs × n_vars = 214449 × 17532\n",
       "  2 modalities\n",
       "    rna:\t214449 x 17472\n",
       "      obs:\t'barcodes', 'n_genes', 'n_counts', 'sample', 'batch'\n",
       "      var:\t'n_cells'\n",
       "    cNMF:\t214449 x 60\n",
       "      obs:\t'sample', 'batch'\n",
       "      uns:\t'var_names'\n",
       "      varm:\t'loadings', 'loadings_zscore'"
      ]
     },
     "execution_count": 359,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mdata = mudata.read_h5mu(path_mdata)\n",
    "mdata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 360,
   "metadata": {},
   "outputs": [],
   "source": [
    "mdata.mod[\"cNMF\"].var_names = pd.Index([\"K60_\" + x for x in mdata.mod[\"cNMF\"].var_names])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Add variance explained"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 361,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>VarianceExplained</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ProgramID</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>K60_1</th>\n",
       "      <td>0.000643</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>K60_2</th>\n",
       "      <td>0.000708</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>K60_3</th>\n",
       "      <td>0.000297</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>K60_4</th>\n",
       "      <td>0.000783</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>K60_5</th>\n",
       "      <td>0.000116</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           VarianceExplained\n",
       "ProgramID                   \n",
       "K60_1               0.000643\n",
       "K60_2               0.000708\n",
       "K60_3               0.000297\n",
       "K60_4               0.000783\n",
       "K60_5               0.000116"
      ]
     },
     "execution_count": 361,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mdata.mod[\"cNMF\"].var = mdata.mod[\"cNMF\"].var.merge(explained_variance, left_index=True, right_on=\"ProgramID\").set_index(\"ProgramID\")\n",
    "mdata.mod[\"cNMF\"].var.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Add GSEA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 362,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'Description_GSEA': 'Description', 'pvalue_GSEA': 'pvalue', 'p.adjust_GSEA': 'p.adjust', 'qvalue_GSEA': 'qvalue'}\n"
     ]
    }
   ],
   "source": [
    "insert_enrichment(\n",
    "    mdata, \n",
    "    enrichment_gsea, \n",
    "    library=\"GSEA\", \n",
    "    prog_key=\"cNMF\",\n",
    "    geneset_index=\"ID\", \n",
    "    program_index=\"ProgramID\",\n",
    "    varmap_name_prefix=\"gsea_varmap\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 363,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'Description': 'Description_GSEA', 'pvalue': 'pvalue_GSEA', 'p.adjust': 'p.adjust_GSEA', 'qvalue': 'qvalue_GSEA'}\n",
      "Description Description_GSEA\n",
      "pvalue pvalue_GSEA\n",
      "p.adjust p.adjust_GSEA\n",
      "qvalue qvalue_GSEA\n"
     ]
    }
   ],
   "source": [
    "inversed_df = extract_enrichment(\n",
    "    mdata, \n",
    "    library=\"GSEA\", \n",
    "    prog_key=\"cNMF\",\n",
    "    geneset_index=\"ID\", \n",
    "    program_index=\"ProgramID\",\n",
    "    varmap_name_prefix=\"gsea_varmap\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 364,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ProgramID</th>\n",
       "      <th>ID</th>\n",
       "      <th>Description</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>p.adjust</th>\n",
       "      <th>qvalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5119</th>\n",
       "      <td>K60_3</td>\n",
       "      <td>GO:0000018</td>\n",
       "      <td>regulation of DNA recombination</td>\n",
       "      <td>2.037010e-15</td>\n",
       "      <td>6.587377e-14</td>\n",
       "      <td>5.380346e-14</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     ProgramID          ID                      Description        pvalue  \\\n",
       "5119     K60_3  GO:0000018  regulation of DNA recombination  2.037010e-15   \n",
       "\n",
       "          p.adjust        qvalue  \n",
       "5119  6.587377e-14  5.380346e-14  "
      ]
     },
     "execution_count": 364,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Find index of GO:0000018 in varmap and of K60_3 in var_names\n",
    "geneset_index = mdata[\"cNMF\"].uns[\"gsea_varmap_GSEA\"].tolist().index(\"GO:0000018\")\n",
    "prog_index = mdata[\"cNMF\"].var_names.tolist().index(\"K60_3\")\n",
    "enrichment_gsea[(enrichment_gsea[\"ID\"] == \"GO:0000018\") & (enrichment_gsea[\"ProgramID\"] == \"K60_3\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 365,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(5.38034585993897e-14, 'regulation of DNA recombination')"
      ]
     },
     "execution_count": 365,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Find values in mdata.varm\n",
    "mdata[\"cNMF\"].varm[\"qvalue_GSEA\"][prog_index, geneset_index], mdata[\"cNMF\"].varm[\"Description_GSEA\"][prog_index, geneset_index]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 366,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>ProgramID</th>\n",
       "      <th>Description</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>p.adjust</th>\n",
       "      <th>qvalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>11676</th>\n",
       "      <td>GO:0000018</td>\n",
       "      <td>K60_3</td>\n",
       "      <td>regulation of DNA recombination</td>\n",
       "      <td>2.037010e-15</td>\n",
       "      <td>6.587377e-14</td>\n",
       "      <td>5.380346e-14</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               ID ProgramID                      Description        pvalue  \\\n",
       "11676  GO:0000018     K60_3  regulation of DNA recombination  2.037010e-15   \n",
       "\n",
       "           p.adjust        qvalue  \n",
       "11676  6.587377e-14  5.380346e-14  "
      ]
     },
     "execution_count": 366,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Find values of GO:0000018 K60_3 in original dataframe\n",
    "inversed_df[(inversed_df[\"ID\"] == \"GO:0000018\") & (inversed_df[\"ProgramID\"] == \"K60_3\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 367,
   "metadata": {},
   "outputs": [],
   "source": [
    "sorted_df = enrichment_gsea.sort_values(by=[\"ID\", \"ProgramID\"])\n",
    "sorted_inversed_df = inversed_df.sort_values(by=[\"ID\", \"ProgramID\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 368,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check if the pvalue column, Description, ID, ProgramID is the same\n",
    "assert sorted_df[\"pvalue\"].tolist() == sorted_inversed_df[\"pvalue\"].tolist()\n",
    "assert sorted_df[\"qvalue\"].tolist() == sorted_inversed_df[\"qvalue\"].tolist()\n",
    "assert sorted_df[\"Description\"].tolist() == sorted_inversed_df[\"Description\"].tolist()\n",
    "assert sorted_df[\"ID\"].tolist() == sorted_inversed_df[\"ID\"].tolist()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Add motif enrichment analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 369,
   "metadata": {},
   "outputs": [],
   "source": [
    "enrichment_motif[\"EPType-TFMotif\"] = enrichment_motif[\"EPType\"] + \"-\" + enrichment_motif[\"TFMotif\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 370,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'EPType_Motif': 'EPType', 'TFMotif_Motif': 'TFMotif', 'PValue_Motif': 'PValue', 'FDR_Motif': 'FDR', 'Enrichment_Motif': 'Enrichment'}\n"
     ]
    }
   ],
   "source": [
    "insert_enrichment(\n",
    "    mdata, \n",
    "    enrichment_motif, \n",
    "    library=\"Motif\", \n",
    "    prog_key=\"cNMF\",\n",
    "    geneset_index=\"EPType-TFMotif\",\n",
    "    program_index=\"ProgramID\",\n",
    "    varmap_name_prefix=\"motif_varmap\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 371,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'EPType': 'EPType_Motif', 'TFMotif': 'TFMotif_Motif', 'PValue': 'PValue_Motif', 'FDR': 'FDR_Motif', 'Enrichment': 'Enrichment_Motif'}\n",
      "EPType EPType_Motif\n",
      "TFMotif TFMotif_Motif\n",
      "PValue PValue_Motif\n",
      "FDR FDR_Motif\n",
      "Enrichment Enrichment_Motif\n"
     ]
    }
   ],
   "source": [
    "inversed_df = extract_enrichment(\n",
    "    mdata, \n",
    "    library=\"Motif\", \n",
    "    prog_key=\"cNMF\",\n",
    "    geneset_index=\"EPType-TFMotif\",\n",
    "    program_index=\"ProgramID\",\n",
    "    varmap_name_prefix=\"motif_varmap\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Add GWAS trait enrichment analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 373,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'program_name_GWAS': 'program_name', 'Gene_set_GWAS': 'Gene_set', 'P-value_GWAS': 'P-value', 'Adjusted_P-value_GWAS': 'Adjusted P-value', 'Odds_Ratio_GWAS': 'Odds Ratio', 'Combined_Score_GWAS': 'Combined Score', 'Genes_GWAS': 'Genes', 'overlap_numerator_GWAS': 'overlap_numerator', 'overlap_denominator_GWAS': 'overlap_denominator'}\n"
     ]
    }
   ],
   "source": [
    "insert_enrichment(\n",
    "    mdata, \n",
    "    enrichment_gwas, \n",
    "    library=\"GWAS\", \n",
    "    prog_key=\"cNMF\",\n",
    "    geneset_index=\"Term\",\n",
    "    program_index=\"ProgramID\",\n",
    "    varmap_name_prefix=\"gwas_varmap\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 374,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'program_name': 'program_name_GWAS', 'Gene_set': 'Gene_set_GWAS', 'P-value': 'P-value_GWAS', 'Adjusted_P-value': 'Adjusted_P-value_GWAS', 'Odds_Ratio': 'Odds_Ratio_GWAS', 'Combined_Score': 'Combined_Score_GWAS', 'Genes': 'Genes_GWAS', 'overlap_numerator': 'overlap_numerator_GWAS', 'overlap_denominator': 'overlap_denominator_GWAS'}\n",
      "program_name program_name_GWAS\n",
      "Gene_set Gene_set_GWAS\n",
      "P-value P-value_GWAS\n",
      "Adjusted_P-value Adjusted_P-value_GWAS\n",
      "Odds_Ratio Odds_Ratio_GWAS\n",
      "Combined_Score Combined_Score_GWAS\n",
      "Genes Genes_GWAS\n",
      "overlap_numerator overlap_numerator_GWAS\n",
      "overlap_denominator overlap_denominator_GWAS\n"
     ]
    }
   ],
   "source": [
    "inversed_df = extract_enrichment(\n",
    "    mdata, \n",
    "    library=\"GWAS\", \n",
    "    prog_key=\"cNMF\",\n",
    "    geneset_index=\"Term\",\n",
    "    program_index=\"ProgramID\",\n",
    "    varmap_name_prefix=\"gwas_varmap\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Add more modalities"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 385,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre>MuData object with n_obs × n_vars = 4000 × 17562\n",
       "  29 modalities\n",
       "    rna:\t4000 x 17472\n",
       "      obs:\t&#x27;barcodes&#x27;, &#x27;n_genes&#x27;, &#x27;n_counts&#x27;, &#x27;sample&#x27;, &#x27;batch&#x27;\n",
       "      var:\t&#x27;n_cells&#x27;\n",
       "    cNMF:\t4000 x 60\n",
       "      obs:\t&#x27;sample&#x27;, &#x27;batch&#x27;\n",
       "      var:\t&#x27;VarianceExplained&#x27;\n",
       "      uns:\t&#x27;var_names&#x27;, &#x27;gsea_varmap_GSEA&#x27;, &#x27;motif_varmap_Motif&#x27;, &#x27;gwas_varmap_GWAS&#x27;\n",
       "      varm:\t&#x27;loadings&#x27;, &#x27;loadings_zscore&#x27;, &#x27;Description_GSEA&#x27;, &#x27;pvalue_GSEA&#x27;, &#x27;p.adjust_GSEA&#x27;, &#x27;qvalue_GSEA&#x27;, &#x27;EPType_Motif&#x27;, &#x27;TFMotif_Motif&#x27;, &#x27;PValue_Motif&#x27;, &#x27;FDR_Motif&#x27;, &#x27;Enrichment_Motif&#x27;, &#x27;program_name_GWAS&#x27;, &#x27;Gene_set_GWAS&#x27;, &#x27;P-value_GWAS&#x27;, &#x27;Adjusted_P-value_GWAS&#x27;, &#x27;Odds_Ratio_GWAS&#x27;, &#x27;Combined_Score_GWAS&#x27;, &#x27;Genes_GWAS&#x27;, &#x27;overlap_numerator_GWAS&#x27;, &#x27;overlap_denominator_GWAS&#x27;\n",
       "    cNMF_30:\t4000 x 30\n",
       "      obs:\t&#x27;sample&#x27;, &#x27;batch&#x27;\n",
       "      var:\t&#x27;VarianceExplained&#x27;\n",
       "      uns:\t&#x27;var_names&#x27;, &#x27;gsea_varmap_GSEA&#x27;, &#x27;motif_varmap_Motif&#x27;, &#x27;gwas_varmap_GWAS&#x27;\n",
       "      varm:\t&#x27;loadings&#x27;, &#x27;loadings_zscore&#x27;, &#x27;Description_GSEA&#x27;, &#x27;pvalue_GSEA&#x27;, &#x27;p.adjust_GSEA&#x27;, &#x27;qvalue_GSEA&#x27;, &#x27;EPType_Motif&#x27;, &#x27;TFMotif_Motif&#x27;, &#x27;PValue_Motif&#x27;, &#x27;FDR_Motif&#x27;, &#x27;Enrichment_Motif&#x27;, &#x27;program_name_GWAS&#x27;, &#x27;Gene_set_GWAS&#x27;, &#x27;P-value_GWAS&#x27;, &#x27;Adjusted_P-value_GWAS&#x27;, &#x27;Odds_Ratio_GWAS&#x27;, &#x27;Combined_Score_GWAS&#x27;, &#x27;Genes_GWAS&#x27;, &#x27;overlap_numerator_GWAS&#x27;, &#x27;overlap_denominator_GWAS&#x27;\n",
       "    cNMF_3:\t4000 x 3\n",
       "      obs:\t&#x27;sample&#x27;, &#x27;batch&#x27;\n",
       "      var:\t&#x27;VarianceExplained&#x27;\n",
       "      uns:\t&#x27;var_names&#x27;, &#x27;gsea_varmap_GSEA&#x27;, &#x27;motif_varmap_Motif&#x27;, &#x27;gwas_varmap_GWAS&#x27;\n",
       "      varm:\t&#x27;loadings&#x27;, &#x27;loadings_zscore&#x27;, &#x27;Description_GSEA&#x27;, &#x27;pvalue_GSEA&#x27;, &#x27;p.adjust_GSEA&#x27;, &#x27;qvalue_GSEA&#x27;, &#x27;EPType_Motif&#x27;, &#x27;TFMotif_Motif&#x27;, &#x27;PValue_Motif&#x27;, &#x27;FDR_Motif&#x27;, &#x27;Enrichment_Motif&#x27;, &#x27;program_name_GWAS&#x27;, &#x27;Gene_set_GWAS&#x27;, &#x27;P-value_GWAS&#x27;, &#x27;Adjusted_P-value_GWAS&#x27;, &#x27;Odds_Ratio_GWAS&#x27;, &#x27;Combined_Score_GWAS&#x27;, &#x27;Genes_GWAS&#x27;, &#x27;overlap_numerator_GWAS&#x27;, &#x27;overlap_denominator_GWAS&#x27;\n",
       "    cNMF_4:\t4000 x 4\n",
       "      obs:\t&#x27;sample&#x27;, &#x27;batch&#x27;\n",
       "      var:\t&#x27;VarianceExplained&#x27;\n",
       "      uns:\t&#x27;var_names&#x27;, &#x27;gsea_varmap_GSEA&#x27;, &#x27;motif_varmap_Motif&#x27;, &#x27;gwas_varmap_GWAS&#x27;\n",
       "      varm:\t&#x27;loadings&#x27;, &#x27;loadings_zscore&#x27;, &#x27;Description_GSEA&#x27;, &#x27;pvalue_GSEA&#x27;, &#x27;p.adjust_GSEA&#x27;, &#x27;qvalue_GSEA&#x27;, &#x27;EPType_Motif&#x27;, &#x27;TFMotif_Motif&#x27;, &#x27;PValue_Motif&#x27;, &#x27;FDR_Motif&#x27;, &#x27;Enrichment_Motif&#x27;, &#x27;program_name_GWAS&#x27;, &#x27;Gene_set_GWAS&#x27;, &#x27;P-value_GWAS&#x27;, &#x27;Adjusted_P-value_GWAS&#x27;, &#x27;Odds_Ratio_GWAS&#x27;, &#x27;Combined_Score_GWAS&#x27;, &#x27;Genes_GWAS&#x27;, &#x27;overlap_numerator_GWAS&#x27;, &#x27;overlap_denominator_GWAS&#x27;\n",
       "    cNMF_5:\t4000 x 5\n",
       "      obs:\t&#x27;sample&#x27;, &#x27;batch&#x27;\n",
       "      var:\t&#x27;VarianceExplained&#x27;\n",
       "      uns:\t&#x27;var_names&#x27;, &#x27;gsea_varmap_GSEA&#x27;, &#x27;motif_varmap_Motif&#x27;, &#x27;gwas_varmap_GWAS&#x27;\n",
       "      varm:\t&#x27;loadings&#x27;, &#x27;loadings_zscore&#x27;, &#x27;Description_GSEA&#x27;, &#x27;pvalue_GSEA&#x27;, &#x27;p.adjust_GSEA&#x27;, &#x27;qvalue_GSEA&#x27;, &#x27;EPType_Motif&#x27;, &#x27;TFMotif_Motif&#x27;, &#x27;PValue_Motif&#x27;, &#x27;FDR_Motif&#x27;, &#x27;Enrichment_Motif&#x27;, &#x27;program_name_GWAS&#x27;, &#x27;Gene_set_GWAS&#x27;, &#x27;P-value_GWAS&#x27;, &#x27;Adjusted_P-value_GWAS&#x27;, &#x27;Odds_Ratio_GWAS&#x27;, &#x27;Combined_Score_GWAS&#x27;, &#x27;Genes_GWAS&#x27;, &#x27;overlap_numerator_GWAS&#x27;, &#x27;overlap_denominator_GWAS&#x27;\n",
       "    cNMF_6:\t4000 x 6\n",
       "      obs:\t&#x27;sample&#x27;, &#x27;batch&#x27;\n",
       "      var:\t&#x27;VarianceExplained&#x27;\n",
       "      uns:\t&#x27;var_names&#x27;, &#x27;gsea_varmap_GSEA&#x27;, &#x27;motif_varmap_Motif&#x27;, &#x27;gwas_varmap_GWAS&#x27;\n",
       "      varm:\t&#x27;loadings&#x27;, &#x27;loadings_zscore&#x27;, &#x27;Description_GSEA&#x27;, &#x27;pvalue_GSEA&#x27;, &#x27;p.adjust_GSEA&#x27;, &#x27;qvalue_GSEA&#x27;, &#x27;EPType_Motif&#x27;, &#x27;TFMotif_Motif&#x27;, &#x27;PValue_Motif&#x27;, &#x27;FDR_Motif&#x27;, &#x27;Enrichment_Motif&#x27;, &#x27;program_name_GWAS&#x27;, &#x27;Gene_set_GWAS&#x27;, &#x27;P-value_GWAS&#x27;, &#x27;Adjusted_P-value_GWAS&#x27;, &#x27;Odds_Ratio_GWAS&#x27;, &#x27;Combined_Score_GWAS&#x27;, &#x27;Genes_GWAS&#x27;, &#x27;overlap_numerator_GWAS&#x27;, &#x27;overlap_denominator_GWAS&#x27;\n",
       "    cNMF_7:\t4000 x 7\n",
       "      obs:\t&#x27;sample&#x27;, &#x27;batch&#x27;\n",
       "      var:\t&#x27;VarianceExplained&#x27;\n",
       "      uns:\t&#x27;var_names&#x27;, &#x27;gsea_varmap_GSEA&#x27;, &#x27;motif_varmap_Motif&#x27;, &#x27;gwas_varmap_GWAS&#x27;\n",
       "      varm:\t&#x27;loadings&#x27;, &#x27;loadings_zscore&#x27;, &#x27;Description_GSEA&#x27;, &#x27;pvalue_GSEA&#x27;, &#x27;p.adjust_GSEA&#x27;, &#x27;qvalue_GSEA&#x27;, &#x27;EPType_Motif&#x27;, &#x27;TFMotif_Motif&#x27;, &#x27;PValue_Motif&#x27;, &#x27;FDR_Motif&#x27;, &#x27;Enrichment_Motif&#x27;, &#x27;program_name_GWAS&#x27;, &#x27;Gene_set_GWAS&#x27;, &#x27;P-value_GWAS&#x27;, &#x27;Adjusted_P-value_GWAS&#x27;, &#x27;Odds_Ratio_GWAS&#x27;, &#x27;Combined_Score_GWAS&#x27;, &#x27;Genes_GWAS&#x27;, &#x27;overlap_numerator_GWAS&#x27;, &#x27;overlap_denominator_GWAS&#x27;\n",
       "    cNMF_8:\t4000 x 8\n",
       "      obs:\t&#x27;sample&#x27;, &#x27;batch&#x27;\n",
       "      var:\t&#x27;VarianceExplained&#x27;\n",
       "      uns:\t&#x27;var_names&#x27;, &#x27;gsea_varmap_GSEA&#x27;, &#x27;motif_varmap_Motif&#x27;, &#x27;gwas_varmap_GWAS&#x27;\n",
       "      varm:\t&#x27;loadings&#x27;, &#x27;loadings_zscore&#x27;, &#x27;Description_GSEA&#x27;, &#x27;pvalue_GSEA&#x27;, &#x27;p.adjust_GSEA&#x27;, &#x27;qvalue_GSEA&#x27;, &#x27;EPType_Motif&#x27;, &#x27;TFMotif_Motif&#x27;, &#x27;PValue_Motif&#x27;, &#x27;FDR_Motif&#x27;, &#x27;Enrichment_Motif&#x27;, &#x27;program_name_GWAS&#x27;, &#x27;Gene_set_GWAS&#x27;, &#x27;P-value_GWAS&#x27;, &#x27;Adjusted_P-value_GWAS&#x27;, &#x27;Odds_Ratio_GWAS&#x27;, &#x27;Combined_Score_GWAS&#x27;, &#x27;Genes_GWAS&#x27;, &#x27;overlap_numerator_GWAS&#x27;, &#x27;overlap_denominator_GWAS&#x27;\n",
       "    cNMF_9:\t4000 x 9\n",
       "      obs:\t&#x27;sample&#x27;, &#x27;batch&#x27;\n",
       "      var:\t&#x27;VarianceExplained&#x27;\n",
       "      uns:\t&#x27;var_names&#x27;, &#x27;gsea_varmap_GSEA&#x27;, &#x27;motif_varmap_Motif&#x27;, &#x27;gwas_varmap_GWAS&#x27;\n",
       "      varm:\t&#x27;loadings&#x27;, &#x27;loadings_zscore&#x27;, &#x27;Description_GSEA&#x27;, &#x27;pvalue_GSEA&#x27;, &#x27;p.adjust_GSEA&#x27;, &#x27;qvalue_GSEA&#x27;, &#x27;EPType_Motif&#x27;, &#x27;TFMotif_Motif&#x27;, &#x27;PValue_Motif&#x27;, &#x27;FDR_Motif&#x27;, &#x27;Enrichment_Motif&#x27;, &#x27;program_name_GWAS&#x27;, &#x27;Gene_set_GWAS&#x27;, &#x27;P-value_GWAS&#x27;, &#x27;Adjusted_P-value_GWAS&#x27;, &#x27;Odds_Ratio_GWAS&#x27;, &#x27;Combined_Score_GWAS&#x27;, &#x27;Genes_GWAS&#x27;, &#x27;overlap_numerator_GWAS&#x27;, &#x27;overlap_denominator_GWAS&#x27;\n",
       "    cNMF_10:\t4000 x 10\n",
       "      obs:\t&#x27;sample&#x27;, &#x27;batch&#x27;\n",
       "      var:\t&#x27;VarianceExplained&#x27;\n",
       "      uns:\t&#x27;var_names&#x27;, &#x27;gsea_varmap_GSEA&#x27;, &#x27;motif_varmap_Motif&#x27;, &#x27;gwas_varmap_GWAS&#x27;\n",
       "      varm:\t&#x27;loadings&#x27;, &#x27;loadings_zscore&#x27;, &#x27;Description_GSEA&#x27;, &#x27;pvalue_GSEA&#x27;, &#x27;p.adjust_GSEA&#x27;, &#x27;qvalue_GSEA&#x27;, &#x27;EPType_Motif&#x27;, &#x27;TFMotif_Motif&#x27;, &#x27;PValue_Motif&#x27;, &#x27;FDR_Motif&#x27;, &#x27;Enrichment_Motif&#x27;, &#x27;program_name_GWAS&#x27;, &#x27;Gene_set_GWAS&#x27;, &#x27;P-value_GWAS&#x27;, &#x27;Adjusted_P-value_GWAS&#x27;, &#x27;Odds_Ratio_GWAS&#x27;, &#x27;Combined_Score_GWAS&#x27;, &#x27;Genes_GWAS&#x27;, &#x27;overlap_numerator_GWAS&#x27;, &#x27;overlap_denominator_GWAS&#x27;\n",
       "    cNMF_11:\t4000 x 11\n",
       "      obs:\t&#x27;sample&#x27;, &#x27;batch&#x27;\n",
       "      var:\t&#x27;VarianceExplained&#x27;\n",
       "      uns:\t&#x27;var_names&#x27;, &#x27;gsea_varmap_GSEA&#x27;, &#x27;motif_varmap_Motif&#x27;, &#x27;gwas_varmap_GWAS&#x27;\n",
       "      varm:\t&#x27;loadings&#x27;, &#x27;loadings_zscore&#x27;, &#x27;Description_GSEA&#x27;, &#x27;pvalue_GSEA&#x27;, &#x27;p.adjust_GSEA&#x27;, &#x27;qvalue_GSEA&#x27;, &#x27;EPType_Motif&#x27;, &#x27;TFMotif_Motif&#x27;, &#x27;PValue_Motif&#x27;, &#x27;FDR_Motif&#x27;, &#x27;Enrichment_Motif&#x27;, &#x27;program_name_GWAS&#x27;, &#x27;Gene_set_GWAS&#x27;, &#x27;P-value_GWAS&#x27;, &#x27;Adjusted_P-value_GWAS&#x27;, &#x27;Odds_Ratio_GWAS&#x27;, &#x27;Combined_Score_GWAS&#x27;, &#x27;Genes_GWAS&#x27;, &#x27;overlap_numerator_GWAS&#x27;, &#x27;overlap_denominator_GWAS&#x27;\n",
       "    cNMF_12:\t4000 x 12\n",
       "      obs:\t&#x27;sample&#x27;, &#x27;batch&#x27;\n",
       "      var:\t&#x27;VarianceExplained&#x27;\n",
       "      uns:\t&#x27;var_names&#x27;, &#x27;gsea_varmap_GSEA&#x27;, &#x27;motif_varmap_Motif&#x27;, &#x27;gwas_varmap_GWAS&#x27;\n",
       "      varm:\t&#x27;loadings&#x27;, &#x27;loadings_zscore&#x27;, &#x27;Description_GSEA&#x27;, &#x27;pvalue_GSEA&#x27;, &#x27;p.adjust_GSEA&#x27;, &#x27;qvalue_GSEA&#x27;, &#x27;EPType_Motif&#x27;, &#x27;TFMotif_Motif&#x27;, &#x27;PValue_Motif&#x27;, &#x27;FDR_Motif&#x27;, &#x27;Enrichment_Motif&#x27;, &#x27;program_name_GWAS&#x27;, &#x27;Gene_set_GWAS&#x27;, &#x27;P-value_GWAS&#x27;, &#x27;Adjusted_P-value_GWAS&#x27;, &#x27;Odds_Ratio_GWAS&#x27;, &#x27;Combined_Score_GWAS&#x27;, &#x27;Genes_GWAS&#x27;, &#x27;overlap_numerator_GWAS&#x27;, &#x27;overlap_denominator_GWAS&#x27;\n",
       "    cNMF_13:\t4000 x 13\n",
       "      obs:\t&#x27;sample&#x27;, &#x27;batch&#x27;\n",
       "      var:\t&#x27;VarianceExplained&#x27;\n",
       "      uns:\t&#x27;var_names&#x27;, &#x27;gsea_varmap_GSEA&#x27;, &#x27;motif_varmap_Motif&#x27;, &#x27;gwas_varmap_GWAS&#x27;\n",
       "      varm:\t&#x27;loadings&#x27;, &#x27;loadings_zscore&#x27;, &#x27;Description_GSEA&#x27;, &#x27;pvalue_GSEA&#x27;, &#x27;p.adjust_GSEA&#x27;, &#x27;qvalue_GSEA&#x27;, &#x27;EPType_Motif&#x27;, &#x27;TFMotif_Motif&#x27;, &#x27;PValue_Motif&#x27;, &#x27;FDR_Motif&#x27;, &#x27;Enrichment_Motif&#x27;, &#x27;program_name_GWAS&#x27;, &#x27;Gene_set_GWAS&#x27;, &#x27;P-value_GWAS&#x27;, &#x27;Adjusted_P-value_GWAS&#x27;, &#x27;Odds_Ratio_GWAS&#x27;, &#x27;Combined_Score_GWAS&#x27;, &#x27;Genes_GWAS&#x27;, &#x27;overlap_numerator_GWAS&#x27;, &#x27;overlap_denominator_GWAS&#x27;\n",
       "    cNMF_14:\t4000 x 14\n",
       "      obs:\t&#x27;sample&#x27;, &#x27;batch&#x27;\n",
       "      var:\t&#x27;VarianceExplained&#x27;\n",
       "      uns:\t&#x27;var_names&#x27;, &#x27;gsea_varmap_GSEA&#x27;, &#x27;motif_varmap_Motif&#x27;, &#x27;gwas_varmap_GWAS&#x27;\n",
       "      varm:\t&#x27;loadings&#x27;, &#x27;loadings_zscore&#x27;, &#x27;Description_GSEA&#x27;, &#x27;pvalue_GSEA&#x27;, &#x27;p.adjust_GSEA&#x27;, &#x27;qvalue_GSEA&#x27;, &#x27;EPType_Motif&#x27;, &#x27;TFMotif_Motif&#x27;, &#x27;PValue_Motif&#x27;, &#x27;FDR_Motif&#x27;, &#x27;Enrichment_Motif&#x27;, &#x27;program_name_GWAS&#x27;, &#x27;Gene_set_GWAS&#x27;, &#x27;P-value_GWAS&#x27;, &#x27;Adjusted_P-value_GWAS&#x27;, &#x27;Odds_Ratio_GWAS&#x27;, &#x27;Combined_Score_GWAS&#x27;, &#x27;Genes_GWAS&#x27;, &#x27;overlap_numerator_GWAS&#x27;, &#x27;overlap_denominator_GWAS&#x27;\n",
       "    cNMF_15:\t4000 x 15\n",
       "      obs:\t&#x27;sample&#x27;, &#x27;batch&#x27;\n",
       "      var:\t&#x27;VarianceExplained&#x27;\n",
       "      uns:\t&#x27;var_names&#x27;, &#x27;gsea_varmap_GSEA&#x27;, &#x27;motif_varmap_Motif&#x27;, &#x27;gwas_varmap_GWAS&#x27;\n",
       "      varm:\t&#x27;loadings&#x27;, &#x27;loadings_zscore&#x27;, &#x27;Description_GSEA&#x27;, &#x27;pvalue_GSEA&#x27;, &#x27;p.adjust_GSEA&#x27;, &#x27;qvalue_GSEA&#x27;, &#x27;EPType_Motif&#x27;, &#x27;TFMotif_Motif&#x27;, &#x27;PValue_Motif&#x27;, &#x27;FDR_Motif&#x27;, &#x27;Enrichment_Motif&#x27;, &#x27;program_name_GWAS&#x27;, &#x27;Gene_set_GWAS&#x27;, &#x27;P-value_GWAS&#x27;, &#x27;Adjusted_P-value_GWAS&#x27;, &#x27;Odds_Ratio_GWAS&#x27;, &#x27;Combined_Score_GWAS&#x27;, &#x27;Genes_GWAS&#x27;, &#x27;overlap_numerator_GWAS&#x27;, &#x27;overlap_denominator_GWAS&#x27;\n",
       "    cNMF_17:\t4000 x 17\n",
       "      obs:\t&#x27;sample&#x27;, &#x27;batch&#x27;\n",
       "      var:\t&#x27;VarianceExplained&#x27;\n",
       "      uns:\t&#x27;var_names&#x27;, &#x27;gsea_varmap_GSEA&#x27;, &#x27;motif_varmap_Motif&#x27;, &#x27;gwas_varmap_GWAS&#x27;\n",
       "      varm:\t&#x27;loadings&#x27;, &#x27;loadings_zscore&#x27;, &#x27;Description_GSEA&#x27;, &#x27;pvalue_GSEA&#x27;, &#x27;p.adjust_GSEA&#x27;, &#x27;qvalue_GSEA&#x27;, &#x27;EPType_Motif&#x27;, &#x27;TFMotif_Motif&#x27;, &#x27;PValue_Motif&#x27;, &#x27;FDR_Motif&#x27;, &#x27;Enrichment_Motif&#x27;, &#x27;program_name_GWAS&#x27;, &#x27;Gene_set_GWAS&#x27;, &#x27;P-value_GWAS&#x27;, &#x27;Adjusted_P-value_GWAS&#x27;, &#x27;Odds_Ratio_GWAS&#x27;, &#x27;Combined_Score_GWAS&#x27;, &#x27;Genes_GWAS&#x27;, &#x27;overlap_numerator_GWAS&#x27;, &#x27;overlap_denominator_GWAS&#x27;\n",
       "    cNMF_19:\t4000 x 19\n",
       "      obs:\t&#x27;sample&#x27;, &#x27;batch&#x27;\n",
       "      var:\t&#x27;VarianceExplained&#x27;\n",
       "      uns:\t&#x27;var_names&#x27;, &#x27;gsea_varmap_GSEA&#x27;, &#x27;motif_varmap_Motif&#x27;, &#x27;gwas_varmap_GWAS&#x27;\n",
       "      varm:\t&#x27;loadings&#x27;, &#x27;loadings_zscore&#x27;, &#x27;Description_GSEA&#x27;, &#x27;pvalue_GSEA&#x27;, &#x27;p.adjust_GSEA&#x27;, &#x27;qvalue_GSEA&#x27;, &#x27;EPType_Motif&#x27;, &#x27;TFMotif_Motif&#x27;, &#x27;PValue_Motif&#x27;, &#x27;FDR_Motif&#x27;, &#x27;Enrichment_Motif&#x27;, &#x27;program_name_GWAS&#x27;, &#x27;Gene_set_GWAS&#x27;, &#x27;P-value_GWAS&#x27;, &#x27;Adjusted_P-value_GWAS&#x27;, &#x27;Odds_Ratio_GWAS&#x27;, &#x27;Combined_Score_GWAS&#x27;, &#x27;Genes_GWAS&#x27;, &#x27;overlap_numerator_GWAS&#x27;, &#x27;overlap_denominator_GWAS&#x27;\n",
       "    cNMF_21:\t4000 x 21\n",
       "      obs:\t&#x27;sample&#x27;, &#x27;batch&#x27;\n",
       "      var:\t&#x27;VarianceExplained&#x27;\n",
       "      uns:\t&#x27;var_names&#x27;, &#x27;gsea_varmap_GSEA&#x27;, &#x27;motif_varmap_Motif&#x27;, &#x27;gwas_varmap_GWAS&#x27;\n",
       "      varm:\t&#x27;loadings&#x27;, &#x27;loadings_zscore&#x27;, &#x27;Description_GSEA&#x27;, &#x27;pvalue_GSEA&#x27;, &#x27;p.adjust_GSEA&#x27;, &#x27;qvalue_GSEA&#x27;, &#x27;EPType_Motif&#x27;, &#x27;TFMotif_Motif&#x27;, &#x27;PValue_Motif&#x27;, &#x27;FDR_Motif&#x27;, &#x27;Enrichment_Motif&#x27;, &#x27;program_name_GWAS&#x27;, &#x27;Gene_set_GWAS&#x27;, &#x27;P-value_GWAS&#x27;, &#x27;Adjusted_P-value_GWAS&#x27;, &#x27;Odds_Ratio_GWAS&#x27;, &#x27;Combined_Score_GWAS&#x27;, &#x27;Genes_GWAS&#x27;, &#x27;overlap_numerator_GWAS&#x27;, &#x27;overlap_denominator_GWAS&#x27;\n",
       "    cNMF_23:\t4000 x 23\n",
       "      obs:\t&#x27;sample&#x27;, &#x27;batch&#x27;\n",
       "      var:\t&#x27;VarianceExplained&#x27;\n",
       "      uns:\t&#x27;var_names&#x27;, &#x27;gsea_varmap_GSEA&#x27;, &#x27;motif_varmap_Motif&#x27;, &#x27;gwas_varmap_GWAS&#x27;\n",
       "      varm:\t&#x27;loadings&#x27;, &#x27;loadings_zscore&#x27;, &#x27;Description_GSEA&#x27;, &#x27;pvalue_GSEA&#x27;, &#x27;p.adjust_GSEA&#x27;, &#x27;qvalue_GSEA&#x27;, &#x27;EPType_Motif&#x27;, &#x27;TFMotif_Motif&#x27;, &#x27;PValue_Motif&#x27;, &#x27;FDR_Motif&#x27;, &#x27;Enrichment_Motif&#x27;, &#x27;program_name_GWAS&#x27;, &#x27;Gene_set_GWAS&#x27;, &#x27;P-value_GWAS&#x27;, &#x27;Adjusted_P-value_GWAS&#x27;, &#x27;Odds_Ratio_GWAS&#x27;, &#x27;Combined_Score_GWAS&#x27;, &#x27;Genes_GWAS&#x27;, &#x27;overlap_numerator_GWAS&#x27;, &#x27;overlap_denominator_GWAS&#x27;\n",
       "    cNMF_25:\t4000 x 25\n",
       "      obs:\t&#x27;sample&#x27;, &#x27;batch&#x27;\n",
       "      var:\t&#x27;VarianceExplained&#x27;\n",
       "      uns:\t&#x27;var_names&#x27;, &#x27;gsea_varmap_GSEA&#x27;, &#x27;motif_varmap_Motif&#x27;, &#x27;gwas_varmap_GWAS&#x27;\n",
       "      varm:\t&#x27;loadings&#x27;, &#x27;loadings_zscore&#x27;, &#x27;Description_GSEA&#x27;, &#x27;pvalue_GSEA&#x27;, &#x27;p.adjust_GSEA&#x27;, &#x27;qvalue_GSEA&#x27;, &#x27;EPType_Motif&#x27;, &#x27;TFMotif_Motif&#x27;, &#x27;PValue_Motif&#x27;, &#x27;FDR_Motif&#x27;, &#x27;Enrichment_Motif&#x27;, &#x27;program_name_GWAS&#x27;, &#x27;Gene_set_GWAS&#x27;, &#x27;P-value_GWAS&#x27;, &#x27;Adjusted_P-value_GWAS&#x27;, &#x27;Odds_Ratio_GWAS&#x27;, &#x27;Combined_Score_GWAS&#x27;, &#x27;Genes_GWAS&#x27;, &#x27;overlap_numerator_GWAS&#x27;, &#x27;overlap_denominator_GWAS&#x27;\n",
       "    cNMF_27:\t4000 x 27\n",
       "      obs:\t&#x27;sample&#x27;, &#x27;batch&#x27;\n",
       "      var:\t&#x27;VarianceExplained&#x27;\n",
       "      uns:\t&#x27;var_names&#x27;, &#x27;gsea_varmap_GSEA&#x27;, &#x27;motif_varmap_Motif&#x27;, &#x27;gwas_varmap_GWAS&#x27;\n",
       "      varm:\t&#x27;loadings&#x27;, &#x27;loadings_zscore&#x27;, &#x27;Description_GSEA&#x27;, &#x27;pvalue_GSEA&#x27;, &#x27;p.adjust_GSEA&#x27;, &#x27;qvalue_GSEA&#x27;, &#x27;EPType_Motif&#x27;, &#x27;TFMotif_Motif&#x27;, &#x27;PValue_Motif&#x27;, &#x27;FDR_Motif&#x27;, &#x27;Enrichment_Motif&#x27;, &#x27;program_name_GWAS&#x27;, &#x27;Gene_set_GWAS&#x27;, &#x27;P-value_GWAS&#x27;, &#x27;Adjusted_P-value_GWAS&#x27;, &#x27;Odds_Ratio_GWAS&#x27;, &#x27;Combined_Score_GWAS&#x27;, &#x27;Genes_GWAS&#x27;, &#x27;overlap_numerator_GWAS&#x27;, &#x27;overlap_denominator_GWAS&#x27;\n",
       "    cNMF_29:\t4000 x 29\n",
       "      obs:\t&#x27;sample&#x27;, &#x27;batch&#x27;\n",
       "      var:\t&#x27;VarianceExplained&#x27;\n",
       "      uns:\t&#x27;var_names&#x27;, &#x27;gsea_varmap_GSEA&#x27;, &#x27;motif_varmap_Motif&#x27;, &#x27;gwas_varmap_GWAS&#x27;\n",
       "      varm:\t&#x27;loadings&#x27;, &#x27;loadings_zscore&#x27;, &#x27;Description_GSEA&#x27;, &#x27;pvalue_GSEA&#x27;, &#x27;p.adjust_GSEA&#x27;, &#x27;qvalue_GSEA&#x27;, &#x27;EPType_Motif&#x27;, &#x27;TFMotif_Motif&#x27;, &#x27;PValue_Motif&#x27;, &#x27;FDR_Motif&#x27;, &#x27;Enrichment_Motif&#x27;, &#x27;program_name_GWAS&#x27;, &#x27;Gene_set_GWAS&#x27;, &#x27;P-value_GWAS&#x27;, &#x27;Adjusted_P-value_GWAS&#x27;, &#x27;Odds_Ratio_GWAS&#x27;, &#x27;Combined_Score_GWAS&#x27;, &#x27;Genes_GWAS&#x27;, &#x27;overlap_numerator_GWAS&#x27;, &#x27;overlap_denominator_GWAS&#x27;\n",
       "    cNMF_35:\t4000 x 35\n",
       "      obs:\t&#x27;sample&#x27;, &#x27;batch&#x27;\n",
       "      var:\t&#x27;VarianceExplained&#x27;\n",
       "      uns:\t&#x27;var_names&#x27;, &#x27;gsea_varmap_GSEA&#x27;, &#x27;motif_varmap_Motif&#x27;, &#x27;gwas_varmap_GWAS&#x27;\n",
       "      varm:\t&#x27;loadings&#x27;, &#x27;loadings_zscore&#x27;, &#x27;Description_GSEA&#x27;, &#x27;pvalue_GSEA&#x27;, &#x27;p.adjust_GSEA&#x27;, &#x27;qvalue_GSEA&#x27;, &#x27;EPType_Motif&#x27;, &#x27;TFMotif_Motif&#x27;, &#x27;PValue_Motif&#x27;, &#x27;FDR_Motif&#x27;, &#x27;Enrichment_Motif&#x27;, &#x27;program_name_GWAS&#x27;, &#x27;Gene_set_GWAS&#x27;, &#x27;P-value_GWAS&#x27;, &#x27;Adjusted_P-value_GWAS&#x27;, &#x27;Odds_Ratio_GWAS&#x27;, &#x27;Combined_Score_GWAS&#x27;, &#x27;Genes_GWAS&#x27;, &#x27;overlap_numerator_GWAS&#x27;, &#x27;overlap_denominator_GWAS&#x27;\n",
       "    cNMF_40:\t4000 x 40\n",
       "      obs:\t&#x27;sample&#x27;, &#x27;batch&#x27;\n",
       "      var:\t&#x27;VarianceExplained&#x27;\n",
       "      uns:\t&#x27;var_names&#x27;, &#x27;gsea_varmap_GSEA&#x27;, &#x27;motif_varmap_Motif&#x27;, &#x27;gwas_varmap_GWAS&#x27;\n",
       "      varm:\t&#x27;loadings&#x27;, &#x27;loadings_zscore&#x27;, &#x27;Description_GSEA&#x27;, &#x27;pvalue_GSEA&#x27;, &#x27;p.adjust_GSEA&#x27;, &#x27;qvalue_GSEA&#x27;, &#x27;EPType_Motif&#x27;, &#x27;TFMotif_Motif&#x27;, &#x27;PValue_Motif&#x27;, &#x27;FDR_Motif&#x27;, &#x27;Enrichment_Motif&#x27;, &#x27;program_name_GWAS&#x27;, &#x27;Gene_set_GWAS&#x27;, &#x27;P-value_GWAS&#x27;, &#x27;Adjusted_P-value_GWAS&#x27;, &#x27;Odds_Ratio_GWAS&#x27;, &#x27;Combined_Score_GWAS&#x27;, &#x27;Genes_GWAS&#x27;, &#x27;overlap_numerator_GWAS&#x27;, &#x27;overlap_denominator_GWAS&#x27;\n",
       "    cNMF_45:\t4000 x 45\n",
       "      obs:\t&#x27;sample&#x27;, &#x27;batch&#x27;\n",
       "      var:\t&#x27;VarianceExplained&#x27;\n",
       "      uns:\t&#x27;var_names&#x27;, &#x27;gsea_varmap_GSEA&#x27;, &#x27;motif_varmap_Motif&#x27;, &#x27;gwas_varmap_GWAS&#x27;\n",
       "      varm:\t&#x27;loadings&#x27;, &#x27;loadings_zscore&#x27;, &#x27;Description_GSEA&#x27;, &#x27;pvalue_GSEA&#x27;, &#x27;p.adjust_GSEA&#x27;, &#x27;qvalue_GSEA&#x27;, &#x27;EPType_Motif&#x27;, &#x27;TFMotif_Motif&#x27;, &#x27;PValue_Motif&#x27;, &#x27;FDR_Motif&#x27;, &#x27;Enrichment_Motif&#x27;, &#x27;program_name_GWAS&#x27;, &#x27;Gene_set_GWAS&#x27;, &#x27;P-value_GWAS&#x27;, &#x27;Adjusted_P-value_GWAS&#x27;, &#x27;Odds_Ratio_GWAS&#x27;, &#x27;Combined_Score_GWAS&#x27;, &#x27;Genes_GWAS&#x27;, &#x27;overlap_numerator_GWAS&#x27;, &#x27;overlap_denominator_GWAS&#x27;\n",
       "    cNMF_50:\t4000 x 50\n",
       "      obs:\t&#x27;sample&#x27;, &#x27;batch&#x27;\n",
       "      var:\t&#x27;VarianceExplained&#x27;\n",
       "      uns:\t&#x27;var_names&#x27;, &#x27;gsea_varmap_GSEA&#x27;, &#x27;motif_varmap_Motif&#x27;, &#x27;gwas_varmap_GWAS&#x27;\n",
       "      varm:\t&#x27;loadings&#x27;, &#x27;loadings_zscore&#x27;, &#x27;Description_GSEA&#x27;, &#x27;pvalue_GSEA&#x27;, &#x27;p.adjust_GSEA&#x27;, &#x27;qvalue_GSEA&#x27;, &#x27;EPType_Motif&#x27;, &#x27;TFMotif_Motif&#x27;, &#x27;PValue_Motif&#x27;, &#x27;FDR_Motif&#x27;, &#x27;Enrichment_Motif&#x27;, &#x27;program_name_GWAS&#x27;, &#x27;Gene_set_GWAS&#x27;, &#x27;P-value_GWAS&#x27;, &#x27;Adjusted_P-value_GWAS&#x27;, &#x27;Odds_Ratio_GWAS&#x27;, &#x27;Combined_Score_GWAS&#x27;, &#x27;Genes_GWAS&#x27;, &#x27;overlap_numerator_GWAS&#x27;, &#x27;overlap_denominator_GWAS&#x27;\n",
       "    cNMF_55:\t4000 x 55\n",
       "      obs:\t&#x27;sample&#x27;, &#x27;batch&#x27;\n",
       "      var:\t&#x27;VarianceExplained&#x27;\n",
       "      uns:\t&#x27;var_names&#x27;, &#x27;gsea_varmap_GSEA&#x27;, &#x27;motif_varmap_Motif&#x27;, &#x27;gwas_varmap_GWAS&#x27;\n",
       "      varm:\t&#x27;loadings&#x27;, &#x27;loadings_zscore&#x27;, &#x27;Description_GSEA&#x27;, &#x27;pvalue_GSEA&#x27;, &#x27;p.adjust_GSEA&#x27;, &#x27;qvalue_GSEA&#x27;, &#x27;EPType_Motif&#x27;, &#x27;TFMotif_Motif&#x27;, &#x27;PValue_Motif&#x27;, &#x27;FDR_Motif&#x27;, &#x27;Enrichment_Motif&#x27;, &#x27;program_name_GWAS&#x27;, &#x27;Gene_set_GWAS&#x27;, &#x27;P-value_GWAS&#x27;, &#x27;Adjusted_P-value_GWAS&#x27;, &#x27;Odds_Ratio_GWAS&#x27;, &#x27;Combined_Score_GWAS&#x27;, &#x27;Genes_GWAS&#x27;, &#x27;overlap_numerator_GWAS&#x27;, &#x27;overlap_denominator_GWAS&#x27;\n",
       "    cNMF_60:\t4000 x 60\n",
       "      obs:\t&#x27;sample&#x27;, &#x27;batch&#x27;\n",
       "      var:\t&#x27;VarianceExplained&#x27;\n",
       "      uns:\t&#x27;var_names&#x27;, &#x27;gsea_varmap_GSEA&#x27;, &#x27;motif_varmap_Motif&#x27;, &#x27;gwas_varmap_GWAS&#x27;\n",
       "      varm:\t&#x27;loadings&#x27;, &#x27;loadings_zscore&#x27;, &#x27;Description_GSEA&#x27;, &#x27;pvalue_GSEA&#x27;, &#x27;p.adjust_GSEA&#x27;, &#x27;qvalue_GSEA&#x27;, &#x27;EPType_Motif&#x27;, &#x27;TFMotif_Motif&#x27;, &#x27;PValue_Motif&#x27;, &#x27;FDR_Motif&#x27;, &#x27;Enrichment_Motif&#x27;, &#x27;program_name_GWAS&#x27;, &#x27;Gene_set_GWAS&#x27;, &#x27;P-value_GWAS&#x27;, &#x27;Adjusted_P-value_GWAS&#x27;, &#x27;Odds_Ratio_GWAS&#x27;, &#x27;Combined_Score_GWAS&#x27;, &#x27;Genes_GWAS&#x27;, &#x27;overlap_numerator_GWAS&#x27;, &#x27;overlap_denominator_GWAS&#x27;</pre>"
      ],
      "text/plain": [
       "MuData object with n_obs × n_vars = 4000 × 17562\n",
       "  29 modalities\n",
       "    rna:\t4000 x 17472\n",
       "      obs:\t'barcodes', 'n_genes', 'n_counts', 'sample', 'batch'\n",
       "      var:\t'n_cells'\n",
       "    cNMF:\t4000 x 60\n",
       "      obs:\t'sample', 'batch'\n",
       "      var:\t'VarianceExplained'\n",
       "      uns:\t'var_names', 'gsea_varmap_GSEA', 'motif_varmap_Motif', 'gwas_varmap_GWAS'\n",
       "      varm:\t'loadings', 'loadings_zscore', 'Description_GSEA', 'pvalue_GSEA', 'p.adjust_GSEA', 'qvalue_GSEA', 'EPType_Motif', 'TFMotif_Motif', 'PValue_Motif', 'FDR_Motif', 'Enrichment_Motif', 'program_name_GWAS', 'Gene_set_GWAS', 'P-value_GWAS', 'Adjusted_P-value_GWAS', 'Odds_Ratio_GWAS', 'Combined_Score_GWAS', 'Genes_GWAS', 'overlap_numerator_GWAS', 'overlap_denominator_GWAS'\n",
       "    cNMF_30:\t4000 x 30\n",
       "      obs:\t'sample', 'batch'\n",
       "      var:\t'VarianceExplained'\n",
       "      uns:\t'var_names', 'gsea_varmap_GSEA', 'motif_varmap_Motif', 'gwas_varmap_GWAS'\n",
       "      varm:\t'loadings', 'loadings_zscore', 'Description_GSEA', 'pvalue_GSEA', 'p.adjust_GSEA', 'qvalue_GSEA', 'EPType_Motif', 'TFMotif_Motif', 'PValue_Motif', 'FDR_Motif', 'Enrichment_Motif', 'program_name_GWAS', 'Gene_set_GWAS', 'P-value_GWAS', 'Adjusted_P-value_GWAS', 'Odds_Ratio_GWAS', 'Combined_Score_GWAS', 'Genes_GWAS', 'overlap_numerator_GWAS', 'overlap_denominator_GWAS'\n",
       "    cNMF_3:\t4000 x 3\n",
       "      obs:\t'sample', 'batch'\n",
       "      var:\t'VarianceExplained'\n",
       "      uns:\t'var_names', 'gsea_varmap_GSEA', 'motif_varmap_Motif', 'gwas_varmap_GWAS'\n",
       "      varm:\t'loadings', 'loadings_zscore', 'Description_GSEA', 'pvalue_GSEA', 'p.adjust_GSEA', 'qvalue_GSEA', 'EPType_Motif', 'TFMotif_Motif', 'PValue_Motif', 'FDR_Motif', 'Enrichment_Motif', 'program_name_GWAS', 'Gene_set_GWAS', 'P-value_GWAS', 'Adjusted_P-value_GWAS', 'Odds_Ratio_GWAS', 'Combined_Score_GWAS', 'Genes_GWAS', 'overlap_numerator_GWAS', 'overlap_denominator_GWAS'\n",
       "    cNMF_4:\t4000 x 4\n",
       "      obs:\t'sample', 'batch'\n",
       "      var:\t'VarianceExplained'\n",
       "      uns:\t'var_names', 'gsea_varmap_GSEA', 'motif_varmap_Motif', 'gwas_varmap_GWAS'\n",
       "      varm:\t'loadings', 'loadings_zscore', 'Description_GSEA', 'pvalue_GSEA', 'p.adjust_GSEA', 'qvalue_GSEA', 'EPType_Motif', 'TFMotif_Motif', 'PValue_Motif', 'FDR_Motif', 'Enrichment_Motif', 'program_name_GWAS', 'Gene_set_GWAS', 'P-value_GWAS', 'Adjusted_P-value_GWAS', 'Odds_Ratio_GWAS', 'Combined_Score_GWAS', 'Genes_GWAS', 'overlap_numerator_GWAS', 'overlap_denominator_GWAS'\n",
       "    cNMF_5:\t4000 x 5\n",
       "      obs:\t'sample', 'batch'\n",
       "      var:\t'VarianceExplained'\n",
       "      uns:\t'var_names', 'gsea_varmap_GSEA', 'motif_varmap_Motif', 'gwas_varmap_GWAS'\n",
       "      varm:\t'loadings', 'loadings_zscore', 'Description_GSEA', 'pvalue_GSEA', 'p.adjust_GSEA', 'qvalue_GSEA', 'EPType_Motif', 'TFMotif_Motif', 'PValue_Motif', 'FDR_Motif', 'Enrichment_Motif', 'program_name_GWAS', 'Gene_set_GWAS', 'P-value_GWAS', 'Adjusted_P-value_GWAS', 'Odds_Ratio_GWAS', 'Combined_Score_GWAS', 'Genes_GWAS', 'overlap_numerator_GWAS', 'overlap_denominator_GWAS'\n",
       "    cNMF_6:\t4000 x 6\n",
       "      obs:\t'sample', 'batch'\n",
       "      var:\t'VarianceExplained'\n",
       "      uns:\t'var_names', 'gsea_varmap_GSEA', 'motif_varmap_Motif', 'gwas_varmap_GWAS'\n",
       "      varm:\t'loadings', 'loadings_zscore', 'Description_GSEA', 'pvalue_GSEA', 'p.adjust_GSEA', 'qvalue_GSEA', 'EPType_Motif', 'TFMotif_Motif', 'PValue_Motif', 'FDR_Motif', 'Enrichment_Motif', 'program_name_GWAS', 'Gene_set_GWAS', 'P-value_GWAS', 'Adjusted_P-value_GWAS', 'Odds_Ratio_GWAS', 'Combined_Score_GWAS', 'Genes_GWAS', 'overlap_numerator_GWAS', 'overlap_denominator_GWAS'\n",
       "    cNMF_7:\t4000 x 7\n",
       "      obs:\t'sample', 'batch'\n",
       "      var:\t'VarianceExplained'\n",
       "      uns:\t'var_names', 'gsea_varmap_GSEA', 'motif_varmap_Motif', 'gwas_varmap_GWAS'\n",
       "      varm:\t'loadings', 'loadings_zscore', 'Description_GSEA', 'pvalue_GSEA', 'p.adjust_GSEA', 'qvalue_GSEA', 'EPType_Motif', 'TFMotif_Motif', 'PValue_Motif', 'FDR_Motif', 'Enrichment_Motif', 'program_name_GWAS', 'Gene_set_GWAS', 'P-value_GWAS', 'Adjusted_P-value_GWAS', 'Odds_Ratio_GWAS', 'Combined_Score_GWAS', 'Genes_GWAS', 'overlap_numerator_GWAS', 'overlap_denominator_GWAS'\n",
       "    cNMF_8:\t4000 x 8\n",
       "      obs:\t'sample', 'batch'\n",
       "      var:\t'VarianceExplained'\n",
       "      uns:\t'var_names', 'gsea_varmap_GSEA', 'motif_varmap_Motif', 'gwas_varmap_GWAS'\n",
       "      varm:\t'loadings', 'loadings_zscore', 'Description_GSEA', 'pvalue_GSEA', 'p.adjust_GSEA', 'qvalue_GSEA', 'EPType_Motif', 'TFMotif_Motif', 'PValue_Motif', 'FDR_Motif', 'Enrichment_Motif', 'program_name_GWAS', 'Gene_set_GWAS', 'P-value_GWAS', 'Adjusted_P-value_GWAS', 'Odds_Ratio_GWAS', 'Combined_Score_GWAS', 'Genes_GWAS', 'overlap_numerator_GWAS', 'overlap_denominator_GWAS'\n",
       "    cNMF_9:\t4000 x 9\n",
       "      obs:\t'sample', 'batch'\n",
       "      var:\t'VarianceExplained'\n",
       "      uns:\t'var_names', 'gsea_varmap_GSEA', 'motif_varmap_Motif', 'gwas_varmap_GWAS'\n",
       "      varm:\t'loadings', 'loadings_zscore', 'Description_GSEA', 'pvalue_GSEA', 'p.adjust_GSEA', 'qvalue_GSEA', 'EPType_Motif', 'TFMotif_Motif', 'PValue_Motif', 'FDR_Motif', 'Enrichment_Motif', 'program_name_GWAS', 'Gene_set_GWAS', 'P-value_GWAS', 'Adjusted_P-value_GWAS', 'Odds_Ratio_GWAS', 'Combined_Score_GWAS', 'Genes_GWAS', 'overlap_numerator_GWAS', 'overlap_denominator_GWAS'\n",
       "    cNMF_10:\t4000 x 10\n",
       "      obs:\t'sample', 'batch'\n",
       "      var:\t'VarianceExplained'\n",
       "      uns:\t'var_names', 'gsea_varmap_GSEA', 'motif_varmap_Motif', 'gwas_varmap_GWAS'\n",
       "      varm:\t'loadings', 'loadings_zscore', 'Description_GSEA', 'pvalue_GSEA', 'p.adjust_GSEA', 'qvalue_GSEA', 'EPType_Motif', 'TFMotif_Motif', 'PValue_Motif', 'FDR_Motif', 'Enrichment_Motif', 'program_name_GWAS', 'Gene_set_GWAS', 'P-value_GWAS', 'Adjusted_P-value_GWAS', 'Odds_Ratio_GWAS', 'Combined_Score_GWAS', 'Genes_GWAS', 'overlap_numerator_GWAS', 'overlap_denominator_GWAS'\n",
       "    cNMF_11:\t4000 x 11\n",
       "      obs:\t'sample', 'batch'\n",
       "      var:\t'VarianceExplained'\n",
       "      uns:\t'var_names', 'gsea_varmap_GSEA', 'motif_varmap_Motif', 'gwas_varmap_GWAS'\n",
       "      varm:\t'loadings', 'loadings_zscore', 'Description_GSEA', 'pvalue_GSEA', 'p.adjust_GSEA', 'qvalue_GSEA', 'EPType_Motif', 'TFMotif_Motif', 'PValue_Motif', 'FDR_Motif', 'Enrichment_Motif', 'program_name_GWAS', 'Gene_set_GWAS', 'P-value_GWAS', 'Adjusted_P-value_GWAS', 'Odds_Ratio_GWAS', 'Combined_Score_GWAS', 'Genes_GWAS', 'overlap_numerator_GWAS', 'overlap_denominator_GWAS'\n",
       "    cNMF_12:\t4000 x 12\n",
       "      obs:\t'sample', 'batch'\n",
       "      var:\t'VarianceExplained'\n",
       "      uns:\t'var_names', 'gsea_varmap_GSEA', 'motif_varmap_Motif', 'gwas_varmap_GWAS'\n",
       "      varm:\t'loadings', 'loadings_zscore', 'Description_GSEA', 'pvalue_GSEA', 'p.adjust_GSEA', 'qvalue_GSEA', 'EPType_Motif', 'TFMotif_Motif', 'PValue_Motif', 'FDR_Motif', 'Enrichment_Motif', 'program_name_GWAS', 'Gene_set_GWAS', 'P-value_GWAS', 'Adjusted_P-value_GWAS', 'Odds_Ratio_GWAS', 'Combined_Score_GWAS', 'Genes_GWAS', 'overlap_numerator_GWAS', 'overlap_denominator_GWAS'\n",
       "    cNMF_13:\t4000 x 13\n",
       "      obs:\t'sample', 'batch'\n",
       "      var:\t'VarianceExplained'\n",
       "      uns:\t'var_names', 'gsea_varmap_GSEA', 'motif_varmap_Motif', 'gwas_varmap_GWAS'\n",
       "      varm:\t'loadings', 'loadings_zscore', 'Description_GSEA', 'pvalue_GSEA', 'p.adjust_GSEA', 'qvalue_GSEA', 'EPType_Motif', 'TFMotif_Motif', 'PValue_Motif', 'FDR_Motif', 'Enrichment_Motif', 'program_name_GWAS', 'Gene_set_GWAS', 'P-value_GWAS', 'Adjusted_P-value_GWAS', 'Odds_Ratio_GWAS', 'Combined_Score_GWAS', 'Genes_GWAS', 'overlap_numerator_GWAS', 'overlap_denominator_GWAS'\n",
       "    cNMF_14:\t4000 x 14\n",
       "      obs:\t'sample', 'batch'\n",
       "      var:\t'VarianceExplained'\n",
       "      uns:\t'var_names', 'gsea_varmap_GSEA', 'motif_varmap_Motif', 'gwas_varmap_GWAS'\n",
       "      varm:\t'loadings', 'loadings_zscore', 'Description_GSEA', 'pvalue_GSEA', 'p.adjust_GSEA', 'qvalue_GSEA', 'EPType_Motif', 'TFMotif_Motif', 'PValue_Motif', 'FDR_Motif', 'Enrichment_Motif', 'program_name_GWAS', 'Gene_set_GWAS', 'P-value_GWAS', 'Adjusted_P-value_GWAS', 'Odds_Ratio_GWAS', 'Combined_Score_GWAS', 'Genes_GWAS', 'overlap_numerator_GWAS', 'overlap_denominator_GWAS'\n",
       "    cNMF_15:\t4000 x 15\n",
       "      obs:\t'sample', 'batch'\n",
       "      var:\t'VarianceExplained'\n",
       "      uns:\t'var_names', 'gsea_varmap_GSEA', 'motif_varmap_Motif', 'gwas_varmap_GWAS'\n",
       "      varm:\t'loadings', 'loadings_zscore', 'Description_GSEA', 'pvalue_GSEA', 'p.adjust_GSEA', 'qvalue_GSEA', 'EPType_Motif', 'TFMotif_Motif', 'PValue_Motif', 'FDR_Motif', 'Enrichment_Motif', 'program_name_GWAS', 'Gene_set_GWAS', 'P-value_GWAS', 'Adjusted_P-value_GWAS', 'Odds_Ratio_GWAS', 'Combined_Score_GWAS', 'Genes_GWAS', 'overlap_numerator_GWAS', 'overlap_denominator_GWAS'\n",
       "    cNMF_17:\t4000 x 17\n",
       "      obs:\t'sample', 'batch'\n",
       "      var:\t'VarianceExplained'\n",
       "      uns:\t'var_names', 'gsea_varmap_GSEA', 'motif_varmap_Motif', 'gwas_varmap_GWAS'\n",
       "      varm:\t'loadings', 'loadings_zscore', 'Description_GSEA', 'pvalue_GSEA', 'p.adjust_GSEA', 'qvalue_GSEA', 'EPType_Motif', 'TFMotif_Motif', 'PValue_Motif', 'FDR_Motif', 'Enrichment_Motif', 'program_name_GWAS', 'Gene_set_GWAS', 'P-value_GWAS', 'Adjusted_P-value_GWAS', 'Odds_Ratio_GWAS', 'Combined_Score_GWAS', 'Genes_GWAS', 'overlap_numerator_GWAS', 'overlap_denominator_GWAS'\n",
       "    cNMF_19:\t4000 x 19\n",
       "      obs:\t'sample', 'batch'\n",
       "      var:\t'VarianceExplained'\n",
       "      uns:\t'var_names', 'gsea_varmap_GSEA', 'motif_varmap_Motif', 'gwas_varmap_GWAS'\n",
       "      varm:\t'loadings', 'loadings_zscore', 'Description_GSEA', 'pvalue_GSEA', 'p.adjust_GSEA', 'qvalue_GSEA', 'EPType_Motif', 'TFMotif_Motif', 'PValue_Motif', 'FDR_Motif', 'Enrichment_Motif', 'program_name_GWAS', 'Gene_set_GWAS', 'P-value_GWAS', 'Adjusted_P-value_GWAS', 'Odds_Ratio_GWAS', 'Combined_Score_GWAS', 'Genes_GWAS', 'overlap_numerator_GWAS', 'overlap_denominator_GWAS'\n",
       "    cNMF_21:\t4000 x 21\n",
       "      obs:\t'sample', 'batch'\n",
       "      var:\t'VarianceExplained'\n",
       "      uns:\t'var_names', 'gsea_varmap_GSEA', 'motif_varmap_Motif', 'gwas_varmap_GWAS'\n",
       "      varm:\t'loadings', 'loadings_zscore', 'Description_GSEA', 'pvalue_GSEA', 'p.adjust_GSEA', 'qvalue_GSEA', 'EPType_Motif', 'TFMotif_Motif', 'PValue_Motif', 'FDR_Motif', 'Enrichment_Motif', 'program_name_GWAS', 'Gene_set_GWAS', 'P-value_GWAS', 'Adjusted_P-value_GWAS', 'Odds_Ratio_GWAS', 'Combined_Score_GWAS', 'Genes_GWAS', 'overlap_numerator_GWAS', 'overlap_denominator_GWAS'\n",
       "    cNMF_23:\t4000 x 23\n",
       "      obs:\t'sample', 'batch'\n",
       "      var:\t'VarianceExplained'\n",
       "      uns:\t'var_names', 'gsea_varmap_GSEA', 'motif_varmap_Motif', 'gwas_varmap_GWAS'\n",
       "      varm:\t'loadings', 'loadings_zscore', 'Description_GSEA', 'pvalue_GSEA', 'p.adjust_GSEA', 'qvalue_GSEA', 'EPType_Motif', 'TFMotif_Motif', 'PValue_Motif', 'FDR_Motif', 'Enrichment_Motif', 'program_name_GWAS', 'Gene_set_GWAS', 'P-value_GWAS', 'Adjusted_P-value_GWAS', 'Odds_Ratio_GWAS', 'Combined_Score_GWAS', 'Genes_GWAS', 'overlap_numerator_GWAS', 'overlap_denominator_GWAS'\n",
       "    cNMF_25:\t4000 x 25\n",
       "      obs:\t'sample', 'batch'\n",
       "      var:\t'VarianceExplained'\n",
       "      uns:\t'var_names', 'gsea_varmap_GSEA', 'motif_varmap_Motif', 'gwas_varmap_GWAS'\n",
       "      varm:\t'loadings', 'loadings_zscore', 'Description_GSEA', 'pvalue_GSEA', 'p.adjust_GSEA', 'qvalue_GSEA', 'EPType_Motif', 'TFMotif_Motif', 'PValue_Motif', 'FDR_Motif', 'Enrichment_Motif', 'program_name_GWAS', 'Gene_set_GWAS', 'P-value_GWAS', 'Adjusted_P-value_GWAS', 'Odds_Ratio_GWAS', 'Combined_Score_GWAS', 'Genes_GWAS', 'overlap_numerator_GWAS', 'overlap_denominator_GWAS'\n",
       "    cNMF_27:\t4000 x 27\n",
       "      obs:\t'sample', 'batch'\n",
       "      var:\t'VarianceExplained'\n",
       "      uns:\t'var_names', 'gsea_varmap_GSEA', 'motif_varmap_Motif', 'gwas_varmap_GWAS'\n",
       "      varm:\t'loadings', 'loadings_zscore', 'Description_GSEA', 'pvalue_GSEA', 'p.adjust_GSEA', 'qvalue_GSEA', 'EPType_Motif', 'TFMotif_Motif', 'PValue_Motif', 'FDR_Motif', 'Enrichment_Motif', 'program_name_GWAS', 'Gene_set_GWAS', 'P-value_GWAS', 'Adjusted_P-value_GWAS', 'Odds_Ratio_GWAS', 'Combined_Score_GWAS', 'Genes_GWAS', 'overlap_numerator_GWAS', 'overlap_denominator_GWAS'\n",
       "    cNMF_29:\t4000 x 29\n",
       "      obs:\t'sample', 'batch'\n",
       "      var:\t'VarianceExplained'\n",
       "      uns:\t'var_names', 'gsea_varmap_GSEA', 'motif_varmap_Motif', 'gwas_varmap_GWAS'\n",
       "      varm:\t'loadings', 'loadings_zscore', 'Description_GSEA', 'pvalue_GSEA', 'p.adjust_GSEA', 'qvalue_GSEA', 'EPType_Motif', 'TFMotif_Motif', 'PValue_Motif', 'FDR_Motif', 'Enrichment_Motif', 'program_name_GWAS', 'Gene_set_GWAS', 'P-value_GWAS', 'Adjusted_P-value_GWAS', 'Odds_Ratio_GWAS', 'Combined_Score_GWAS', 'Genes_GWAS', 'overlap_numerator_GWAS', 'overlap_denominator_GWAS'\n",
       "    cNMF_35:\t4000 x 35\n",
       "      obs:\t'sample', 'batch'\n",
       "      var:\t'VarianceExplained'\n",
       "      uns:\t'var_names', 'gsea_varmap_GSEA', 'motif_varmap_Motif', 'gwas_varmap_GWAS'\n",
       "      varm:\t'loadings', 'loadings_zscore', 'Description_GSEA', 'pvalue_GSEA', 'p.adjust_GSEA', 'qvalue_GSEA', 'EPType_Motif', 'TFMotif_Motif', 'PValue_Motif', 'FDR_Motif', 'Enrichment_Motif', 'program_name_GWAS', 'Gene_set_GWAS', 'P-value_GWAS', 'Adjusted_P-value_GWAS', 'Odds_Ratio_GWAS', 'Combined_Score_GWAS', 'Genes_GWAS', 'overlap_numerator_GWAS', 'overlap_denominator_GWAS'\n",
       "    cNMF_40:\t4000 x 40\n",
       "      obs:\t'sample', 'batch'\n",
       "      var:\t'VarianceExplained'\n",
       "      uns:\t'var_names', 'gsea_varmap_GSEA', 'motif_varmap_Motif', 'gwas_varmap_GWAS'\n",
       "      varm:\t'loadings', 'loadings_zscore', 'Description_GSEA', 'pvalue_GSEA', 'p.adjust_GSEA', 'qvalue_GSEA', 'EPType_Motif', 'TFMotif_Motif', 'PValue_Motif', 'FDR_Motif', 'Enrichment_Motif', 'program_name_GWAS', 'Gene_set_GWAS', 'P-value_GWAS', 'Adjusted_P-value_GWAS', 'Odds_Ratio_GWAS', 'Combined_Score_GWAS', 'Genes_GWAS', 'overlap_numerator_GWAS', 'overlap_denominator_GWAS'\n",
       "    cNMF_45:\t4000 x 45\n",
       "      obs:\t'sample', 'batch'\n",
       "      var:\t'VarianceExplained'\n",
       "      uns:\t'var_names', 'gsea_varmap_GSEA', 'motif_varmap_Motif', 'gwas_varmap_GWAS'\n",
       "      varm:\t'loadings', 'loadings_zscore', 'Description_GSEA', 'pvalue_GSEA', 'p.adjust_GSEA', 'qvalue_GSEA', 'EPType_Motif', 'TFMotif_Motif', 'PValue_Motif', 'FDR_Motif', 'Enrichment_Motif', 'program_name_GWAS', 'Gene_set_GWAS', 'P-value_GWAS', 'Adjusted_P-value_GWAS', 'Odds_Ratio_GWAS', 'Combined_Score_GWAS', 'Genes_GWAS', 'overlap_numerator_GWAS', 'overlap_denominator_GWAS'\n",
       "    cNMF_50:\t4000 x 50\n",
       "      obs:\t'sample', 'batch'\n",
       "      var:\t'VarianceExplained'\n",
       "      uns:\t'var_names', 'gsea_varmap_GSEA', 'motif_varmap_Motif', 'gwas_varmap_GWAS'\n",
       "      varm:\t'loadings', 'loadings_zscore', 'Description_GSEA', 'pvalue_GSEA', 'p.adjust_GSEA', 'qvalue_GSEA', 'EPType_Motif', 'TFMotif_Motif', 'PValue_Motif', 'FDR_Motif', 'Enrichment_Motif', 'program_name_GWAS', 'Gene_set_GWAS', 'P-value_GWAS', 'Adjusted_P-value_GWAS', 'Odds_Ratio_GWAS', 'Combined_Score_GWAS', 'Genes_GWAS', 'overlap_numerator_GWAS', 'overlap_denominator_GWAS'\n",
       "    cNMF_55:\t4000 x 55\n",
       "      obs:\t'sample', 'batch'\n",
       "      var:\t'VarianceExplained'\n",
       "      uns:\t'var_names', 'gsea_varmap_GSEA', 'motif_varmap_Motif', 'gwas_varmap_GWAS'\n",
       "      varm:\t'loadings', 'loadings_zscore', 'Description_GSEA', 'pvalue_GSEA', 'p.adjust_GSEA', 'qvalue_GSEA', 'EPType_Motif', 'TFMotif_Motif', 'PValue_Motif', 'FDR_Motif', 'Enrichment_Motif', 'program_name_GWAS', 'Gene_set_GWAS', 'P-value_GWAS', 'Adjusted_P-value_GWAS', 'Odds_Ratio_GWAS', 'Combined_Score_GWAS', 'Genes_GWAS', 'overlap_numerator_GWAS', 'overlap_denominator_GWAS'\n",
       "    cNMF_60:\t4000 x 60\n",
       "      obs:\t'sample', 'batch'\n",
       "      var:\t'VarianceExplained'\n",
       "      uns:\t'var_names', 'gsea_varmap_GSEA', 'motif_varmap_Motif', 'gwas_varmap_GWAS'\n",
       "      varm:\t'loadings', 'loadings_zscore', 'Description_GSEA', 'pvalue_GSEA', 'p.adjust_GSEA', 'qvalue_GSEA', 'EPType_Motif', 'TFMotif_Motif', 'PValue_Motif', 'FDR_Motif', 'Enrichment_Motif', 'program_name_GWAS', 'Gene_set_GWAS', 'P-value_GWAS', 'Adjusted_P-value_GWAS', 'Odds_Ratio_GWAS', 'Combined_Score_GWAS', 'Genes_GWAS', 'overlap_numerator_GWAS', 'overlap_denominator_GWAS'"
      ]
     },
     "execution_count": 385,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create an identical \"cNMF\" modality that just has the first k var_names\n",
    "num_progs = [3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 17, 19, 21, 23, 25, 27, 29, 30, 35, 40, 45, 50, 55, 60]\n",
    "for k in num_progs:\n",
    "    mdata.mod[f\"cNMF_{k}\"] = mdata.mod[\"cNMF\"][:, :k].copy()\n",
    "mdata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Write"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 376,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([['intrinsic apoptotic signaling pathway in response to endoplasmic reticulum stress',\n",
       "        'response to endoplasmic reticulum stress',\n",
       "        'neutral amino acid transmembrane transporter activity', ...,\n",
       "        nan, nan, nan],\n",
       "       [nan, nan, nan, ..., nan, nan, nan],\n",
       "       [nan, nan, nan, ..., nan, nan, nan],\n",
       "       ...,\n",
       "       [nan, nan, nan, ..., nan, nan, nan],\n",
       "       [nan, nan, nan, ..., nan, nan, nan],\n",
       "       [nan, nan, nan, ..., nan, nan, nan]], dtype=object)"
      ]
     },
     "execution_count": 376,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mdata[\"cNMF\"].varm[\"Description_GSEA\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 377,
   "metadata": {},
   "outputs": [],
   "source": [
    "# We need to convert each value in varm from an array of objects to an array of strings\n",
    "for key in mdata[\"cNMF\"].varm.keys():\n",
    "    # If type is object\n",
    "    if mdata[\"cNMF\"].varm[key].dtype == \"O\":\n",
    "        mdata[\"cNMF\"].varm[key] = mdata[\"cNMF\"].varm[key].astype(str)\n",
    "        mdata[\"cNMF_30\"].varm[key] = mdata[\"cNMF_30\"].varm[key].astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 378,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([['intrinsic apoptotic signaling pathway in response to endoplasmic reticulum stress',\n",
       "        'response to endoplasmic reticulum stress',\n",
       "        'neutral amino acid transmembrane transporter activity', ...,\n",
       "        'nan', 'nan', 'nan'],\n",
       "       ['nan', 'nan', 'nan', ..., 'nan', 'nan', 'nan'],\n",
       "       ['nan', 'nan', 'nan', ..., 'nan', 'nan', 'nan'],\n",
       "       ...,\n",
       "       ['nan', 'nan', 'nan', ..., 'nan', 'nan', 'nan'],\n",
       "       ['nan', 'nan', 'nan', ..., 'nan', 'nan', 'nan'],\n",
       "       ['nan', 'nan', 'nan', ..., 'nan', 'nan', 'nan']], dtype='<U193')"
      ]
     },
     "execution_count": 378,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mdata[\"cNMF\"].varm[\"Description_GSEA\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 379,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['GO:0070059', 'GO:0034976', 'GO:0015175', 'GO:0036499', 'GO:0006986',\n",
       "       'GO:0006984', 'GO:0034620', 'GO:0036003', 'GO:0015179', 'GO:0015171',\n",
       "       ...\n",
       "       'GO:0061900', 'GO:0032461', 'GO:0061333', 'GO:0044232', 'GO:0016461',\n",
       "       'GO:0005640', 'GO:0044233', 'GO:0097440', 'GO:0003382', 'GO:1904754'],\n",
       "      dtype='object', length=4461)"
      ]
     },
     "execution_count": 379,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mdata[\"cNMF\"].uns[\"gsea_varmap_GSEA\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 380,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "var_names\n",
      "gsea_varmap_GSEA\n",
      "motif_varmap_Motif\n",
      "gwas_varmap_GWAS\n"
     ]
    }
   ],
   "source": [
    "# We need to convert each value in uns from an array of objects to an array of strings\n",
    "# use string_array = np.array([str(element) for element in object_array])a\n",
    "for key in mdata[\"cNMF\"].uns.keys():\n",
    "    # If type is object\n",
    "    if mdata[\"cNMF\"].uns[key].dtype == \"O\":\n",
    "        print(key)\n",
    "        mdata[\"cNMF\"].uns[key] = np.array([str(element) for element in mdata[\"cNMF\"].uns[key]])\n",
    "        mdata[\"cNMF_30\"].uns[key] = np.array([str(element) for element in mdata[\"cNMF_30\"].uns[key]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 381,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['GO:0070059', 'GO:0034976', 'GO:0015175', ..., 'GO:0097440',\n",
       "       'GO:0003382', 'GO:1904754'], dtype='<U10')"
      ]
     },
     "execution_count": 381,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mdata[\"cNMF\"].uns[\"gsea_varmap_GSEA\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 382,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/cellar/users/aklie/opt/miniconda3/envs/test_celloracle/lib/python3.10/site-packages/mudata/_core/mudata.py:491: UserWarning: Cannot join columns with the same name because var_names are intersecting.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "mdata.write_h5mu(path_out)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Downsample version"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 383,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/cellar/users/aklie/opt/miniconda3/envs/test_celloracle/lib/python3.10/site-packages/mudata/_core/mudata.py:491: UserWarning: Cannot join columns with the same name because var_names are intersecting.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<pre>MuData object with n_obs × n_vars = 4000 × 17562\n",
       "  3 modalities\n",
       "    rna:\t4000 x 17472\n",
       "      obs:\t&#x27;barcodes&#x27;, &#x27;n_genes&#x27;, &#x27;n_counts&#x27;, &#x27;sample&#x27;, &#x27;batch&#x27;\n",
       "      var:\t&#x27;n_cells&#x27;\n",
       "    cNMF:\t4000 x 60\n",
       "      obs:\t&#x27;sample&#x27;, &#x27;batch&#x27;\n",
       "      var:\t&#x27;VarianceExplained&#x27;\n",
       "      uns:\t&#x27;var_names&#x27;, &#x27;gsea_varmap_GSEA&#x27;, &#x27;motif_varmap_Motif&#x27;, &#x27;gwas_varmap_GWAS&#x27;\n",
       "      varm:\t&#x27;loadings&#x27;, &#x27;loadings_zscore&#x27;, &#x27;Description_GSEA&#x27;, &#x27;pvalue_GSEA&#x27;, &#x27;p.adjust_GSEA&#x27;, &#x27;qvalue_GSEA&#x27;, &#x27;EPType_Motif&#x27;, &#x27;TFMotif_Motif&#x27;, &#x27;PValue_Motif&#x27;, &#x27;FDR_Motif&#x27;, &#x27;Enrichment_Motif&#x27;, &#x27;program_name_GWAS&#x27;, &#x27;Gene_set_GWAS&#x27;, &#x27;P-value_GWAS&#x27;, &#x27;Adjusted_P-value_GWAS&#x27;, &#x27;Odds_Ratio_GWAS&#x27;, &#x27;Combined_Score_GWAS&#x27;, &#x27;Genes_GWAS&#x27;, &#x27;overlap_numerator_GWAS&#x27;, &#x27;overlap_denominator_GWAS&#x27;\n",
       "    cNMF_30:\t4000 x 30\n",
       "      obs:\t&#x27;sample&#x27;, &#x27;batch&#x27;\n",
       "      var:\t&#x27;VarianceExplained&#x27;\n",
       "      uns:\t&#x27;var_names&#x27;, &#x27;gsea_varmap_GSEA&#x27;, &#x27;motif_varmap_Motif&#x27;, &#x27;gwas_varmap_GWAS&#x27;\n",
       "      varm:\t&#x27;loadings&#x27;, &#x27;loadings_zscore&#x27;, &#x27;Description_GSEA&#x27;, &#x27;pvalue_GSEA&#x27;, &#x27;p.adjust_GSEA&#x27;, &#x27;qvalue_GSEA&#x27;, &#x27;EPType_Motif&#x27;, &#x27;TFMotif_Motif&#x27;, &#x27;PValue_Motif&#x27;, &#x27;FDR_Motif&#x27;, &#x27;Enrichment_Motif&#x27;, &#x27;program_name_GWAS&#x27;, &#x27;Gene_set_GWAS&#x27;, &#x27;P-value_GWAS&#x27;, &#x27;Adjusted_P-value_GWAS&#x27;, &#x27;Odds_Ratio_GWAS&#x27;, &#x27;Combined_Score_GWAS&#x27;, &#x27;Genes_GWAS&#x27;, &#x27;overlap_numerator_GWAS&#x27;, &#x27;overlap_denominator_GWAS&#x27;</pre>"
      ],
      "text/plain": [
       "MuData object with n_obs × n_vars = 4000 × 17562\n",
       "  3 modalities\n",
       "    rna:\t4000 x 17472\n",
       "      obs:\t'barcodes', 'n_genes', 'n_counts', 'sample', 'batch'\n",
       "      var:\t'n_cells'\n",
       "    cNMF:\t4000 x 60\n",
       "      obs:\t'sample', 'batch'\n",
       "      var:\t'VarianceExplained'\n",
       "      uns:\t'var_names', 'gsea_varmap_GSEA', 'motif_varmap_Motif', 'gwas_varmap_GWAS'\n",
       "      varm:\t'loadings', 'loadings_zscore', 'Description_GSEA', 'pvalue_GSEA', 'p.adjust_GSEA', 'qvalue_GSEA', 'EPType_Motif', 'TFMotif_Motif', 'PValue_Motif', 'FDR_Motif', 'Enrichment_Motif', 'program_name_GWAS', 'Gene_set_GWAS', 'P-value_GWAS', 'Adjusted_P-value_GWAS', 'Odds_Ratio_GWAS', 'Combined_Score_GWAS', 'Genes_GWAS', 'overlap_numerator_GWAS', 'overlap_denominator_GWAS'\n",
       "    cNMF_30:\t4000 x 30\n",
       "      obs:\t'sample', 'batch'\n",
       "      var:\t'VarianceExplained'\n",
       "      uns:\t'var_names', 'gsea_varmap_GSEA', 'motif_varmap_Motif', 'gwas_varmap_GWAS'\n",
       "      varm:\t'loadings', 'loadings_zscore', 'Description_GSEA', 'pvalue_GSEA', 'p.adjust_GSEA', 'qvalue_GSEA', 'EPType_Motif', 'TFMotif_Motif', 'PValue_Motif', 'FDR_Motif', 'Enrichment_Motif', 'program_name_GWAS', 'Gene_set_GWAS', 'P-value_GWAS', 'Adjusted_P-value_GWAS', 'Odds_Ratio_GWAS', 'Combined_Score_GWAS', 'Genes_GWAS', 'overlap_numerator_GWAS', 'overlap_denominator_GWAS'"
      ]
     },
     "execution_count": 383,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sel_idx = []\n",
    "for batch in mdata['cNMF'].obs['batch'].unique():\n",
    "    for samp in mdata['cNMF'].obs['sample'].unique():\n",
    "        mdata_obs_ = mdata['cNMF'].obs.loc[(mdata['cNMF'].obs['batch']==batch) & \\\n",
    "                                           (mdata['cNMF'].obs['sample']==samp)]\n",
    "\n",
    "        sel_idx.extend(mdata_obs_.iloc[:200].index.tolist())\n",
    "\n",
    "mdata = mdata[sel_idx].copy()\n",
    "mdata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 386,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/cellar/users/aklie/opt/miniconda3/envs/test_celloracle/lib/python3.10/site-packages/mudata/_core/mudata.py:491: UserWarning: Cannot join columns with the same name because var_names are intersecting.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "mdata.write_h5mu(path_out.replace(\".h5mu\", \"_small.h5mu\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Display"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 388,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".hl-dim, .hl-size, .hl-values, .hl-types, .hl-dims {\n",
       "  color: #777777;\n",
       "}\n",
       ".hl-dim::before, .hl-size::before {\n",
       "  content: \"\\00a0\\00a0\\00a0\";\n",
       "}\n",
       ".hl-values {\n",
       "  font-family: monospace;\n",
       "}\n",
       ".hl-file {\n",
       "  background-color: #EEEEEE;\n",
       "  border-radius: .5rem;\n",
       "  padding: .2rem .4rem;\n",
       "  color: #555555;\n",
       "}\n",
       ".hl-empty {\n",
       "  color: #999999;\n",
       "}\n",
       ".hl-import {\n",
       "  color: #777777;\n",
       "}\n",
       ".block-mod {\n",
       "  display: block;\n",
       "  margin: 0 2rem;\n",
       "}\n",
       ".title {\n",
       "  display: inline-block;\n",
       "  font-weight: 600;\n",
       "  color: #555555;\n",
       "}\n",
       ".title-mod {\n",
       "  font-size: 1.2rem;\n",
       "  color: #04b374;\n",
       "  padding: 0 .5rem;\n",
       "}\n",
       ".title-attr {\n",
       "  font-size: 1.0rem;\n",
       "  padding-top: .2rem;\n",
       "}\n",
       "summary {\n",
       "  cursor: pointer;\n",
       "  list-style: none;\n",
       "}\n",
       "summary::-webkit-details-marker {\n",
       "  display: none;\n",
       "}\n",
       "details > summary::before {\n",
       "  content: '⊕';\n",
       "}\n",
       "details[open] > summary::before {\n",
       "  content: '⊖';\n",
       "}\n",
       "table tr {\n",
       "  background-color: transparent !important;\n",
       "}\n",
       "table tr:hover {\n",
       "  background-color: #04b37433 !important;\n",
       "}\n",
       ".col-index {\n",
       "  text-align: left !important;\n",
       "}\n",
       ".summary-mod {\n",
       "  margin-left: -2rem;\n",
       "}\n",
       ".summary-mod:hover {\n",
       "  background-color: #04b37411;\n",
       "}\n",
       ".summary-mod::before {\n",
       "  color: #04b374;\n",
       "  content: '●';\n",
       "}\n",
       "details[open] > .summary-mod::before {\n",
       "  content: '○';\n",
       "}\n",
       "</style><span>MuData object <span class='hl-dim'>4000 obs &times; 18125 var in 29 modalities</span></span><br><details><summary><div class='title title-attr'>Metadata</div><span class='hl-dim'>.obs</span><span class='hl-size'>61 elements</span></summary><div><table><tr>\n",
       "                                <td class='col-index'>rna:barcodes</td>  <td class='hl-types'>object</td>  <td class='hl-values'>ANKEF1:GAAGGGACATCATTCACGCCT:AAACCCAAGAAGTCAT-scRNAseq_2kG_11AMDox_1...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>rna:n_genes</td>  <td class='hl-types'>int64</td>  <td class='hl-values'>3609,1904,4473,3306,3570,3256,3225,2933,3045,3827,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>rna:n_counts</td>  <td class='hl-types'>float32</td>  <td class='hl-values'>12474.00,5743.00,21423.00,13837.00,13278.00,11613....</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>rna:sample</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>rna:batch</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>cNMF:sample</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>cNMF:batch</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>cNMF_30:sample</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>cNMF_30:batch</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>cNMF_3:sample</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>cNMF_3:batch</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>cNMF_4:sample</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>cNMF_4:batch</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>cNMF_5:sample</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>cNMF_5:batch</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>cNMF_6:sample</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>cNMF_6:batch</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>cNMF_7:sample</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>cNMF_7:batch</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>cNMF_8:sample</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>cNMF_8:batch</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>cNMF_9:sample</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>cNMF_9:batch</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>cNMF_10:sample</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>cNMF_10:batch</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>cNMF_11:sample</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>cNMF_11:batch</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>cNMF_12:sample</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>cNMF_12:batch</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>cNMF_13:sample</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>cNMF_13:batch</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>cNMF_14:sample</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>cNMF_14:batch</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>cNMF_15:sample</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>cNMF_15:batch</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>cNMF_17:sample</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>cNMF_17:batch</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>cNMF_19:sample</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>cNMF_19:batch</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>cNMF_21:sample</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>cNMF_21:batch</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>cNMF_23:sample</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>cNMF_23:batch</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>cNMF_25:sample</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>cNMF_25:batch</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>cNMF_27:sample</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>cNMF_27:batch</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>cNMF_29:sample</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>cNMF_29:batch</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>cNMF_35:sample</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>cNMF_35:batch</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>cNMF_40:sample</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>cNMF_40:batch</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>cNMF_45:sample</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>cNMF_45:batch</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>cNMF_50:sample</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>cNMF_50:batch</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>cNMF_55:sample</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>cNMF_55:batch</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>cNMF_60:sample</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>cNMF_60:batch</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11...</td>\n",
       "                            </tr></table></div></details><details><summary><div class='title title-attr'>Embeddings & mappings</div><span class='hl-dim'>.obsm</span><span class='hl-size'>29 elements</span></summary><div><table><tr>\n",
       "                                       <td class='col-index'>cNMF</td>  <td class='hl-types'>bool</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'></td>\n",
       "                                   </tr>\n",
       "<tr>\n",
       "                                       <td class='col-index'>rna</td>  <td class='hl-types'>bool</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'></td>\n",
       "                                   </tr>\n",
       "<tr>\n",
       "                                       <td class='col-index'>cNMF_30</td>  <td class='hl-types'>bool</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'></td>\n",
       "                                   </tr>\n",
       "<tr>\n",
       "                                       <td class='col-index'>cNMF_3</td>  <td class='hl-types'>bool</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'></td>\n",
       "                                   </tr>\n",
       "<tr>\n",
       "                                       <td class='col-index'>cNMF_4</td>  <td class='hl-types'>bool</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'></td>\n",
       "                                   </tr>\n",
       "<tr>\n",
       "                                       <td class='col-index'>cNMF_5</td>  <td class='hl-types'>bool</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'></td>\n",
       "                                   </tr>\n",
       "<tr>\n",
       "                                       <td class='col-index'>cNMF_6</td>  <td class='hl-types'>bool</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'></td>\n",
       "                                   </tr>\n",
       "<tr>\n",
       "                                       <td class='col-index'>cNMF_7</td>  <td class='hl-types'>bool</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'></td>\n",
       "                                   </tr>\n",
       "<tr>\n",
       "                                       <td class='col-index'>cNMF_8</td>  <td class='hl-types'>bool</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'></td>\n",
       "                                   </tr>\n",
       "<tr>\n",
       "                                       <td class='col-index'>cNMF_9</td>  <td class='hl-types'>bool</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'></td>\n",
       "                                   </tr>\n",
       "<tr>\n",
       "                                       <td class='col-index'>cNMF_10</td>  <td class='hl-types'>bool</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'></td>\n",
       "                                   </tr>\n",
       "<tr>\n",
       "                                       <td class='col-index'>cNMF_11</td>  <td class='hl-types'>bool</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'></td>\n",
       "                                   </tr>\n",
       "<tr>\n",
       "                                       <td class='col-index'>cNMF_12</td>  <td class='hl-types'>bool</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'></td>\n",
       "                                   </tr>\n",
       "<tr>\n",
       "                                       <td class='col-index'>cNMF_13</td>  <td class='hl-types'>bool</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'></td>\n",
       "                                   </tr>\n",
       "<tr>\n",
       "                                       <td class='col-index'>cNMF_14</td>  <td class='hl-types'>bool</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'></td>\n",
       "                                   </tr>\n",
       "<tr>\n",
       "                                       <td class='col-index'>cNMF_15</td>  <td class='hl-types'>bool</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'></td>\n",
       "                                   </tr>\n",
       "<tr>\n",
       "                                       <td class='col-index'>cNMF_17</td>  <td class='hl-types'>bool</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'></td>\n",
       "                                   </tr>\n",
       "<tr>\n",
       "                                       <td class='col-index'>cNMF_19</td>  <td class='hl-types'>bool</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'></td>\n",
       "                                   </tr>\n",
       "<tr>\n",
       "                                       <td class='col-index'>cNMF_21</td>  <td class='hl-types'>bool</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'></td>\n",
       "                                   </tr>\n",
       "<tr>\n",
       "                                       <td class='col-index'>cNMF_23</td>  <td class='hl-types'>bool</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'></td>\n",
       "                                   </tr>\n",
       "<tr>\n",
       "                                       <td class='col-index'>cNMF_25</td>  <td class='hl-types'>bool</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'></td>\n",
       "                                   </tr>\n",
       "<tr>\n",
       "                                       <td class='col-index'>cNMF_27</td>  <td class='hl-types'>bool</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'></td>\n",
       "                                   </tr>\n",
       "<tr>\n",
       "                                       <td class='col-index'>cNMF_29</td>  <td class='hl-types'>bool</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'></td>\n",
       "                                   </tr>\n",
       "<tr>\n",
       "                                       <td class='col-index'>cNMF_35</td>  <td class='hl-types'>bool</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'></td>\n",
       "                                   </tr>\n",
       "<tr>\n",
       "                                       <td class='col-index'>cNMF_40</td>  <td class='hl-types'>bool</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'></td>\n",
       "                                   </tr>\n",
       "<tr>\n",
       "                                       <td class='col-index'>cNMF_45</td>  <td class='hl-types'>bool</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'></td>\n",
       "                                   </tr>\n",
       "<tr>\n",
       "                                       <td class='col-index'>cNMF_50</td>  <td class='hl-types'>bool</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'></td>\n",
       "                                   </tr>\n",
       "<tr>\n",
       "                                       <td class='col-index'>cNMF_55</td>  <td class='hl-types'>bool</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'></td>\n",
       "                                   </tr>\n",
       "<tr>\n",
       "                                       <td class='col-index'>cNMF_60</td>  <td class='hl-types'>bool</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'></td>\n",
       "                                   </tr></table></div></details><details><summary><div class='title title-attr'>Distances</div><span class='hl-dim'>.obsp</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No distances</span></details><div class='block-mod'><div><details><summary class='summary-mod'><div class='title title-mod'>rna</div><span class='hl-dim'>4000 &times 17472</span></summary><span>AnnData object <span class='hl-dim'>4000 obs &times; 17472 var</span></span><br><div class='title title-attr'>Matrix</div><span class='hl-dim'>.X</span><div>\n",
       "            <span class='hl-types'>float32</span> <span>&nbsp;&nbsp;&nbsp;<span class='hl-import'>scipy.sparse._csr.</span>csr_matrix</span>\n",
       "         </table></div><details><summary><div class='title title-attr'>Layers</div><span class='hl-dim'>.layers</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No layers</span></details><details><summary><div class='title title-attr'>Metadata</div><span class='hl-dim'>.obs</span><span class='hl-size'>5 elements</span></summary><div><table><tr>\n",
       "                                <td class='col-index'>barcodes</td>  <td class='hl-types'>object</td>  <td class='hl-values'>ANKEF1:GAAGGGACATCATTCACGCCT:AAACCCAAGAAGTCAT-scRNAseq_2kG_11AMDox_1...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>n_genes</td>  <td class='hl-types'>int64</td>  <td class='hl-values'>3609,1904,4473,3306,3570,3256,3225,2933,3045,3827,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>n_counts</td>  <td class='hl-types'>float32</td>  <td class='hl-values'>12474.00,5743.00,21423.00,13837.00,13278.00,11613....</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>sample</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>batch</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11...</td>\n",
       "                            </tr></table></div></details><details><summary><div class='title title-attr'>Embeddings</div><span class='hl-dim'>.obsm</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No embeddings</span></details><details><summary><div class='title title-attr'>Distances</div><span class='hl-dim'>.obsp</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No distances</span></details><details><summary><div class='title title-attr'>Miscellaneous</div><span class='hl-dim'>.uns</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No miscellaneous</span></details></details></div></div><div class='block-mod'><div><details><summary class='summary-mod'><div class='title title-mod'>cNMF</div><span class='hl-dim'>4000 &times 60</span></summary><span>AnnData object <span class='hl-dim'>4000 obs &times; 60 var</span></span><br><div class='title title-attr'>Matrix</div><span class='hl-dim'>.X</span><div>\n",
       "            <span class='hl-types'>float64</span> <span>&nbsp;&nbsp;&nbsp;<span class='hl-import'>numpy.</span>ndarray</span>\n",
       "         </table></div><details><summary><div class='title title-attr'>Layers</div><span class='hl-dim'>.layers</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No layers</span></details><details><summary><div class='title title-attr'>Metadata</div><span class='hl-dim'>.obs</span><span class='hl-size'>2 elements</span></summary><div><table><tr>\n",
       "                                <td class='col-index'>sample</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>batch</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11...</td>\n",
       "                            </tr></table></div></details><details><summary><div class='title title-attr'>Embeddings</div><span class='hl-dim'>.obsm</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No embeddings</span></details><details><summary><div class='title title-attr'>Distances</div><span class='hl-dim'>.obsp</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No distances</span></details><details><summary><div class='title title-attr'>Miscellaneous</div><span class='hl-dim'>.uns</span><span class='hl-size'>4 elements</span></summary><div><table><tr>\n",
       "                                <td class='col-index'>var_names</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>17472 elements</td>  <td class='hl-values'>FAM87B:ENSG00000177757,FAM41C:ENSG00000230368,SAMD...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>gsea_varmap_GSEA</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>4461 elements</td>  <td class='hl-values'>GO:0070059,GO:0034976,GO:0015175,GO:0036499,GO:000...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>motif_varmap_Motif</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>1267 elements</td>  <td class='hl-values'>Promoter-AHR,Promoter-AIRE,Promoter-ALX1,Promoter-...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>gwas_varmap_GWAS</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>630 elements</td>  <td class='hl-values'>EFO_0000275,EFO_0000280,EFO_0000284,EFO_0000305,EF...</td>\n",
       "                            </tr></table></div></details></details></div></div><div class='block-mod'><div><details><summary class='summary-mod'><div class='title title-mod'>cNMF_30</div><span class='hl-dim'>4000 &times 30</span></summary><span>AnnData object <span class='hl-dim'>4000 obs &times; 30 var</span></span><br><div class='title title-attr'>Matrix</div><span class='hl-dim'>.X</span><div>\n",
       "            <span class='hl-types'>float64</span> <span>&nbsp;&nbsp;&nbsp;<span class='hl-import'>numpy.</span>ndarray</span>\n",
       "         </table></div><details><summary><div class='title title-attr'>Layers</div><span class='hl-dim'>.layers</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No layers</span></details><details><summary><div class='title title-attr'>Metadata</div><span class='hl-dim'>.obs</span><span class='hl-size'>2 elements</span></summary><div><table><tr>\n",
       "                                <td class='col-index'>sample</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>batch</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11...</td>\n",
       "                            </tr></table></div></details><details><summary><div class='title title-attr'>Embeddings</div><span class='hl-dim'>.obsm</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No embeddings</span></details><details><summary><div class='title title-attr'>Distances</div><span class='hl-dim'>.obsp</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No distances</span></details><details><summary><div class='title title-attr'>Miscellaneous</div><span class='hl-dim'>.uns</span><span class='hl-size'>4 elements</span></summary><div><table><tr>\n",
       "                                <td class='col-index'>var_names</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>17472 elements</td>  <td class='hl-values'>FAM87B:ENSG00000177757,FAM41C:ENSG00000230368,SAMD...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>gsea_varmap_GSEA</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>4461 elements</td>  <td class='hl-values'>GO:0070059,GO:0034976,GO:0015175,GO:0036499,GO:000...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>motif_varmap_Motif</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>1267 elements</td>  <td class='hl-values'>Promoter-AHR,Promoter-AIRE,Promoter-ALX1,Promoter-...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>gwas_varmap_GWAS</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>630 elements</td>  <td class='hl-values'>EFO_0000275,EFO_0000280,EFO_0000284,EFO_0000305,EF...</td>\n",
       "                            </tr></table></div></details></details></div></div><div class='block-mod'><div><details><summary class='summary-mod'><div class='title title-mod'>cNMF_3</div><span class='hl-dim'>4000 &times 3</span></summary><span>AnnData object <span class='hl-dim'>4000 obs &times; 3 var</span></span><br><div class='title title-attr'>Matrix</div><span class='hl-dim'>.X</span><div>\n",
       "            <span class='hl-types'>float64</span> <span>&nbsp;&nbsp;&nbsp;<span class='hl-import'>numpy.</span>ndarray</span>\n",
       "         </table></div><details><summary><div class='title title-attr'>Layers</div><span class='hl-dim'>.layers</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No layers</span></details><details><summary><div class='title title-attr'>Metadata</div><span class='hl-dim'>.obs</span><span class='hl-size'>2 elements</span></summary><div><table><tr>\n",
       "                                <td class='col-index'>sample</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>batch</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11...</td>\n",
       "                            </tr></table></div></details><details><summary><div class='title title-attr'>Embeddings</div><span class='hl-dim'>.obsm</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No embeddings</span></details><details><summary><div class='title title-attr'>Distances</div><span class='hl-dim'>.obsp</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No distances</span></details><details><summary><div class='title title-attr'>Miscellaneous</div><span class='hl-dim'>.uns</span><span class='hl-size'>4 elements</span></summary><div><table><tr>\n",
       "                                <td class='col-index'>var_names</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>17472 elements</td>  <td class='hl-values'>FAM87B:ENSG00000177757,FAM41C:ENSG00000230368,SAMD...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>gsea_varmap_GSEA</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>4461 elements</td>  <td class='hl-values'>GO:0070059,GO:0034976,GO:0015175,GO:0036499,GO:000...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>motif_varmap_Motif</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>1267 elements</td>  <td class='hl-values'>Promoter-AHR,Promoter-AIRE,Promoter-ALX1,Promoter-...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>gwas_varmap_GWAS</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>630 elements</td>  <td class='hl-values'>EFO_0000275,EFO_0000280,EFO_0000284,EFO_0000305,EF...</td>\n",
       "                            </tr></table></div></details></details></div></div><div class='block-mod'><div><details><summary class='summary-mod'><div class='title title-mod'>cNMF_4</div><span class='hl-dim'>4000 &times 4</span></summary><span>AnnData object <span class='hl-dim'>4000 obs &times; 4 var</span></span><br><div class='title title-attr'>Matrix</div><span class='hl-dim'>.X</span><div>\n",
       "            <span class='hl-types'>float64</span> <span>&nbsp;&nbsp;&nbsp;<span class='hl-import'>numpy.</span>ndarray</span>\n",
       "         </table></div><details><summary><div class='title title-attr'>Layers</div><span class='hl-dim'>.layers</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No layers</span></details><details><summary><div class='title title-attr'>Metadata</div><span class='hl-dim'>.obs</span><span class='hl-size'>2 elements</span></summary><div><table><tr>\n",
       "                                <td class='col-index'>sample</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>batch</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11...</td>\n",
       "                            </tr></table></div></details><details><summary><div class='title title-attr'>Embeddings</div><span class='hl-dim'>.obsm</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No embeddings</span></details><details><summary><div class='title title-attr'>Distances</div><span class='hl-dim'>.obsp</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No distances</span></details><details><summary><div class='title title-attr'>Miscellaneous</div><span class='hl-dim'>.uns</span><span class='hl-size'>4 elements</span></summary><div><table><tr>\n",
       "                                <td class='col-index'>var_names</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>17472 elements</td>  <td class='hl-values'>FAM87B:ENSG00000177757,FAM41C:ENSG00000230368,SAMD...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>gsea_varmap_GSEA</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>4461 elements</td>  <td class='hl-values'>GO:0070059,GO:0034976,GO:0015175,GO:0036499,GO:000...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>motif_varmap_Motif</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>1267 elements</td>  <td class='hl-values'>Promoter-AHR,Promoter-AIRE,Promoter-ALX1,Promoter-...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>gwas_varmap_GWAS</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>630 elements</td>  <td class='hl-values'>EFO_0000275,EFO_0000280,EFO_0000284,EFO_0000305,EF...</td>\n",
       "                            </tr></table></div></details></details></div></div><div class='block-mod'><div><details><summary class='summary-mod'><div class='title title-mod'>cNMF_5</div><span class='hl-dim'>4000 &times 5</span></summary><span>AnnData object <span class='hl-dim'>4000 obs &times; 5 var</span></span><br><div class='title title-attr'>Matrix</div><span class='hl-dim'>.X</span><div>\n",
       "            <span class='hl-types'>float64</span> <span>&nbsp;&nbsp;&nbsp;<span class='hl-import'>numpy.</span>ndarray</span>\n",
       "         </table></div><details><summary><div class='title title-attr'>Layers</div><span class='hl-dim'>.layers</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No layers</span></details><details><summary><div class='title title-attr'>Metadata</div><span class='hl-dim'>.obs</span><span class='hl-size'>2 elements</span></summary><div><table><tr>\n",
       "                                <td class='col-index'>sample</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>batch</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11...</td>\n",
       "                            </tr></table></div></details><details><summary><div class='title title-attr'>Embeddings</div><span class='hl-dim'>.obsm</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No embeddings</span></details><details><summary><div class='title title-attr'>Distances</div><span class='hl-dim'>.obsp</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No distances</span></details><details><summary><div class='title title-attr'>Miscellaneous</div><span class='hl-dim'>.uns</span><span class='hl-size'>4 elements</span></summary><div><table><tr>\n",
       "                                <td class='col-index'>var_names</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>17472 elements</td>  <td class='hl-values'>FAM87B:ENSG00000177757,FAM41C:ENSG00000230368,SAMD...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>gsea_varmap_GSEA</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>4461 elements</td>  <td class='hl-values'>GO:0070059,GO:0034976,GO:0015175,GO:0036499,GO:000...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>motif_varmap_Motif</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>1267 elements</td>  <td class='hl-values'>Promoter-AHR,Promoter-AIRE,Promoter-ALX1,Promoter-...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>gwas_varmap_GWAS</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>630 elements</td>  <td class='hl-values'>EFO_0000275,EFO_0000280,EFO_0000284,EFO_0000305,EF...</td>\n",
       "                            </tr></table></div></details></details></div></div><div class='block-mod'><div><details><summary class='summary-mod'><div class='title title-mod'>cNMF_6</div><span class='hl-dim'>4000 &times 6</span></summary><span>AnnData object <span class='hl-dim'>4000 obs &times; 6 var</span></span><br><div class='title title-attr'>Matrix</div><span class='hl-dim'>.X</span><div>\n",
       "            <span class='hl-types'>float64</span> <span>&nbsp;&nbsp;&nbsp;<span class='hl-import'>numpy.</span>ndarray</span>\n",
       "         </table></div><details><summary><div class='title title-attr'>Layers</div><span class='hl-dim'>.layers</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No layers</span></details><details><summary><div class='title title-attr'>Metadata</div><span class='hl-dim'>.obs</span><span class='hl-size'>2 elements</span></summary><div><table><tr>\n",
       "                                <td class='col-index'>sample</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>batch</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11...</td>\n",
       "                            </tr></table></div></details><details><summary><div class='title title-attr'>Embeddings</div><span class='hl-dim'>.obsm</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No embeddings</span></details><details><summary><div class='title title-attr'>Distances</div><span class='hl-dim'>.obsp</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No distances</span></details><details><summary><div class='title title-attr'>Miscellaneous</div><span class='hl-dim'>.uns</span><span class='hl-size'>4 elements</span></summary><div><table><tr>\n",
       "                                <td class='col-index'>var_names</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>17472 elements</td>  <td class='hl-values'>FAM87B:ENSG00000177757,FAM41C:ENSG00000230368,SAMD...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>gsea_varmap_GSEA</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>4461 elements</td>  <td class='hl-values'>GO:0070059,GO:0034976,GO:0015175,GO:0036499,GO:000...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>motif_varmap_Motif</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>1267 elements</td>  <td class='hl-values'>Promoter-AHR,Promoter-AIRE,Promoter-ALX1,Promoter-...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>gwas_varmap_GWAS</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>630 elements</td>  <td class='hl-values'>EFO_0000275,EFO_0000280,EFO_0000284,EFO_0000305,EF...</td>\n",
       "                            </tr></table></div></details></details></div></div><div class='block-mod'><div><details><summary class='summary-mod'><div class='title title-mod'>cNMF_7</div><span class='hl-dim'>4000 &times 7</span></summary><span>AnnData object <span class='hl-dim'>4000 obs &times; 7 var</span></span><br><div class='title title-attr'>Matrix</div><span class='hl-dim'>.X</span><div>\n",
       "            <span class='hl-types'>float64</span> <span>&nbsp;&nbsp;&nbsp;<span class='hl-import'>numpy.</span>ndarray</span>\n",
       "         </table></div><details><summary><div class='title title-attr'>Layers</div><span class='hl-dim'>.layers</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No layers</span></details><details><summary><div class='title title-attr'>Metadata</div><span class='hl-dim'>.obs</span><span class='hl-size'>2 elements</span></summary><div><table><tr>\n",
       "                                <td class='col-index'>sample</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>batch</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11...</td>\n",
       "                            </tr></table></div></details><details><summary><div class='title title-attr'>Embeddings</div><span class='hl-dim'>.obsm</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No embeddings</span></details><details><summary><div class='title title-attr'>Distances</div><span class='hl-dim'>.obsp</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No distances</span></details><details><summary><div class='title title-attr'>Miscellaneous</div><span class='hl-dim'>.uns</span><span class='hl-size'>4 elements</span></summary><div><table><tr>\n",
       "                                <td class='col-index'>var_names</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>17472 elements</td>  <td class='hl-values'>FAM87B:ENSG00000177757,FAM41C:ENSG00000230368,SAMD...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>gsea_varmap_GSEA</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>4461 elements</td>  <td class='hl-values'>GO:0070059,GO:0034976,GO:0015175,GO:0036499,GO:000...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>motif_varmap_Motif</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>1267 elements</td>  <td class='hl-values'>Promoter-AHR,Promoter-AIRE,Promoter-ALX1,Promoter-...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>gwas_varmap_GWAS</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>630 elements</td>  <td class='hl-values'>EFO_0000275,EFO_0000280,EFO_0000284,EFO_0000305,EF...</td>\n",
       "                            </tr></table></div></details></details></div></div><div class='block-mod'><div><details><summary class='summary-mod'><div class='title title-mod'>cNMF_8</div><span class='hl-dim'>4000 &times 8</span></summary><span>AnnData object <span class='hl-dim'>4000 obs &times; 8 var</span></span><br><div class='title title-attr'>Matrix</div><span class='hl-dim'>.X</span><div>\n",
       "            <span class='hl-types'>float64</span> <span>&nbsp;&nbsp;&nbsp;<span class='hl-import'>numpy.</span>ndarray</span>\n",
       "         </table></div><details><summary><div class='title title-attr'>Layers</div><span class='hl-dim'>.layers</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No layers</span></details><details><summary><div class='title title-attr'>Metadata</div><span class='hl-dim'>.obs</span><span class='hl-size'>2 elements</span></summary><div><table><tr>\n",
       "                                <td class='col-index'>sample</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>batch</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11...</td>\n",
       "                            </tr></table></div></details><details><summary><div class='title title-attr'>Embeddings</div><span class='hl-dim'>.obsm</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No embeddings</span></details><details><summary><div class='title title-attr'>Distances</div><span class='hl-dim'>.obsp</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No distances</span></details><details><summary><div class='title title-attr'>Miscellaneous</div><span class='hl-dim'>.uns</span><span class='hl-size'>4 elements</span></summary><div><table><tr>\n",
       "                                <td class='col-index'>var_names</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>17472 elements</td>  <td class='hl-values'>FAM87B:ENSG00000177757,FAM41C:ENSG00000230368,SAMD...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>gsea_varmap_GSEA</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>4461 elements</td>  <td class='hl-values'>GO:0070059,GO:0034976,GO:0015175,GO:0036499,GO:000...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>motif_varmap_Motif</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>1267 elements</td>  <td class='hl-values'>Promoter-AHR,Promoter-AIRE,Promoter-ALX1,Promoter-...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>gwas_varmap_GWAS</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>630 elements</td>  <td class='hl-values'>EFO_0000275,EFO_0000280,EFO_0000284,EFO_0000305,EF...</td>\n",
       "                            </tr></table></div></details></details></div></div><div class='block-mod'><div><details><summary class='summary-mod'><div class='title title-mod'>cNMF_9</div><span class='hl-dim'>4000 &times 9</span></summary><span>AnnData object <span class='hl-dim'>4000 obs &times; 9 var</span></span><br><div class='title title-attr'>Matrix</div><span class='hl-dim'>.X</span><div>\n",
       "            <span class='hl-types'>float64</span> <span>&nbsp;&nbsp;&nbsp;<span class='hl-import'>numpy.</span>ndarray</span>\n",
       "         </table></div><details><summary><div class='title title-attr'>Layers</div><span class='hl-dim'>.layers</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No layers</span></details><details><summary><div class='title title-attr'>Metadata</div><span class='hl-dim'>.obs</span><span class='hl-size'>2 elements</span></summary><div><table><tr>\n",
       "                                <td class='col-index'>sample</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>batch</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11...</td>\n",
       "                            </tr></table></div></details><details><summary><div class='title title-attr'>Embeddings</div><span class='hl-dim'>.obsm</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No embeddings</span></details><details><summary><div class='title title-attr'>Distances</div><span class='hl-dim'>.obsp</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No distances</span></details><details><summary><div class='title title-attr'>Miscellaneous</div><span class='hl-dim'>.uns</span><span class='hl-size'>4 elements</span></summary><div><table><tr>\n",
       "                                <td class='col-index'>var_names</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>17472 elements</td>  <td class='hl-values'>FAM87B:ENSG00000177757,FAM41C:ENSG00000230368,SAMD...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>gsea_varmap_GSEA</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>4461 elements</td>  <td class='hl-values'>GO:0070059,GO:0034976,GO:0015175,GO:0036499,GO:000...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>motif_varmap_Motif</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>1267 elements</td>  <td class='hl-values'>Promoter-AHR,Promoter-AIRE,Promoter-ALX1,Promoter-...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>gwas_varmap_GWAS</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>630 elements</td>  <td class='hl-values'>EFO_0000275,EFO_0000280,EFO_0000284,EFO_0000305,EF...</td>\n",
       "                            </tr></table></div></details></details></div></div><div class='block-mod'><div><details><summary class='summary-mod'><div class='title title-mod'>cNMF_10</div><span class='hl-dim'>4000 &times 10</span></summary><span>AnnData object <span class='hl-dim'>4000 obs &times; 10 var</span></span><br><div class='title title-attr'>Matrix</div><span class='hl-dim'>.X</span><div>\n",
       "            <span class='hl-types'>float64</span> <span>&nbsp;&nbsp;&nbsp;<span class='hl-import'>numpy.</span>ndarray</span>\n",
       "         </table></div><details><summary><div class='title title-attr'>Layers</div><span class='hl-dim'>.layers</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No layers</span></details><details><summary><div class='title title-attr'>Metadata</div><span class='hl-dim'>.obs</span><span class='hl-size'>2 elements</span></summary><div><table><tr>\n",
       "                                <td class='col-index'>sample</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>batch</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11...</td>\n",
       "                            </tr></table></div></details><details><summary><div class='title title-attr'>Embeddings</div><span class='hl-dim'>.obsm</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No embeddings</span></details><details><summary><div class='title title-attr'>Distances</div><span class='hl-dim'>.obsp</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No distances</span></details><details><summary><div class='title title-attr'>Miscellaneous</div><span class='hl-dim'>.uns</span><span class='hl-size'>4 elements</span></summary><div><table><tr>\n",
       "                                <td class='col-index'>var_names</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>17472 elements</td>  <td class='hl-values'>FAM87B:ENSG00000177757,FAM41C:ENSG00000230368,SAMD...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>gsea_varmap_GSEA</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>4461 elements</td>  <td class='hl-values'>GO:0070059,GO:0034976,GO:0015175,GO:0036499,GO:000...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>motif_varmap_Motif</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>1267 elements</td>  <td class='hl-values'>Promoter-AHR,Promoter-AIRE,Promoter-ALX1,Promoter-...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>gwas_varmap_GWAS</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>630 elements</td>  <td class='hl-values'>EFO_0000275,EFO_0000280,EFO_0000284,EFO_0000305,EF...</td>\n",
       "                            </tr></table></div></details></details></div></div><div class='block-mod'><div><details><summary class='summary-mod'><div class='title title-mod'>cNMF_11</div><span class='hl-dim'>4000 &times 11</span></summary><span>AnnData object <span class='hl-dim'>4000 obs &times; 11 var</span></span><br><div class='title title-attr'>Matrix</div><span class='hl-dim'>.X</span><div>\n",
       "            <span class='hl-types'>float64</span> <span>&nbsp;&nbsp;&nbsp;<span class='hl-import'>numpy.</span>ndarray</span>\n",
       "         </table></div><details><summary><div class='title title-attr'>Layers</div><span class='hl-dim'>.layers</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No layers</span></details><details><summary><div class='title title-attr'>Metadata</div><span class='hl-dim'>.obs</span><span class='hl-size'>2 elements</span></summary><div><table><tr>\n",
       "                                <td class='col-index'>sample</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>batch</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11...</td>\n",
       "                            </tr></table></div></details><details><summary><div class='title title-attr'>Embeddings</div><span class='hl-dim'>.obsm</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No embeddings</span></details><details><summary><div class='title title-attr'>Distances</div><span class='hl-dim'>.obsp</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No distances</span></details><details><summary><div class='title title-attr'>Miscellaneous</div><span class='hl-dim'>.uns</span><span class='hl-size'>4 elements</span></summary><div><table><tr>\n",
       "                                <td class='col-index'>var_names</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>17472 elements</td>  <td class='hl-values'>FAM87B:ENSG00000177757,FAM41C:ENSG00000230368,SAMD...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>gsea_varmap_GSEA</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>4461 elements</td>  <td class='hl-values'>GO:0070059,GO:0034976,GO:0015175,GO:0036499,GO:000...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>motif_varmap_Motif</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>1267 elements</td>  <td class='hl-values'>Promoter-AHR,Promoter-AIRE,Promoter-ALX1,Promoter-...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>gwas_varmap_GWAS</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>630 elements</td>  <td class='hl-values'>EFO_0000275,EFO_0000280,EFO_0000284,EFO_0000305,EF...</td>\n",
       "                            </tr></table></div></details></details></div></div><div class='block-mod'><div><details><summary class='summary-mod'><div class='title title-mod'>cNMF_12</div><span class='hl-dim'>4000 &times 12</span></summary><span>AnnData object <span class='hl-dim'>4000 obs &times; 12 var</span></span><br><div class='title title-attr'>Matrix</div><span class='hl-dim'>.X</span><div>\n",
       "            <span class='hl-types'>float64</span> <span>&nbsp;&nbsp;&nbsp;<span class='hl-import'>numpy.</span>ndarray</span>\n",
       "         </table></div><details><summary><div class='title title-attr'>Layers</div><span class='hl-dim'>.layers</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No layers</span></details><details><summary><div class='title title-attr'>Metadata</div><span class='hl-dim'>.obs</span><span class='hl-size'>2 elements</span></summary><div><table><tr>\n",
       "                                <td class='col-index'>sample</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>batch</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11...</td>\n",
       "                            </tr></table></div></details><details><summary><div class='title title-attr'>Embeddings</div><span class='hl-dim'>.obsm</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No embeddings</span></details><details><summary><div class='title title-attr'>Distances</div><span class='hl-dim'>.obsp</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No distances</span></details><details><summary><div class='title title-attr'>Miscellaneous</div><span class='hl-dim'>.uns</span><span class='hl-size'>4 elements</span></summary><div><table><tr>\n",
       "                                <td class='col-index'>var_names</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>17472 elements</td>  <td class='hl-values'>FAM87B:ENSG00000177757,FAM41C:ENSG00000230368,SAMD...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>gsea_varmap_GSEA</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>4461 elements</td>  <td class='hl-values'>GO:0070059,GO:0034976,GO:0015175,GO:0036499,GO:000...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>motif_varmap_Motif</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>1267 elements</td>  <td class='hl-values'>Promoter-AHR,Promoter-AIRE,Promoter-ALX1,Promoter-...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>gwas_varmap_GWAS</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>630 elements</td>  <td class='hl-values'>EFO_0000275,EFO_0000280,EFO_0000284,EFO_0000305,EF...</td>\n",
       "                            </tr></table></div></details></details></div></div><div class='block-mod'><div><details><summary class='summary-mod'><div class='title title-mod'>cNMF_13</div><span class='hl-dim'>4000 &times 13</span></summary><span>AnnData object <span class='hl-dim'>4000 obs &times; 13 var</span></span><br><div class='title title-attr'>Matrix</div><span class='hl-dim'>.X</span><div>\n",
       "            <span class='hl-types'>float64</span> <span>&nbsp;&nbsp;&nbsp;<span class='hl-import'>numpy.</span>ndarray</span>\n",
       "         </table></div><details><summary><div class='title title-attr'>Layers</div><span class='hl-dim'>.layers</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No layers</span></details><details><summary><div class='title title-attr'>Metadata</div><span class='hl-dim'>.obs</span><span class='hl-size'>2 elements</span></summary><div><table><tr>\n",
       "                                <td class='col-index'>sample</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>batch</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11...</td>\n",
       "                            </tr></table></div></details><details><summary><div class='title title-attr'>Embeddings</div><span class='hl-dim'>.obsm</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No embeddings</span></details><details><summary><div class='title title-attr'>Distances</div><span class='hl-dim'>.obsp</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No distances</span></details><details><summary><div class='title title-attr'>Miscellaneous</div><span class='hl-dim'>.uns</span><span class='hl-size'>4 elements</span></summary><div><table><tr>\n",
       "                                <td class='col-index'>var_names</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>17472 elements</td>  <td class='hl-values'>FAM87B:ENSG00000177757,FAM41C:ENSG00000230368,SAMD...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>gsea_varmap_GSEA</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>4461 elements</td>  <td class='hl-values'>GO:0070059,GO:0034976,GO:0015175,GO:0036499,GO:000...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>motif_varmap_Motif</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>1267 elements</td>  <td class='hl-values'>Promoter-AHR,Promoter-AIRE,Promoter-ALX1,Promoter-...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>gwas_varmap_GWAS</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>630 elements</td>  <td class='hl-values'>EFO_0000275,EFO_0000280,EFO_0000284,EFO_0000305,EF...</td>\n",
       "                            </tr></table></div></details></details></div></div><div class='block-mod'><div><details><summary class='summary-mod'><div class='title title-mod'>cNMF_14</div><span class='hl-dim'>4000 &times 14</span></summary><span>AnnData object <span class='hl-dim'>4000 obs &times; 14 var</span></span><br><div class='title title-attr'>Matrix</div><span class='hl-dim'>.X</span><div>\n",
       "            <span class='hl-types'>float64</span> <span>&nbsp;&nbsp;&nbsp;<span class='hl-import'>numpy.</span>ndarray</span>\n",
       "         </table></div><details><summary><div class='title title-attr'>Layers</div><span class='hl-dim'>.layers</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No layers</span></details><details><summary><div class='title title-attr'>Metadata</div><span class='hl-dim'>.obs</span><span class='hl-size'>2 elements</span></summary><div><table><tr>\n",
       "                                <td class='col-index'>sample</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>batch</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11...</td>\n",
       "                            </tr></table></div></details><details><summary><div class='title title-attr'>Embeddings</div><span class='hl-dim'>.obsm</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No embeddings</span></details><details><summary><div class='title title-attr'>Distances</div><span class='hl-dim'>.obsp</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No distances</span></details><details><summary><div class='title title-attr'>Miscellaneous</div><span class='hl-dim'>.uns</span><span class='hl-size'>4 elements</span></summary><div><table><tr>\n",
       "                                <td class='col-index'>var_names</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>17472 elements</td>  <td class='hl-values'>FAM87B:ENSG00000177757,FAM41C:ENSG00000230368,SAMD...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>gsea_varmap_GSEA</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>4461 elements</td>  <td class='hl-values'>GO:0070059,GO:0034976,GO:0015175,GO:0036499,GO:000...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>motif_varmap_Motif</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>1267 elements</td>  <td class='hl-values'>Promoter-AHR,Promoter-AIRE,Promoter-ALX1,Promoter-...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>gwas_varmap_GWAS</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>630 elements</td>  <td class='hl-values'>EFO_0000275,EFO_0000280,EFO_0000284,EFO_0000305,EF...</td>\n",
       "                            </tr></table></div></details></details></div></div><div class='block-mod'><div><details><summary class='summary-mod'><div class='title title-mod'>cNMF_15</div><span class='hl-dim'>4000 &times 15</span></summary><span>AnnData object <span class='hl-dim'>4000 obs &times; 15 var</span></span><br><div class='title title-attr'>Matrix</div><span class='hl-dim'>.X</span><div>\n",
       "            <span class='hl-types'>float64</span> <span>&nbsp;&nbsp;&nbsp;<span class='hl-import'>numpy.</span>ndarray</span>\n",
       "         </table></div><details><summary><div class='title title-attr'>Layers</div><span class='hl-dim'>.layers</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No layers</span></details><details><summary><div class='title title-attr'>Metadata</div><span class='hl-dim'>.obs</span><span class='hl-size'>2 elements</span></summary><div><table><tr>\n",
       "                                <td class='col-index'>sample</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>batch</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11...</td>\n",
       "                            </tr></table></div></details><details><summary><div class='title title-attr'>Embeddings</div><span class='hl-dim'>.obsm</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No embeddings</span></details><details><summary><div class='title title-attr'>Distances</div><span class='hl-dim'>.obsp</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No distances</span></details><details><summary><div class='title title-attr'>Miscellaneous</div><span class='hl-dim'>.uns</span><span class='hl-size'>4 elements</span></summary><div><table><tr>\n",
       "                                <td class='col-index'>var_names</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>17472 elements</td>  <td class='hl-values'>FAM87B:ENSG00000177757,FAM41C:ENSG00000230368,SAMD...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>gsea_varmap_GSEA</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>4461 elements</td>  <td class='hl-values'>GO:0070059,GO:0034976,GO:0015175,GO:0036499,GO:000...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>motif_varmap_Motif</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>1267 elements</td>  <td class='hl-values'>Promoter-AHR,Promoter-AIRE,Promoter-ALX1,Promoter-...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>gwas_varmap_GWAS</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>630 elements</td>  <td class='hl-values'>EFO_0000275,EFO_0000280,EFO_0000284,EFO_0000305,EF...</td>\n",
       "                            </tr></table></div></details></details></div></div><div class='block-mod'><div><details><summary class='summary-mod'><div class='title title-mod'>cNMF_17</div><span class='hl-dim'>4000 &times 17</span></summary><span>AnnData object <span class='hl-dim'>4000 obs &times; 17 var</span></span><br><div class='title title-attr'>Matrix</div><span class='hl-dim'>.X</span><div>\n",
       "            <span class='hl-types'>float64</span> <span>&nbsp;&nbsp;&nbsp;<span class='hl-import'>numpy.</span>ndarray</span>\n",
       "         </table></div><details><summary><div class='title title-attr'>Layers</div><span class='hl-dim'>.layers</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No layers</span></details><details><summary><div class='title title-attr'>Metadata</div><span class='hl-dim'>.obs</span><span class='hl-size'>2 elements</span></summary><div><table><tr>\n",
       "                                <td class='col-index'>sample</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>batch</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11...</td>\n",
       "                            </tr></table></div></details><details><summary><div class='title title-attr'>Embeddings</div><span class='hl-dim'>.obsm</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No embeddings</span></details><details><summary><div class='title title-attr'>Distances</div><span class='hl-dim'>.obsp</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No distances</span></details><details><summary><div class='title title-attr'>Miscellaneous</div><span class='hl-dim'>.uns</span><span class='hl-size'>4 elements</span></summary><div><table><tr>\n",
       "                                <td class='col-index'>var_names</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>17472 elements</td>  <td class='hl-values'>FAM87B:ENSG00000177757,FAM41C:ENSG00000230368,SAMD...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>gsea_varmap_GSEA</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>4461 elements</td>  <td class='hl-values'>GO:0070059,GO:0034976,GO:0015175,GO:0036499,GO:000...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>motif_varmap_Motif</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>1267 elements</td>  <td class='hl-values'>Promoter-AHR,Promoter-AIRE,Promoter-ALX1,Promoter-...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>gwas_varmap_GWAS</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>630 elements</td>  <td class='hl-values'>EFO_0000275,EFO_0000280,EFO_0000284,EFO_0000305,EF...</td>\n",
       "                            </tr></table></div></details></details></div></div><div class='block-mod'><div><details><summary class='summary-mod'><div class='title title-mod'>cNMF_19</div><span class='hl-dim'>4000 &times 19</span></summary><span>AnnData object <span class='hl-dim'>4000 obs &times; 19 var</span></span><br><div class='title title-attr'>Matrix</div><span class='hl-dim'>.X</span><div>\n",
       "            <span class='hl-types'>float64</span> <span>&nbsp;&nbsp;&nbsp;<span class='hl-import'>numpy.</span>ndarray</span>\n",
       "         </table></div><details><summary><div class='title title-attr'>Layers</div><span class='hl-dim'>.layers</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No layers</span></details><details><summary><div class='title title-attr'>Metadata</div><span class='hl-dim'>.obs</span><span class='hl-size'>2 elements</span></summary><div><table><tr>\n",
       "                                <td class='col-index'>sample</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>batch</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11...</td>\n",
       "                            </tr></table></div></details><details><summary><div class='title title-attr'>Embeddings</div><span class='hl-dim'>.obsm</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No embeddings</span></details><details><summary><div class='title title-attr'>Distances</div><span class='hl-dim'>.obsp</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No distances</span></details><details><summary><div class='title title-attr'>Miscellaneous</div><span class='hl-dim'>.uns</span><span class='hl-size'>4 elements</span></summary><div><table><tr>\n",
       "                                <td class='col-index'>var_names</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>17472 elements</td>  <td class='hl-values'>FAM87B:ENSG00000177757,FAM41C:ENSG00000230368,SAMD...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>gsea_varmap_GSEA</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>4461 elements</td>  <td class='hl-values'>GO:0070059,GO:0034976,GO:0015175,GO:0036499,GO:000...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>motif_varmap_Motif</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>1267 elements</td>  <td class='hl-values'>Promoter-AHR,Promoter-AIRE,Promoter-ALX1,Promoter-...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>gwas_varmap_GWAS</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>630 elements</td>  <td class='hl-values'>EFO_0000275,EFO_0000280,EFO_0000284,EFO_0000305,EF...</td>\n",
       "                            </tr></table></div></details></details></div></div><div class='block-mod'><div><details><summary class='summary-mod'><div class='title title-mod'>cNMF_21</div><span class='hl-dim'>4000 &times 21</span></summary><span>AnnData object <span class='hl-dim'>4000 obs &times; 21 var</span></span><br><div class='title title-attr'>Matrix</div><span class='hl-dim'>.X</span><div>\n",
       "            <span class='hl-types'>float64</span> <span>&nbsp;&nbsp;&nbsp;<span class='hl-import'>numpy.</span>ndarray</span>\n",
       "         </table></div><details><summary><div class='title title-attr'>Layers</div><span class='hl-dim'>.layers</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No layers</span></details><details><summary><div class='title title-attr'>Metadata</div><span class='hl-dim'>.obs</span><span class='hl-size'>2 elements</span></summary><div><table><tr>\n",
       "                                <td class='col-index'>sample</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>batch</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11...</td>\n",
       "                            </tr></table></div></details><details><summary><div class='title title-attr'>Embeddings</div><span class='hl-dim'>.obsm</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No embeddings</span></details><details><summary><div class='title title-attr'>Distances</div><span class='hl-dim'>.obsp</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No distances</span></details><details><summary><div class='title title-attr'>Miscellaneous</div><span class='hl-dim'>.uns</span><span class='hl-size'>4 elements</span></summary><div><table><tr>\n",
       "                                <td class='col-index'>var_names</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>17472 elements</td>  <td class='hl-values'>FAM87B:ENSG00000177757,FAM41C:ENSG00000230368,SAMD...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>gsea_varmap_GSEA</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>4461 elements</td>  <td class='hl-values'>GO:0070059,GO:0034976,GO:0015175,GO:0036499,GO:000...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>motif_varmap_Motif</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>1267 elements</td>  <td class='hl-values'>Promoter-AHR,Promoter-AIRE,Promoter-ALX1,Promoter-...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>gwas_varmap_GWAS</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>630 elements</td>  <td class='hl-values'>EFO_0000275,EFO_0000280,EFO_0000284,EFO_0000305,EF...</td>\n",
       "                            </tr></table></div></details></details></div></div><div class='block-mod'><div><details><summary class='summary-mod'><div class='title title-mod'>cNMF_23</div><span class='hl-dim'>4000 &times 23</span></summary><span>AnnData object <span class='hl-dim'>4000 obs &times; 23 var</span></span><br><div class='title title-attr'>Matrix</div><span class='hl-dim'>.X</span><div>\n",
       "            <span class='hl-types'>float64</span> <span>&nbsp;&nbsp;&nbsp;<span class='hl-import'>numpy.</span>ndarray</span>\n",
       "         </table></div><details><summary><div class='title title-attr'>Layers</div><span class='hl-dim'>.layers</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No layers</span></details><details><summary><div class='title title-attr'>Metadata</div><span class='hl-dim'>.obs</span><span class='hl-size'>2 elements</span></summary><div><table><tr>\n",
       "                                <td class='col-index'>sample</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>batch</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11...</td>\n",
       "                            </tr></table></div></details><details><summary><div class='title title-attr'>Embeddings</div><span class='hl-dim'>.obsm</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No embeddings</span></details><details><summary><div class='title title-attr'>Distances</div><span class='hl-dim'>.obsp</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No distances</span></details><details><summary><div class='title title-attr'>Miscellaneous</div><span class='hl-dim'>.uns</span><span class='hl-size'>4 elements</span></summary><div><table><tr>\n",
       "                                <td class='col-index'>var_names</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>17472 elements</td>  <td class='hl-values'>FAM87B:ENSG00000177757,FAM41C:ENSG00000230368,SAMD...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>gsea_varmap_GSEA</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>4461 elements</td>  <td class='hl-values'>GO:0070059,GO:0034976,GO:0015175,GO:0036499,GO:000...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>motif_varmap_Motif</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>1267 elements</td>  <td class='hl-values'>Promoter-AHR,Promoter-AIRE,Promoter-ALX1,Promoter-...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>gwas_varmap_GWAS</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>630 elements</td>  <td class='hl-values'>EFO_0000275,EFO_0000280,EFO_0000284,EFO_0000305,EF...</td>\n",
       "                            </tr></table></div></details></details></div></div><div class='block-mod'><div><details><summary class='summary-mod'><div class='title title-mod'>cNMF_25</div><span class='hl-dim'>4000 &times 25</span></summary><span>AnnData object <span class='hl-dim'>4000 obs &times; 25 var</span></span><br><div class='title title-attr'>Matrix</div><span class='hl-dim'>.X</span><div>\n",
       "            <span class='hl-types'>float64</span> <span>&nbsp;&nbsp;&nbsp;<span class='hl-import'>numpy.</span>ndarray</span>\n",
       "         </table></div><details><summary><div class='title title-attr'>Layers</div><span class='hl-dim'>.layers</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No layers</span></details><details><summary><div class='title title-attr'>Metadata</div><span class='hl-dim'>.obs</span><span class='hl-size'>2 elements</span></summary><div><table><tr>\n",
       "                                <td class='col-index'>sample</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>batch</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11...</td>\n",
       "                            </tr></table></div></details><details><summary><div class='title title-attr'>Embeddings</div><span class='hl-dim'>.obsm</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No embeddings</span></details><details><summary><div class='title title-attr'>Distances</div><span class='hl-dim'>.obsp</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No distances</span></details><details><summary><div class='title title-attr'>Miscellaneous</div><span class='hl-dim'>.uns</span><span class='hl-size'>4 elements</span></summary><div><table><tr>\n",
       "                                <td class='col-index'>var_names</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>17472 elements</td>  <td class='hl-values'>FAM87B:ENSG00000177757,FAM41C:ENSG00000230368,SAMD...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>gsea_varmap_GSEA</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>4461 elements</td>  <td class='hl-values'>GO:0070059,GO:0034976,GO:0015175,GO:0036499,GO:000...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>motif_varmap_Motif</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>1267 elements</td>  <td class='hl-values'>Promoter-AHR,Promoter-AIRE,Promoter-ALX1,Promoter-...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>gwas_varmap_GWAS</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>630 elements</td>  <td class='hl-values'>EFO_0000275,EFO_0000280,EFO_0000284,EFO_0000305,EF...</td>\n",
       "                            </tr></table></div></details></details></div></div><div class='block-mod'><div><details><summary class='summary-mod'><div class='title title-mod'>cNMF_27</div><span class='hl-dim'>4000 &times 27</span></summary><span>AnnData object <span class='hl-dim'>4000 obs &times; 27 var</span></span><br><div class='title title-attr'>Matrix</div><span class='hl-dim'>.X</span><div>\n",
       "            <span class='hl-types'>float64</span> <span>&nbsp;&nbsp;&nbsp;<span class='hl-import'>numpy.</span>ndarray</span>\n",
       "         </table></div><details><summary><div class='title title-attr'>Layers</div><span class='hl-dim'>.layers</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No layers</span></details><details><summary><div class='title title-attr'>Metadata</div><span class='hl-dim'>.obs</span><span class='hl-size'>2 elements</span></summary><div><table><tr>\n",
       "                                <td class='col-index'>sample</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>batch</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11...</td>\n",
       "                            </tr></table></div></details><details><summary><div class='title title-attr'>Embeddings</div><span class='hl-dim'>.obsm</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No embeddings</span></details><details><summary><div class='title title-attr'>Distances</div><span class='hl-dim'>.obsp</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No distances</span></details><details><summary><div class='title title-attr'>Miscellaneous</div><span class='hl-dim'>.uns</span><span class='hl-size'>4 elements</span></summary><div><table><tr>\n",
       "                                <td class='col-index'>var_names</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>17472 elements</td>  <td class='hl-values'>FAM87B:ENSG00000177757,FAM41C:ENSG00000230368,SAMD...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>gsea_varmap_GSEA</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>4461 elements</td>  <td class='hl-values'>GO:0070059,GO:0034976,GO:0015175,GO:0036499,GO:000...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>motif_varmap_Motif</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>1267 elements</td>  <td class='hl-values'>Promoter-AHR,Promoter-AIRE,Promoter-ALX1,Promoter-...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>gwas_varmap_GWAS</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>630 elements</td>  <td class='hl-values'>EFO_0000275,EFO_0000280,EFO_0000284,EFO_0000305,EF...</td>\n",
       "                            </tr></table></div></details></details></div></div><div class='block-mod'><div><details><summary class='summary-mod'><div class='title title-mod'>cNMF_29</div><span class='hl-dim'>4000 &times 29</span></summary><span>AnnData object <span class='hl-dim'>4000 obs &times; 29 var</span></span><br><div class='title title-attr'>Matrix</div><span class='hl-dim'>.X</span><div>\n",
       "            <span class='hl-types'>float64</span> <span>&nbsp;&nbsp;&nbsp;<span class='hl-import'>numpy.</span>ndarray</span>\n",
       "         </table></div><details><summary><div class='title title-attr'>Layers</div><span class='hl-dim'>.layers</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No layers</span></details><details><summary><div class='title title-attr'>Metadata</div><span class='hl-dim'>.obs</span><span class='hl-size'>2 elements</span></summary><div><table><tr>\n",
       "                                <td class='col-index'>sample</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>batch</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11...</td>\n",
       "                            </tr></table></div></details><details><summary><div class='title title-attr'>Embeddings</div><span class='hl-dim'>.obsm</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No embeddings</span></details><details><summary><div class='title title-attr'>Distances</div><span class='hl-dim'>.obsp</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No distances</span></details><details><summary><div class='title title-attr'>Miscellaneous</div><span class='hl-dim'>.uns</span><span class='hl-size'>4 elements</span></summary><div><table><tr>\n",
       "                                <td class='col-index'>var_names</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>17472 elements</td>  <td class='hl-values'>FAM87B:ENSG00000177757,FAM41C:ENSG00000230368,SAMD...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>gsea_varmap_GSEA</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>4461 elements</td>  <td class='hl-values'>GO:0070059,GO:0034976,GO:0015175,GO:0036499,GO:000...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>motif_varmap_Motif</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>1267 elements</td>  <td class='hl-values'>Promoter-AHR,Promoter-AIRE,Promoter-ALX1,Promoter-...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>gwas_varmap_GWAS</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>630 elements</td>  <td class='hl-values'>EFO_0000275,EFO_0000280,EFO_0000284,EFO_0000305,EF...</td>\n",
       "                            </tr></table></div></details></details></div></div><div class='block-mod'><div><details><summary class='summary-mod'><div class='title title-mod'>cNMF_35</div><span class='hl-dim'>4000 &times 35</span></summary><span>AnnData object <span class='hl-dim'>4000 obs &times; 35 var</span></span><br><div class='title title-attr'>Matrix</div><span class='hl-dim'>.X</span><div>\n",
       "            <span class='hl-types'>float64</span> <span>&nbsp;&nbsp;&nbsp;<span class='hl-import'>numpy.</span>ndarray</span>\n",
       "         </table></div><details><summary><div class='title title-attr'>Layers</div><span class='hl-dim'>.layers</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No layers</span></details><details><summary><div class='title title-attr'>Metadata</div><span class='hl-dim'>.obs</span><span class='hl-size'>2 elements</span></summary><div><table><tr>\n",
       "                                <td class='col-index'>sample</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>batch</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11...</td>\n",
       "                            </tr></table></div></details><details><summary><div class='title title-attr'>Embeddings</div><span class='hl-dim'>.obsm</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No embeddings</span></details><details><summary><div class='title title-attr'>Distances</div><span class='hl-dim'>.obsp</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No distances</span></details><details><summary><div class='title title-attr'>Miscellaneous</div><span class='hl-dim'>.uns</span><span class='hl-size'>4 elements</span></summary><div><table><tr>\n",
       "                                <td class='col-index'>var_names</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>17472 elements</td>  <td class='hl-values'>FAM87B:ENSG00000177757,FAM41C:ENSG00000230368,SAMD...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>gsea_varmap_GSEA</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>4461 elements</td>  <td class='hl-values'>GO:0070059,GO:0034976,GO:0015175,GO:0036499,GO:000...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>motif_varmap_Motif</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>1267 elements</td>  <td class='hl-values'>Promoter-AHR,Promoter-AIRE,Promoter-ALX1,Promoter-...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>gwas_varmap_GWAS</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>630 elements</td>  <td class='hl-values'>EFO_0000275,EFO_0000280,EFO_0000284,EFO_0000305,EF...</td>\n",
       "                            </tr></table></div></details></details></div></div><div class='block-mod'><div><details><summary class='summary-mod'><div class='title title-mod'>cNMF_40</div><span class='hl-dim'>4000 &times 40</span></summary><span>AnnData object <span class='hl-dim'>4000 obs &times; 40 var</span></span><br><div class='title title-attr'>Matrix</div><span class='hl-dim'>.X</span><div>\n",
       "            <span class='hl-types'>float64</span> <span>&nbsp;&nbsp;&nbsp;<span class='hl-import'>numpy.</span>ndarray</span>\n",
       "         </table></div><details><summary><div class='title title-attr'>Layers</div><span class='hl-dim'>.layers</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No layers</span></details><details><summary><div class='title title-attr'>Metadata</div><span class='hl-dim'>.obs</span><span class='hl-size'>2 elements</span></summary><div><table><tr>\n",
       "                                <td class='col-index'>sample</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>batch</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11...</td>\n",
       "                            </tr></table></div></details><details><summary><div class='title title-attr'>Embeddings</div><span class='hl-dim'>.obsm</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No embeddings</span></details><details><summary><div class='title title-attr'>Distances</div><span class='hl-dim'>.obsp</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No distances</span></details><details><summary><div class='title title-attr'>Miscellaneous</div><span class='hl-dim'>.uns</span><span class='hl-size'>4 elements</span></summary><div><table><tr>\n",
       "                                <td class='col-index'>var_names</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>17472 elements</td>  <td class='hl-values'>FAM87B:ENSG00000177757,FAM41C:ENSG00000230368,SAMD...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>gsea_varmap_GSEA</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>4461 elements</td>  <td class='hl-values'>GO:0070059,GO:0034976,GO:0015175,GO:0036499,GO:000...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>motif_varmap_Motif</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>1267 elements</td>  <td class='hl-values'>Promoter-AHR,Promoter-AIRE,Promoter-ALX1,Promoter-...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>gwas_varmap_GWAS</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>630 elements</td>  <td class='hl-values'>EFO_0000275,EFO_0000280,EFO_0000284,EFO_0000305,EF...</td>\n",
       "                            </tr></table></div></details></details></div></div><div class='block-mod'><div><details><summary class='summary-mod'><div class='title title-mod'>cNMF_45</div><span class='hl-dim'>4000 &times 45</span></summary><span>AnnData object <span class='hl-dim'>4000 obs &times; 45 var</span></span><br><div class='title title-attr'>Matrix</div><span class='hl-dim'>.X</span><div>\n",
       "            <span class='hl-types'>float64</span> <span>&nbsp;&nbsp;&nbsp;<span class='hl-import'>numpy.</span>ndarray</span>\n",
       "         </table></div><details><summary><div class='title title-attr'>Layers</div><span class='hl-dim'>.layers</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No layers</span></details><details><summary><div class='title title-attr'>Metadata</div><span class='hl-dim'>.obs</span><span class='hl-size'>2 elements</span></summary><div><table><tr>\n",
       "                                <td class='col-index'>sample</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>batch</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11...</td>\n",
       "                            </tr></table></div></details><details><summary><div class='title title-attr'>Embeddings</div><span class='hl-dim'>.obsm</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No embeddings</span></details><details><summary><div class='title title-attr'>Distances</div><span class='hl-dim'>.obsp</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No distances</span></details><details><summary><div class='title title-attr'>Miscellaneous</div><span class='hl-dim'>.uns</span><span class='hl-size'>4 elements</span></summary><div><table><tr>\n",
       "                                <td class='col-index'>var_names</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>17472 elements</td>  <td class='hl-values'>FAM87B:ENSG00000177757,FAM41C:ENSG00000230368,SAMD...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>gsea_varmap_GSEA</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>4461 elements</td>  <td class='hl-values'>GO:0070059,GO:0034976,GO:0015175,GO:0036499,GO:000...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>motif_varmap_Motif</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>1267 elements</td>  <td class='hl-values'>Promoter-AHR,Promoter-AIRE,Promoter-ALX1,Promoter-...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>gwas_varmap_GWAS</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>630 elements</td>  <td class='hl-values'>EFO_0000275,EFO_0000280,EFO_0000284,EFO_0000305,EF...</td>\n",
       "                            </tr></table></div></details></details></div></div><div class='block-mod'><div><details><summary class='summary-mod'><div class='title title-mod'>cNMF_50</div><span class='hl-dim'>4000 &times 50</span></summary><span>AnnData object <span class='hl-dim'>4000 obs &times; 50 var</span></span><br><div class='title title-attr'>Matrix</div><span class='hl-dim'>.X</span><div>\n",
       "            <span class='hl-types'>float64</span> <span>&nbsp;&nbsp;&nbsp;<span class='hl-import'>numpy.</span>ndarray</span>\n",
       "         </table></div><details><summary><div class='title title-attr'>Layers</div><span class='hl-dim'>.layers</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No layers</span></details><details><summary><div class='title title-attr'>Metadata</div><span class='hl-dim'>.obs</span><span class='hl-size'>2 elements</span></summary><div><table><tr>\n",
       "                                <td class='col-index'>sample</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>batch</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11...</td>\n",
       "                            </tr></table></div></details><details><summary><div class='title title-attr'>Embeddings</div><span class='hl-dim'>.obsm</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No embeddings</span></details><details><summary><div class='title title-attr'>Distances</div><span class='hl-dim'>.obsp</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No distances</span></details><details><summary><div class='title title-attr'>Miscellaneous</div><span class='hl-dim'>.uns</span><span class='hl-size'>4 elements</span></summary><div><table><tr>\n",
       "                                <td class='col-index'>var_names</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>17472 elements</td>  <td class='hl-values'>FAM87B:ENSG00000177757,FAM41C:ENSG00000230368,SAMD...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>gsea_varmap_GSEA</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>4461 elements</td>  <td class='hl-values'>GO:0070059,GO:0034976,GO:0015175,GO:0036499,GO:000...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>motif_varmap_Motif</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>1267 elements</td>  <td class='hl-values'>Promoter-AHR,Promoter-AIRE,Promoter-ALX1,Promoter-...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>gwas_varmap_GWAS</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>630 elements</td>  <td class='hl-values'>EFO_0000275,EFO_0000280,EFO_0000284,EFO_0000305,EF...</td>\n",
       "                            </tr></table></div></details></details></div></div><div class='block-mod'><div><details><summary class='summary-mod'><div class='title title-mod'>cNMF_55</div><span class='hl-dim'>4000 &times 55</span></summary><span>AnnData object <span class='hl-dim'>4000 obs &times; 55 var</span></span><br><div class='title title-attr'>Matrix</div><span class='hl-dim'>.X</span><div>\n",
       "            <span class='hl-types'>float64</span> <span>&nbsp;&nbsp;&nbsp;<span class='hl-import'>numpy.</span>ndarray</span>\n",
       "         </table></div><details><summary><div class='title title-attr'>Layers</div><span class='hl-dim'>.layers</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No layers</span></details><details><summary><div class='title title-attr'>Metadata</div><span class='hl-dim'>.obs</span><span class='hl-size'>2 elements</span></summary><div><table><tr>\n",
       "                                <td class='col-index'>sample</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>batch</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11...</td>\n",
       "                            </tr></table></div></details><details><summary><div class='title title-attr'>Embeddings</div><span class='hl-dim'>.obsm</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No embeddings</span></details><details><summary><div class='title title-attr'>Distances</div><span class='hl-dim'>.obsp</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No distances</span></details><details><summary><div class='title title-attr'>Miscellaneous</div><span class='hl-dim'>.uns</span><span class='hl-size'>4 elements</span></summary><div><table><tr>\n",
       "                                <td class='col-index'>var_names</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>17472 elements</td>  <td class='hl-values'>FAM87B:ENSG00000177757,FAM41C:ENSG00000230368,SAMD...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>gsea_varmap_GSEA</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>4461 elements</td>  <td class='hl-values'>GO:0070059,GO:0034976,GO:0015175,GO:0036499,GO:000...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>motif_varmap_Motif</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>1267 elements</td>  <td class='hl-values'>Promoter-AHR,Promoter-AIRE,Promoter-ALX1,Promoter-...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>gwas_varmap_GWAS</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>630 elements</td>  <td class='hl-values'>EFO_0000275,EFO_0000280,EFO_0000284,EFO_0000305,EF...</td>\n",
       "                            </tr></table></div></details></details></div></div><div class='block-mod'><div><details><summary class='summary-mod'><div class='title title-mod'>cNMF_60</div><span class='hl-dim'>4000 &times 60</span></summary><span>AnnData object <span class='hl-dim'>4000 obs &times; 60 var</span></span><br><div class='title title-attr'>Matrix</div><span class='hl-dim'>.X</span><div>\n",
       "            <span class='hl-types'>float64</span> <span>&nbsp;&nbsp;&nbsp;<span class='hl-import'>numpy.</span>ndarray</span>\n",
       "         </table></div><details><summary><div class='title title-attr'>Layers</div><span class='hl-dim'>.layers</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No layers</span></details><details><summary><div class='title title-attr'>Metadata</div><span class='hl-dim'>.obs</span><span class='hl-size'>2 elements</span></summary><div><table><tr>\n",
       "                                <td class='col-index'>sample</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,11AMDox_1,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>batch</td>  <td class='hl-types'>category</td>  <td class='hl-values'>11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11AMDox,11...</td>\n",
       "                            </tr></table></div></details><details><summary><div class='title title-attr'>Embeddings</div><span class='hl-dim'>.obsm</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No embeddings</span></details><details><summary><div class='title title-attr'>Distances</div><span class='hl-dim'>.obsp</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No distances</span></details><details><summary><div class='title title-attr'>Miscellaneous</div><span class='hl-dim'>.uns</span><span class='hl-size'>4 elements</span></summary><div><table><tr>\n",
       "                                <td class='col-index'>var_names</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>17472 elements</td>  <td class='hl-values'>FAM87B:ENSG00000177757,FAM41C:ENSG00000230368,SAMD...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>gsea_varmap_GSEA</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>4461 elements</td>  <td class='hl-values'>GO:0070059,GO:0034976,GO:0015175,GO:0036499,GO:000...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>motif_varmap_Motif</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>1267 elements</td>  <td class='hl-values'>Promoter-AHR,Promoter-AIRE,Promoter-ALX1,Promoter-...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>gwas_varmap_GWAS</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>630 elements</td>  <td class='hl-values'>EFO_0000275,EFO_0000280,EFO_0000284,EFO_0000305,EF...</td>\n",
       "                            </tr></table></div></details></details></div></div><br/>"
      ],
      "text/plain": [
       "MuData object with n_obs × n_vars = 4000 × 18125\n",
       "  29 modalities\n",
       "    rna:\t4000 x 17472\n",
       "      obs:\t'barcodes', 'n_genes', 'n_counts', 'sample', 'batch'\n",
       "      var:\t'n_cells'\n",
       "    cNMF:\t4000 x 60\n",
       "      obs:\t'sample', 'batch'\n",
       "      var:\t'VarianceExplained'\n",
       "      uns:\t'var_names', 'gsea_varmap_GSEA', 'motif_varmap_Motif', 'gwas_varmap_GWAS'\n",
       "      varm:\t'loadings', 'loadings_zscore', 'Description_GSEA', 'pvalue_GSEA', 'p.adjust_GSEA', 'qvalue_GSEA', 'EPType_Motif', 'TFMotif_Motif', 'PValue_Motif', 'FDR_Motif', 'Enrichment_Motif', 'program_name_GWAS', 'Gene_set_GWAS', 'P-value_GWAS', 'Adjusted_P-value_GWAS', 'Odds_Ratio_GWAS', 'Combined_Score_GWAS', 'Genes_GWAS', 'overlap_numerator_GWAS', 'overlap_denominator_GWAS'\n",
       "    cNMF_30:\t4000 x 30\n",
       "      obs:\t'sample', 'batch'\n",
       "      var:\t'VarianceExplained'\n",
       "      uns:\t'var_names', 'gsea_varmap_GSEA', 'motif_varmap_Motif', 'gwas_varmap_GWAS'\n",
       "      varm:\t'loadings', 'loadings_zscore', 'Description_GSEA', 'pvalue_GSEA', 'p.adjust_GSEA', 'qvalue_GSEA', 'EPType_Motif', 'TFMotif_Motif', 'PValue_Motif', 'FDR_Motif', 'Enrichment_Motif', 'program_name_GWAS', 'Gene_set_GWAS', 'P-value_GWAS', 'Adjusted_P-value_GWAS', 'Odds_Ratio_GWAS', 'Combined_Score_GWAS', 'Genes_GWAS', 'overlap_numerator_GWAS', 'overlap_denominator_GWAS'\n",
       "    cNMF_3:\t4000 x 3\n",
       "      obs:\t'sample', 'batch'\n",
       "      var:\t'VarianceExplained'\n",
       "      uns:\t'var_names', 'gsea_varmap_GSEA', 'motif_varmap_Motif', 'gwas_varmap_GWAS'\n",
       "      varm:\t'loadings', 'loadings_zscore', 'Description_GSEA', 'pvalue_GSEA', 'p.adjust_GSEA', 'qvalue_GSEA', 'EPType_Motif', 'TFMotif_Motif', 'PValue_Motif', 'FDR_Motif', 'Enrichment_Motif', 'program_name_GWAS', 'Gene_set_GWAS', 'P-value_GWAS', 'Adjusted_P-value_GWAS', 'Odds_Ratio_GWAS', 'Combined_Score_GWAS', 'Genes_GWAS', 'overlap_numerator_GWAS', 'overlap_denominator_GWAS'\n",
       "    cNMF_4:\t4000 x 4\n",
       "      obs:\t'sample', 'batch'\n",
       "      var:\t'VarianceExplained'\n",
       "      uns:\t'var_names', 'gsea_varmap_GSEA', 'motif_varmap_Motif', 'gwas_varmap_GWAS'\n",
       "      varm:\t'loadings', 'loadings_zscore', 'Description_GSEA', 'pvalue_GSEA', 'p.adjust_GSEA', 'qvalue_GSEA', 'EPType_Motif', 'TFMotif_Motif', 'PValue_Motif', 'FDR_Motif', 'Enrichment_Motif', 'program_name_GWAS', 'Gene_set_GWAS', 'P-value_GWAS', 'Adjusted_P-value_GWAS', 'Odds_Ratio_GWAS', 'Combined_Score_GWAS', 'Genes_GWAS', 'overlap_numerator_GWAS', 'overlap_denominator_GWAS'\n",
       "    cNMF_5:\t4000 x 5\n",
       "      obs:\t'sample', 'batch'\n",
       "      var:\t'VarianceExplained'\n",
       "      uns:\t'var_names', 'gsea_varmap_GSEA', 'motif_varmap_Motif', 'gwas_varmap_GWAS'\n",
       "      varm:\t'loadings', 'loadings_zscore', 'Description_GSEA', 'pvalue_GSEA', 'p.adjust_GSEA', 'qvalue_GSEA', 'EPType_Motif', 'TFMotif_Motif', 'PValue_Motif', 'FDR_Motif', 'Enrichment_Motif', 'program_name_GWAS', 'Gene_set_GWAS', 'P-value_GWAS', 'Adjusted_P-value_GWAS', 'Odds_Ratio_GWAS', 'Combined_Score_GWAS', 'Genes_GWAS', 'overlap_numerator_GWAS', 'overlap_denominator_GWAS'\n",
       "    cNMF_6:\t4000 x 6\n",
       "      obs:\t'sample', 'batch'\n",
       "      var:\t'VarianceExplained'\n",
       "      uns:\t'var_names', 'gsea_varmap_GSEA', 'motif_varmap_Motif', 'gwas_varmap_GWAS'\n",
       "      varm:\t'loadings', 'loadings_zscore', 'Description_GSEA', 'pvalue_GSEA', 'p.adjust_GSEA', 'qvalue_GSEA', 'EPType_Motif', 'TFMotif_Motif', 'PValue_Motif', 'FDR_Motif', 'Enrichment_Motif', 'program_name_GWAS', 'Gene_set_GWAS', 'P-value_GWAS', 'Adjusted_P-value_GWAS', 'Odds_Ratio_GWAS', 'Combined_Score_GWAS', 'Genes_GWAS', 'overlap_numerator_GWAS', 'overlap_denominator_GWAS'\n",
       "    cNMF_7:\t4000 x 7\n",
       "      obs:\t'sample', 'batch'\n",
       "      var:\t'VarianceExplained'\n",
       "      uns:\t'var_names', 'gsea_varmap_GSEA', 'motif_varmap_Motif', 'gwas_varmap_GWAS'\n",
       "      varm:\t'loadings', 'loadings_zscore', 'Description_GSEA', 'pvalue_GSEA', 'p.adjust_GSEA', 'qvalue_GSEA', 'EPType_Motif', 'TFMotif_Motif', 'PValue_Motif', 'FDR_Motif', 'Enrichment_Motif', 'program_name_GWAS', 'Gene_set_GWAS', 'P-value_GWAS', 'Adjusted_P-value_GWAS', 'Odds_Ratio_GWAS', 'Combined_Score_GWAS', 'Genes_GWAS', 'overlap_numerator_GWAS', 'overlap_denominator_GWAS'\n",
       "    cNMF_8:\t4000 x 8\n",
       "      obs:\t'sample', 'batch'\n",
       "      var:\t'VarianceExplained'\n",
       "      uns:\t'var_names', 'gsea_varmap_GSEA', 'motif_varmap_Motif', 'gwas_varmap_GWAS'\n",
       "      varm:\t'loadings', 'loadings_zscore', 'Description_GSEA', 'pvalue_GSEA', 'p.adjust_GSEA', 'qvalue_GSEA', 'EPType_Motif', 'TFMotif_Motif', 'PValue_Motif', 'FDR_Motif', 'Enrichment_Motif', 'program_name_GWAS', 'Gene_set_GWAS', 'P-value_GWAS', 'Adjusted_P-value_GWAS', 'Odds_Ratio_GWAS', 'Combined_Score_GWAS', 'Genes_GWAS', 'overlap_numerator_GWAS', 'overlap_denominator_GWAS'\n",
       "    cNMF_9:\t4000 x 9\n",
       "      obs:\t'sample', 'batch'\n",
       "      var:\t'VarianceExplained'\n",
       "      uns:\t'var_names', 'gsea_varmap_GSEA', 'motif_varmap_Motif', 'gwas_varmap_GWAS'\n",
       "      varm:\t'loadings', 'loadings_zscore', 'Description_GSEA', 'pvalue_GSEA', 'p.adjust_GSEA', 'qvalue_GSEA', 'EPType_Motif', 'TFMotif_Motif', 'PValue_Motif', 'FDR_Motif', 'Enrichment_Motif', 'program_name_GWAS', 'Gene_set_GWAS', 'P-value_GWAS', 'Adjusted_P-value_GWAS', 'Odds_Ratio_GWAS', 'Combined_Score_GWAS', 'Genes_GWAS', 'overlap_numerator_GWAS', 'overlap_denominator_GWAS'\n",
       "    cNMF_10:\t4000 x 10\n",
       "      obs:\t'sample', 'batch'\n",
       "      var:\t'VarianceExplained'\n",
       "      uns:\t'var_names', 'gsea_varmap_GSEA', 'motif_varmap_Motif', 'gwas_varmap_GWAS'\n",
       "      varm:\t'loadings', 'loadings_zscore', 'Description_GSEA', 'pvalue_GSEA', 'p.adjust_GSEA', 'qvalue_GSEA', 'EPType_Motif', 'TFMotif_Motif', 'PValue_Motif', 'FDR_Motif', 'Enrichment_Motif', 'program_name_GWAS', 'Gene_set_GWAS', 'P-value_GWAS', 'Adjusted_P-value_GWAS', 'Odds_Ratio_GWAS', 'Combined_Score_GWAS', 'Genes_GWAS', 'overlap_numerator_GWAS', 'overlap_denominator_GWAS'\n",
       "    cNMF_11:\t4000 x 11\n",
       "      obs:\t'sample', 'batch'\n",
       "      var:\t'VarianceExplained'\n",
       "      uns:\t'var_names', 'gsea_varmap_GSEA', 'motif_varmap_Motif', 'gwas_varmap_GWAS'\n",
       "      varm:\t'loadings', 'loadings_zscore', 'Description_GSEA', 'pvalue_GSEA', 'p.adjust_GSEA', 'qvalue_GSEA', 'EPType_Motif', 'TFMotif_Motif', 'PValue_Motif', 'FDR_Motif', 'Enrichment_Motif', 'program_name_GWAS', 'Gene_set_GWAS', 'P-value_GWAS', 'Adjusted_P-value_GWAS', 'Odds_Ratio_GWAS', 'Combined_Score_GWAS', 'Genes_GWAS', 'overlap_numerator_GWAS', 'overlap_denominator_GWAS'\n",
       "    cNMF_12:\t4000 x 12\n",
       "      obs:\t'sample', 'batch'\n",
       "      var:\t'VarianceExplained'\n",
       "      uns:\t'var_names', 'gsea_varmap_GSEA', 'motif_varmap_Motif', 'gwas_varmap_GWAS'\n",
       "      varm:\t'loadings', 'loadings_zscore', 'Description_GSEA', 'pvalue_GSEA', 'p.adjust_GSEA', 'qvalue_GSEA', 'EPType_Motif', 'TFMotif_Motif', 'PValue_Motif', 'FDR_Motif', 'Enrichment_Motif', 'program_name_GWAS', 'Gene_set_GWAS', 'P-value_GWAS', 'Adjusted_P-value_GWAS', 'Odds_Ratio_GWAS', 'Combined_Score_GWAS', 'Genes_GWAS', 'overlap_numerator_GWAS', 'overlap_denominator_GWAS'\n",
       "    cNMF_13:\t4000 x 13\n",
       "      obs:\t'sample', 'batch'\n",
       "      var:\t'VarianceExplained'\n",
       "      uns:\t'var_names', 'gsea_varmap_GSEA', 'motif_varmap_Motif', 'gwas_varmap_GWAS'\n",
       "      varm:\t'loadings', 'loadings_zscore', 'Description_GSEA', 'pvalue_GSEA', 'p.adjust_GSEA', 'qvalue_GSEA', 'EPType_Motif', 'TFMotif_Motif', 'PValue_Motif', 'FDR_Motif', 'Enrichment_Motif', 'program_name_GWAS', 'Gene_set_GWAS', 'P-value_GWAS', 'Adjusted_P-value_GWAS', 'Odds_Ratio_GWAS', 'Combined_Score_GWAS', 'Genes_GWAS', 'overlap_numerator_GWAS', 'overlap_denominator_GWAS'\n",
       "    cNMF_14:\t4000 x 14\n",
       "      obs:\t'sample', 'batch'\n",
       "      var:\t'VarianceExplained'\n",
       "      uns:\t'var_names', 'gsea_varmap_GSEA', 'motif_varmap_Motif', 'gwas_varmap_GWAS'\n",
       "      varm:\t'loadings', 'loadings_zscore', 'Description_GSEA', 'pvalue_GSEA', 'p.adjust_GSEA', 'qvalue_GSEA', 'EPType_Motif', 'TFMotif_Motif', 'PValue_Motif', 'FDR_Motif', 'Enrichment_Motif', 'program_name_GWAS', 'Gene_set_GWAS', 'P-value_GWAS', 'Adjusted_P-value_GWAS', 'Odds_Ratio_GWAS', 'Combined_Score_GWAS', 'Genes_GWAS', 'overlap_numerator_GWAS', 'overlap_denominator_GWAS'\n",
       "    cNMF_15:\t4000 x 15\n",
       "      obs:\t'sample', 'batch'\n",
       "      var:\t'VarianceExplained'\n",
       "      uns:\t'var_names', 'gsea_varmap_GSEA', 'motif_varmap_Motif', 'gwas_varmap_GWAS'\n",
       "      varm:\t'loadings', 'loadings_zscore', 'Description_GSEA', 'pvalue_GSEA', 'p.adjust_GSEA', 'qvalue_GSEA', 'EPType_Motif', 'TFMotif_Motif', 'PValue_Motif', 'FDR_Motif', 'Enrichment_Motif', 'program_name_GWAS', 'Gene_set_GWAS', 'P-value_GWAS', 'Adjusted_P-value_GWAS', 'Odds_Ratio_GWAS', 'Combined_Score_GWAS', 'Genes_GWAS', 'overlap_numerator_GWAS', 'overlap_denominator_GWAS'\n",
       "    cNMF_17:\t4000 x 17\n",
       "      obs:\t'sample', 'batch'\n",
       "      var:\t'VarianceExplained'\n",
       "      uns:\t'var_names', 'gsea_varmap_GSEA', 'motif_varmap_Motif', 'gwas_varmap_GWAS'\n",
       "      varm:\t'loadings', 'loadings_zscore', 'Description_GSEA', 'pvalue_GSEA', 'p.adjust_GSEA', 'qvalue_GSEA', 'EPType_Motif', 'TFMotif_Motif', 'PValue_Motif', 'FDR_Motif', 'Enrichment_Motif', 'program_name_GWAS', 'Gene_set_GWAS', 'P-value_GWAS', 'Adjusted_P-value_GWAS', 'Odds_Ratio_GWAS', 'Combined_Score_GWAS', 'Genes_GWAS', 'overlap_numerator_GWAS', 'overlap_denominator_GWAS'\n",
       "    cNMF_19:\t4000 x 19\n",
       "      obs:\t'sample', 'batch'\n",
       "      var:\t'VarianceExplained'\n",
       "      uns:\t'var_names', 'gsea_varmap_GSEA', 'motif_varmap_Motif', 'gwas_varmap_GWAS'\n",
       "      varm:\t'loadings', 'loadings_zscore', 'Description_GSEA', 'pvalue_GSEA', 'p.adjust_GSEA', 'qvalue_GSEA', 'EPType_Motif', 'TFMotif_Motif', 'PValue_Motif', 'FDR_Motif', 'Enrichment_Motif', 'program_name_GWAS', 'Gene_set_GWAS', 'P-value_GWAS', 'Adjusted_P-value_GWAS', 'Odds_Ratio_GWAS', 'Combined_Score_GWAS', 'Genes_GWAS', 'overlap_numerator_GWAS', 'overlap_denominator_GWAS'\n",
       "    cNMF_21:\t4000 x 21\n",
       "      obs:\t'sample', 'batch'\n",
       "      var:\t'VarianceExplained'\n",
       "      uns:\t'var_names', 'gsea_varmap_GSEA', 'motif_varmap_Motif', 'gwas_varmap_GWAS'\n",
       "      varm:\t'loadings', 'loadings_zscore', 'Description_GSEA', 'pvalue_GSEA', 'p.adjust_GSEA', 'qvalue_GSEA', 'EPType_Motif', 'TFMotif_Motif', 'PValue_Motif', 'FDR_Motif', 'Enrichment_Motif', 'program_name_GWAS', 'Gene_set_GWAS', 'P-value_GWAS', 'Adjusted_P-value_GWAS', 'Odds_Ratio_GWAS', 'Combined_Score_GWAS', 'Genes_GWAS', 'overlap_numerator_GWAS', 'overlap_denominator_GWAS'\n",
       "    cNMF_23:\t4000 x 23\n",
       "      obs:\t'sample', 'batch'\n",
       "      var:\t'VarianceExplained'\n",
       "      uns:\t'var_names', 'gsea_varmap_GSEA', 'motif_varmap_Motif', 'gwas_varmap_GWAS'\n",
       "      varm:\t'loadings', 'loadings_zscore', 'Description_GSEA', 'pvalue_GSEA', 'p.adjust_GSEA', 'qvalue_GSEA', 'EPType_Motif', 'TFMotif_Motif', 'PValue_Motif', 'FDR_Motif', 'Enrichment_Motif', 'program_name_GWAS', 'Gene_set_GWAS', 'P-value_GWAS', 'Adjusted_P-value_GWAS', 'Odds_Ratio_GWAS', 'Combined_Score_GWAS', 'Genes_GWAS', 'overlap_numerator_GWAS', 'overlap_denominator_GWAS'\n",
       "    cNMF_25:\t4000 x 25\n",
       "      obs:\t'sample', 'batch'\n",
       "      var:\t'VarianceExplained'\n",
       "      uns:\t'var_names', 'gsea_varmap_GSEA', 'motif_varmap_Motif', 'gwas_varmap_GWAS'\n",
       "      varm:\t'loadings', 'loadings_zscore', 'Description_GSEA', 'pvalue_GSEA', 'p.adjust_GSEA', 'qvalue_GSEA', 'EPType_Motif', 'TFMotif_Motif', 'PValue_Motif', 'FDR_Motif', 'Enrichment_Motif', 'program_name_GWAS', 'Gene_set_GWAS', 'P-value_GWAS', 'Adjusted_P-value_GWAS', 'Odds_Ratio_GWAS', 'Combined_Score_GWAS', 'Genes_GWAS', 'overlap_numerator_GWAS', 'overlap_denominator_GWAS'\n",
       "    cNMF_27:\t4000 x 27\n",
       "      obs:\t'sample', 'batch'\n",
       "      var:\t'VarianceExplained'\n",
       "      uns:\t'var_names', 'gsea_varmap_GSEA', 'motif_varmap_Motif', 'gwas_varmap_GWAS'\n",
       "      varm:\t'loadings', 'loadings_zscore', 'Description_GSEA', 'pvalue_GSEA', 'p.adjust_GSEA', 'qvalue_GSEA', 'EPType_Motif', 'TFMotif_Motif', 'PValue_Motif', 'FDR_Motif', 'Enrichment_Motif', 'program_name_GWAS', 'Gene_set_GWAS', 'P-value_GWAS', 'Adjusted_P-value_GWAS', 'Odds_Ratio_GWAS', 'Combined_Score_GWAS', 'Genes_GWAS', 'overlap_numerator_GWAS', 'overlap_denominator_GWAS'\n",
       "    cNMF_29:\t4000 x 29\n",
       "      obs:\t'sample', 'batch'\n",
       "      var:\t'VarianceExplained'\n",
       "      uns:\t'var_names', 'gsea_varmap_GSEA', 'motif_varmap_Motif', 'gwas_varmap_GWAS'\n",
       "      varm:\t'loadings', 'loadings_zscore', 'Description_GSEA', 'pvalue_GSEA', 'p.adjust_GSEA', 'qvalue_GSEA', 'EPType_Motif', 'TFMotif_Motif', 'PValue_Motif', 'FDR_Motif', 'Enrichment_Motif', 'program_name_GWAS', 'Gene_set_GWAS', 'P-value_GWAS', 'Adjusted_P-value_GWAS', 'Odds_Ratio_GWAS', 'Combined_Score_GWAS', 'Genes_GWAS', 'overlap_numerator_GWAS', 'overlap_denominator_GWAS'\n",
       "    cNMF_35:\t4000 x 35\n",
       "      obs:\t'sample', 'batch'\n",
       "      var:\t'VarianceExplained'\n",
       "      uns:\t'var_names', 'gsea_varmap_GSEA', 'motif_varmap_Motif', 'gwas_varmap_GWAS'\n",
       "      varm:\t'loadings', 'loadings_zscore', 'Description_GSEA', 'pvalue_GSEA', 'p.adjust_GSEA', 'qvalue_GSEA', 'EPType_Motif', 'TFMotif_Motif', 'PValue_Motif', 'FDR_Motif', 'Enrichment_Motif', 'program_name_GWAS', 'Gene_set_GWAS', 'P-value_GWAS', 'Adjusted_P-value_GWAS', 'Odds_Ratio_GWAS', 'Combined_Score_GWAS', 'Genes_GWAS', 'overlap_numerator_GWAS', 'overlap_denominator_GWAS'\n",
       "    cNMF_40:\t4000 x 40\n",
       "      obs:\t'sample', 'batch'\n",
       "      var:\t'VarianceExplained'\n",
       "      uns:\t'var_names', 'gsea_varmap_GSEA', 'motif_varmap_Motif', 'gwas_varmap_GWAS'\n",
       "      varm:\t'loadings', 'loadings_zscore', 'Description_GSEA', 'pvalue_GSEA', 'p.adjust_GSEA', 'qvalue_GSEA', 'EPType_Motif', 'TFMotif_Motif', 'PValue_Motif', 'FDR_Motif', 'Enrichment_Motif', 'program_name_GWAS', 'Gene_set_GWAS', 'P-value_GWAS', 'Adjusted_P-value_GWAS', 'Odds_Ratio_GWAS', 'Combined_Score_GWAS', 'Genes_GWAS', 'overlap_numerator_GWAS', 'overlap_denominator_GWAS'\n",
       "    cNMF_45:\t4000 x 45\n",
       "      obs:\t'sample', 'batch'\n",
       "      var:\t'VarianceExplained'\n",
       "      uns:\t'var_names', 'gsea_varmap_GSEA', 'motif_varmap_Motif', 'gwas_varmap_GWAS'\n",
       "      varm:\t'loadings', 'loadings_zscore', 'Description_GSEA', 'pvalue_GSEA', 'p.adjust_GSEA', 'qvalue_GSEA', 'EPType_Motif', 'TFMotif_Motif', 'PValue_Motif', 'FDR_Motif', 'Enrichment_Motif', 'program_name_GWAS', 'Gene_set_GWAS', 'P-value_GWAS', 'Adjusted_P-value_GWAS', 'Odds_Ratio_GWAS', 'Combined_Score_GWAS', 'Genes_GWAS', 'overlap_numerator_GWAS', 'overlap_denominator_GWAS'\n",
       "    cNMF_50:\t4000 x 50\n",
       "      obs:\t'sample', 'batch'\n",
       "      var:\t'VarianceExplained'\n",
       "      uns:\t'var_names', 'gsea_varmap_GSEA', 'motif_varmap_Motif', 'gwas_varmap_GWAS'\n",
       "      varm:\t'loadings', 'loadings_zscore', 'Description_GSEA', 'pvalue_GSEA', 'p.adjust_GSEA', 'qvalue_GSEA', 'EPType_Motif', 'TFMotif_Motif', 'PValue_Motif', 'FDR_Motif', 'Enrichment_Motif', 'program_name_GWAS', 'Gene_set_GWAS', 'P-value_GWAS', 'Adjusted_P-value_GWAS', 'Odds_Ratio_GWAS', 'Combined_Score_GWAS', 'Genes_GWAS', 'overlap_numerator_GWAS', 'overlap_denominator_GWAS'\n",
       "    cNMF_55:\t4000 x 55\n",
       "      obs:\t'sample', 'batch'\n",
       "      var:\t'VarianceExplained'\n",
       "      uns:\t'var_names', 'gsea_varmap_GSEA', 'motif_varmap_Motif', 'gwas_varmap_GWAS'\n",
       "      varm:\t'loadings', 'loadings_zscore', 'Description_GSEA', 'pvalue_GSEA', 'p.adjust_GSEA', 'qvalue_GSEA', 'EPType_Motif', 'TFMotif_Motif', 'PValue_Motif', 'FDR_Motif', 'Enrichment_Motif', 'program_name_GWAS', 'Gene_set_GWAS', 'P-value_GWAS', 'Adjusted_P-value_GWAS', 'Odds_Ratio_GWAS', 'Combined_Score_GWAS', 'Genes_GWAS', 'overlap_numerator_GWAS', 'overlap_denominator_GWAS'\n",
       "    cNMF_60:\t4000 x 60\n",
       "      obs:\t'sample', 'batch'\n",
       "      var:\t'VarianceExplained'\n",
       "      uns:\t'var_names', 'gsea_varmap_GSEA', 'motif_varmap_Motif', 'gwas_varmap_GWAS'\n",
       "      varm:\t'loadings', 'loadings_zscore', 'Description_GSEA', 'pvalue_GSEA', 'p.adjust_GSEA', 'qvalue_GSEA', 'EPType_Motif', 'TFMotif_Motif', 'PValue_Motif', 'FDR_Motif', 'Enrichment_Motif', 'program_name_GWAS', 'Gene_set_GWAS', 'P-value_GWAS', 'Adjusted_P-value_GWAS', 'Odds_Ratio_GWAS', 'Combined_Score_GWAS', 'Genes_GWAS', 'overlap_numerator_GWAS', 'overlap_denominator_GWAS'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "with mudata.set_options(display_style = \"html\", display_html_expand = 0b000):\n",
    "    display(mdata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 390,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 4000 × 3\n",
       "    obs: 'sample', 'batch'\n",
       "    var: 'VarianceExplained'\n",
       "    uns: 'var_names', 'gsea_varmap_GSEA', 'motif_varmap_Motif', 'gwas_varmap_GWAS'\n",
       "    varm: 'loadings', 'loadings_zscore', 'Description_GSEA', 'pvalue_GSEA', 'p.adjust_GSEA', 'qvalue_GSEA', 'EPType_Motif', 'TFMotif_Motif', 'PValue_Motif', 'FDR_Motif', 'Enrichment_Motif', 'program_name_GWAS', 'Gene_set_GWAS', 'P-value_GWAS', 'Adjusted_P-value_GWAS', 'Odds_Ratio_GWAS', 'Combined_Score_GWAS', 'Genes_GWAS', 'overlap_numerator_GWAS', 'overlap_denominator_GWAS'"
      ]
     },
     "execution_count": 390,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mdata[\"cNMF_3\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.10 test_celloracle",
   "language": "python",
   "name": "test_scenicplus-py310"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
