{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "eb9afaa8",
   "metadata": {},
   "source": [
    "# How to Query OpenTargets Genetics for L2G Scores"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b7a3eb23",
   "metadata": {},
   "source": [
    "### 1. Create Google Cloud Credentials\n",
    "\n",
    "This step explains how to setup a Google Cloud Service Account to use Google BigQuery. Querying the first 1TB/month of public BigQuery is free, but there are fees beyond that, so you will need to link payment information to your credentials to run queries. Unless you are doing very large queries, you should not reach the 1TB limit.\n",
    "\n",
    "1. Go to https://console.cloud.google.com/\n",
    "2. If you do not have a project already, create one.\n",
    "3. Go to https://console.cloud.google.com/iam-admin/serviceaccounts & click `+ CREATE SERVICE ACCOUNT`\n",
    "4. Enter a descriptive Service Account name like \"groberts-query-opentargets\" & set the role to \"owner\"\n",
    "5. Once back on the Service Accounts main page, click the \"Actions\" menu button > \"Manage Keys\"\n",
    "6. `ADD KEY` > `Create New Key` > Select`JSON`.\n",
    "7. A `.json` file will automatically download. **Be careful with your `.json`. You will need to call this file, but DO NOT upload it to GitHub as these are your private credentials and malicious actors could use your credentials to rack up expensive queries**\n",
    "\n",
    "Once your service account is setup, the query below shows you how to access data in the BigQuery database\n",
    "`bigquery-public-data.open_targets_genetics`.  You can view the table schema in the OpenTargets Genetics database by going to \n",
    "https://console.cloud.google.com/bigquery?p=bigquery-public-data&d=open_targets_genetics&page=dataset .\n",
    "\n",
    "Note that there is a separate database for the OpenTargets Platform `bigquery-public-data.open_targets_platform` that contains additional information that you may wish to map to the OpenTargets Genetics data you are pulling in, for example known drugs that target GWAS genes."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c2939f42",
   "metadata": {},
   "source": [
    "## 2. Run the Query\n",
    "\n",
    "This demonstrates how to pull in the GWAS lead variants and genes with an L2G score of at least 0.2 across all GWAS with at least 3,000 cases and at least 10 independent genome-wide significant loci. It processes ~0.95GB of data in ~1 minute and returns 41,686 gene-GWAS connection results.\n",
    "\n",
    "This query can be modified to be more restrictive and thereby be return results much more quickly."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "305aaefd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Setup credentials for BigQuery (.json generated in 7 steps above)\n",
    "\n",
    "from google.cloud import bigquery\n",
    "credentials_path = '/home/robertg1/.ssh/test-bigquery-ot-956f8a01208f.json' #Do not push this file to GitHub!\n",
    "client = bigquery.Client.from_service_account_json(credentials_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "7088039c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Basic Imports\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b8bbfb81",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study_id</th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>study_id_1</th>\n",
       "      <th>ancestry_initial</th>\n",
       "      <th>ancestry_replication</th>\n",
       "      <th>n_cases</th>\n",
       "      <th>n_initial</th>\n",
       "      <th>...</th>\n",
       "      <th>has_sumstats</th>\n",
       "      <th>num_assoc_loci</th>\n",
       "      <th>source</th>\n",
       "      <th>trait_reported</th>\n",
       "      <th>trait_efos</th>\n",
       "      <th>trait_category</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>y_proba_full_model</th>\n",
       "      <th>pval</th>\n",
       "      <th>rn</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>41253</th>\n",
       "      <td>FINNGEN_R6_ABDOM_HERNIA</td>\n",
       "      <td>1</td>\n",
       "      <td>219570796</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>FINNGEN_R6_ABDOM_HERNIA</td>\n",
       "      <td>{'list': [{'element': 'European=260405'}]}</td>\n",
       "      <td>{'list': []}</td>\n",
       "      <td>9245</td>\n",
       "      <td>260405</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>11</td>\n",
       "      <td>FINNGEN</td>\n",
       "      <td>Hernia of abodminal wall</td>\n",
       "      <td>{'list': [{'element': 'EFO_1001866'}]}</td>\n",
       "      <td>gastrointestinal disease</td>\n",
       "      <td>ZC3H11B</td>\n",
       "      <td>0.685505</td>\n",
       "      <td>5.520000e-43</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36613</th>\n",
       "      <td>FINNGEN_R6_ABDOM_HERNIA</td>\n",
       "      <td>7</td>\n",
       "      <td>134907837</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>FINNGEN_R6_ABDOM_HERNIA</td>\n",
       "      <td>{'list': [{'element': 'European=260405'}]}</td>\n",
       "      <td>{'list': []}</td>\n",
       "      <td>9245</td>\n",
       "      <td>260405</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>11</td>\n",
       "      <td>FINNGEN</td>\n",
       "      <td>Hernia of abodminal wall</td>\n",
       "      <td>{'list': [{'element': 'EFO_1001866'}]}</td>\n",
       "      <td>gastrointestinal disease</td>\n",
       "      <td>CALD1</td>\n",
       "      <td>0.595480</td>\n",
       "      <td>1.000000e-15</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41252</th>\n",
       "      <td>FINNGEN_R6_ABDOM_HERNIA</td>\n",
       "      <td>7</td>\n",
       "      <td>134907837</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>FINNGEN_R6_ABDOM_HERNIA</td>\n",
       "      <td>{'list': [{'element': 'European=260405'}]}</td>\n",
       "      <td>{'list': []}</td>\n",
       "      <td>9245</td>\n",
       "      <td>260405</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>11</td>\n",
       "      <td>FINNGEN</td>\n",
       "      <td>Hernia of abodminal wall</td>\n",
       "      <td>{'list': [{'element': 'EFO_1001866'}]}</td>\n",
       "      <td>gastrointestinal disease</td>\n",
       "      <td>AGBL3</td>\n",
       "      <td>0.217192</td>\n",
       "      <td>1.000000e-15</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36873</th>\n",
       "      <td>FINNGEN_R6_ABDOM_HERNIA</td>\n",
       "      <td>2</td>\n",
       "      <td>19576027</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>FINNGEN_R6_ABDOM_HERNIA</td>\n",
       "      <td>{'list': [{'element': 'European=260405'}]}</td>\n",
       "      <td>{'list': []}</td>\n",
       "      <td>9245</td>\n",
       "      <td>260405</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>11</td>\n",
       "      <td>FINNGEN</td>\n",
       "      <td>Hernia of abodminal wall</td>\n",
       "      <td>{'list': [{'element': 'EFO_1001866'}]}</td>\n",
       "      <td>gastrointestinal disease</td>\n",
       "      <td>OSR1</td>\n",
       "      <td>0.551340</td>\n",
       "      <td>1.320000e-13</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41357</th>\n",
       "      <td>FINNGEN_R6_ABDOM_HERNIA</td>\n",
       "      <td>12</td>\n",
       "      <td>77829818</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>FINNGEN_R6_ABDOM_HERNIA</td>\n",
       "      <td>{'list': [{'element': 'European=260405'}]}</td>\n",
       "      <td>{'list': []}</td>\n",
       "      <td>9245</td>\n",
       "      <td>260405</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>11</td>\n",
       "      <td>FINNGEN</td>\n",
       "      <td>Hernia of abodminal wall</td>\n",
       "      <td>{'list': [{'element': 'EFO_1001866'}]}</td>\n",
       "      <td>gastrointestinal disease</td>\n",
       "      <td>NAV3</td>\n",
       "      <td>0.732569</td>\n",
       "      <td>7.730000e-11</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23885</th>\n",
       "      <td>SAIGE_740_1</td>\n",
       "      <td>1</td>\n",
       "      <td>102938828</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>SAIGE_740_1</td>\n",
       "      <td>{'list': [{'element': 'European=398213'}]}</td>\n",
       "      <td>{'list': []}</td>\n",
       "      <td>17691</td>\n",
       "      <td>398213</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>10</td>\n",
       "      <td>SAIGE</td>\n",
       "      <td>Osteoarthritis; localized</td>\n",
       "      <td>{'list': [{'element': 'EFO_0002506'}]}</td>\n",
       "      <td>Uncategorised</td>\n",
       "      <td>COL11A1</td>\n",
       "      <td>0.663457</td>\n",
       "      <td>4.860000e-10</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22739</th>\n",
       "      <td>SAIGE_740_1</td>\n",
       "      <td>6</td>\n",
       "      <td>75478614</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>SAIGE_740_1</td>\n",
       "      <td>{'list': [{'element': 'European=398213'}]}</td>\n",
       "      <td>{'list': []}</td>\n",
       "      <td>17691</td>\n",
       "      <td>398213</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>10</td>\n",
       "      <td>SAIGE</td>\n",
       "      <td>Osteoarthritis; localized</td>\n",
       "      <td>{'list': [{'element': 'EFO_0002506'}]}</td>\n",
       "      <td>Uncategorised</td>\n",
       "      <td>FILIP1</td>\n",
       "      <td>0.754302</td>\n",
       "      <td>4.710000e-09</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22703</th>\n",
       "      <td>SAIGE_740_1</td>\n",
       "      <td>3</td>\n",
       "      <td>52422957</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SAIGE_740_1</td>\n",
       "      <td>{'list': [{'element': 'European=398213'}]}</td>\n",
       "      <td>{'list': []}</td>\n",
       "      <td>17691</td>\n",
       "      <td>398213</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>10</td>\n",
       "      <td>SAIGE</td>\n",
       "      <td>Osteoarthritis; localized</td>\n",
       "      <td>{'list': [{'element': 'EFO_0002506'}]}</td>\n",
       "      <td>Uncategorised</td>\n",
       "      <td>ITIH4</td>\n",
       "      <td>0.242676</td>\n",
       "      <td>1.420000e-08</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22565</th>\n",
       "      <td>SAIGE_740_1</td>\n",
       "      <td>6</td>\n",
       "      <td>44991647</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>SAIGE_740_1</td>\n",
       "      <td>{'list': [{'element': 'European=398213'}]}</td>\n",
       "      <td>{'list': []}</td>\n",
       "      <td>17691</td>\n",
       "      <td>398213</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>10</td>\n",
       "      <td>SAIGE</td>\n",
       "      <td>Osteoarthritis; localized</td>\n",
       "      <td>{'list': [{'element': 'EFO_0002506'}]}</td>\n",
       "      <td>Uncategorised</td>\n",
       "      <td>RUNX2</td>\n",
       "      <td>0.379980</td>\n",
       "      <td>2.630000e-08</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23922</th>\n",
       "      <td>SAIGE_740_1</td>\n",
       "      <td>6</td>\n",
       "      <td>44991647</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>SAIGE_740_1</td>\n",
       "      <td>{'list': [{'element': 'European=398213'}]}</td>\n",
       "      <td>{'list': []}</td>\n",
       "      <td>17691</td>\n",
       "      <td>398213</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>10</td>\n",
       "      <td>SAIGE</td>\n",
       "      <td>Osteoarthritis; localized</td>\n",
       "      <td>{'list': [{'element': 'EFO_0002506'}]}</td>\n",
       "      <td>Uncategorised</td>\n",
       "      <td>SUPT3H</td>\n",
       "      <td>0.391780</td>\n",
       "      <td>2.630000e-08</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>41686 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      study_id chrom        pos ref alt  \\\n",
       "41253  FINNGEN_R6_ABDOM_HERNIA     1  219570796   C   A   \n",
       "36613  FINNGEN_R6_ABDOM_HERNIA     7  134907837   G   A   \n",
       "41252  FINNGEN_R6_ABDOM_HERNIA     7  134907837   G   A   \n",
       "36873  FINNGEN_R6_ABDOM_HERNIA     2   19576027   C   A   \n",
       "41357  FINNGEN_R6_ABDOM_HERNIA    12   77829818   G   A   \n",
       "...                        ...   ...        ...  ..  ..   \n",
       "23885              SAIGE_740_1     1  102938828   T   C   \n",
       "22739              SAIGE_740_1     6   75478614   C   A   \n",
       "22703              SAIGE_740_1     3   52422957   A   G   \n",
       "22565              SAIGE_740_1     6   44991647   A   T   \n",
       "23922              SAIGE_740_1     6   44991647   A   T   \n",
       "\n",
       "                    study_id_1                            ancestry_initial  \\\n",
       "41253  FINNGEN_R6_ABDOM_HERNIA  {'list': [{'element': 'European=260405'}]}   \n",
       "36613  FINNGEN_R6_ABDOM_HERNIA  {'list': [{'element': 'European=260405'}]}   \n",
       "41252  FINNGEN_R6_ABDOM_HERNIA  {'list': [{'element': 'European=260405'}]}   \n",
       "36873  FINNGEN_R6_ABDOM_HERNIA  {'list': [{'element': 'European=260405'}]}   \n",
       "41357  FINNGEN_R6_ABDOM_HERNIA  {'list': [{'element': 'European=260405'}]}   \n",
       "...                        ...                                         ...   \n",
       "23885              SAIGE_740_1  {'list': [{'element': 'European=398213'}]}   \n",
       "22739              SAIGE_740_1  {'list': [{'element': 'European=398213'}]}   \n",
       "22703              SAIGE_740_1  {'list': [{'element': 'European=398213'}]}   \n",
       "22565              SAIGE_740_1  {'list': [{'element': 'European=398213'}]}   \n",
       "23922              SAIGE_740_1  {'list': [{'element': 'European=398213'}]}   \n",
       "\n",
       "      ancestry_replication  n_cases  n_initial  ...  has_sumstats  \\\n",
       "41253         {'list': []}     9245     260405  ...          True   \n",
       "36613         {'list': []}     9245     260405  ...          True   \n",
       "41252         {'list': []}     9245     260405  ...          True   \n",
       "36873         {'list': []}     9245     260405  ...          True   \n",
       "41357         {'list': []}     9245     260405  ...          True   \n",
       "...                    ...      ...        ...  ...           ...   \n",
       "23885         {'list': []}    17691     398213  ...          True   \n",
       "22739         {'list': []}    17691     398213  ...          True   \n",
       "22703         {'list': []}    17691     398213  ...          True   \n",
       "22565         {'list': []}    17691     398213  ...          True   \n",
       "23922         {'list': []}    17691     398213  ...          True   \n",
       "\n",
       "      num_assoc_loci   source             trait_reported  \\\n",
       "41253             11  FINNGEN   Hernia of abodminal wall   \n",
       "36613             11  FINNGEN   Hernia of abodminal wall   \n",
       "41252             11  FINNGEN   Hernia of abodminal wall   \n",
       "36873             11  FINNGEN   Hernia of abodminal wall   \n",
       "41357             11  FINNGEN   Hernia of abodminal wall   \n",
       "...              ...      ...                        ...   \n",
       "23885             10    SAIGE  Osteoarthritis; localized   \n",
       "22739             10    SAIGE  Osteoarthritis; localized   \n",
       "22703             10    SAIGE  Osteoarthritis; localized   \n",
       "22565             10    SAIGE  Osteoarthritis; localized   \n",
       "23922             10    SAIGE  Osteoarthritis; localized   \n",
       "\n",
       "                                   trait_efos            trait_category  \\\n",
       "41253  {'list': [{'element': 'EFO_1001866'}]}  gastrointestinal disease   \n",
       "36613  {'list': [{'element': 'EFO_1001866'}]}  gastrointestinal disease   \n",
       "41252  {'list': [{'element': 'EFO_1001866'}]}  gastrointestinal disease   \n",
       "36873  {'list': [{'element': 'EFO_1001866'}]}  gastrointestinal disease   \n",
       "41357  {'list': [{'element': 'EFO_1001866'}]}  gastrointestinal disease   \n",
       "...                                       ...                       ...   \n",
       "23885  {'list': [{'element': 'EFO_0002506'}]}             Uncategorised   \n",
       "22739  {'list': [{'element': 'EFO_0002506'}]}             Uncategorised   \n",
       "22703  {'list': [{'element': 'EFO_0002506'}]}             Uncategorised   \n",
       "22565  {'list': [{'element': 'EFO_0002506'}]}             Uncategorised   \n",
       "23922  {'list': [{'element': 'EFO_0002506'}]}             Uncategorised   \n",
       "\n",
       "       gene_name  y_proba_full_model          pval rn  \n",
       "41253    ZC3H11B            0.685505  5.520000e-43  1  \n",
       "36613      CALD1            0.595480  1.000000e-15  1  \n",
       "41252      AGBL3            0.217192  1.000000e-15  1  \n",
       "36873       OSR1            0.551340  1.320000e-13  1  \n",
       "41357       NAV3            0.732569  7.730000e-11  1  \n",
       "...          ...                 ...           ... ..  \n",
       "23885    COL11A1            0.663457  4.860000e-10  1  \n",
       "22739     FILIP1            0.754302  4.710000e-09  1  \n",
       "22703      ITIH4            0.242676  1.420000e-08  1  \n",
       "22565      RUNX2            0.379980  2.630000e-08  1  \n",
       "23922     SUPT3H            0.391780  2.630000e-08  1  \n",
       "\n",
       "[41686 rows x 26 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Optional: Define a list of GWAS Catalog IDs/SAIGE/FINNGEN/NEAL2 GWAS IDs to keep\n",
    "#           This can be useful in selecting the highest quality/latest GWAS for each trait.\n",
    "study_ids_to_keep = [\"GCST005523\", \"FINNGEN_R6_ABDOM_HERNIA\"]\n",
    "\n",
    "# Construct parameterized query\n",
    "query = '''\n",
    "WITH ranked_genes AS (\n",
    "    SELECT locus2gene.study_id, \n",
    "        locus2gene.chrom, locus2gene.pos, locus2gene.ref, locus2gene.alt, \n",
    "        study_metadata.*, \n",
    "        genes.gene_name, locus2gene.y_proba_full_model,\n",
    "        lead_variants.pval,\n",
    "        ROW_NUMBER() OVER(PARTITION BY locus2gene.study_id, genes.gene_name ORDER BY lead_variants.pval) AS rn\n",
    "    \n",
    "    FROM `bigquery-public-data.open_targets_genetics.locus2gene` AS locus2gene\n",
    "    \n",
    "    -- Get GWAS metadata\n",
    "    INNER JOIN `bigquery-public-data.open_targets_genetics.studies` AS study_metadata\n",
    "    ON locus2gene.study_id = study_metadata.study_id\n",
    "    \n",
    "    -- Get HGNC IDs\n",
    "    INNER JOIN `bigquery-public-data.open_targets_genetics.genes` AS genes\n",
    "    ON locus2gene.gene_id = genes.gene_id\n",
    "    \n",
    "    -- Get lead variant P-values\n",
    "    INNER JOIN `bigquery-public-data.open_targets_genetics.variant_disease` AS lead_variants\n",
    "    ON locus2gene.pos = lead_variants.lead_pos\n",
    "        AND locus2gene.chrom = lead_variants.lead_chrom\n",
    "        AND locus2gene.study_id = lead_variants.study_id\n",
    "    \n",
    "        -- Optional: Filter on the study IDs in the list study_ids_to_keep\n",
    "        -- WHERE locus2gene.study_id IN UNNEST(@study_ids_to_keep)\n",
    "        \n",
    "        -- Optional: filter on a disease\n",
    "        -- AND LOWER(study_metadata.trait_reported) LIKE '%coeliac%'\n",
    "        \n",
    "        -- Remove the \"raw\" Neale lab results -- I'm not sure what this is\n",
    "        AND locus2gene.study_id NOT LIKE '%raw%'\n",
    "        \n",
    "        -- Filter to a l2g score threshold\n",
    "        AND locus2gene.y_proba_full_model > 0.2\n",
    "        \n",
    "        -- Filter to number of associated loci of at least 10\n",
    "        AND study_metadata.num_assoc_loci >= 10\n",
    "        \n",
    "        -- Filter to n_cases of at least 3000\n",
    "        AND study_metadata.n_cases >= 3000\n",
    ")\n",
    "\n",
    "SELECT * FROM ranked_genes WHERE rn = 1;\n",
    "'''\n",
    "\n",
    "# Set query parameters\n",
    "job_config = bigquery.QueryJobConfig(\n",
    "    query_parameters=[\n",
    "        bigquery.ArrayQueryParameter(\"study_ids_to_keep\", \"STRING\", study_ids_to_keep)\n",
    "    ]\n",
    ")\n",
    "\n",
    "# Run the query\n",
    "# Note - as defined, this query returns \n",
    "query_job = client.query(query, job_config=job_config)\n",
    "\n",
    "# Convert the query results to a Pandas dataframe\n",
    "l2g = query_job.to_dataframe()\n",
    "l2g.sort_values(by=['study_id', 'pval'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d6d57300",
   "metadata": {},
   "source": [
    "## 3. Save the Output of the Query"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "75cdaf7a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(41686, 26)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "l2g.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "34f2cad2",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['study_id', 'chrom', 'pos', 'ref', 'alt', 'study_id_1',\n",
       "       'ancestry_initial', 'ancestry_replication', 'n_cases', 'n_initial',\n",
       "       'n_replication', 'pmid', 'pub_author', 'pub_date', 'pub_journal',\n",
       "       'pub_title', 'has_sumstats', 'num_assoc_loci', 'source',\n",
       "       'trait_reported', 'trait_efos', 'trait_category', 'gene_name',\n",
       "       'y_proba_full_model', 'pval', 'rn'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "l2g.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "bf034990",
   "metadata": {},
   "outputs": [],
   "source": [
    "l2g.to_csv('gwas_data/OpenTargets_L2G_noQC.csv.gz', compression=\"gzip\", index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "cell_programs",
   "language": "python",
   "name": "cell_programs"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
