{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Set-up"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import yaml\n",
    "import logging\n",
    "import mudata\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from tqdm.auto import tqdm\n",
    "from typing import List, Dict, Tuple, Union, Optional, Literal, Mapping\n",
    "\n",
    "from scipy.stats import pearsonr, spearmanr, kendalltau\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "\n",
    "from tangermeme.io import read_meme, extract_loci\n",
    "from tangermeme.tools.fimo import fimo\n",
    "\n",
    "from joblib import Parallel, delayed\n",
    "from tqdm.auto import tqdm\n",
    "\n",
    "# Change path to wherever you have repo locally\n",
    "sys.path.append('/cellar/users/aklie/opt/gene_program_evaluation')\n",
    "\n",
    "from src.evaluation import (\n",
    "    compute_categorical_association,\n",
    "    compute_geneset_enrichment,\n",
    "    compute_trait_enrichment,\n",
    "    compute_perturbation_association,\n",
    "    compute_explained_variance_ratio,\n",
    "    compute_motif_enrichment\n",
    ")\n",
    "from src.evaluation.enrichment_trait import process_enrichment_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# I/O paths\n",
    "path_config = \"/cellar/users/aklie/opt/gene_program_evaluation/examples/evaluation/iPSC_EC/cNMF_30/evaluation_pipeline.yml\"\n",
    "config = yaml.safe_load(open(path_config))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## I/O"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'path_mdata': '/cellar/users/aklie/opt/gene_program_evaluation/examples/inference/iPSC_EC/cNMF/cNMF_30_0.2_gene_names.h5mu',\n",
       " 'path_out': '/cellar/users/aklie/opt/gene_program_evaluation/examples/evaluation/iPSC_EC/cNMF_30',\n",
       " 'data_key': 'rna',\n",
       " 'prog_key': 'cNMF'}"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "io_config = config['io']\n",
    "io_config"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/cellar/users/aklie/opt/miniconda3/envs/test_celloracle/lib/python3.10/site-packages/anndata/_core/anndata.py:522: FutureWarning: The dtype argument is deprecated and will be removed in late 2024.\n",
      "  warnings.warn(\n",
      "/cellar/users/aklie/opt/miniconda3/envs/test_celloracle/lib/python3.10/site-packages/anndata/_core/anndata.py:522: FutureWarning: The dtype argument is deprecated and will be removed in late 2024.\n",
      "  warnings.warn(\n",
      "/cellar/users/aklie/opt/miniconda3/envs/test_celloracle/lib/python3.10/site-packages/anndata/_core/anndata.py:1899: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/cellar/users/aklie/opt/miniconda3/envs/test_celloracle/lib/python3.10/site-packages/mudata/_core/mudata.py:479: UserWarning: var_names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<pre>MuData object with n_obs × n_vars = 92284 × 5481\n",
       "  2 modalities\n",
       "    rna:\t92284 x 5451\n",
       "      obs:\t&#x27;sample&#x27;, &#x27;species&#x27;, &#x27;gene_count&#x27;, &#x27;tscp_count&#x27;, &#x27;mread_count&#x27;, &#x27;leiden&#x27;, &#x27;n_counts&#x27;\n",
       "      obsm:\t&#x27;X_pca&#x27;, &#x27;X_umap&#x27;\n",
       "      layers:\t&#x27;norm10k&#x27;\n",
       "    cNMF:\t92284 x 30\n",
       "      obs:\t&#x27;sample&#x27;, &#x27;species&#x27;, &#x27;gene_count&#x27;, &#x27;tscp_count&#x27;, &#x27;mread_count&#x27;, &#x27;leiden&#x27;, &#x27;n_counts&#x27;\n",
       "      uns:\t&#x27;guide_names&#x27;, &#x27;guide_targets&#x27;, &#x27;var_names&#x27;\n",
       "      obsm:\t&#x27;guide_assignment&#x27;\n",
       "      varm:\t&#x27;loadings&#x27;</pre>"
      ],
      "text/plain": [
       "MuData object with n_obs × n_vars = 92284 × 5481\n",
       "  2 modalities\n",
       "    rna:\t92284 x 5451\n",
       "      obs:\t'sample', 'species', 'gene_count', 'tscp_count', 'mread_count', 'leiden', 'n_counts'\n",
       "      obsm:\t'X_pca', 'X_umap'\n",
       "      layers:\t'norm10k'\n",
       "    cNMF:\t92284 x 30\n",
       "      obs:\t'sample', 'species', 'gene_count', 'tscp_count', 'mread_count', 'leiden', 'n_counts'\n",
       "      uns:\t'guide_names', 'guide_targets', 'var_names'\n",
       "      obsm:\t'guide_assignment'\n",
       "      varm:\t'loadings'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load mdata\n",
    "path_mdata = io_config['path_mdata']\n",
    "mdata = mudata.read(path_mdata)\n",
    "mdata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "prog_key = \"cNMF\"\n",
    "data_key = \"rna\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Motif enrichment testing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'motif_file': '/cellar/users/aklie/opt/gene_program_evaluation/src/tests/test_data/motifs.meme',\n",
       " 'seq_file': '/cellar/users/aklie/data/ref/genomes/hg38/hg38.fa',\n",
       " 'loci_file': '/cellar/users/aklie/opt/gene_program_evaluation/src/tests/test_data/p2g_links_minimal_2024_09_02.txt',\n",
       " 'output_loc': '/cellar/users/aklie/opt/gene_program_evaluation/examples/evaluation/iPSC_EC/cNMF_30/motif_counts',\n",
       " 'use_previous': False,\n",
       " 'sig': 0.05,\n",
       " 'num_genes': None,\n",
       " 'correlation': 'pearsonr',\n",
       " 'n_jobs': -1,\n",
       " 'inplace': False}"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "motif_enrichment_config = config['motif_enrichment']\n",
    "motif_enrichment_config"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "5451"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_names = mdata[prog_key].uns['var_names']\n",
    "len(gene_names)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read coordinate/bed/loci"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_loci(\n",
    "    path_loci: os.PathLike\n",
    ") -> pd.DataFrame:\n",
    "    \"\"\"Read loci to gene links file\n",
    "\n",
    "    Read promoter/enhancer to gene links tab delimited file with the following column headers:\n",
    "        - chromosome: chromosome of the locus\n",
    "        - start: start position of the locus\n",
    "        - end: end position of the locus\n",
    "        - seq_name: name of the locus\n",
    "        - seq_class {promoter, enhancer}: class of the locus\n",
    "        - seq_score: score of the locus\n",
    "        - gene_name: name of the gene linked to the locus\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    path_loci : str\n",
    "        Path to the coordinates file.\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    loci : pd.DataFrame\n",
    "        DataFrame containing the coordinates.\n",
    "    \"\"\"\n",
    "\n",
    "    # Read formatted coords file\n",
    "    loci = pd.read_csv(path_loci, sep='\\t')\n",
    "\n",
    "    # Enforce header format\n",
    "    expected_headers = [\n",
    "        'chromosome', \n",
    "        'start', \n",
    "        'end', \n",
    "        'seq_name', \n",
    "        'seq_class', \n",
    "        'seq_score', \n",
    "        'gene_name'\n",
    "    ]\n",
    "\n",
    "    # Check for expected headers\n",
    "    for col in expected_headers:\n",
    "        try: assert col in loci.columns\n",
    "        except: raise ValueError('Coordinate file is not formatted correctly')\n",
    "\n",
    "    # TODO: Support multiple classes at once\n",
    "    try: assert len(loci['seq_class']==1)\n",
    "    except: raise ValueError('Coordinate file contains multiple sequence classes')\n",
    "\n",
    "    return loci"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "motif_enrichment_config[\"coords_file\"] = \"/cellar/users/aklie/opt/gene_program_evaluation/src/tests/test_data/p2g_links_minimal_2024_09_02.txt\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>seq_name</th>\n",
       "      <th>seq_class</th>\n",
       "      <th>seq_score</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr12</td>\n",
       "      <td>52299565</td>\n",
       "      <td>52301301</td>\n",
       "      <td>promoter|chr12:52299415-52301451</td>\n",
       "      <td>promoter</td>\n",
       "      <td>56.971568</td>\n",
       "      <td>ACVRL1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr12</td>\n",
       "      <td>52299565</td>\n",
       "      <td>52301301</td>\n",
       "      <td>promoter|chr12:52299415-52301451</td>\n",
       "      <td>promoter</td>\n",
       "      <td>56.971568</td>\n",
       "      <td>ANKRD33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr12</td>\n",
       "      <td>57028808</td>\n",
       "      <td>57030874</td>\n",
       "      <td>promoter|chr12:57028658-57031024</td>\n",
       "      <td>promoter</td>\n",
       "      <td>41.677500</td>\n",
       "      <td>BAZ2A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr12</td>\n",
       "      <td>57028808</td>\n",
       "      <td>57030874</td>\n",
       "      <td>promoter|chr12:57028658-57031024</td>\n",
       "      <td>promoter</td>\n",
       "      <td>41.677500</td>\n",
       "      <td>ATP5B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr12</td>\n",
       "      <td>89745534</td>\n",
       "      <td>89746814</td>\n",
       "      <td>promoter|chr12:89745384-89746964</td>\n",
       "      <td>promoter</td>\n",
       "      <td>43.190770</td>\n",
       "      <td>DUSP6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chromosome     start       end                          seq_name seq_class  \\\n",
       "0      chr12  52299565  52301301  promoter|chr12:52299415-52301451  promoter   \n",
       "1      chr12  52299565  52301301  promoter|chr12:52299415-52301451  promoter   \n",
       "2      chr12  57028808  57030874  promoter|chr12:57028658-57031024  promoter   \n",
       "3      chr12  57028808  57030874  promoter|chr12:57028658-57031024  promoter   \n",
       "4      chr12  89745534  89746814  promoter|chr12:89745384-89746964  promoter   \n",
       "\n",
       "   seq_score gene_name  \n",
       "0  56.971568    ACVRL1  \n",
       "1  56.971568   ANKRD33  \n",
       "2  41.677500     BAZ2A  \n",
       "3  41.677500     ATP5B  \n",
       "4  43.190770     DUSP6  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "loci = read_loci(motif_enrichment_config[\"coords_file\"])\n",
    "loci.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "matching_gene_names = np.intersect1d(loci['gene_name'].unique(), gene_names)\n",
    "len(matching_gene_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>seq_name</th>\n",
       "      <th>seq_class</th>\n",
       "      <th>seq_score</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr12</td>\n",
       "      <td>52299565</td>\n",
       "      <td>52301301</td>\n",
       "      <td>promoter|chr12:52299415-52301451</td>\n",
       "      <td>promoter</td>\n",
       "      <td>56.971568</td>\n",
       "      <td>ACVRL1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr12</td>\n",
       "      <td>89745534</td>\n",
       "      <td>89746814</td>\n",
       "      <td>promoter|chr12:89745384-89746964</td>\n",
       "      <td>promoter</td>\n",
       "      <td>43.190770</td>\n",
       "      <td>DUSP6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr13</td>\n",
       "      <td>21099338</td>\n",
       "      <td>21101201</td>\n",
       "      <td>promoter|chr13:21099188-21101351</td>\n",
       "      <td>promoter</td>\n",
       "      <td>40.740980</td>\n",
       "      <td>CRYL1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chromosome     start       end                          seq_name seq_class  \\\n",
       "0      chr12  52299565  52301301  promoter|chr12:52299415-52301451  promoter   \n",
       "4      chr12  89745534  89746814  promoter|chr12:89745384-89746964  promoter   \n",
       "9      chr13  21099338  21101201  promoter|chr13:21099188-21101351  promoter   \n",
       "\n",
       "   seq_score gene_name  \n",
       "0  56.971568    ACVRL1  \n",
       "4  43.190770     DUSP6  \n",
       "9  40.740980     CRYL1  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Filter loci to only include gene names that are in the gene program\n",
    "loci_= loci[loci['gene_name'].isin(matching_gene_names)]\n",
    "loci_.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "torch.Size([3, 4, 1000])"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X = extract_loci(\n",
    "    loci_,\n",
    "    motif_enrichment_config[\"seq_file\"], \n",
    "    in_window=1000\n",
    ").float()\n",
    "X.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read in motifs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['AHR.H12CORE.0.P.B',\n",
       " 'AHRR.H12CORE.0.P.C',\n",
       " 'ALX1.H12CORE.0.SM.B',\n",
       " 'ALX3.H12CORE.0.SM.B',\n",
       " 'FOXF1.H12CORE.0.P.C']"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pwms = read_meme(motif_enrichment_config['motif_file'])\n",
    "list(pwms.keys())[:5]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Motif matching with FIMO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "def perform_motif_match(\n",
    "    loci: pd.DataFrame,\n",
    "    sequences: os.PathLike,\n",
    "    pwms: Mapping[str, np.ndarray],\n",
    "    in_window: int=1000,\n",
    "    threshold: float=1e-4,\n",
    "    eps: float=1e-4,\n",
    "    reverse_complement: bool=True,\n",
    "    output_loc: os.PathLike=None\n",
    "):\n",
    "        \"\"\"Score motif matches\n",
    "\n",
    "        Perform motif matching on sequences linked to genes\n",
    "        via enhancer/promoter coordinates.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        loci : pd.DataFrame\n",
    "            DataFrame containing sequence coordinates.\n",
    "        sequences : os.PathLike\n",
    "            Path to FASTA formatted genomic sequence.\n",
    "        pwms : Mapping[str, np.ndarray]\n",
    "            Dictionary of PWMs where keys are motif names and values are PWMs.\n",
    "        in_window : int\n",
    "            Window size to extract sequences around center of loci.\n",
    "        threshold : float\n",
    "            Threshold for motif matching.\n",
    "        output_loc : os.PathLike\n",
    "            Path to directory to store motif matches for individual motifs.\n",
    "        \"\"\"\n",
    "        # Perform motif matching\n",
    "        X = extract_loci(loci, sequences, in_window=in_window).float()\n",
    "        hits = fimo(pwms, X, threshold=threshold, eps=eps, reverse_complement=reverse_complement)\n",
    "\n",
    "        # Create motif match dataframe\n",
    "        motif_match_df = pd.DataFrame()\n",
    "        for i, hit in enumerate(hits):\n",
    "            annotated_hit = hit.merge(loci[[\"chromosome\", \"seq_name\", \"seq_class\", \"gene_name\"]], left_on=\"sequence_name\", right_index=True).drop(columns=[\"sequence_name\"])\n",
    "            annotated_hit[\"motif_name\"] = list(pwms.keys())[i]\n",
    "            annotated_hit = annotated_hit[[\"chromosome\", \"start\", \"end\", \"strand\", \"motif_name\", \"score\", \"p-value\", \"seq_name\", \"seq_class\", \"gene_name\"]]\n",
    "            if output_loc is not None:\n",
    "                annotated_hit.to_csv(os.path.join(output_loc, f\"motif_match_{list(pwms.keys())[i]}.txt\"), sep='\\t', index=False)\n",
    "            motif_match_df = pd.concat([motif_match_df, annotated_hit])\n",
    "        motif_match_df[\"adj_pval\"] = multipletests(motif_match_df[\"p-value\"], method=\"fdr_bh\")[1]\n",
    "        motif_match_df.reset_index(drop=True, inplace=True)\n",
    "        return motif_match_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>motif_name</th>\n",
       "      <th>score</th>\n",
       "      <th>p-value</th>\n",
       "      <th>seq_name</th>\n",
       "      <th>seq_class</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>adj_pval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr12</td>\n",
       "      <td>405</td>\n",
       "      <td>425</td>\n",
       "      <td>-</td>\n",
       "      <td>ALX1.H12CORE.0.SM.B</td>\n",
       "      <td>9.742607</td>\n",
       "      <td>0.000076</td>\n",
       "      <td>promoter|chr12:52299415-52301451</td>\n",
       "      <td>promoter</td>\n",
       "      <td>ACVRL1</td>\n",
       "      <td>0.000076</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr12</td>\n",
       "      <td>443</td>\n",
       "      <td>455</td>\n",
       "      <td>-</td>\n",
       "      <td>FOXF1.H12CORE.0.P.C</td>\n",
       "      <td>11.801019</td>\n",
       "      <td>0.000068</td>\n",
       "      <td>promoter|chr12:52299415-52301451</td>\n",
       "      <td>promoter</td>\n",
       "      <td>ACVRL1</td>\n",
       "      <td>0.000076</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chromosome start  end strand           motif_name      score   p-value  \\\n",
       "0      chr12   405  425      -  ALX1.H12CORE.0.SM.B   9.742607  0.000076   \n",
       "1      chr12   443  455      -  FOXF1.H12CORE.0.P.C  11.801019  0.000068   \n",
       "\n",
       "                           seq_name seq_class gene_name  adj_pval  \n",
       "0  promoter|chr12:52299415-52301451  promoter    ACVRL1  0.000076  \n",
       "1  promoter|chr12:52299415-52301451  promoter    ACVRL1  0.000076  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "motif_match_df = perform_motif_match(loci_, motif_enrichment_config[\"seq_file\"], pwms, in_window=1000, threshold=1e-4)\n",
    "motif_match_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Compute counts of motifs for each gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_motif_instances(\n",
    "    motif_match_df: pd.DataFrame,\n",
    "    motif_var: str = 'motif_name',\n",
    "    sig: float=0.05,\n",
    "    sig_var: str='adj_pval',\n",
    "    gene_names: Optional[np.ndarray]=None\n",
    "):\n",
    "    \"\"\"Count motif instances per gene (via enahncer/promoter linking)\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    motif_match_df : pd.DataFrame\n",
    "        DataFrame containing motif matches.\n",
    "    motif_var : str\n",
    "        Column name for motif names. Default is 'motif_name'.\n",
    "    sig : float\n",
    "        Significance threshold for motif matches. Default is 0.05.\n",
    "    sig_var : str\n",
    "        Column name for significance values. Default is 'adj_pval'.\n",
    "    gene_names : np.ndarray\n",
    "        Array of gene names. Default is None.\n",
    "    \n",
    "    Returns\n",
    "    -------\n",
    "    motif_count_df : pd.DataFrame\n",
    "        DataFrame containing\n",
    "    \"\"\"\n",
    "\n",
    "    # Count up significant occurences of motif\n",
    "    motif_match_df_ = motif_match_df.loc[motif_match_df[sig_var] < sig]\n",
    "    motif_match_df_ = motif_match_df.value_counts(subset=['gene_name', motif_var]).reset_index()\n",
    "    motif_match_df_.columns = ['gene_name', motif_var, 'motif_count']\n",
    "    motif_match_df_ = motif_match_df_.pivot(index='gene_name', columns=motif_var, values='motif_count')\n",
    "    motif_count_df = pd.DataFrame(index=gene_names, columns=motif_match_df_.columns)\n",
    "    motif_count_df.loc[motif_match_df_.index.values] = motif_match_df_ # Gene names should match as this point\n",
    "    return motif_count_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "motif_name\n",
       "ALX1.H12CORE.0.SM.B    1\n",
       "FOXF1.H12CORE.0.P.C    1\n",
       "dtype: object"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "motif_count_df = compute_motif_instances(\n",
    "    motif_match_df, \n",
    "    motif_var='motif_name', \n",
    "    sig=0.05, \n",
    "    sig_var='adj_pval', \n",
    "    gene_names=gene_names\n",
    ")\n",
    "motif_count_df.sum(axis=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Motif enrichment statistics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "def perform_correlation(\n",
    "    motif_count_df: pd.DataFrame,\n",
    "    prog_genes: pd.DataFrame,\n",
    "    motif_enrich_stat_df: pd.DataFrame,\n",
    "    motif_enrich_pval_df: pd.DataFrame,\n",
    "    motif_idx: int,\n",
    "    prog_idx: int,\n",
    "    correlation: str='pearsonr'\n",
    "):\n",
    "    \"\"\"Compute motif enrichment as correlation b/w gene weights/ranks and motif counts\n",
    "    \n",
    "    Perform pearson correlation test for motif count enrichment vs gene loadings\n",
    "    If loadings are dichotomized then this is equivalent to a point biserial correlation test.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    motif_count_df : pd.DataFrame\n",
    "        DataFrame containing motif counts.\n",
    "    prog_genes : pd.DataFrame\n",
    "        DataFrame containing gene program loadings.\n",
    "    motif_enrich_stat_df : pd.DataFrame\n",
    "        DataFrame to store correlation statistics.\n",
    "    motif_enrich_pval_df : pd.DataFrame\n",
    "        DataFrame to store correlation p-values.\n",
    "    motif_idx : int\n",
    "        Index of motif.\n",
    "    prog_idx : int\n",
    "        Index of gene program.\n",
    "    correlation : {'pearsonr','spearmanr','kendalltau'}\n",
    "        Type of correlation to perform. Default is 'pearsonr'.\n",
    "    \"\"\"\n",
    "\n",
    "    loadings = prog_genes.iloc[prog_idx].values.flatten()\n",
    "    counts = motif_count_df.T.iloc[motif_idx].fillna(0).values.flatten()\n",
    "    \n",
    "    if correlation=='pearsonr':\n",
    "        stat, pval = pearsonr(loadings, counts)\n",
    "    elif correlation=='spearmanr':\n",
    "        stat, pval = spearmanr(loadings, counts)\n",
    "    elif correlation=='kendalltau':\n",
    "        stat, pval = kendalltau(loadings, counts)\n",
    "\n",
    "    motif_enrich_stat_df.iloc[prog_idx, motif_idx]  = stat\n",
    "    motif_enrich_pval_df.iloc[prog_idx, motif_idx]  = pval\n",
    "\n",
    "\n",
    "def compute_motif_enrichment_(\n",
    "    mdata: mudata.MuData,\n",
    "    motif_count_df: pd.DataFrame,\n",
    "    prog_key: str='prog',\n",
    "    gene_names: Optional[np.ndarray]=None,\n",
    "    weighted: bool=True,\n",
    "    num_genes: Optional[int]=None,\n",
    "    correlation: str='pearsonr',\n",
    "    n_jobs: int=1\n",
    "):\n",
    "    \"\"\"Count up motif ocurrences and perform diff. test\n",
    "\n",
    "    Perform motif enrichment using gene program loadings and\n",
    "    motif counts linked to genes via motif scanning of\n",
    "    linked enhancer/promoter sequences.\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    mdata : MuData\n",
    "        MuData object containing anndata of program scores and cell-level metadata.\n",
    "    motif_count_df : pd.DataFrame\n",
    "        DataFrame containing motif counts.\n",
    "    prog_key : str\n",
    "        Key for the anndata object in the mudata object. Default is 'prog'.\n",
    "    gene_names : np.ndarray\n",
    "        Array of gene names. Default is None.\n",
    "    weighted : bool\n",
    "        Use weighted loadings. Default is True.\n",
    "    num_genes : int\n",
    "        Number of genes threshold to dichotomize loadings. Default is None.\n",
    "    correlation : {'pearsonr','spearmanr','kendalltau'}\n",
    "        Type of correlation to perform. Default is 'pearsonr'.\n",
    "    n_jobs : int\n",
    "        Number of threads to run processes on. Default is 1.\n",
    "    \"\"\"\n",
    "\n",
    "    # Both weighted and num_genes cannot be set\n",
    "    if num_genes is not None and weighted:\n",
    "        raise ValueError('Will not use weighted when num_genes specified.')\n",
    "\n",
    "    loadings = pd.DataFrame(data=mdata[prog_key].varm['loadings'],\n",
    "                            index=mdata[prog_key].var_names,\n",
    "                            columns=gene_names)\n",
    "    # FIXME: Causes expansion due to duplications in gene_names\n",
    "    # Ensure index matches b/w loadings and counts\n",
    "    # loadings = loadings.loc[:, motif_count_df.index.values]\n",
    "\n",
    "    # Binary matrix \n",
    "    if not weighted:\n",
    "        prog_genes = (loadings.rank(axis=1)<num_genes).astype(int)\n",
    "    elif weighted and num_genes is None:\n",
    "        prog_genes = loadings\n",
    "    elif weighted and isinstance(num_genes, int):\n",
    "        prog_genes = (loadings.rank(axis=1)<num_genes).astype(int)\n",
    "        prog_genes *= loadings\n",
    "    else:\n",
    "        raise ValueError('num_genes not specified correctly.')\n",
    "\n",
    "    # Use pearson correlation\n",
    "    # If dichotomized then this is equivalent to point biserial correlation\n",
    "    motif_enrich_stat_df = pd.DataFrame(index=mdata[prog_key].var_names,\n",
    "                                        columns=motif_count_df.columns.values)\n",
    "    motif_enrich_pval_df = pd.DataFrame(index=mdata[prog_key].var_names,\n",
    "                                        columns=motif_count_df.columns.values)\n",
    "\n",
    "    # FIXME: If n_jobs>1 then parallel processes dont seem to terminate.\n",
    "    # Perform test in parallel across motifs and programs\n",
    "    Parallel(n_jobs=n_jobs, \n",
    "             backend='threading')(delayed(perform_correlation)(motif_count_df,\n",
    "                                                               prog_genes, \n",
    "                                                               motif_enrich_stat_df,\n",
    "                                                               motif_enrich_pval_df,\n",
    "                                                               motif_idx, \n",
    "                                                               prog_idx,\n",
    "                                                               correlation=correlation) \\\n",
    "                                                            for motif_idx in tqdm(range(motif_count_df.columns.values.shape[-1]),\n",
    "                                                                                     desc='Computing motif enrichment',\n",
    "                                                                                     unit='motifs') \\\n",
    "                                                            for prog_idx in range(mdata[prog_key].var_names.shape[0]))\n",
    "\n",
    "    return motif_enrich_stat_df, motif_enrich_pval_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "d6fb6dddd74540c2bba3c330bb031f38",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Computing motif enrichment:   0%|          | 0/2 [00:00<?, ?motifs/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "motif_enrich_stat_df, motif_enrich_pval_df = compute_motif_enrichment_(\n",
    "    mdata,\n",
    "    motif_count_df=motif_count_df,\n",
    "    prog_key=prog_key,\n",
    "    gene_names=gene_names,\n",
    "    num_genes=None,\n",
    "    n_jobs=1,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Wrapper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_motif_enrichment(\n",
    "    mdata: Union[mudata.MuData, os.PathLike],\n",
    "    prog_key: str='prog',\n",
    "    data_key: str='data',\n",
    "    motif_file: Optional[os.PathLike]=None,\n",
    "    seq_file: Optional[os.PathLike]=None,\n",
    "    loci_file: Optional[os.PathLike]=None,\n",
    "    output_loc: Optional[os.PathLike]=None,\n",
    "    window: int=1000,\n",
    "    threshold: float=1e-4,\n",
    "    eps: float=1e-4,\n",
    "    reverse_complement: bool=True,\n",
    "    sig: float=0.05,\n",
    "    num_genes: Optional[int]=None,\n",
    "    correlation: str='pearsonr',\n",
    "    n_jobs: int=1,\n",
    "    inplace: bool=True,\n",
    "    **kwargs\n",
    "):\n",
    "    \n",
    "    \"\"\"Compute motif enrichment in enhancers or promoters associated with a gene\n",
    "    \n",
    "    Perform motif enrichment using gene program loadings and\n",
    "    motif counts linked to genes via motif scanning of\n",
    "    linked enhancer/promoter sequences.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    mdata : MuData\n",
    "        mudata object containing anndata of program scores and cell-level metadata.\n",
    "    prog_key: \n",
    "        index for the anndata object (mdata[prog_key]) in the mudata object.\n",
    "    data_key: str\n",
    "        index of the genomic data anndata object (mdata[data_key]) in the mudata object.\n",
    "    motif_file: str\n",
    "        path to motif file formatted in MEME format.\n",
    "    seq_file: str\n",
    "        path to FASTA formatted genomic sequence.\n",
    "    loci_file: str\n",
    "        path to enhancer/promoter gene links file with sequence coordinates.\n",
    "        Tab delimited file with the following column headers:\n",
    "        chr, start, end, seq_name, seq_class {promoter, enhancer}, \n",
    "        seq_score, gene_name.\n",
    "    output_loc: str\n",
    "        path to directory to store motif - gene counts.\n",
    "    sig: (0,1] (default: 0.05)\n",
    "        significance level for inferring a motif match.\n",
    "    num_genes: int (default: None)\n",
    "        number of genes threshold to dichtomize loadings.\n",
    "    correlation: {'pearsonr','spearmanr','kendalltau'} (default: 'peasronsr')\n",
    "        correlation type to use to compute motif enirchments.\n",
    "        Use kendalltau when expecting enrichment/de-enrichment at both ends.\n",
    "    use_previous: bool (default: True)\n",
    "        if outplot is provided try to load motif matches from previous run.\n",
    "    n_jobs: int (default: 1)\n",
    "        number of threads to run processes on.\n",
    "    inplace: Bool (default: True)\n",
    "        update the mudata object inplace or return a copy\n",
    "       \n",
    "    Returns\n",
    "    -------\n",
    "    if not inplace:\n",
    "        motif_match_df,\n",
    "        motif_count_df.loc[gene_names].values,\n",
    "        motif_enrichment_df\n",
    "    else:\n",
    "        None, edits mdata in place\n",
    "    \"\"\"\n",
    "\n",
    "    # Read in mudata if it is provided as a path\n",
    "    frompath=False\n",
    "    if isinstance(mdata, str):\n",
    "        if os.path.exists(mdata):\n",
    "            mdata = mudata.read(mdata)\n",
    "            if inplace:\n",
    "                logging.warning('Changed to inplace=False since path was provided')\n",
    "                inplace=False\n",
    "            frompath=True\n",
    "        else: raise ValueError('Incorrect mudata specification.')\n",
    "    \n",
    "    if not inplace and not frompath:\n",
    "        mdata = mudata.MuData({prog_key: mdata[prog_key].copy(),\n",
    "                               data_key: mdata[data_key].copy()})\n",
    "\n",
    "    # Get gene names in MuData\n",
    "    if 'var_names' in mdata[prog_key].uns.keys():\n",
    "        gene_names = mdata[prog_key].uns['var_names']\n",
    "    else:\n",
    "        try: assert mdata[prog_key].varm['loadings'].shape[1]==mdata[data_key].var.shape[0]\n",
    "        except: raise ValueError('Different number of genes present in data and program loadings')\n",
    "        gene_names = mdata[data_key].var_names\n",
    "    \n",
    "    # \n",
    "    if ':ens' in gene_names[0].lower():\n",
    "        gene_names = [name.split(':')[0] for name in gene_names]\n",
    "\n",
    "    # Check if output loc exists\n",
    "    if output_loc is not None:\n",
    "        try: os.makedirs(output_loc, exist_ok=True)\n",
    "        except: raise ValueError('Output location does not exist.')\n",
    "\n",
    "    # If num_genes specified then cannot be weighted\n",
    "    if num_genes is None:\n",
    "        weighted=True\n",
    "    elif isinstance(num_genes, int):\n",
    "        weighted=False\n",
    "\n",
    "    # Intake motif file path or in memory\n",
    "    if isinstance(motif_file, str) and os.path.exists(motif_file):\n",
    "        pwms = read_meme(motif_file)\n",
    "    else:\n",
    "        raise ValueError('Motif file not found.')\n",
    "    \n",
    "    # Intake coord file path or in memory\n",
    "    if isinstance(loci_file, str) and os.path.exists(loci_file):\n",
    "        loci = read_loci(loci_file)\n",
    "    else:\n",
    "        raise ValueError('Coordinate file not found.')\n",
    "\n",
    "    # Valid genes\n",
    "    matching_gene_names = np.intersect1d(loci['gene_name'].unique(), gene_names)\n",
    "    print(f'Number of matching genes: {len(matching_gene_names)}')\n",
    "    try: assert len(matching_gene_names) > 0\n",
    "    except: raise ValueError('No matching genes b/w data and coordinate files')\n",
    "\n",
    "    # Compute motif matching\n",
    "    loci_ = loci[loci['gene_name'].isin(matching_gene_names)]\n",
    "    print(f'Number of loci: {loci_.shape[0]}')\n",
    "    motif_match_df = perform_motif_match(\n",
    "        loci=loci_,\n",
    "        sequences=seq_file,\n",
    "        pwms=pwms,\n",
    "        in_window=window,\n",
    "        threshold=threshold,\n",
    "        eps=eps,\n",
    "        reverse_complement=reverse_complement,\n",
    "        output_loc=output_loc\n",
    "    )\n",
    "\n",
    "    # Count motif enrichment\n",
    "    motif_count_df = compute_motif_instances(\n",
    "        motif_match_df,\n",
    "        motif_var='motif_name',\n",
    "        sig=sig,\n",
    "        sig_var='adj_pval',\n",
    "        gene_names=gene_names\n",
    "    )\n",
    "    motif_enrich_stat_df, motif_enrich_pval_df = compute_motif_enrichment_(\n",
    "        mdata,\n",
    "        motif_count_df,\n",
    "        prog_key=prog_key,\n",
    "        gene_names=gene_names,\n",
    "        weighted=weighted,\n",
    "        num_genes=num_genes,\n",
    "        n_jobs=n_jobs)\n",
    "                         \n",
    "    # Store motif counts\n",
    "    if inplace:\n",
    "        mdata[prog_key].uns['motif_counts'] = motif_count_df.loc[gene_names].values\n",
    "        mdata[prog_key].uns['motif_names'] = motif_count_df.columns.values\n",
    "\n",
    "        mdata[prog_key].varm['motif_enrich_{}_stat'.format(correlation)] = motif_enrich_stat_df.values\n",
    "        mdata[prog_key].varm['motif_enrich_{}_pval'.format(correlation)] = motif_enrich_pval_df.values\n",
    "        mdata[prog_key].uns['motif_names'] = motif_count_df.columns.values\n",
    "\n",
    "    else:\n",
    "\n",
    "        motif_enrich_stat_df = motif_enrich_stat_df.reset_index().melt(id_vars='index',\n",
    "                                                                       var_name='motif', \n",
    "                                                                       value_name='stat')\n",
    "        motif_enrich_stat_df = motif_enrich_stat_df.set_index(['index', 'motif'])\n",
    "\n",
    "        motif_enrich_pval_df = motif_enrich_pval_df.reset_index().melt(id_vars='index',\n",
    "                                                                       var_name='motif', \n",
    "                                                                       value_name='pval')\n",
    "        motif_enrich_pval_df = motif_enrich_pval_df.set_index(['index', 'motif'])\n",
    "\n",
    "        motif_enrichment_df = motif_enrich_stat_df.merge(motif_enrich_pval_df,\n",
    "                                                        left_index=True, \n",
    "                                                        right_index=True)\n",
    "        motif_enrichment_df = motif_enrichment_df.reset_index()\n",
    "        motif_enrichment_df['program_name'] = motif_enrichment_df['index']\n",
    "        motif_enrichment_df.drop('index', axis=1, inplace=True)\n",
    "        motif_enrichment_df = motif_enrichment_df.sort_values(['program_name', 'pval'])\n",
    "\n",
    "        motif_count_df.columns.name=''\n",
    "\n",
    "        return (motif_match_df,\n",
    "                motif_count_df.loc[gene_names].fillna(0),\n",
    "                motif_enrichment_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/cellar/users/aklie/opt/miniconda3/envs/test_celloracle/lib/python3.10/site-packages/anndata/_core/anndata.py:1899: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/cellar/users/aklie/opt/miniconda3/envs/test_celloracle/lib/python3.10/site-packages/mudata/_core/mudata.py:479: UserWarning: var_names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of matching genes: 3\n",
      "Number of loci: 3\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "20df27072b004982b70b28a26e789957",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Computing motif enrichment:   0%|          | 0/2 [00:00<?, ?motifs/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Run motif enrichment and save results\n",
    "motif_match_minimal_df, motif_count_minimal_df, motif_enrichment_minimal_df = compute_motif_enrichment(\n",
    "    mdata, \n",
    "    prog_key=prog_key,\n",
    "    data_key=data_key,\n",
    "    motif_file=motif_enrichment_config['motif_file'],\n",
    "    seq_file=motif_enrichment_config['seq_file'],\n",
    "    loci_file=\"/cellar/users/aklie/opt/gene_program_evaluation/src/tests/test_data/p2g_links_minimal_2024_09_02.txt\",\n",
    "    output_loc=None,\n",
    "    window=1000,\n",
    "    threshold=1e-4,\n",
    "    eps=1e-4,\n",
    "    reverse_complement=True,\n",
    "    sig=0.05,\n",
    "    num_genes=None,\n",
    "    correlation='pearsonr',\n",
    "    n_jobs=1,\n",
    "    inplace=False\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>motif</th>\n",
       "      <th>stat</th>\n",
       "      <th>pval</th>\n",
       "      <th>program_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ALX1.H12CORE.0.SM.B</td>\n",
       "      <td>-0.007562</td>\n",
       "      <td>0.576735</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>FOXF1.H12CORE.0.P.C</td>\n",
       "      <td>-0.007562</td>\n",
       "      <td>0.576735</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ALX1.H12CORE.0.SM.B</td>\n",
       "      <td>0.007341</td>\n",
       "      <td>0.587885</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>FOXF1.H12CORE.0.P.C</td>\n",
       "      <td>0.007341</td>\n",
       "      <td>0.587885</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ALX1.H12CORE.0.SM.B</td>\n",
       "      <td>0.054421</td>\n",
       "      <td>0.000058</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>FOXF1.H12CORE.0.P.C</td>\n",
       "      <td>0.054421</td>\n",
       "      <td>0.000058</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ALX1.H12CORE.0.SM.B</td>\n",
       "      <td>-0.006529</td>\n",
       "      <td>0.629873</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>FOXF1.H12CORE.0.P.C</td>\n",
       "      <td>-0.006529</td>\n",
       "      <td>0.629873</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ALX1.H12CORE.0.SM.B</td>\n",
       "      <td>0.027282</td>\n",
       "      <td>0.043993</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>FOXF1.H12CORE.0.P.C</td>\n",
       "      <td>0.027282</td>\n",
       "      <td>0.043993</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ALX1.H12CORE.0.SM.B</td>\n",
       "      <td>0.035017</td>\n",
       "      <td>0.009723</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>FOXF1.H12CORE.0.P.C</td>\n",
       "      <td>0.035017</td>\n",
       "      <td>0.009723</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ALX1.H12CORE.0.SM.B</td>\n",
       "      <td>0.013686</td>\n",
       "      <td>0.312365</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>FOXF1.H12CORE.0.P.C</td>\n",
       "      <td>0.013686</td>\n",
       "      <td>0.312365</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ALX1.H12CORE.0.SM.B</td>\n",
       "      <td>-0.003615</td>\n",
       "      <td>0.789594</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>FOXF1.H12CORE.0.P.C</td>\n",
       "      <td>-0.003615</td>\n",
       "      <td>0.789594</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ALX1.H12CORE.0.SM.B</td>\n",
       "      <td>-0.005528</td>\n",
       "      <td>0.683261</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>FOXF1.H12CORE.0.P.C</td>\n",
       "      <td>-0.005528</td>\n",
       "      <td>0.683261</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ALX1.H12CORE.0.SM.B</td>\n",
       "      <td>-0.003137</td>\n",
       "      <td>0.816859</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>FOXF1.H12CORE.0.P.C</td>\n",
       "      <td>-0.003137</td>\n",
       "      <td>0.816859</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ALX1.H12CORE.0.SM.B</td>\n",
       "      <td>-0.005974</td>\n",
       "      <td>0.659239</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>FOXF1.H12CORE.0.P.C</td>\n",
       "      <td>-0.005974</td>\n",
       "      <td>0.659239</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ALX1.H12CORE.0.SM.B</td>\n",
       "      <td>-0.006308</td>\n",
       "      <td>0.64148</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>FOXF1.H12CORE.0.P.C</td>\n",
       "      <td>-0.006308</td>\n",
       "      <td>0.64148</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ALX1.H12CORE.0.SM.B</td>\n",
       "      <td>-0.009248</td>\n",
       "      <td>0.494819</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>FOXF1.H12CORE.0.P.C</td>\n",
       "      <td>-0.009248</td>\n",
       "      <td>0.494819</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>ALX1.H12CORE.0.SM.B</td>\n",
       "      <td>-0.005756</td>\n",
       "      <td>0.670934</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>FOXF1.H12CORE.0.P.C</td>\n",
       "      <td>-0.005756</td>\n",
       "      <td>0.670934</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>ALX1.H12CORE.0.SM.B</td>\n",
       "      <td>-0.004982</td>\n",
       "      <td>0.713086</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>FOXF1.H12CORE.0.P.C</td>\n",
       "      <td>-0.004982</td>\n",
       "      <td>0.713086</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>ALX1.H12CORE.0.SM.B</td>\n",
       "      <td>-0.005821</td>\n",
       "      <td>0.667429</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>FOXF1.H12CORE.0.P.C</td>\n",
       "      <td>-0.005821</td>\n",
       "      <td>0.667429</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>ALX1.H12CORE.0.SM.B</td>\n",
       "      <td>-0.005676</td>\n",
       "      <td>0.675216</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>FOXF1.H12CORE.0.P.C</td>\n",
       "      <td>-0.005676</td>\n",
       "      <td>0.675216</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>ALX1.H12CORE.0.SM.B</td>\n",
       "      <td>-0.005332</td>\n",
       "      <td>0.693891</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>FOXF1.H12CORE.0.P.C</td>\n",
       "      <td>-0.005332</td>\n",
       "      <td>0.693891</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>ALX1.H12CORE.0.SM.B</td>\n",
       "      <td>-0.004735</td>\n",
       "      <td>0.726708</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>FOXF1.H12CORE.0.P.C</td>\n",
       "      <td>-0.004735</td>\n",
       "      <td>0.726708</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>ALX1.H12CORE.0.SM.B</td>\n",
       "      <td>-0.004908</td>\n",
       "      <td>0.717141</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>FOXF1.H12CORE.0.P.C</td>\n",
       "      <td>-0.004908</td>\n",
       "      <td>0.717141</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>ALX1.H12CORE.0.SM.B</td>\n",
       "      <td>-0.005134</td>\n",
       "      <td>0.704701</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>FOXF1.H12CORE.0.P.C</td>\n",
       "      <td>-0.005134</td>\n",
       "      <td>0.704701</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>ALX1.H12CORE.0.SM.B</td>\n",
       "      <td>-0.004398</td>\n",
       "      <td>0.745462</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>FOXF1.H12CORE.0.P.C</td>\n",
       "      <td>-0.004398</td>\n",
       "      <td>0.745462</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>ALX1.H12CORE.0.SM.B</td>\n",
       "      <td>-0.004707</td>\n",
       "      <td>0.728243</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>FOXF1.H12CORE.0.P.C</td>\n",
       "      <td>-0.004707</td>\n",
       "      <td>0.728243</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ALX1.H12CORE.0.SM.B</td>\n",
       "      <td>-0.005264</td>\n",
       "      <td>0.697582</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>FOXF1.H12CORE.0.P.C</td>\n",
       "      <td>-0.005264</td>\n",
       "      <td>0.697582</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ALX1.H12CORE.0.SM.B</td>\n",
       "      <td>-0.006496</td>\n",
       "      <td>0.631579</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>FOXF1.H12CORE.0.P.C</td>\n",
       "      <td>-0.006496</td>\n",
       "      <td>0.631579</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ALX1.H12CORE.0.SM.B</td>\n",
       "      <td>-0.006912</td>\n",
       "      <td>0.609899</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>FOXF1.H12CORE.0.P.C</td>\n",
       "      <td>-0.006912</td>\n",
       "      <td>0.609899</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ALX1.H12CORE.0.SM.B</td>\n",
       "      <td>-0.006783</td>\n",
       "      <td>0.616604</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>FOXF1.H12CORE.0.P.C</td>\n",
       "      <td>-0.006783</td>\n",
       "      <td>0.616604</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ALX1.H12CORE.0.SM.B</td>\n",
       "      <td>-0.007924</td>\n",
       "      <td>0.558601</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>FOXF1.H12CORE.0.P.C</td>\n",
       "      <td>-0.007924</td>\n",
       "      <td>0.558601</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ALX1.H12CORE.0.SM.B</td>\n",
       "      <td>-0.005727</td>\n",
       "      <td>0.672504</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>FOXF1.H12CORE.0.P.C</td>\n",
       "      <td>-0.005727</td>\n",
       "      <td>0.672504</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ALX1.H12CORE.0.SM.B</td>\n",
       "      <td>0.031572</td>\n",
       "      <td>0.019752</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>FOXF1.H12CORE.0.P.C</td>\n",
       "      <td>0.031572</td>\n",
       "      <td>0.019752</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  motif      stat      pval program_name\n",
       "0   ALX1.H12CORE.0.SM.B -0.007562  0.576735            0\n",
       "30  FOXF1.H12CORE.0.P.C -0.007562  0.576735            0\n",
       "1   ALX1.H12CORE.0.SM.B  0.007341  0.587885            1\n",
       "31  FOXF1.H12CORE.0.P.C  0.007341  0.587885            1\n",
       "10  ALX1.H12CORE.0.SM.B  0.054421  0.000058           10\n",
       "40  FOXF1.H12CORE.0.P.C  0.054421  0.000058           10\n",
       "11  ALX1.H12CORE.0.SM.B -0.006529  0.629873           11\n",
       "41  FOXF1.H12CORE.0.P.C -0.006529  0.629873           11\n",
       "12  ALX1.H12CORE.0.SM.B  0.027282  0.043993           12\n",
       "42  FOXF1.H12CORE.0.P.C  0.027282  0.043993           12\n",
       "13  ALX1.H12CORE.0.SM.B  0.035017  0.009723           13\n",
       "43  FOXF1.H12CORE.0.P.C  0.035017  0.009723           13\n",
       "14  ALX1.H12CORE.0.SM.B  0.013686  0.312365           14\n",
       "44  FOXF1.H12CORE.0.P.C  0.013686  0.312365           14\n",
       "15  ALX1.H12CORE.0.SM.B -0.003615  0.789594           15\n",
       "45  FOXF1.H12CORE.0.P.C -0.003615  0.789594           15\n",
       "16  ALX1.H12CORE.0.SM.B -0.005528  0.683261           16\n",
       "46  FOXF1.H12CORE.0.P.C -0.005528  0.683261           16\n",
       "17  ALX1.H12CORE.0.SM.B -0.003137  0.816859           17\n",
       "47  FOXF1.H12CORE.0.P.C -0.003137  0.816859           17\n",
       "18  ALX1.H12CORE.0.SM.B -0.005974  0.659239           18\n",
       "48  FOXF1.H12CORE.0.P.C -0.005974  0.659239           18\n",
       "19  ALX1.H12CORE.0.SM.B -0.006308   0.64148           19\n",
       "49  FOXF1.H12CORE.0.P.C -0.006308   0.64148           19\n",
       "2   ALX1.H12CORE.0.SM.B -0.009248  0.494819            2\n",
       "32  FOXF1.H12CORE.0.P.C -0.009248  0.494819            2\n",
       "20  ALX1.H12CORE.0.SM.B -0.005756  0.670934           20\n",
       "50  FOXF1.H12CORE.0.P.C -0.005756  0.670934           20\n",
       "21  ALX1.H12CORE.0.SM.B -0.004982  0.713086           21\n",
       "51  FOXF1.H12CORE.0.P.C -0.004982  0.713086           21\n",
       "22  ALX1.H12CORE.0.SM.B -0.005821  0.667429           22\n",
       "52  FOXF1.H12CORE.0.P.C -0.005821  0.667429           22\n",
       "23  ALX1.H12CORE.0.SM.B -0.005676  0.675216           23\n",
       "53  FOXF1.H12CORE.0.P.C -0.005676  0.675216           23\n",
       "24  ALX1.H12CORE.0.SM.B -0.005332  0.693891           24\n",
       "54  FOXF1.H12CORE.0.P.C -0.005332  0.693891           24\n",
       "25  ALX1.H12CORE.0.SM.B -0.004735  0.726708           25\n",
       "55  FOXF1.H12CORE.0.P.C -0.004735  0.726708           25\n",
       "26  ALX1.H12CORE.0.SM.B -0.004908  0.717141           26\n",
       "56  FOXF1.H12CORE.0.P.C -0.004908  0.717141           26\n",
       "27  ALX1.H12CORE.0.SM.B -0.005134  0.704701           27\n",
       "57  FOXF1.H12CORE.0.P.C -0.005134  0.704701           27\n",
       "28  ALX1.H12CORE.0.SM.B -0.004398  0.745462           28\n",
       "58  FOXF1.H12CORE.0.P.C -0.004398  0.745462           28\n",
       "29  ALX1.H12CORE.0.SM.B -0.004707  0.728243           29\n",
       "59  FOXF1.H12CORE.0.P.C -0.004707  0.728243           29\n",
       "3   ALX1.H12CORE.0.SM.B -0.005264  0.697582            3\n",
       "33  FOXF1.H12CORE.0.P.C -0.005264  0.697582            3\n",
       "4   ALX1.H12CORE.0.SM.B -0.006496  0.631579            4\n",
       "34  FOXF1.H12CORE.0.P.C -0.006496  0.631579            4\n",
       "5   ALX1.H12CORE.0.SM.B -0.006912  0.609899            5\n",
       "35  FOXF1.H12CORE.0.P.C -0.006912  0.609899            5\n",
       "6   ALX1.H12CORE.0.SM.B -0.006783  0.616604            6\n",
       "36  FOXF1.H12CORE.0.P.C -0.006783  0.616604            6\n",
       "7   ALX1.H12CORE.0.SM.B -0.007924  0.558601            7\n",
       "37  FOXF1.H12CORE.0.P.C -0.007924  0.558601            7\n",
       "8   ALX1.H12CORE.0.SM.B -0.005727  0.672504            8\n",
       "38  FOXF1.H12CORE.0.P.C -0.005727  0.672504            8\n",
       "9   ALX1.H12CORE.0.SM.B  0.031572  0.019752            9\n",
       "39  FOXF1.H12CORE.0.P.C  0.031572  0.019752            9"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "motif_enrichment_minimal_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>motif_name</th>\n",
       "      <th>score</th>\n",
       "      <th>p-value</th>\n",
       "      <th>seq_name</th>\n",
       "      <th>seq_class</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>adj_pval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr12</td>\n",
       "      <td>405</td>\n",
       "      <td>425</td>\n",
       "      <td>-</td>\n",
       "      <td>ALX1.H12CORE.0.SM.B</td>\n",
       "      <td>9.742607</td>\n",
       "      <td>0.000076</td>\n",
       "      <td>promoter|chr12:52299415-52301451</td>\n",
       "      <td>promoter</td>\n",
       "      <td>ACVRL1</td>\n",
       "      <td>0.000076</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr12</td>\n",
       "      <td>443</td>\n",
       "      <td>455</td>\n",
       "      <td>-</td>\n",
       "      <td>FOXF1.H12CORE.0.P.C</td>\n",
       "      <td>11.801019</td>\n",
       "      <td>0.000068</td>\n",
       "      <td>promoter|chr12:52299415-52301451</td>\n",
       "      <td>promoter</td>\n",
       "      <td>ACVRL1</td>\n",
       "      <td>0.000076</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chromosome start  end strand           motif_name      score   p-value  \\\n",
       "0      chr12   405  425      -  ALX1.H12CORE.0.SM.B   9.742607  0.000076   \n",
       "1      chr12   443  455      -  FOXF1.H12CORE.0.P.C  11.801019  0.000068   \n",
       "\n",
       "                           seq_name seq_class gene_name  adj_pval  \n",
       "0  promoter|chr12:52299415-52301451  promoter    ACVRL1  0.000076  \n",
       "1  promoter|chr12:52299415-52301451  promoter    ACVRL1  0.000076  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "motif_match_minimal_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/cellar/users/aklie/opt/miniconda3/envs/test_celloracle/lib/python3.10/site-packages/anndata/_core/anndata.py:1899: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/cellar/users/aklie/opt/miniconda3/envs/test_celloracle/lib/python3.10/site-packages/mudata/_core/mudata.py:479: UserWarning: var_names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of matching genes: 4420\n",
      "Number of loci: 15692\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "cb41919f349e40f6b4eacf1fc1b41797",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Computing motif enrichment:   0%|          | 0/8 [00:00<?, ?motifs/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Run motif enrichment and save results\n",
    "motif_match_D0_df, motif_count_D0_df, motif_enrichment_D0_df = compute_motif_enrichment(\n",
    "    mdata, \n",
    "    prog_key=prog_key,\n",
    "    data_key=data_key,\n",
    "    motif_file=motif_enrichment_config['motif_file'],\n",
    "    seq_file=motif_enrichment_config['seq_file'],\n",
    "    loci_file=\"/cellar/users/aklie/opt/gene_program_evaluation/examples/datasets/iPSC_EC/EnhancerPredictions_D0_2024_09_07.txt\",\n",
    "    output_loc=None,\n",
    "    window=1000,\n",
    "    threshold=1e-4,\n",
    "    eps=1e-4,\n",
    "    reverse_complement=True,\n",
    "    sig=0.05,\n",
    "    num_genes=None,\n",
    "    correlation='pearsonr',\n",
    "    n_jobs=1,\n",
    "    inplace=False\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>motif_name</th>\n",
       "      <th>score</th>\n",
       "      <th>p-value</th>\n",
       "      <th>seq_name</th>\n",
       "      <th>seq_class</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>adj_pval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr16</td>\n",
       "      <td>607</td>\n",
       "      <td>617</td>\n",
       "      <td>+</td>\n",
       "      <td>AHR.H12CORE.0.P.B</td>\n",
       "      <td>14.323496</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>genic|chr16:2715057-2715968</td>\n",
       "      <td>Enchancer</td>\n",
       "      <td>PRSS27</td>\n",
       "      <td>0.000073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr16</td>\n",
       "      <td>611</td>\n",
       "      <td>621</td>\n",
       "      <td>+</td>\n",
       "      <td>AHR.H12CORE.0.P.B</td>\n",
       "      <td>14.598420</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>genic|chr16:2715057-2715968</td>\n",
       "      <td>Enchancer</td>\n",
       "      <td>PRSS27</td>\n",
       "      <td>0.000073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr16</td>\n",
       "      <td>57</td>\n",
       "      <td>67</td>\n",
       "      <td>-</td>\n",
       "      <td>AHR.H12CORE.0.P.B</td>\n",
       "      <td>14.376087</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>genic|chr16:2715057-2715968</td>\n",
       "      <td>Enchancer</td>\n",
       "      <td>PRSS27</td>\n",
       "      <td>0.000073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr16</td>\n",
       "      <td>542</td>\n",
       "      <td>552</td>\n",
       "      <td>-</td>\n",
       "      <td>AHR.H12CORE.0.P.B</td>\n",
       "      <td>12.360946</td>\n",
       "      <td>0.000037</td>\n",
       "      <td>genic|chr16:2715057-2715968</td>\n",
       "      <td>Enchancer</td>\n",
       "      <td>PRSS27</td>\n",
       "      <td>0.000087</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr16</td>\n",
       "      <td>287</td>\n",
       "      <td>297</td>\n",
       "      <td>+</td>\n",
       "      <td>AHR.H12CORE.0.P.B</td>\n",
       "      <td>11.207113</td>\n",
       "      <td>0.000088</td>\n",
       "      <td>intergenic|chr16:2905249-2906117</td>\n",
       "      <td>Enchancer</td>\n",
       "      <td>PRSS27</td>\n",
       "      <td>0.000096</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5652</th>\n",
       "      <td>chr17</td>\n",
       "      <td>89</td>\n",
       "      <td>105</td>\n",
       "      <td>-</td>\n",
       "      <td>TEAD4.H12CORE.1.P.B</td>\n",
       "      <td>13.177800</td>\n",
       "      <td>0.000020</td>\n",
       "      <td>intergenic|chr17:427213-427713</td>\n",
       "      <td>Enchancer</td>\n",
       "      <td>C17orf97</td>\n",
       "      <td>0.000080</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5653</th>\n",
       "      <td>chr17</td>\n",
       "      <td>818</td>\n",
       "      <td>834</td>\n",
       "      <td>-</td>\n",
       "      <td>TEAD4.H12CORE.1.P.B</td>\n",
       "      <td>12.383566</td>\n",
       "      <td>0.000034</td>\n",
       "      <td>genic|chr17:2023558-2024058</td>\n",
       "      <td>Enchancer</td>\n",
       "      <td>RTN4RL1</td>\n",
       "      <td>0.000086</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5654</th>\n",
       "      <td>chr17</td>\n",
       "      <td>710</td>\n",
       "      <td>726</td>\n",
       "      <td>-</td>\n",
       "      <td>TEAD4.H12CORE.1.P.B</td>\n",
       "      <td>11.962195</td>\n",
       "      <td>0.000045</td>\n",
       "      <td>genic|chr17:4895716-4896445</td>\n",
       "      <td>Enchancer</td>\n",
       "      <td>C17orf107</td>\n",
       "      <td>0.000089</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5655</th>\n",
       "      <td>chr17</td>\n",
       "      <td>318</td>\n",
       "      <td>334</td>\n",
       "      <td>-</td>\n",
       "      <td>TEAD4.H12CORE.1.P.B</td>\n",
       "      <td>12.134159</td>\n",
       "      <td>0.000039</td>\n",
       "      <td>intergenic|chr17:7280453-7280953</td>\n",
       "      <td>Enchancer</td>\n",
       "      <td>EIF5A</td>\n",
       "      <td>0.000088</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5656</th>\n",
       "      <td>chr17</td>\n",
       "      <td>402</td>\n",
       "      <td>418</td>\n",
       "      <td>-</td>\n",
       "      <td>TEAD4.H12CORE.1.P.B</td>\n",
       "      <td>12.634568</td>\n",
       "      <td>0.000028</td>\n",
       "      <td>genic|chr17:7656916-7657696</td>\n",
       "      <td>Enchancer</td>\n",
       "      <td>ATP1B2</td>\n",
       "      <td>0.000084</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5657 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     chromosome  start  end strand           motif_name      score   p-value  \\\n",
       "0         chr16    607  617      +    AHR.H12CORE.0.P.B  14.323496  0.000006   \n",
       "1         chr16    611  621      +    AHR.H12CORE.0.P.B  14.598420  0.000004   \n",
       "2         chr16     57   67      -    AHR.H12CORE.0.P.B  14.376087  0.000006   \n",
       "3         chr16    542  552      -    AHR.H12CORE.0.P.B  12.360946  0.000037   \n",
       "4         chr16    287  297      +    AHR.H12CORE.0.P.B  11.207113  0.000088   \n",
       "...         ...    ...  ...    ...                  ...        ...       ...   \n",
       "5652      chr17     89  105      -  TEAD4.H12CORE.1.P.B  13.177800  0.000020   \n",
       "5653      chr17    818  834      -  TEAD4.H12CORE.1.P.B  12.383566  0.000034   \n",
       "5654      chr17    710  726      -  TEAD4.H12CORE.1.P.B  11.962195  0.000045   \n",
       "5655      chr17    318  334      -  TEAD4.H12CORE.1.P.B  12.134159  0.000039   \n",
       "5656      chr17    402  418      -  TEAD4.H12CORE.1.P.B  12.634568  0.000028   \n",
       "\n",
       "                              seq_name  seq_class  gene_name  adj_pval  \n",
       "0          genic|chr16:2715057-2715968  Enchancer     PRSS27  0.000073  \n",
       "1          genic|chr16:2715057-2715968  Enchancer     PRSS27  0.000073  \n",
       "2          genic|chr16:2715057-2715968  Enchancer     PRSS27  0.000073  \n",
       "3          genic|chr16:2715057-2715968  Enchancer     PRSS27  0.000087  \n",
       "4     intergenic|chr16:2905249-2906117  Enchancer     PRSS27  0.000096  \n",
       "...                                ...        ...        ...       ...  \n",
       "5652    intergenic|chr17:427213-427713  Enchancer   C17orf97  0.000080  \n",
       "5653       genic|chr17:2023558-2024058  Enchancer    RTN4RL1  0.000086  \n",
       "5654       genic|chr17:4895716-4896445  Enchancer  C17orf107  0.000089  \n",
       "5655  intergenic|chr17:7280453-7280953  Enchancer      EIF5A  0.000088  \n",
       "5656       genic|chr17:7656916-7657696  Enchancer     ATP1B2  0.000084  \n",
       "\n",
       "[5657 rows x 11 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "motif_match_D0_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# DONE!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.10 test_celloracle",
   "language": "python",
   "name": "test_scenicplus-py310"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
