{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Set-up"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import yaml\n",
    "import logging\n",
    "import mudata\n",
    "import pandas as pd\n",
    "\n",
    "# Change path to wherever you have repo locally\n",
    "sys.path.append('/cellar/users/aklie/opt/gene_program_evaluation')\n",
    "\n",
    "from src.evaluation import (\n",
    "    compute_categorical_association,\n",
    "    compute_geneset_enrichment,\n",
    "    compute_trait_enrichment,\n",
    "    compute_perturbation_association,\n",
    "    compute_explained_variance_ratio,\n",
    "    compute_motif_enrichment\n",
    ")\n",
    "from src.evaluation.enrichment_trait import process_enrichment_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# I/O paths\n",
    "path_config = \"/cellar/users/aklie/opt/gene_program_evaluation/examples/evaluation/iPSC_EC/cNMF_30/evaluation_pipeline.yml\"\n",
    "config = yaml.safe_load(open(path_config))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Motif enrichment testing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import argparse\n",
    "\n",
    "import mudata\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "from scipy.stats import pearsonr, spearmanr, kendalltau\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "\n",
    "from tangermeme.io import read_meme, extract_loci\n",
    "from tangermeme.tools.fimo import FIMO\n",
    "\n",
    "from joblib import Parallel, delayed\n",
    "from tqdm.auto import tqdm\n",
    "from typing import Union, Mapping, Optional\n",
    "\n",
    "\n",
    "def read_coords_file(\n",
    "    coords_file_loc: os.PathLike\n",
    "):\n",
    "\n",
    "    \"\"\"Read coordinates from tab formatted file\n",
    "\n",
    "    Read promoter/enhancer to gene links tab \n",
    "    delimited file with the following column headers:\n",
    "    chr, start, end, seq_name, seq_class {promoter, enhancer}, \n",
    "    seq_score, gene_name\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    coords_file_loc : str\n",
    "        Path to the coordinates file.\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    records : pd.DataFrame\n",
    "        DataFrame containing the coordinates.\n",
    "    \"\"\"\n",
    "\n",
    "    # Read formatted coords file\n",
    "    records = pd.read_csv(coords_file_loc, sep='\\t')\n",
    "\n",
    "    # Enforce header format\n",
    "    expected_headers = ['chromosome', 'start', 'end', 'seq_name', \n",
    "                        'seq_class', 'seq_score', 'gene_name']\n",
    "    for col in expected_headers:\n",
    "        try: assert col in records.columns\n",
    "        except: raise ValueError('Coordinate file is not formatted correctly')\n",
    "\n",
    "    # TODO: Support multiple classes at once\n",
    "    try: assert len(records['seq_class']==1)\n",
    "    except: raise ValueError('Coordinate file contains multiple sequence classes')\n",
    "\n",
    "    return records\n",
    "\n",
    "\n",
    "def perform_motif_match(\n",
    "    loci: Union[pd.DataFrame, os.PathLike],\n",
    "    sequences: os.PathLike,\n",
    "    pwms: Mapping[str, np.ndarray],\n",
    "    in_window: int=1000,\n",
    "    threshold: float=1e-4,\n",
    "    output_loc: Optional[os.PathLike]=None\n",
    "):\n",
    "        \"\"\"Score motif matches\n",
    "\n",
    "        Perform motif matching on sequences linked to genes\n",
    "        via enhancer/promoter coordinates.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        loci : pd.DataFrame\n",
    "            DataFrame containing sequence coordinates.\n",
    "        sequences : os.PathLike\n",
    "            Path to FASTA formatted genomic sequence.\n",
    "        pwms : Mapping[str, np.ndarray]\n",
    "            Dictionary of PWMs where keys are motif names and values are PWMs.\n",
    "        in_window : int\n",
    "            Window size to extract sequences around center of loci.\n",
    "        threshold : float\n",
    "            Threshold for motif matching.\n",
    "        output_loc : os.PathLike\n",
    "            Path to directory to store motif matches for individual motifs.\n",
    "        \"\"\"\n",
    "        # Perform motif matching\n",
    "        model = FIMO(pwms)\n",
    "        X = extract_loci(loci, sequences, in_window=in_window).float()\n",
    "        print(X.shape)\n",
    "        y = model(X)\n",
    "        hits = model.hits(X, threshold=threshold)\n",
    "\n",
    "        # Create motif match dataframe\n",
    "        motif_match_df = pd.DataFrame()\n",
    "        for i, hit in enumerate(hits):\n",
    "            annotated_hit = hit.merge(loci[[\"chromosome\", \"seq_name\", \"seq_class\", \"gene_name\"]], left_on=\"example_idx\", right_index=True).drop(columns=[\"example_idx\"])\n",
    "            annotated_hit[\"motif_name\"] = list(pwms.keys())[i]\n",
    "            annotated_hit = annotated_hit[[\"chromosome\", \"start\", \"end\", \"strand\", \"motif_name\", \"score\", \"p-value\", \"seq_name\", \"seq_class\", \"gene_name\"]]\n",
    "            if output_loc is not None:\n",
    "                annotated_hit.to_csv(os.path.join(motif_enrichment_config['output_loc'], f\"motif_match_{list(pwms.keys())[i]}.txt\"), sep='\\t', index=False)\n",
    "            motif_match_df = pd.concat([motif_match_df, annotated_hit])\n",
    "        motif_match_df[\"adj_pval\"] = multipletests(motif_match_df[\"p-value\"], method=\"fdr_bh\")[1]\n",
    "        return motif_match_df\n",
    "\n",
    "\n",
    "def compute_motif_instances(\n",
    "    motif_match_df: pd.DataFrame,\n",
    "    motif_var: str = 'motif_name',\n",
    "    sig: float=0.05,\n",
    "    sig_var: str='adj_pval',\n",
    "    gene_names: Optional[np.ndarray]=None\n",
    "):\n",
    "\n",
    "    \"\"\"Count motif instances per gene (via enahncer/promoter linking)\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    motif_match_df : pd.DataFrame\n",
    "        DataFrame containing motif matches.\n",
    "    motif_var : str\n",
    "        Column name for motif names. Default is 'motif_name'.\n",
    "    sig : float\n",
    "        Significance threshold for motif matches. Default is 0.05.\n",
    "    sig_var : str\n",
    "        Column name for significance values. Default is 'adj_pval'.\n",
    "    gene_names : np.ndarray\n",
    "        Array of gene names. Default is None.\n",
    "    \n",
    "    Returns\n",
    "    -------\n",
    "    motif_count_df : pd.DataFrame\n",
    "        DataFrame containing\n",
    "    \"\"\"\n",
    "\n",
    "    # Count up significant occurences of motif\n",
    "    motif_match_df_ = motif_match_df.loc[motif_match_df[sig_var] < sig]\n",
    "    motif_match_df_ = motif_match_df.value_counts(subset=['gene_name', motif_var]).reset_index()\n",
    "    motif_match_df_.columns = ['gene_name', motif_var, 'motif_count']\n",
    "    motif_match_df_ = motif_match_df_.pivot(index='gene_name', columns=motif_var, values='motif_count')\n",
    "    motif_count_df = pd.DataFrame(index=gene_names, columns=motif_match_df_.columns)\n",
    "    motif_count_df.loc[motif_match_df_.index.values] = motif_match_df_ # Gene names should match as this point\n",
    "    return motif_count_df\n",
    "\n",
    "\n",
    "def perform_correlation(\n",
    "    motif_count_df: pd.DataFrame,\n",
    "    prog_genes: pd.DataFrame,\n",
    "    motif_enrich_stat_df: pd.DataFrame,\n",
    "    motif_enrich_pval_df: pd.DataFrame,\n",
    "    motif_idx: int,\n",
    "    prog_idx: int,\n",
    "    correlation: str='pearsonr'\n",
    "):\n",
    "    \"\"\"Compute motif enrichment as correlation b/w gene weights/ranks and motif counts\n",
    "    \n",
    "    Perform pearson correlation test for motif count enrichment vs gene loadings\n",
    "    If loadings are dichotomized then this is equivalent to a point biserial correlation test.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    motif_count_df : pd.DataFrame\n",
    "        DataFrame containing motif counts.\n",
    "    prog_genes : pd.DataFrame\n",
    "        DataFrame containing gene program loadings.\n",
    "    motif_enrich_stat_df : pd.DataFrame\n",
    "        DataFrame to store correlation statistics.\n",
    "    motif_enrich_pval_df : pd.DataFrame\n",
    "        DataFrame to store correlation p-values.\n",
    "    motif_idx : int\n",
    "        Index of motif.\n",
    "    prog_idx : int\n",
    "        Index of gene program.\n",
    "    correlation : {'pearsonr','spearmanr','kendalltau'}\n",
    "        Type of correlation to perform. Default is 'pearsonr'.\n",
    "    \"\"\"\n",
    "\n",
    "    loadings = prog_genes.iloc[prog_idx].values.flatten()\n",
    "    counts = motif_count_df.T.iloc[motif_idx].fillna(0).values.flatten()\n",
    "    \n",
    "    if correlation=='pearsonr':\n",
    "        stat, pval = pearsonr(loadings, counts)\n",
    "    elif correlation=='spearmanr':\n",
    "        stat, pval = spearmanr(loadings, counts)\n",
    "    elif correlation=='kendalltau':\n",
    "        stat, pval = kendalltau(loadings, counts)\n",
    "\n",
    "    motif_enrich_stat_df.iloc[prog_idx, motif_idx]  = stat\n",
    "    motif_enrich_pval_df.iloc[prog_idx, motif_idx]  = pval\n",
    "\n",
    "\n",
    "def compute_motif_enrichment_(\n",
    "    mdata: mudata.MuData,\n",
    "    motif_count_df: pd.DataFrame,\n",
    "    prog_key: str='prog',\n",
    "    gene_names: Optional[np.ndarray]=None,\n",
    "    weighted: bool=True,\n",
    "    num_genes: Optional[int]=None,\n",
    "    correlation: str='pearsonr',\n",
    "    n_jobs: int=1\n",
    "):\n",
    "    \"\"\"Count up motif ocurrences and perform diff. test\n",
    "\n",
    "    Perform motif enrichment using gene program loadings and\n",
    "    motif counts linked to genes via motif scanning of\n",
    "    linked enhancer/promoter sequences.\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    mdata : MuData\n",
    "        MuData object containing anndata of program scores and cell-level metadata.\n",
    "    motif_count_df : pd.DataFrame\n",
    "        DataFrame containing motif counts.\n",
    "    prog_key : str\n",
    "        Key for the anndata object in the mudata object. Default is 'prog'.\n",
    "    gene_names : np.ndarray\n",
    "        Array of gene names. Default is None.\n",
    "    weighted : bool\n",
    "        Use weighted loadings. Default is True.\n",
    "    num_genes : int\n",
    "        Number of genes threshold to dichotomize loadings. Default is None.\n",
    "    correlation : {'pearsonr','spearmanr','kendalltau'}\n",
    "        Type of correlation to perform. Default is 'pearsonr'.\n",
    "    n_jobs : int\n",
    "        Number of threads to run processes on. Default is 1.\n",
    "    \"\"\"\n",
    "\n",
    "    # Both weighted and num_genes cannot be set\n",
    "    if num_genes is not None and weighted:\n",
    "        raise ValueError('Will not use weighted when num_genes specified.')\n",
    "\n",
    "    loadings = pd.DataFrame(data=mdata[prog_key].varm['loadings'],\n",
    "                            index=mdata[prog_key].var_names,\n",
    "                            columns=gene_names)\n",
    "    # FIXME: Causes expansion due to duplications in gene_names\n",
    "    # Ensure index matches b/w loadings and counts\n",
    "    # loadings = loadings.loc[:, motif_count_df.index.values]\n",
    "\n",
    "    # Binary matrix \n",
    "    if not weighted:\n",
    "        prog_genes = (loadings.rank(axis=1)<num_genes).astype(int)\n",
    "    elif weighted and num_genes is None:\n",
    "        prog_genes = loadings\n",
    "    elif weighted and isinstance(num_genes, int):\n",
    "        prog_genes = (loadings.rank(axis=1)<num_genes).astype(int)\n",
    "        prog_genes *= loadings\n",
    "    else:\n",
    "        raise ValueError('num_genes not specified correctly.')\n",
    "\n",
    "    # Use pearson correlation\n",
    "    # If dichotomized then this is equivalent to point biserial correlation\n",
    "    motif_enrich_stat_df = pd.DataFrame(index=mdata[prog_key].var_names,\n",
    "                                        columns=motif_count_df.columns.values)\n",
    "    motif_enrich_pval_df = pd.DataFrame(index=mdata[prog_key].var_names,\n",
    "                                        columns=motif_count_df.columns.values)\n",
    "\n",
    "    # FIXME: If n_jobs>1 then parallel processes dont seem to terminate.\n",
    "    # Perform test in parallel across motifs and programs\n",
    "    Parallel(n_jobs=n_jobs, \n",
    "             backend='threading')(delayed(perform_correlation)(motif_count_df,\n",
    "                                                               prog_genes, \n",
    "                                                               motif_enrich_stat_df,\n",
    "                                                               motif_enrich_pval_df,\n",
    "                                                               motif_idx, \n",
    "                                                               prog_idx,\n",
    "                                                               correlation=correlation) \\\n",
    "                                                            for motif_idx in tqdm(range(motif_count_df.columns.values.shape[-1]),\n",
    "                                                                                     desc='Computing motif enrichment',\n",
    "                                                                                     unit='motifs') \\\n",
    "                                                            for prog_idx in range(mdata[prog_key].var_names.shape[0]))\n",
    "\n",
    "    return motif_enrich_stat_df, motif_enrich_pval_df\n",
    "\n",
    "\n",
    "def compute_motif_enrichment(\n",
    "    mdata: Union[mudata.MuData, os.PathLike],\n",
    "    prog_key: str='prog',\n",
    "    data_key: str='data',\n",
    "    motif_file: Optional[os.PathLike]=None,\n",
    "    seq_file: Optional[os.PathLike]=None,\n",
    "    coords_file: Optional[os.PathLike]=None,\n",
    "    output_loc: Optional[os.PathLike]=None,\n",
    "    window: int=1000,\n",
    "    threshold: float=1e-4,\n",
    "    sig: float=0.05,\n",
    "    num_genes: Optional[int]=None,\n",
    "    correlation: str='pearsonr',\n",
    "    n_jobs: int=1,\n",
    "    inplace: bool=True,\n",
    "    **kwargs\n",
    "):\n",
    "    \n",
    "    \"\"\"Compute motif enrichment in enhancers or promoters associated with a gene\n",
    "    \n",
    "    Perform motif enrichment using gene program loadings and\n",
    "    motif counts linked to genes via motif scanning of\n",
    "    linked enhancer/promoter sequences.\n",
    "\n",
    "    ARGS\n",
    "        mdata : MuData\n",
    "            mudata object containing anndata of program scores and cell-level metadata.\n",
    "        prog_key: \n",
    "            index for the anndata object (mdata[prog_key]) in the mudata object.\n",
    "        data_key: str\n",
    "            index of the genomic data anndata object (mdata[data_key]) in the mudata object.\n",
    "        motif_file: str\n",
    "            path to motif file formatted in MEME format.\n",
    "        seq_file: str\n",
    "            path to FASTA formatted genomic sequence.\n",
    "        coords_file: str\n",
    "            path to enhancer/promoter gene links file with sequence coordinates.\n",
    "            Tab delimited file with the following column headers:\n",
    "            chr, start, end, seq_name, seq_class {promoter, enhancer}, \n",
    "            seq_score, gene_name.\n",
    "        output_loc: str\n",
    "            path to directory to store motif - gene counts.\n",
    "        sig: (0,1] (default: 0.05)\n",
    "            significance level for inferring a motif match.\n",
    "        num_genes: int (default: None)\n",
    "            number of genes threshold to dichtomize loadings.\n",
    "        correlation: {'pearsonr','spearmanr','kendalltau'} (default: 'peasronsr')\n",
    "            correlation type to use to compute motif enirchments.\n",
    "            Use kendalltau when expecting enrichment/de-enrichment at both ends.\n",
    "        use_previous: bool (default: True)\n",
    "            if outplot is provided try to load motif matches from previous run.\n",
    "        n_jobs: int (default: 1)\n",
    "            number of threads to run processes on.\n",
    "        inplace: Bool (default: True)\n",
    "            update the mudata object inplace or return a copy\n",
    "       \n",
    "    ---\n",
    "        if not inplace:\n",
    "            RETURNS\n",
    "                motif_match_df,\n",
    "                motif_count_df.loc[gene_names].values,\n",
    "                motif_enrichment_df\n",
    "        else:\n",
    "            UPDATES\n",
    "\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    # Read in mudata if it is provided as a path\n",
    "    frompath=False\n",
    "    if isinstance(mdata, str):\n",
    "        if os.path.exists(mdata):\n",
    "            mdata = mudata.read(mdata)\n",
    "            if inplace:\n",
    "                logging.warning('Changed to inplace=False since path was provided')\n",
    "                inplace=False\n",
    "            frompath=True\n",
    "        else: raise ValueError('Incorrect mudata specification.')\n",
    "    \n",
    "    if not inplace and not frompath:\n",
    "        mdata = mudata.MuData({prog_key: mdata[prog_key].copy(),\n",
    "                               data_key: mdata[data_key].copy()})\n",
    "\n",
    "    if 'var_names' in mdata[prog_key].uns.keys():\n",
    "        gene_names = mdata[prog_key].uns['var_names']\n",
    "    else:\n",
    "        try: assert mdata[prog_key].varm['loadings'].shape[1]==mdata[data_key].var.shape[0]\n",
    "        except: raise ValueError('Different number of genes present in data and program loadings')\n",
    "        gene_names = mdata[data_key].var_names\n",
    "    \n",
    "    # \n",
    "    if ':ens' in gene_names[0].lower():\n",
    "        gene_names = [name.split(':')[0] for name in gene_names]\n",
    "\n",
    "    # Check if output loc exists\n",
    "    if output_loc is not None:\n",
    "        try: os.makedirs(output_loc, exist_ok=True)\n",
    "        except: raise ValueError('Output location does not exist.')\n",
    "\n",
    "    # If num_genes specified then cannot be weighted\n",
    "    if num_genes is None:\n",
    "        weighted=True\n",
    "    elif isinstance(num_genes, int):\n",
    "        weighted=False\n",
    "\n",
    "    # Intake motif file path or in memory\n",
    "    if isinstance(motif_file, str) and os.path.exists(motif_file):\n",
    "        pwms = read_meme(motif_file)\n",
    "    else:\n",
    "        raise ValueError('Motif file not found.')\n",
    "    \n",
    "    # Intake coord file path or in memory\n",
    "    if isinstance(coords_file, str) and os.path.exists(coords_file):\n",
    "        coords = read_coords_file(coords_file)\n",
    "    else:\n",
    "        raise ValueError('Coordinate file not found.')\n",
    "\n",
    "    # Valid genes\n",
    "    matching_gene_names = np.intersect1d(coords['gene_name'].unique(), gene_names)\n",
    "    try: assert len(matching_gene_names) > 0\n",
    "    except: raise ValueError('No matching genes b/w data and coordinate files')\n",
    "\n",
    "    # Compute motif matching\n",
    "    coords = coords[coords['gene_name'].isin(matching_gene_names)]\n",
    "    motif_match_df = perform_motif_match(\n",
    "        loci=coords,\n",
    "        sequences=seq_file,\n",
    "        pwms=pwms,\n",
    "        in_window=window,\n",
    "        threshold=threshold,\n",
    "        output_loc=output_loc\n",
    "    )\n",
    "\n",
    "    # Count motif enrichment\n",
    "    motif_count_df = compute_motif_instances(\n",
    "        motif_match_df,\n",
    "        motif_var='motif_name',\n",
    "        sig=sig,\n",
    "        sig_var='adj_pval',\n",
    "        gene_names=gene_names\n",
    "    )\n",
    "    motif_enrich_stat_df, motif_enrich_pval_df = compute_motif_enrichment_(\n",
    "        mdata,\n",
    "        motif_count_df,\n",
    "        prog_key=prog_key,\n",
    "        gene_names=gene_names,\n",
    "        weighted=weighted,\n",
    "        num_genes=num_genes,\n",
    "        n_jobs=n_jobs)\n",
    "                         \n",
    "    # Store motif counts\n",
    "    mdata[prog_key].uns['motif_counts'] = motif_count_df.loc[gene_names].values\n",
    "    mdata[prog_key].uns['motif_names'] = motif_count_df.columns.values\n",
    "\n",
    "    mdata[prog_key].varm['motif_enrich_{}_stat'.format(correlation)] = motif_enrich_stat_df.values\n",
    "    mdata[prog_key].varm['motif_enrich_{}_pval'.format(correlation)] = motif_enrich_pval_df.values\n",
    "    mdata[prog_key].uns['motif_names'] = motif_count_df.columns.values\n",
    "\n",
    "    if not inplace: \n",
    "\n",
    "        motif_enrich_stat_df = motif_enrich_stat_df.reset_index().melt(id_vars='index',\n",
    "                                                                       var_name='motif', \n",
    "                                                                       value_name='stat')\n",
    "        motif_enrich_stat_df = motif_enrich_stat_df.set_index(['index', 'motif'])\n",
    "\n",
    "        motif_enrich_pval_df = motif_enrich_pval_df.reset_index().melt(id_vars='index',\n",
    "                                                                       var_name='motif', \n",
    "                                                                       value_name='pval')\n",
    "        motif_enrich_pval_df = motif_enrich_pval_df.set_index(['index', 'motif'])\n",
    "\n",
    "        motif_enrichment_df = motif_enrich_stat_df.merge(motif_enrich_pval_df,\n",
    "                                                        left_index=True, \n",
    "                                                        right_index=True)\n",
    "        motif_enrichment_df = motif_enrichment_df.reset_index()\n",
    "        motif_enrichment_df['program_name'] = motif_enrichment_df['index']\n",
    "        motif_enrichment_df.drop('index', axis=1, inplace=True)\n",
    "\n",
    "        motif_count_df.columns.name=''\n",
    "\n",
    "        return (motif_match_df,\n",
    "                motif_count_df.loc[gene_names].dropna(),\n",
    "                motif_enrichment_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/cellar/users/aklie/opt/miniconda3/envs/test_celloracle/lib/python3.10/site-packages/anndata/_core/anndata.py:1899: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/cellar/users/aklie/opt/miniconda3/envs/test_celloracle/lib/python3.10/site-packages/mudata/_core/mudata.py:479: UserWarning: var_names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "torch.Size([15692, 4, 1000])\n"
     ]
    },
    {
     "ename": "",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31mThe Kernel crashed while executing code in the current cell or a previous cell. \n",
      "\n",
      "\u001b[1;31mPlease review the code in the cell(s) to identify a possible cause of the failure. \n",
      "\n",
      "\u001b[1;31mClick <a href='https://aka.ms/vscodeJupyterKernelCrash'>here</a> for more info. \n",
      "\n",
      "\u001b[1;31mView Jupyter <a href='command:jupyter.viewOutput'>log</a> for further details."
     ]
    },
    {
     "ename": "",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31mCanceled future for execute_request message before replies were done"
     ]
    },
    {
     "ename": "",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31mCanceled future for execute_request message before replies were done. \n",
      "\n",
      "\u001b[1;31mView Jupyter <a href='command:jupyter.viewOutput'>log</a> for further details."
     ]
    }
   ],
   "source": [
    "# Run motif enrichment and save results\n",
    "motif_enrichment_config = config['motif_enrichment']\n",
    "motif_match_df, motif_count_df, motif_enrichment_df = compute_motif_enrichment(\n",
    "    mdata, \n",
    "    prog_key=prog_key,\n",
    "    data_key=data_key,\n",
    "    **motif_enrichment_config,\n",
    ")\n",
    "motif_match_df.to_csv(os.path.join(path_out, 'motif_enrichment_matches.txt'), sep='\\t', index=False)\n",
    "motif_count_df.to_csv(os.path.join(path_out, 'motif_enrichment_counts.txt'), sep='\\t', index=False)\n",
    "motif_enrichment_df.to_csv(os.path.join(path_out, 'motif_enrichment.txt'), sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run motif enrichment and save results\n",
    "motif_enrichment_config = config['motif_enrichment']\n",
    "motif_match_df, motif_count_df, motif_enrichment_df = compute_motif_enrichment(\n",
    "    mdata, \n",
    "    prog_key=prog_key,\n",
    "    data_key=data_key,\n",
    "    **motif_enrichment_config,\n",
    ")\n",
    "motif_match_df.to_csv(os.path.join(path_out, 'motif_enrichment_matches.txt'), sep='\\t', index=False)\n",
    "motif_count_df.to_csv(os.path.join(path_out, 'motif_enrichment_counts.txt'), sep='\\t', index=False)\n",
    "motif_enrichment_df.to_csv(os.path.join(path_out, 'motif_enrichment.txt'), sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pwms = read_meme(motif_enrichment_config['motif_file'])\n",
    "model = FIMO(pwms)\n",
    "X = extract_loci(loci_, motif_enrichment_config[\"seq_file\"], in_window=1000).float()\n",
    "y = model(X)\n",
    "hits = model.hits(X, threshold=0.001)\n",
    "motif_match_df = pd.DataFrame()\n",
    "for i, hit in enumerate(hits):\n",
    "    annotated_hit = hit.merge(loci_[[\"chromosome\", \"seq_name\", \"seq_class\", \"gene_name\"]], left_on=\"example_idx\", right_index=True).drop(columns=[\"example_idx\"])\n",
    "    annotated_hit[\"motif_name\"] = list(pwms.keys())[i]\n",
    "    annotated_hit = annotated_hit[[\"chromosome\", \"start\", \"end\", \"strand\", \"motif_name\", \"score\", \"p-value\", \"seq_name\", \"seq_class\", \"gene_name\"]]\n",
    "    annotated_hit.to_csv(os.path.join(motif_enrichment_config['output_loc'], f\"motif_match_{list(pwms.keys())[i]}.txt\"), sep='\\t', index=False)\n",
    "    motif_match_df = pd.concat([motif_match_df, annotated_hit])\n",
    "motif_match_df[\"adj_pval\"] = multipletests(motif_match_df[\"p-value\"], method=\"fdr_bh\")[1]\n",
    "motif_count_df = compute_motif_instances(motif_match_df, gene_names=gene_names)\n",
    "motif_enrich_stat_df, motif_enrich_pval_df = compute_motif_enrichment_(\n",
    "    mdata,\n",
    "    motif_count_df=motif_count_df,\n",
    "    prog_key=prog_key,\n",
    "    gene_names=gene_names,\n",
    "    num_genes=None,\n",
    "    n_jobs=1,\n",
    ")\n",
    "motif_enrich_stat_df = motif_enrich_stat_df.reset_index().melt(id_vars='index', var_name='motif', value_name='stat')\n",
    "motif_enrich_stat_df = motif_enrich_stat_df.set_index(['index', 'motif'])\n",
    "motif_enrich_pval_df = motif_enrich_pval_df.reset_index().melt(id_vars='index', var_name='motif', value_name='pval')\n",
    "motif_enrich_pval_df = motif_enrich_pval_df.set_index(['index', 'motif'])\n",
    "motif_enrichment_df = motif_enrich_stat_df.merge(motif_enrich_pval_df, left_index=True, right_index=True)\n",
    "motif_enrichment_df = motif_enrichment_df.reset_index()\n",
    "motif_enrichment_df['program_name'] = motif_enrichment_df['index']\n",
    "motif_enrichment_df.drop('index', axis=1, inplace=True)\n",
    "motif_count_df.columns.name = ''"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
