{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import glob\n",
    "import sys\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "sys.path.append('/cellar/users/aklie/opt/gene_program_evaluation')\n",
    "sys.path.append('/cellar/users/aklie/opt/gene_program_evaluation/dashapp/')\n",
    "import mudata\n",
    "import scanpy as sc\n",
    "\n",
    "from data_processing import count_unique"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def count(categorical_var, count_var, dataframe):\n",
    "    counts_df = dataframe.value_counts([categorical_var, count_var])\n",
    "    counts_df = counts_df.groupby(categorical_var).sum()\n",
    "    counts_df = counts_df.sort_values(ascending=False)\n",
    "    counts_df = pd.DataFrame(counts_df.reset_index().values,\n",
    "                             columns=[categorical_var, count_var])\n",
    "    return counts_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Pipeline outputs\n",
    "path_pipeline_outs = \"/cellar/users/aklie/opt/gene_program_evaluation/dashapp/example_data/iPSC_EC_evaluations\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "`path_pipeline_outs` should be structured something like as follows:\n",
    "    \n",
    "```bash\n",
    "path_pipeline_outs/\n",
    "├── cNMF_60\n",
    "├── cNMF_59\n",
    "├── cNMF_58\n",
    "...\n",
    "├── cNMF_1\n",
    "└── output.mdata\n",
    "```\n",
    "\n",
    "where `output.mdata` contains the MuData object with keys `input` that contains the original input matrix and `cNMF_60`, `cNMF_59`, ..., `cNMF_1` containing the cell and gene loadings for each of the inference runs.\n",
    "\n",
    "The inference runs can be 1 of 4 types:\n",
    "1. A single run of a method with a fixed number of components.\n",
    "2. Cross k-analysis: where the number of components is varied across the same method.\n",
    "3. Cross method analysis: where the number of components is fixed and the method is varied.\n",
    "4. A combination of 2) and 3).\n",
    "\n",
    "The keys of the MuData and the output subdirectories in `path_pipeline_outs` should be named accordingly to the desired type of analysis:\n",
    "1. If one key is present, it can be named anything. In this scenario, the cross run analysis will be omitted.\n",
    "2. A cross k-analysis should share the same base name and be suffixed by an integer that indicates the number of components used in inference. e.g. `cNMF_60`, `cNMF_59`, ..., `cNMF_1`.\n",
    "3. A cross method analysis should have separate names for each method. e.g. `cNMF`, `Topyfic`, etc.\n",
    "4. A combination of 2) and 3) should have the same base name for each method and be suffixed by an integer that indicates the number of components used in inference. e.g. `cNMF_60`, `cNMF_59`, ..., `cNMF_1`, `Topyfic_60`, `Topyfic_59`, ..., `Topyfic_1`."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load MuData"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/cellar/users/aklie/opt/miniconda3/envs/test_celloracle/lib/python3.10/site-packages/anndata/_core/anndata.py:522: FutureWarning: The dtype argument is deprecated and will be removed in late 2024.\n",
      "  warnings.warn(\n",
      "/cellar/users/aklie/opt/miniconda3/envs/test_celloracle/lib/python3.10/site-packages/anndata/_core/anndata.py:522: FutureWarning: The dtype argument is deprecated and will be removed in late 2024.\n",
      "  warnings.warn(\n",
      "/cellar/users/aklie/opt/miniconda3/envs/test_celloracle/lib/python3.10/site-packages/anndata/_core/anndata.py:1899: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/cellar/users/aklie/opt/miniconda3/envs/test_celloracle/lib/python3.10/site-packages/mudata/_core/mudata.py:479: UserWarning: var_names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "try:\n",
    "    path_mdata = os.path.join(path_pipeline_outs, \"cNMF_60_0.2_gene_names.h5mu\")\n",
    "    mdata = mudata.read_h5mu(path_mdata)\n",
    "except:\n",
    "    print(\"Could not load mdata\")\n",
    "    sys.exit(1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Get subdirectories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['/cellar/users/aklie/opt/gene_program_evaluation/dashapp/example_data/iPSC_EC_evaluations/cNMF']"
      ]
     },
     "execution_count": 122,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get all subdirectories of the pipeline outputs\n",
    "subdirs = [x[0] for x in os.walk(path_pipeline_outs)][1:]\n",
    "subdirs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Overview page\n",
    "Capture the output of this display call"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_1182793/3124802129.py:2: DeprecationWarning: Importing display from IPython.core.display is deprecated since IPython 7.14, please import from IPython display\n",
      "  from IPython.core.display import display, HTML\n"
     ]
    }
   ],
   "source": [
    "from io import StringIO\n",
    "from IPython.core.display import display, HTML\n",
    "from contextlib import redirect_stdout"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'<style>\\n.hl-dim, .hl-size, .hl-values, .hl-types, .hl-dims {\\n  color: #777777;\\n}\\n.hl-dim::before, .hl-size::before {\\n  content: \"\\\\00a0\\\\00a0\\\\00a0\";\\n}\\n.hl-values {\\n  font-family: monospace;\\n}\\n.hl-file {\\n  background-color: #EEEEEE;\\n  border-radius: .5rem;\\n  padding: .2rem .4rem;\\n  color: #555555;\\n}\\n.hl-empty {\\n  color: #999999;\\n}\\n.hl-import {\\n  color: #777777;\\n}\\n.block-mod {\\n  display: block;\\n  margin: 0 2rem;\\n}\\n.title {\\n  display: inline-block;\\n  font-weight: 600;\\n  color: #555555;\\n}\\n.title-mod {\\n  font-size: 1.2rem;\\n  color: #04b374;\\n  padding: 0 .5rem;\\n}\\n.title-attr {\\n  font-size: 1.0rem;\\n  padding-top: .2rem;\\n}\\nsummary {\\n  cursor: pointer;\\n  list-style: none;\\n}\\nsummary::-webkit-details-marker {\\n  display: none;\\n}\\ndetails > summary::before {\\n  content: \\'⊕\\';\\n}\\ndetails[open] > summary::before {\\n  content: \\'⊖\\';\\n}\\ntable tr {\\n  background-color: transparent !important;\\n}\\ntable tr:hover {\\n  background-color: #04b37433 !important;\\n}\\n.col-index {\\n  text-align: left !important;\\n}\\n.summary-mod {\\n  margin-left: -2rem;\\n}\\n.summary-mod:hover {\\n  background-color: #04b37411;\\n}\\n.summary-mod::before {\\n  color: #04b374;\\n  content: \\'●\\';\\n}\\ndetails[open] > .summary-mod::before {\\n  content: \\'○\\';\\n}\\n</style><span>MuData object <span class=\\'hl-dim\\'>92284 obs &times; 5511 var in 2 modalities</span></span><br><details><summary><div class=\\'title title-attr\\'>Metadata</div><span class=\\'hl-dim\\'>.obs</span><span class=\\'hl-size\\'>14 elements</span></summary><div><table><tr>\\n                                <td class=\\'col-index\\'>rna:sample</td>  <td class=\\'hl-types\\'>category</td>  <td class=\\'hl-values\\'>D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0...</td>\\n                            </tr>\\n<tr>\\n                                <td class=\\'col-index\\'>rna:species</td>  <td class=\\'hl-types\\'>category</td>  <td class=\\'hl-values\\'>hg38,hg38,hg38,hg38,hg38,hg38,hg38,hg38,hg38,hg38,...</td>\\n                            </tr>\\n<tr>\\n                                <td class=\\'col-index\\'>rna:gene_count</td>  <td class=\\'hl-types\\'>int64</td>  <td class=\\'hl-values\\'>1903,2664,2147,2011,1441,1607,1385,3690,1348,1247,...</td>\\n                            </tr>\\n<tr>\\n                                <td class=\\'col-index\\'>rna:tscp_count</td>  <td class=\\'hl-types\\'>int64</td>  <td class=\\'hl-values\\'>3010,4710,3492,3286,2027,2616,2256,7268,2093,1934,...</td>\\n                            </tr>\\n<tr>\\n                                <td class=\\'col-index\\'>rna:mread_count</td>  <td class=\\'hl-types\\'>int64</td>  <td class=\\'hl-values\\'>4079,6401,4666,4432,2773,3499,3044,9975,2837,2533,...</td>\\n                            </tr>\\n<tr>\\n                                <td class=\\'col-index\\'>rna:leiden</td>  <td class=\\'hl-types\\'>category</td>  <td class=\\'hl-values\\'>0,2,1,0,0,0,4,5,4,4,4,0,1,3,2,0,2,2,1,1,1,3,0,0,0,...</td>\\n                            </tr>\\n<tr>\\n                                <td class=\\'col-index\\'>rna:n_counts</td>  <td class=\\'hl-types\\'>float32</td>  <td class=\\'hl-values\\'>653.00,997.00,618.00,742.00,430.00,585.00,570.00,1...</td>\\n                            </tr>\\n<tr>\\n                                <td class=\\'col-index\\'>cNMF:sample</td>  <td class=\\'hl-types\\'>category</td>  <td class=\\'hl-values\\'>D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0...</td>\\n                            </tr>\\n<tr>\\n                                <td class=\\'col-index\\'>cNMF:species</td>  <td class=\\'hl-types\\'>category</td>  <td class=\\'hl-values\\'>hg38,hg38,hg38,hg38,hg38,hg38,hg38,hg38,hg38,hg38,...</td>\\n                            </tr>\\n<tr>\\n                                <td class=\\'col-index\\'>cNMF:gene_count</td>  <td class=\\'hl-types\\'>int64</td>  <td class=\\'hl-values\\'>1903,2664,2147,2011,1441,1607,1385,3690,1348,1247,...</td>\\n                            </tr>\\n<tr>\\n                                <td class=\\'col-index\\'>cNMF:tscp_count</td>  <td class=\\'hl-types\\'>int64</td>  <td class=\\'hl-values\\'>3010,4710,3492,3286,2027,2616,2256,7268,2093,1934,...</td>\\n                            </tr>\\n<tr>\\n                                <td class=\\'col-index\\'>cNMF:mread_count</td>  <td class=\\'hl-types\\'>int64</td>  <td class=\\'hl-values\\'>4079,6401,4666,4432,2773,3499,3044,9975,2837,2533,...</td>\\n                            </tr>\\n<tr>\\n                                <td class=\\'col-index\\'>cNMF:leiden</td>  <td class=\\'hl-types\\'>category</td>  <td class=\\'hl-values\\'>0,2,1,0,0,0,4,5,4,4,4,0,1,3,2,0,2,2,1,1,1,3,0,0,0,...</td>\\n                            </tr>\\n<tr>\\n                                <td class=\\'col-index\\'>cNMF:n_counts</td>  <td class=\\'hl-types\\'>float32</td>  <td class=\\'hl-values\\'>653.00,997.00,618.00,742.00,430.00,585.00,570.00,1...</td>\\n                            </tr></table></div></details><details><summary><div class=\\'title title-attr\\'>Embeddings & mappings</div><span class=\\'hl-dim\\'>.obsm</span><span class=\\'hl-size\\'>2 elements</span></summary><div><table><tr>\\n                                       <td class=\\'col-index\\'>cNMF</td>  <td class=\\'hl-types\\'>bool</td>  <td><span class=\\'hl-import\\'>numpy.</span>ndarray</td>  <td class=\\'hl-dims\\'></td>\\n                                   </tr>\\n<tr>\\n                                       <td class=\\'col-index\\'>rna</td>  <td class=\\'hl-types\\'>bool</td>  <td><span class=\\'hl-import\\'>numpy.</span>ndarray</td>  <td class=\\'hl-dims\\'></td>\\n                                   </tr></table></div></details><details><summary><div class=\\'title title-attr\\'>Distances</div><span class=\\'hl-dim\\'>.obsp</span><span class=\\'hl-size\\'>0 elements</span></summary><span class=\\'hl-empty\\'>No distances</span></details><div class=\\'block-mod\\'><div><details><summary class=\\'summary-mod\\'><div class=\\'title title-mod\\'>rna</div><span class=\\'hl-dim\\'>92284 &times 5451</span></summary><span>AnnData object <span class=\\'hl-dim\\'>92284 obs &times; 5451 var</span></span><br><div class=\\'title title-attr\\'>Matrix</div><span class=\\'hl-dim\\'>.X</span><div>\\n            <span class=\\'hl-types\\'>float32</span> <span>&nbsp;&nbsp;&nbsp;<span class=\\'hl-import\\'>scipy.sparse._csr.</span>csr_matrix</span>\\n         </table></div><details><summary><div class=\\'title title-attr\\'>Layers</div><span class=\\'hl-dim\\'>.layers</span><span class=\\'hl-size\\'>1 element</span></summary><div><table><tr>\\n                                       <td class=\\'col-index\\'>norm10k</td>  <td class=\\'hl-types\\'>float32</td>  <td><span class=\\'hl-import\\'>scipy.sparse._csr.</span>csr_matrix</td>  <td class=\\'hl-dims\\'></td>\\n                                   </tr></table></div></details><details><summary><div class=\\'title title-attr\\'>Metadata</div><span class=\\'hl-dim\\'>.obs</span><span class=\\'hl-size\\'>7 elements</span></summary><div><table><tr>\\n                                <td class=\\'col-index\\'>sample</td>  <td class=\\'hl-types\\'>category</td>  <td class=\\'hl-values\\'>D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0...</td>\\n                            </tr>\\n<tr>\\n                                <td class=\\'col-index\\'>species</td>  <td class=\\'hl-types\\'>category</td>  <td class=\\'hl-values\\'>hg38,hg38,hg38,hg38,hg38,hg38,hg38,hg38,hg38,hg38,...</td>\\n                            </tr>\\n<tr>\\n                                <td class=\\'col-index\\'>gene_count</td>  <td class=\\'hl-types\\'>int64</td>  <td class=\\'hl-values\\'>1903,2664,2147,2011,1441,1607,1385,3690,1348,1247,...</td>\\n                            </tr>\\n<tr>\\n                                <td class=\\'col-index\\'>tscp_count</td>  <td class=\\'hl-types\\'>int64</td>  <td class=\\'hl-values\\'>3010,4710,3492,3286,2027,2616,2256,7268,2093,1934,...</td>\\n                            </tr>\\n<tr>\\n                                <td class=\\'col-index\\'>mread_count</td>  <td class=\\'hl-types\\'>int64</td>  <td class=\\'hl-values\\'>4079,6401,4666,4432,2773,3499,3044,9975,2837,2533,...</td>\\n                            </tr>\\n<tr>\\n                                <td class=\\'col-index\\'>leiden</td>  <td class=\\'hl-types\\'>category</td>  <td class=\\'hl-values\\'>0,2,1,0,0,0,4,5,4,4,4,0,1,3,2,0,2,2,1,1,1,3,0,0,0,...</td>\\n                            </tr>\\n<tr>\\n                                <td class=\\'col-index\\'>n_counts</td>  <td class=\\'hl-types\\'>float32</td>  <td class=\\'hl-values\\'>653.00,997.00,618.00,742.00,430.00,585.00,570.00,1...</td>\\n                            </tr></table></div></details><details><summary><div class=\\'title title-attr\\'>Embeddings</div><span class=\\'hl-dim\\'>.obsm</span><span class=\\'hl-size\\'>2 elements</span></summary><div><table><tr>\\n                                       <td class=\\'col-index\\'>X_pca</td>  <td class=\\'hl-types\\'>float64</td>  <td><span class=\\'hl-import\\'>numpy.</span>ndarray</td>  <td class=\\'hl-dims\\'>50 columns</td>\\n                                   </tr>\\n<tr>\\n                                       <td class=\\'col-index\\'>X_umap</td>  <td class=\\'hl-types\\'>float32</td>  <td><span class=\\'hl-import\\'>numpy.</span>ndarray</td>  <td class=\\'hl-dims\\'>2 columns</td>\\n                                   </tr></table></div></details><details><summary><div class=\\'title title-attr\\'>Distances</div><span class=\\'hl-dim\\'>.obsp</span><span class=\\'hl-size\\'>0 elements</span></summary><span class=\\'hl-empty\\'>No distances</span></details><details><summary><div class=\\'title title-attr\\'>Miscellaneous</div><span class=\\'hl-dim\\'>.uns</span><span class=\\'hl-size\\'>1 elements</span></summary><div><table><tr>\\n                                <td class=\\'col-index\\'>sample_colors</td>  <td><span class=\\'hl-import\\'></span>list</td>  <td class=\\'hl-dims\\'>4 elements</td>  <td class=\\'hl-values\\'>#1f77b4,#ff7f0e,#2ca02c,#d62728...</td>\\n                            </tr></table></div></details></details></div></div><div class=\\'block-mod\\'><div><details><summary class=\\'summary-mod\\'><div class=\\'title title-mod\\'>cNMF</div><span class=\\'hl-dim\\'>92284 &times 60</span></summary><span>AnnData object <span class=\\'hl-dim\\'>92284 obs &times; 60 var</span></span><br><div class=\\'title title-attr\\'>Matrix</div><span class=\\'hl-dim\\'>.X</span><div>\\n            <span class=\\'hl-types\\'>float64</span> <span>&nbsp;&nbsp;&nbsp;<span class=\\'hl-import\\'>numpy.</span>ndarray</span>\\n         </table></div><details><summary><div class=\\'title title-attr\\'>Layers</div><span class=\\'hl-dim\\'>.layers</span><span class=\\'hl-size\\'>0 elements</span></summary><span class=\\'hl-empty\\'>No layers</span></details><details><summary><div class=\\'title title-attr\\'>Metadata</div><span class=\\'hl-dim\\'>.obs</span><span class=\\'hl-size\\'>7 elements</span></summary><div><table><tr>\\n                                <td class=\\'col-index\\'>sample</td>  <td class=\\'hl-types\\'>category</td>  <td class=\\'hl-values\\'>D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0...</td>\\n                            </tr>\\n<tr>\\n                                <td class=\\'col-index\\'>species</td>  <td class=\\'hl-types\\'>category</td>  <td class=\\'hl-values\\'>hg38,hg38,hg38,hg38,hg38,hg38,hg38,hg38,hg38,hg38,...</td>\\n                            </tr>\\n<tr>\\n                                <td class=\\'col-index\\'>gene_count</td>  <td class=\\'hl-types\\'>int64</td>  <td class=\\'hl-values\\'>1903,2664,2147,2011,1441,1607,1385,3690,1348,1247,...</td>\\n                            </tr>\\n<tr>\\n                                <td class=\\'col-index\\'>tscp_count</td>  <td class=\\'hl-types\\'>int64</td>  <td class=\\'hl-values\\'>3010,4710,3492,3286,2027,2616,2256,7268,2093,1934,...</td>\\n                            </tr>\\n<tr>\\n                                <td class=\\'col-index\\'>mread_count</td>  <td class=\\'hl-types\\'>int64</td>  <td class=\\'hl-values\\'>4079,6401,4666,4432,2773,3499,3044,9975,2837,2533,...</td>\\n                            </tr>\\n<tr>\\n                                <td class=\\'col-index\\'>leiden</td>  <td class=\\'hl-types\\'>category</td>  <td class=\\'hl-values\\'>0,2,1,0,0,0,4,5,4,4,4,0,1,3,2,0,2,2,1,1,1,3,0,0,0,...</td>\\n                            </tr>\\n<tr>\\n                                <td class=\\'col-index\\'>n_counts</td>  <td class=\\'hl-types\\'>float32</td>  <td class=\\'hl-values\\'>653.00,997.00,618.00,742.00,430.00,585.00,570.00,1...</td>\\n                            </tr></table></div></details><details><summary><div class=\\'title title-attr\\'>Embeddings</div><span class=\\'hl-dim\\'>.obsm</span><span class=\\'hl-size\\'>1 element</span></summary><div><table><tr>\\n                                       <td class=\\'col-index\\'>guide_assignment</td>  <td class=\\'hl-types\\'>int64</td>  <td><span class=\\'hl-import\\'>numpy.</span>ndarray</td>  <td class=\\'hl-dims\\'>1926 columns</td>\\n                                   </tr></table></div></details><details><summary><div class=\\'title title-attr\\'>Distances</div><span class=\\'hl-dim\\'>.obsp</span><span class=\\'hl-size\\'>0 elements</span></summary><span class=\\'hl-empty\\'>No distances</span></details><details><summary><div class=\\'title title-attr\\'>Miscellaneous</div><span class=\\'hl-dim\\'>.uns</span><span class=\\'hl-size\\'>3 elements</span></summary><div><table><tr>\\n                                <td class=\\'col-index\\'>guide_names</td>  <td><span class=\\'hl-import\\'>numpy.</span>ndarray</td>  <td class=\\'hl-dims\\'>1926 elements</td>  <td class=\\'hl-values\\'>ACAA1_-_38178575.23-P1P2,ACAA1_+_38178488.23-P1P2,...</td>\\n                            </tr>\\n<tr>\\n                                <td class=\\'col-index\\'>guide_targets</td>  <td><span class=\\'hl-import\\'>numpy.</span>ndarray</td>  <td class=\\'hl-dims\\'>1926 elements</td>  <td class=\\'hl-values\\'>ACAA1,ACAA1,ACAA1,ACAA1,ACAA1,ACAA1,ACTL7B,ACTL7B,...</td>\\n                            </tr>\\n<tr>\\n                                <td class=\\'col-index\\'>var_names</td>  <td><span class=\\'hl-import\\'>numpy.</span>ndarray</td>  <td class=\\'hl-dims\\'>5451 elements</td>  <td class=\\'hl-values\\'>SEMA3F,CFTR,CYP51A1,HECW1,KLHL13,CASP10,CFLAR,TFPI...</td>\\n                            </tr></table></div></details></details></div></div><br/>'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'<style>\\n.hl-dim, .hl-size, .hl-values, .hl-types, .hl-dims {\\n  color: #777777;\\n}\\n.hl-dim::before, .hl-size::before {\\n  content: \"\\\\00a0\\\\00a0\\\\00a0\";\\n}\\n.hl-values {\\n  font-family: monospace;\\n}\\n.hl-file {\\n  background-color: #EEEEEE;\\n  border-radius: .5rem;\\n  padding: .2rem .4rem;\\n  color: #555555;\\n}\\n.hl-empty {\\n  color: #999999;\\n}\\n.hl-import {\\n  color: #777777;\\n}\\n.block-mod {\\n  display: block;\\n  margin: 0 2rem;\\n}\\n.title {\\n  display: inline-block;\\n  font-weight: 600;\\n  color: #555555;\\n}\\n.title-mod {\\n  font-size: 1.2rem;\\n  color: #04b374;\\n  padding: 0 .5rem;\\n}\\n.title-attr {\\n  font-size: 1.0rem;\\n  padding-top: .2rem;\\n}\\nsummary {\\n  cursor: pointer;\\n  list-style: none;\\n}\\nsummary::-webkit-details-marker {\\n  display: none;\\n}\\ndetails > summary::before {\\n  content: \\'⊕\\';\\n}\\ndetails[open] > summary::before {\\n  content: \\'⊖\\';\\n}\\ntable tr {\\n  background-color: transparent !important;\\n}\\ntable tr:hover {\\n  background-color: #04b37433 !important;\\n}\\n.col-index {\\n  text-align: left !important;\\n}\\n.summary-mod {\\n  margin-left: -2rem;\\n}\\n.summary-mod:hover {\\n  background-color: #04b37411;\\n}\\n.summary-mod::before {\\n  color: #04b374;\\n  content: \\'●\\';\\n}\\ndetails[open] > .summary-mod::before {\\n  content: \\'○\\';\\n}\\n</style><span>MuData object <span class=\\'hl-dim\\'>92284 obs &times; 5511 var in 2 modalities</span></span><br><details><summary><div class=\\'title title-attr\\'>Metadata</div><span class=\\'hl-dim\\'>.obs</span><span class=\\'hl-size\\'>14 elements</span></summary><div><table><tr>\\n                                <td class=\\'col-index\\'>rna:sample</td>  <td class=\\'hl-types\\'>category</td>  <td class=\\'hl-values\\'>D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0...</td>\\n                            </tr>\\n<tr>\\n                                <td class=\\'col-index\\'>rna:species</td>  <td class=\\'hl-types\\'>category</td>  <td class=\\'hl-values\\'>hg38,hg38,hg38,hg38,hg38,hg38,hg38,hg38,hg38,hg38,...</td>\\n                            </tr>\\n<tr>\\n                                <td class=\\'col-index\\'>rna:gene_count</td>  <td class=\\'hl-types\\'>int64</td>  <td class=\\'hl-values\\'>1903,2664,2147,2011,1441,1607,1385,3690,1348,1247,...</td>\\n                            </tr>\\n<tr>\\n                                <td class=\\'col-index\\'>rna:tscp_count</td>  <td class=\\'hl-types\\'>int64</td>  <td class=\\'hl-values\\'>3010,4710,3492,3286,2027,2616,2256,7268,2093,1934,...</td>\\n                            </tr>\\n<tr>\\n                                <td class=\\'col-index\\'>rna:mread_count</td>  <td class=\\'hl-types\\'>int64</td>  <td class=\\'hl-values\\'>4079,6401,4666,4432,2773,3499,3044,9975,2837,2533,...</td>\\n                            </tr>\\n<tr>\\n                                <td class=\\'col-index\\'>rna:leiden</td>  <td class=\\'hl-types\\'>category</td>  <td class=\\'hl-values\\'>0,2,1,0,0,0,4,5,4,4,4,0,1,3,2,0,2,2,1,1,1,3,0,0,0,...</td>\\n                            </tr>\\n<tr>\\n                                <td class=\\'col-index\\'>rna:n_counts</td>  <td class=\\'hl-types\\'>float32</td>  <td class=\\'hl-values\\'>653.00,997.00,618.00,742.00,430.00,585.00,570.00,1...</td>\\n                            </tr>\\n<tr>\\n                                <td class=\\'col-index\\'>cNMF:sample</td>  <td class=\\'hl-types\\'>category</td>  <td class=\\'hl-values\\'>D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0...</td>\\n                            </tr>\\n<tr>\\n                                <td class=\\'col-index\\'>cNMF:species</td>  <td class=\\'hl-types\\'>category</td>  <td class=\\'hl-values\\'>hg38,hg38,hg38,hg38,hg38,hg38,hg38,hg38,hg38,hg38,...</td>\\n                            </tr>\\n<tr>\\n                                <td class=\\'col-index\\'>cNMF:gene_count</td>  <td class=\\'hl-types\\'>int64</td>  <td class=\\'hl-values\\'>1903,2664,2147,2011,1441,1607,1385,3690,1348,1247,...</td>\\n                            </tr>\\n<tr>\\n                                <td class=\\'col-index\\'>cNMF:tscp_count</td>  <td class=\\'hl-types\\'>int64</td>  <td class=\\'hl-values\\'>3010,4710,3492,3286,2027,2616,2256,7268,2093,1934,...</td>\\n                            </tr>\\n<tr>\\n                                <td class=\\'col-index\\'>cNMF:mread_count</td>  <td class=\\'hl-types\\'>int64</td>  <td class=\\'hl-values\\'>4079,6401,4666,4432,2773,3499,3044,9975,2837,2533,...</td>\\n                            </tr>\\n<tr>\\n                                <td class=\\'col-index\\'>cNMF:leiden</td>  <td class=\\'hl-types\\'>category</td>  <td class=\\'hl-values\\'>0,2,1,0,0,0,4,5,4,4,4,0,1,3,2,0,2,2,1,1,1,3,0,0,0,...</td>\\n                            </tr>\\n<tr>\\n                                <td class=\\'col-index\\'>cNMF:n_counts</td>  <td class=\\'hl-types\\'>float32</td>  <td class=\\'hl-values\\'>653.00,997.00,618.00,742.00,430.00,585.00,570.00,1...</td>\\n                            </tr></table></div></details><details><summary><div class=\\'title title-attr\\'>Embeddings & mappings</div><span class=\\'hl-dim\\'>.obsm</span><span class=\\'hl-size\\'>2 elements</span></summary><div><table><tr>\\n                                       <td class=\\'col-index\\'>cNMF</td>  <td class=\\'hl-types\\'>bool</td>  <td><span class=\\'hl-import\\'>numpy.</span>ndarray</td>  <td class=\\'hl-dims\\'></td>\\n                                   </tr>\\n<tr>\\n                                       <td class=\\'col-index\\'>rna</td>  <td class=\\'hl-types\\'>bool</td>  <td><span class=\\'hl-import\\'>numpy.</span>ndarray</td>  <td class=\\'hl-dims\\'></td>\\n                                   </tr></table></div></details><details><summary><div class=\\'title title-attr\\'>Distances</div><span class=\\'hl-dim\\'>.obsp</span><span class=\\'hl-size\\'>0 elements</span></summary><span class=\\'hl-empty\\'>No distances</span></details><div class=\\'block-mod\\'><div><details><summary class=\\'summary-mod\\'><div class=\\'title title-mod\\'>rna</div><span class=\\'hl-dim\\'>92284 &times 5451</span></summary><span>AnnData object <span class=\\'hl-dim\\'>92284 obs &times; 5451 var</span></span><br><div class=\\'title title-attr\\'>Matrix</div><span class=\\'hl-dim\\'>.X</span><div>\\n            <span class=\\'hl-types\\'>float32</span> <span>&nbsp;&nbsp;&nbsp;<span class=\\'hl-import\\'>scipy.sparse._csr.</span>csr_matrix</span>\\n         </table></div><details><summary><div class=\\'title title-attr\\'>Layers</div><span class=\\'hl-dim\\'>.layers</span><span class=\\'hl-size\\'>1 element</span></summary><div><table><tr>\\n                                       <td class=\\'col-index\\'>norm10k</td>  <td class=\\'hl-types\\'>float32</td>  <td><span class=\\'hl-import\\'>scipy.sparse._csr.</span>csr_matrix</td>  <td class=\\'hl-dims\\'></td>\\n                                   </tr></table></div></details><details><summary><div class=\\'title title-attr\\'>Metadata</div><span class=\\'hl-dim\\'>.obs</span><span class=\\'hl-size\\'>7 elements</span></summary><div><table><tr>\\n                                <td class=\\'col-index\\'>sample</td>  <td class=\\'hl-types\\'>category</td>  <td class=\\'hl-values\\'>D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0...</td>\\n                            </tr>\\n<tr>\\n                                <td class=\\'col-index\\'>species</td>  <td class=\\'hl-types\\'>category</td>  <td class=\\'hl-values\\'>hg38,hg38,hg38,hg38,hg38,hg38,hg38,hg38,hg38,hg38,...</td>\\n                            </tr>\\n<tr>\\n                                <td class=\\'col-index\\'>gene_count</td>  <td class=\\'hl-types\\'>int64</td>  <td class=\\'hl-values\\'>1903,2664,2147,2011,1441,1607,1385,3690,1348,1247,...</td>\\n                            </tr>\\n<tr>\\n                                <td class=\\'col-index\\'>tscp_count</td>  <td class=\\'hl-types\\'>int64</td>  <td class=\\'hl-values\\'>3010,4710,3492,3286,2027,2616,2256,7268,2093,1934,...</td>\\n                            </tr>\\n<tr>\\n                                <td class=\\'col-index\\'>mread_count</td>  <td class=\\'hl-types\\'>int64</td>  <td class=\\'hl-values\\'>4079,6401,4666,4432,2773,3499,3044,9975,2837,2533,...</td>\\n                            </tr>\\n<tr>\\n                                <td class=\\'col-index\\'>leiden</td>  <td class=\\'hl-types\\'>category</td>  <td class=\\'hl-values\\'>0,2,1,0,0,0,4,5,4,4,4,0,1,3,2,0,2,2,1,1,1,3,0,0,0,...</td>\\n                            </tr>\\n<tr>\\n                                <td class=\\'col-index\\'>n_counts</td>  <td class=\\'hl-types\\'>float32</td>  <td class=\\'hl-values\\'>653.00,997.00,618.00,742.00,430.00,585.00,570.00,1...</td>\\n                            </tr></table></div></details><details><summary><div class=\\'title title-attr\\'>Embeddings</div><span class=\\'hl-dim\\'>.obsm</span><span class=\\'hl-size\\'>2 elements</span></summary><div><table><tr>\\n                                       <td class=\\'col-index\\'>X_pca</td>  <td class=\\'hl-types\\'>float64</td>  <td><span class=\\'hl-import\\'>numpy.</span>ndarray</td>  <td class=\\'hl-dims\\'>50 columns</td>\\n                                   </tr>\\n<tr>\\n                                       <td class=\\'col-index\\'>X_umap</td>  <td class=\\'hl-types\\'>float32</td>  <td><span class=\\'hl-import\\'>numpy.</span>ndarray</td>  <td class=\\'hl-dims\\'>2 columns</td>\\n                                   </tr></table></div></details><details><summary><div class=\\'title title-attr\\'>Distances</div><span class=\\'hl-dim\\'>.obsp</span><span class=\\'hl-size\\'>0 elements</span></summary><span class=\\'hl-empty\\'>No distances</span></details><details><summary><div class=\\'title title-attr\\'>Miscellaneous</div><span class=\\'hl-dim\\'>.uns</span><span class=\\'hl-size\\'>1 elements</span></summary><div><table><tr>\\n                                <td class=\\'col-index\\'>sample_colors</td>  <td><span class=\\'hl-import\\'></span>list</td>  <td class=\\'hl-dims\\'>4 elements</td>  <td class=\\'hl-values\\'>#1f77b4,#ff7f0e,#2ca02c,#d62728...</td>\\n                            </tr></table></div></details></details></div></div><div class=\\'block-mod\\'><div><details><summary class=\\'summary-mod\\'><div class=\\'title title-mod\\'>cNMF</div><span class=\\'hl-dim\\'>92284 &times 60</span></summary><span>AnnData object <span class=\\'hl-dim\\'>92284 obs &times; 60 var</span></span><br><div class=\\'title title-attr\\'>Matrix</div><span class=\\'hl-dim\\'>.X</span><div>\\n            <span class=\\'hl-types\\'>float64</span> <span>&nbsp;&nbsp;&nbsp;<span class=\\'hl-import\\'>numpy.</span>ndarray</span>\\n         </table></div><details><summary><div class=\\'title title-attr\\'>Layers</div><span class=\\'hl-dim\\'>.layers</span><span class=\\'hl-size\\'>0 elements</span></summary><span class=\\'hl-empty\\'>No layers</span></details><details><summary><div class=\\'title title-attr\\'>Metadata</div><span class=\\'hl-dim\\'>.obs</span><span class=\\'hl-size\\'>7 elements</span></summary><div><table><tr>\\n                                <td class=\\'col-index\\'>sample</td>  <td class=\\'hl-types\\'>category</td>  <td class=\\'hl-values\\'>D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0...</td>\\n                            </tr>\\n<tr>\\n                                <td class=\\'col-index\\'>species</td>  <td class=\\'hl-types\\'>category</td>  <td class=\\'hl-values\\'>hg38,hg38,hg38,hg38,hg38,hg38,hg38,hg38,hg38,hg38,...</td>\\n                            </tr>\\n<tr>\\n                                <td class=\\'col-index\\'>gene_count</td>  <td class=\\'hl-types\\'>int64</td>  <td class=\\'hl-values\\'>1903,2664,2147,2011,1441,1607,1385,3690,1348,1247,...</td>\\n                            </tr>\\n<tr>\\n                                <td class=\\'col-index\\'>tscp_count</td>  <td class=\\'hl-types\\'>int64</td>  <td class=\\'hl-values\\'>3010,4710,3492,3286,2027,2616,2256,7268,2093,1934,...</td>\\n                            </tr>\\n<tr>\\n                                <td class=\\'col-index\\'>mread_count</td>  <td class=\\'hl-types\\'>int64</td>  <td class=\\'hl-values\\'>4079,6401,4666,4432,2773,3499,3044,9975,2837,2533,...</td>\\n                            </tr>\\n<tr>\\n                                <td class=\\'col-index\\'>leiden</td>  <td class=\\'hl-types\\'>category</td>  <td class=\\'hl-values\\'>0,2,1,0,0,0,4,5,4,4,4,0,1,3,2,0,2,2,1,1,1,3,0,0,0,...</td>\\n                            </tr>\\n<tr>\\n                                <td class=\\'col-index\\'>n_counts</td>  <td class=\\'hl-types\\'>float32</td>  <td class=\\'hl-values\\'>653.00,997.00,618.00,742.00,430.00,585.00,570.00,1...</td>\\n                            </tr></table></div></details><details><summary><div class=\\'title title-attr\\'>Embeddings</div><span class=\\'hl-dim\\'>.obsm</span><span class=\\'hl-size\\'>1 element</span></summary><div><table><tr>\\n                                       <td class=\\'col-index\\'>guide_assignment</td>  <td class=\\'hl-types\\'>int64</td>  <td><span class=\\'hl-import\\'>numpy.</span>ndarray</td>  <td class=\\'hl-dims\\'>1926 columns</td>\\n                                   </tr></table></div></details><details><summary><div class=\\'title title-attr\\'>Distances</div><span class=\\'hl-dim\\'>.obsp</span><span class=\\'hl-size\\'>0 elements</span></summary><span class=\\'hl-empty\\'>No distances</span></details><details><summary><div class=\\'title title-attr\\'>Miscellaneous</div><span class=\\'hl-dim\\'>.uns</span><span class=\\'hl-size\\'>3 elements</span></summary><div><table><tr>\\n                                <td class=\\'col-index\\'>guide_names</td>  <td><span class=\\'hl-import\\'>numpy.</span>ndarray</td>  <td class=\\'hl-dims\\'>1926 elements</td>  <td class=\\'hl-values\\'>ACAA1_-_38178575.23-P1P2,ACAA1_+_38178488.23-P1P2,...</td>\\n                            </tr>\\n<tr>\\n                                <td class=\\'col-index\\'>guide_targets</td>  <td><span class=\\'hl-import\\'>numpy.</span>ndarray</td>  <td class=\\'hl-dims\\'>1926 elements</td>  <td class=\\'hl-values\\'>ACAA1,ACAA1,ACAA1,ACAA1,ACAA1,ACAA1,ACTL7B,ACTL7B,...</td>\\n                            </tr>\\n<tr>\\n                                <td class=\\'col-index\\'>var_names</td>  <td><span class=\\'hl-import\\'>numpy.</span>ndarray</td>  <td class=\\'hl-dims\\'>5451 elements</td>  <td class=\\'hl-values\\'>SEMA3F,CFTR,CYP51A1,HECW1,KLHL13,CASP10,CFLAR,TFPI...</td>\\n                            </tr></table></div></details></details></div></div><br/>'"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# write html to file\n",
    "with mudata.set_options(display_style=\"html\", display_html_expand=0b000):\n",
    "    html_rep = mdata._repr_html_(expand=0b000)\n",
    "with open(\"/cellar/users/aklie/opt/gene_program_evaluation/dashapp/mdata.html\", \"w\") as f:\n",
    "    f.write(html_rep)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".hl-dim, .hl-size, .hl-values, .hl-types, .hl-dims {\n",
       "  color: #777777;\n",
       "}\n",
       ".hl-dim::before, .hl-size::before {\n",
       "  content: \"\\00a0\\00a0\\00a0\";\n",
       "}\n",
       ".hl-values {\n",
       "  font-family: monospace;\n",
       "}\n",
       ".hl-file {\n",
       "  background-color: #EEEEEE;\n",
       "  border-radius: .5rem;\n",
       "  padding: .2rem .4rem;\n",
       "  color: #555555;\n",
       "}\n",
       ".hl-empty {\n",
       "  color: #999999;\n",
       "}\n",
       ".hl-import {\n",
       "  color: #777777;\n",
       "}\n",
       ".block-mod {\n",
       "  display: block;\n",
       "  margin: 0 2rem;\n",
       "}\n",
       ".title {\n",
       "  display: inline-block;\n",
       "  font-weight: 600;\n",
       "  color: #555555;\n",
       "}\n",
       ".title-mod {\n",
       "  font-size: 1.2rem;\n",
       "  color: #04b374;\n",
       "  padding: 0 .5rem;\n",
       "}\n",
       ".title-attr {\n",
       "  font-size: 1.0rem;\n",
       "  padding-top: .2rem;\n",
       "}\n",
       "summary {\n",
       "  cursor: pointer;\n",
       "  list-style: none;\n",
       "}\n",
       "summary::-webkit-details-marker {\n",
       "  display: none;\n",
       "}\n",
       "details > summary::before {\n",
       "  content: '⊕';\n",
       "}\n",
       "details[open] > summary::before {\n",
       "  content: '⊖';\n",
       "}\n",
       "table tr {\n",
       "  background-color: transparent !important;\n",
       "}\n",
       "table tr:hover {\n",
       "  background-color: #04b37433 !important;\n",
       "}\n",
       ".col-index {\n",
       "  text-align: left !important;\n",
       "}\n",
       ".summary-mod {\n",
       "  margin-left: -2rem;\n",
       "}\n",
       ".summary-mod:hover {\n",
       "  background-color: #04b37411;\n",
       "}\n",
       ".summary-mod::before {\n",
       "  color: #04b374;\n",
       "  content: '●';\n",
       "}\n",
       "details[open] > .summary-mod::before {\n",
       "  content: '○';\n",
       "}\n",
       "</style><span>MuData object <span class='hl-dim'>92284 obs &times; 5511 var in 2 modalities</span></span><br><details><summary><div class='title title-attr'>Metadata</div><span class='hl-dim'>.obs</span><span class='hl-size'>14 elements</span></summary><div><table><tr>\n",
       "                                <td class='col-index'>rna:sample</td>  <td class='hl-types'>category</td>  <td class='hl-values'>D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>rna:species</td>  <td class='hl-types'>category</td>  <td class='hl-values'>hg38,hg38,hg38,hg38,hg38,hg38,hg38,hg38,hg38,hg38,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>rna:gene_count</td>  <td class='hl-types'>int64</td>  <td class='hl-values'>1903,2664,2147,2011,1441,1607,1385,3690,1348,1247,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>rna:tscp_count</td>  <td class='hl-types'>int64</td>  <td class='hl-values'>3010,4710,3492,3286,2027,2616,2256,7268,2093,1934,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>rna:mread_count</td>  <td class='hl-types'>int64</td>  <td class='hl-values'>4079,6401,4666,4432,2773,3499,3044,9975,2837,2533,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>rna:leiden</td>  <td class='hl-types'>category</td>  <td class='hl-values'>0,2,1,0,0,0,4,5,4,4,4,0,1,3,2,0,2,2,1,1,1,3,0,0,0,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>rna:n_counts</td>  <td class='hl-types'>float32</td>  <td class='hl-values'>653.00,997.00,618.00,742.00,430.00,585.00,570.00,1...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>cNMF:sample</td>  <td class='hl-types'>category</td>  <td class='hl-values'>D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>cNMF:species</td>  <td class='hl-types'>category</td>  <td class='hl-values'>hg38,hg38,hg38,hg38,hg38,hg38,hg38,hg38,hg38,hg38,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>cNMF:gene_count</td>  <td class='hl-types'>int64</td>  <td class='hl-values'>1903,2664,2147,2011,1441,1607,1385,3690,1348,1247,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>cNMF:tscp_count</td>  <td class='hl-types'>int64</td>  <td class='hl-values'>3010,4710,3492,3286,2027,2616,2256,7268,2093,1934,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>cNMF:mread_count</td>  <td class='hl-types'>int64</td>  <td class='hl-values'>4079,6401,4666,4432,2773,3499,3044,9975,2837,2533,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>cNMF:leiden</td>  <td class='hl-types'>category</td>  <td class='hl-values'>0,2,1,0,0,0,4,5,4,4,4,0,1,3,2,0,2,2,1,1,1,3,0,0,0,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>cNMF:n_counts</td>  <td class='hl-types'>float32</td>  <td class='hl-values'>653.00,997.00,618.00,742.00,430.00,585.00,570.00,1...</td>\n",
       "                            </tr></table></div></details><details><summary><div class='title title-attr'>Embeddings & mappings</div><span class='hl-dim'>.obsm</span><span class='hl-size'>2 elements</span></summary><div><table><tr>\n",
       "                                       <td class='col-index'>cNMF</td>  <td class='hl-types'>bool</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'></td>\n",
       "                                   </tr>\n",
       "<tr>\n",
       "                                       <td class='col-index'>rna</td>  <td class='hl-types'>bool</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'></td>\n",
       "                                   </tr></table></div></details><details><summary><div class='title title-attr'>Distances</div><span class='hl-dim'>.obsp</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No distances</span></details><div class='block-mod'><div><details><summary class='summary-mod'><div class='title title-mod'>rna</div><span class='hl-dim'>92284 &times 5451</span></summary><span>AnnData object <span class='hl-dim'>92284 obs &times; 5451 var</span></span><br><div class='title title-attr'>Matrix</div><span class='hl-dim'>.X</span><div>\n",
       "            <span class='hl-types'>float32</span> <span>&nbsp;&nbsp;&nbsp;<span class='hl-import'>scipy.sparse._csr.</span>csr_matrix</span>\n",
       "         </table></div><details><summary><div class='title title-attr'>Layers</div><span class='hl-dim'>.layers</span><span class='hl-size'>1 element</span></summary><div><table><tr>\n",
       "                                       <td class='col-index'>norm10k</td>  <td class='hl-types'>float32</td>  <td><span class='hl-import'>scipy.sparse._csr.</span>csr_matrix</td>  <td class='hl-dims'></td>\n",
       "                                   </tr></table></div></details><details><summary><div class='title title-attr'>Metadata</div><span class='hl-dim'>.obs</span><span class='hl-size'>7 elements</span></summary><div><table><tr>\n",
       "                                <td class='col-index'>sample</td>  <td class='hl-types'>category</td>  <td class='hl-values'>D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>species</td>  <td class='hl-types'>category</td>  <td class='hl-values'>hg38,hg38,hg38,hg38,hg38,hg38,hg38,hg38,hg38,hg38,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>gene_count</td>  <td class='hl-types'>int64</td>  <td class='hl-values'>1903,2664,2147,2011,1441,1607,1385,3690,1348,1247,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>tscp_count</td>  <td class='hl-types'>int64</td>  <td class='hl-values'>3010,4710,3492,3286,2027,2616,2256,7268,2093,1934,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>mread_count</td>  <td class='hl-types'>int64</td>  <td class='hl-values'>4079,6401,4666,4432,2773,3499,3044,9975,2837,2533,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>leiden</td>  <td class='hl-types'>category</td>  <td class='hl-values'>0,2,1,0,0,0,4,5,4,4,4,0,1,3,2,0,2,2,1,1,1,3,0,0,0,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>n_counts</td>  <td class='hl-types'>float32</td>  <td class='hl-values'>653.00,997.00,618.00,742.00,430.00,585.00,570.00,1...</td>\n",
       "                            </tr></table></div></details><details><summary><div class='title title-attr'>Embeddings</div><span class='hl-dim'>.obsm</span><span class='hl-size'>2 elements</span></summary><div><table><tr>\n",
       "                                       <td class='col-index'>X_pca</td>  <td class='hl-types'>float64</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>50 columns</td>\n",
       "                                   </tr>\n",
       "<tr>\n",
       "                                       <td class='col-index'>X_umap</td>  <td class='hl-types'>float32</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>2 columns</td>\n",
       "                                   </tr></table></div></details><details><summary><div class='title title-attr'>Distances</div><span class='hl-dim'>.obsp</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No distances</span></details><details><summary><div class='title title-attr'>Miscellaneous</div><span class='hl-dim'>.uns</span><span class='hl-size'>1 elements</span></summary><div><table><tr>\n",
       "                                <td class='col-index'>sample_colors</td>  <td><span class='hl-import'></span>list</td>  <td class='hl-dims'>4 elements</td>  <td class='hl-values'>#1f77b4,#ff7f0e,#2ca02c,#d62728...</td>\n",
       "                            </tr></table></div></details></details></div></div><div class='block-mod'><div><details><summary class='summary-mod'><div class='title title-mod'>cNMF</div><span class='hl-dim'>92284 &times 60</span></summary><span>AnnData object <span class='hl-dim'>92284 obs &times; 60 var</span></span><br><div class='title title-attr'>Matrix</div><span class='hl-dim'>.X</span><div>\n",
       "            <span class='hl-types'>float64</span> <span>&nbsp;&nbsp;&nbsp;<span class='hl-import'>numpy.</span>ndarray</span>\n",
       "         </table></div><details><summary><div class='title title-attr'>Layers</div><span class='hl-dim'>.layers</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No layers</span></details><details><summary><div class='title title-attr'>Metadata</div><span class='hl-dim'>.obs</span><span class='hl-size'>7 elements</span></summary><div><table><tr>\n",
       "                                <td class='col-index'>sample</td>  <td class='hl-types'>category</td>  <td class='hl-values'>D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>species</td>  <td class='hl-types'>category</td>  <td class='hl-values'>hg38,hg38,hg38,hg38,hg38,hg38,hg38,hg38,hg38,hg38,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>gene_count</td>  <td class='hl-types'>int64</td>  <td class='hl-values'>1903,2664,2147,2011,1441,1607,1385,3690,1348,1247,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>tscp_count</td>  <td class='hl-types'>int64</td>  <td class='hl-values'>3010,4710,3492,3286,2027,2616,2256,7268,2093,1934,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>mread_count</td>  <td class='hl-types'>int64</td>  <td class='hl-values'>4079,6401,4666,4432,2773,3499,3044,9975,2837,2533,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>leiden</td>  <td class='hl-types'>category</td>  <td class='hl-values'>0,2,1,0,0,0,4,5,4,4,4,0,1,3,2,0,2,2,1,1,1,3,0,0,0,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>n_counts</td>  <td class='hl-types'>float32</td>  <td class='hl-values'>653.00,997.00,618.00,742.00,430.00,585.00,570.00,1...</td>\n",
       "                            </tr></table></div></details><details><summary><div class='title title-attr'>Embeddings</div><span class='hl-dim'>.obsm</span><span class='hl-size'>1 element</span></summary><div><table><tr>\n",
       "                                       <td class='col-index'>guide_assignment</td>  <td class='hl-types'>int64</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>1926 columns</td>\n",
       "                                   </tr></table></div></details><details><summary><div class='title title-attr'>Distances</div><span class='hl-dim'>.obsp</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No distances</span></details><details><summary><div class='title title-attr'>Miscellaneous</div><span class='hl-dim'>.uns</span><span class='hl-size'>3 elements</span></summary><div><table><tr>\n",
       "                                <td class='col-index'>guide_names</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>1926 elements</td>  <td class='hl-values'>ACAA1_-_38178575.23-P1P2,ACAA1_+_38178488.23-P1P2,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>guide_targets</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>1926 elements</td>  <td class='hl-values'>ACAA1,ACAA1,ACAA1,ACAA1,ACAA1,ACAA1,ACTL7B,ACTL7B,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>var_names</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>5451 elements</td>  <td class='hl-values'>SEMA3F,CFTR,CYP51A1,HECW1,KLHL13,CASP10,CFLAR,TFPI...</td>\n",
       "                            </tr></table></div></details></details></div></div><br/>"
      ],
      "text/plain": [
       "MuData object with n_obs × n_vars = 92284 × 5511\n",
       "  2 modalities\n",
       "    rna:\t92284 x 5451\n",
       "      obs:\t'sample', 'species', 'gene_count', 'tscp_count', 'mread_count', 'leiden', 'n_counts'\n",
       "      uns:\t'sample_colors'\n",
       "      obsm:\t'X_pca', 'X_umap'\n",
       "      layers:\t'norm10k'\n",
       "    cNMF:\t92284 x 60\n",
       "      obs:\t'sample', 'species', 'gene_count', 'tscp_count', 'mread_count', 'leiden', 'n_counts'\n",
       "      uns:\t'guide_names', 'guide_targets', 'var_names'\n",
       "      obsm:\t'guide_assignment'\n",
       "      varm:\t'loadings'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "with mudata.set_options(display_style=\"html\", display_html_expand=0b000):\n",
    "    display(mdata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".hl-dim, .hl-size, .hl-values, .hl-types, .hl-dims {\n",
       "  color: #777777;\n",
       "}\n",
       ".hl-dim::before, .hl-size::before {\n",
       "  content: \"\\00a0\\00a0\\00a0\";\n",
       "}\n",
       ".hl-values {\n",
       "  font-family: monospace;\n",
       "}\n",
       ".hl-file {\n",
       "  background-color: #EEEEEE;\n",
       "  border-radius: .5rem;\n",
       "  padding: .2rem .4rem;\n",
       "  color: #555555;\n",
       "}\n",
       ".hl-empty {\n",
       "  color: #999999;\n",
       "}\n",
       ".hl-import {\n",
       "  color: #777777;\n",
       "}\n",
       ".block-mod {\n",
       "  display: block;\n",
       "  margin: 0 2rem;\n",
       "}\n",
       ".title {\n",
       "  display: inline-block;\n",
       "  font-weight: 600;\n",
       "  color: #555555;\n",
       "}\n",
       ".title-mod {\n",
       "  font-size: 1.2rem;\n",
       "  color: #04b374;\n",
       "  padding: 0 .5rem;\n",
       "}\n",
       ".title-attr {\n",
       "  font-size: 1.0rem;\n",
       "  padding-top: .2rem;\n",
       "}\n",
       "summary {\n",
       "  cursor: pointer;\n",
       "  list-style: none;\n",
       "}\n",
       "summary::-webkit-details-marker {\n",
       "  display: none;\n",
       "}\n",
       "details > summary::before {\n",
       "  content: '⊕';\n",
       "}\n",
       "details[open] > summary::before {\n",
       "  content: '⊖';\n",
       "}\n",
       "table tr {\n",
       "  background-color: transparent !important;\n",
       "}\n",
       "table tr:hover {\n",
       "  background-color: #04b37433 !important;\n",
       "}\n",
       ".col-index {\n",
       "  text-align: left !important;\n",
       "}\n",
       ".summary-mod {\n",
       "  margin-left: -2rem;\n",
       "}\n",
       ".summary-mod:hover {\n",
       "  background-color: #04b37411;\n",
       "}\n",
       ".summary-mod::before {\n",
       "  color: #04b374;\n",
       "  content: '●';\n",
       "}\n",
       "details[open] > .summary-mod::before {\n",
       "  content: '○';\n",
       "}\n",
       "</style><span>MuData object <span class='hl-dim'>92284 obs &times; 5511 var in 2 modalities</span></span><br><details><summary><div class='title title-attr'>Metadata</div><span class='hl-dim'>.obs</span><span class='hl-size'>14 elements</span></summary><div><table><tr>\n",
       "                                <td class='col-index'>rna:sample</td>  <td class='hl-types'>category</td>  <td class='hl-values'>D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>rna:species</td>  <td class='hl-types'>category</td>  <td class='hl-values'>hg38,hg38,hg38,hg38,hg38,hg38,hg38,hg38,hg38,hg38,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>rna:gene_count</td>  <td class='hl-types'>int64</td>  <td class='hl-values'>1903,2664,2147,2011,1441,1607,1385,3690,1348,1247,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>rna:tscp_count</td>  <td class='hl-types'>int64</td>  <td class='hl-values'>3010,4710,3492,3286,2027,2616,2256,7268,2093,1934,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>rna:mread_count</td>  <td class='hl-types'>int64</td>  <td class='hl-values'>4079,6401,4666,4432,2773,3499,3044,9975,2837,2533,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>rna:leiden</td>  <td class='hl-types'>category</td>  <td class='hl-values'>0,2,1,0,0,0,4,5,4,4,4,0,1,3,2,0,2,2,1,1,1,3,0,0,0,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>rna:n_counts</td>  <td class='hl-types'>float32</td>  <td class='hl-values'>653.00,997.00,618.00,742.00,430.00,585.00,570.00,1...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>cNMF:sample</td>  <td class='hl-types'>category</td>  <td class='hl-values'>D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>cNMF:species</td>  <td class='hl-types'>category</td>  <td class='hl-values'>hg38,hg38,hg38,hg38,hg38,hg38,hg38,hg38,hg38,hg38,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>cNMF:gene_count</td>  <td class='hl-types'>int64</td>  <td class='hl-values'>1903,2664,2147,2011,1441,1607,1385,3690,1348,1247,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>cNMF:tscp_count</td>  <td class='hl-types'>int64</td>  <td class='hl-values'>3010,4710,3492,3286,2027,2616,2256,7268,2093,1934,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>cNMF:mread_count</td>  <td class='hl-types'>int64</td>  <td class='hl-values'>4079,6401,4666,4432,2773,3499,3044,9975,2837,2533,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>cNMF:leiden</td>  <td class='hl-types'>category</td>  <td class='hl-values'>0,2,1,0,0,0,4,5,4,4,4,0,1,3,2,0,2,2,1,1,1,3,0,0,0,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>cNMF:n_counts</td>  <td class='hl-types'>float32</td>  <td class='hl-values'>653.00,997.00,618.00,742.00,430.00,585.00,570.00,1...</td>\n",
       "                            </tr></table></div></details><details><summary><div class='title title-attr'>Embeddings & mappings</div><span class='hl-dim'>.obsm</span><span class='hl-size'>2 elements</span></summary><div><table><tr>\n",
       "                                       <td class='col-index'>cNMF</td>  <td class='hl-types'>bool</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'></td>\n",
       "                                   </tr>\n",
       "<tr>\n",
       "                                       <td class='col-index'>rna</td>  <td class='hl-types'>bool</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'></td>\n",
       "                                   </tr></table></div></details><details><summary><div class='title title-attr'>Distances</div><span class='hl-dim'>.obsp</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No distances</span></details><div class='block-mod'><div><details><summary class='summary-mod'><div class='title title-mod'>cNMF</div><span class='hl-dim'>92284 &times 60</span></summary><span>AnnData object <span class='hl-dim'>92284 obs &times; 60 var</span></span><br><div class='title title-attr'>Matrix</div><span class='hl-dim'>.X</span><div>\n",
       "            <span class='hl-types'>float64</span> <span>&nbsp;&nbsp;&nbsp;<span class='hl-import'>numpy.</span>ndarray</span>\n",
       "         </table></div><details><summary><div class='title title-attr'>Layers</div><span class='hl-dim'>.layers</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No layers</span></details><details><summary><div class='title title-attr'>Metadata</div><span class='hl-dim'>.obs</span><span class='hl-size'>7 elements</span></summary><div><table><tr>\n",
       "                                <td class='col-index'>sample</td>  <td class='hl-types'>category</td>  <td class='hl-values'>D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>species</td>  <td class='hl-types'>category</td>  <td class='hl-values'>hg38,hg38,hg38,hg38,hg38,hg38,hg38,hg38,hg38,hg38,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>gene_count</td>  <td class='hl-types'>int64</td>  <td class='hl-values'>1903,2664,2147,2011,1441,1607,1385,3690,1348,1247,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>tscp_count</td>  <td class='hl-types'>int64</td>  <td class='hl-values'>3010,4710,3492,3286,2027,2616,2256,7268,2093,1934,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>mread_count</td>  <td class='hl-types'>int64</td>  <td class='hl-values'>4079,6401,4666,4432,2773,3499,3044,9975,2837,2533,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>leiden</td>  <td class='hl-types'>category</td>  <td class='hl-values'>0,2,1,0,0,0,4,5,4,4,4,0,1,3,2,0,2,2,1,1,1,3,0,0,0,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>n_counts</td>  <td class='hl-types'>float32</td>  <td class='hl-values'>653.00,997.00,618.00,742.00,430.00,585.00,570.00,1...</td>\n",
       "                            </tr></table></div></details><details><summary><div class='title title-attr'>Embeddings</div><span class='hl-dim'>.obsm</span><span class='hl-size'>1 element</span></summary><div><table><tr>\n",
       "                                       <td class='col-index'>guide_assignment</td>  <td class='hl-types'>int64</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>1926 columns</td>\n",
       "                                   </tr></table></div></details><details><summary><div class='title title-attr'>Distances</div><span class='hl-dim'>.obsp</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No distances</span></details><details><summary><div class='title title-attr'>Miscellaneous</div><span class='hl-dim'>.uns</span><span class='hl-size'>3 elements</span></summary><div><table><tr>\n",
       "                                <td class='col-index'>guide_names</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>1926 elements</td>  <td class='hl-values'>ACAA1_-_38178575.23-P1P2,ACAA1_+_38178488.23-P1P2,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>guide_targets</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>1926 elements</td>  <td class='hl-values'>ACAA1,ACAA1,ACAA1,ACAA1,ACAA1,ACAA1,ACTL7B,ACTL7B,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>var_names</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>5451 elements</td>  <td class='hl-values'>SEMA3F,CFTR,CYP51A1,HECW1,KLHL13,CASP10,CFLAR,TFPI...</td>\n",
       "                            </tr></table></div></details></details></div></div><div class='block-mod'><div><details><summary class='summary-mod'><div class='title title-mod'>rna</div><span class='hl-dim'>92284 &times 5451</span></summary><span>AnnData object <span class='hl-dim'>92284 obs &times; 5451 var</span></span><br><div class='title title-attr'>Matrix</div><span class='hl-dim'>.X</span><div>\n",
       "            <span class='hl-types'>float32</span> <span>&nbsp;&nbsp;&nbsp;<span class='hl-import'>scipy.sparse._csr.</span>csr_matrix</span>\n",
       "         </table></div><details><summary><div class='title title-attr'>Layers</div><span class='hl-dim'>.layers</span><span class='hl-size'>1 element</span></summary><div><table><tr>\n",
       "                                       <td class='col-index'>norm10k</td>  <td class='hl-types'>float32</td>  <td><span class='hl-import'>scipy.sparse._csr.</span>csr_matrix</td>  <td class='hl-dims'></td>\n",
       "                                   </tr></table></div></details><details><summary><div class='title title-attr'>Metadata</div><span class='hl-dim'>.obs</span><span class='hl-size'>7 elements</span></summary><div><table><tr>\n",
       "                                <td class='col-index'>sample</td>  <td class='hl-types'>category</td>  <td class='hl-values'>D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0,D0...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>species</td>  <td class='hl-types'>category</td>  <td class='hl-values'>hg38,hg38,hg38,hg38,hg38,hg38,hg38,hg38,hg38,hg38,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>gene_count</td>  <td class='hl-types'>int64</td>  <td class='hl-values'>1903,2664,2147,2011,1441,1607,1385,3690,1348,1247,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>tscp_count</td>  <td class='hl-types'>int64</td>  <td class='hl-values'>3010,4710,3492,3286,2027,2616,2256,7268,2093,1934,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>mread_count</td>  <td class='hl-types'>int64</td>  <td class='hl-values'>4079,6401,4666,4432,2773,3499,3044,9975,2837,2533,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>leiden</td>  <td class='hl-types'>category</td>  <td class='hl-values'>0,2,1,0,0,0,4,5,4,4,4,0,1,3,2,0,2,2,1,1,1,3,0,0,0,...</td>\n",
       "                            </tr>\n",
       "<tr>\n",
       "                                <td class='col-index'>n_counts</td>  <td class='hl-types'>float32</td>  <td class='hl-values'>653.00,997.00,618.00,742.00,430.00,585.00,570.00,1...</td>\n",
       "                            </tr></table></div></details><details><summary><div class='title title-attr'>Embeddings</div><span class='hl-dim'>.obsm</span><span class='hl-size'>2 elements</span></summary><div><table><tr>\n",
       "                                       <td class='col-index'>X_pca</td>  <td class='hl-types'>float64</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>50 columns</td>\n",
       "                                   </tr>\n",
       "<tr>\n",
       "                                       <td class='col-index'>X_umap</td>  <td class='hl-types'>float32</td>  <td><span class='hl-import'>numpy.</span>ndarray</td>  <td class='hl-dims'>2 columns</td>\n",
       "                                   </tr></table></div></details><details><summary><div class='title title-attr'>Distances</div><span class='hl-dim'>.obsp</span><span class='hl-size'>0 elements</span></summary><span class='hl-empty'>No distances</span></details><details><summary><div class='title title-attr'>Miscellaneous</div><span class='hl-dim'>.uns</span><span class='hl-size'>1 elements</span></summary><div><table><tr>\n",
       "                                <td class='col-index'>sample_colors</td>  <td><span class='hl-import'></span>list</td>  <td class='hl-dims'>4 elements</td>  <td class='hl-values'>#1f77b4,#ff7f0e,#2ca02c,#d62728...</td>\n",
       "                            </tr></table></div></details></details></div></div><br/>"
      ],
      "text/plain": [
       "MuData object with n_obs × n_vars = 92284 × 5511\n",
       "  2 modalities\n",
       "    cNMF:\t92284 x 60\n",
       "      obs:\t'sample', 'species', 'gene_count', 'tscp_count', 'mread_count', 'leiden', 'n_counts'\n",
       "      uns:\t'guide_names', 'guide_targets', 'var_names'\n",
       "      obsm:\t'guide_assignment'\n",
       "      varm:\t'loadings'\n",
       "    rna:\t92284 x 5451\n",
       "      obs:\t'sample', 'species', 'gene_count', 'tscp_count', 'mread_count', 'leiden', 'n_counts'\n",
       "      uns:\t'sample_colors'\n",
       "      obsm:\t'X_pca', 'X_umap'\n",
       "      layers:\t'norm10k'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "with mudata.set_options(display_style=\"html\", display_html_expand=0b000):\n",
    "    display(mdata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['cNMF', 'rna']"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sorted(mdata.mod.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 92284 × 5451\n",
       "    obs: 'sample', 'species', 'gene_count', 'tscp_count', 'mread_count', 'leiden', 'n_counts'\n",
       "    uns: 'sample_colors'\n",
       "    obsm: 'X_pca', 'X_umap'\n",
       "    layers: 'norm10k'"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "input_adata = mdata.mod[\"rna\"]\n",
    "input_adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Single run analysis\n",
    "Here we will go through each subdirectory and corresponding key in the MuData object and extract per program metrics for plotting and evaluation."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Explained variance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 174,
   "metadata": {},
   "outputs": [],
   "source": [
    "# For each subdirectory, find the file explained_variance_ratio.txt\n",
    "# Read it in and\n",
    "# Store this in a dictionary with the subdirectory dir_name as the key\n",
    "explained_variance_ratios = {}\n",
    "for subdir in subdirs:\n",
    "    try:\n",
    "        run_name = os.path.basename(subdir)\n",
    "        df = pd.read_csv(os.path.join(subdir, \"explained_variance_ratio.txt\"), sep=\"\\t\")\n",
    "        df.columns = [\"program_name\", \"explained_variance_ratio\"]\n",
    "        explained_variance_ratios[run_name] = df\n",
    "    except FileNotFoundError:\n",
    "        print(f\"File not found: {subdir}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Gene-set enrichment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 175,
   "metadata": {},
   "outputs": [],
   "source": [
    "# For each subdirectory, find the file geneset_enrichment.txt\n",
    "# Store this dataframe in a dictionary with the key being the subdirectory name\n",
    "gene_set_enrichment_dfs = {}\n",
    "for subdir in subdirs:\n",
    "    try:\n",
    "        run_name = os.path.basename(subdir)\n",
    "        gene_set_enrichment_file = os.path.join(subdir, \"geneset_enrichment.txt\")\n",
    "        gene_set_enrichment_df = pd.read_csv(gene_set_enrichment_file, sep=\"\\t\")\n",
    "        gene_set_enrichment_dfs[run_name] = gene_set_enrichment_df\n",
    "    except FileNotFoundError:\n",
    "        print(f\"File not found: {gene_set_enrichment_file}\")\n",
    "        continue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 176,
   "metadata": {},
   "outputs": [],
   "source": [
    "categorical_var = \"program_name\"\n",
    "count_var = \"Term\"\n",
    "sig_var = \"FDR q-val\"\n",
    "sig_threshold = 0.25"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 177,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Now for each df, calculate the number of unique gene sets that are enriched\n",
    "# Include a FDR q-val cutoff\n",
    "gene_sets_counts = {}\n",
    "unique_gene_sets_counts = {}\n",
    "for run_name, gene_set_enrichment_df in gene_set_enrichment_dfs.items():\n",
    "    # Filter the dataframe based on the FDR q-value\n",
    "    filtered_data = gene_set_enrichment_df[gene_set_enrichment_df[sig_var] < sig_threshold]\n",
    "\n",
    "    # Get the count of all gene sets passing the FDR q-value cutoff for each program\n",
    "    count_df = count(categorical_var=categorical_var, count_var=count_var, dataframe=filtered_data)\n",
    "    gene_sets_counts[run_name] = count_df\n",
    "\n",
    "    # Get the count of unique gene sets passing the FDR q-value cutoff for each program\n",
    "    unique_data = filtered_data.sort_values(by=sig_var)\n",
    "    unique_data = unique_data.drop_duplicates(subset=count_var)\n",
    "    unique_df = count_unique(categorical_var=categorical_var, count_var=count_var, dataframe=unique_data)\n",
    "    unique_gene_sets_counts[run_name] = unique_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Motif enrichment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 178,
   "metadata": {},
   "outputs": [],
   "source": [
    "# For each subdirectory, find the file motif_enrichment.txt\n",
    "# Store this dataframe in a dictionary with the key being the subdirectory name\n",
    "motif_enrichment_dfs = {}\n",
    "for subdir in subdirs:\n",
    "    try:\n",
    "        run_name = os.path.basename(subdir)\n",
    "        motif_enrichment_file = os.path.join(subdir, \"motif_enrichment.txt\")\n",
    "        motif_enrichment_df = pd.read_csv(motif_enrichment_file, sep=\"\\t\")\n",
    "        motif_enrichment_dfs[run_name] = motif_enrichment_df\n",
    "    except FileNotFoundError:\n",
    "        print(f\"File not found: {motif_enrichment_file}\")\n",
    "        continue"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Trait enrichment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 180,
   "metadata": {},
   "outputs": [],
   "source": [
    "# For each subdirectory, find the file trait_enrichment.txt\n",
    "# Store this dataframe in a dictionary with the key being the subdirectory name\n",
    "trait_enrichment_dfs = {}\n",
    "for subdir in subdirs:\n",
    "    try:\n",
    "        run_name = os.path.basename(subdir)\n",
    "        trait_enrichment_file = os.path.join(subdir, \"trait_enrichment.txt\")\n",
    "        trait_enrichment_df = pd.read_csv(trait_enrichment_file, sep=\"\\t\")\n",
    "        trait_enrichment_dfs[run_name] = trait_enrichment_df\n",
    "    except FileNotFoundError:\n",
    "        print(f\"File not found: {trait_enrichment_file}\")\n",
    "        continue"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Cross run analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Cumulative explained variance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [],
   "source": [
    "# For each subdirectory, find the file explained_variance_ratio.txt\n",
    "# Read it in and sum across all the programs to get one value\n",
    "# Store this in a dictionary with the subdirectory dir_name as the key\n",
    "cumulative_explained_variance_ratios = {}\n",
    "for subdir in subdirs:\n",
    "    try:\n",
    "        run_name = os.path.basename(subdir)\n",
    "        df = pd.read_csv(os.path.join(subdir, \"explained_variance_ratio.txt\"), sep=\"\\t\")\n",
    "        cumulative_explained_variance_ratios[run_name] = df[\"explained_variance_ratio_X\"].sum()\n",
    "    except FileNotFoundError:\n",
    "        print(f\"File not found: {subdir}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'cNMF': -3.7118508404785633}"
      ]
     },
     "execution_count": 129,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cumulative_explained_variance_ratios"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Reconstruction error"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'rna': AnnData object with n_obs × n_vars = 92284 × 5451\n",
       "     obs: 'sample', 'species', 'gene_count', 'tscp_count', 'mread_count', 'leiden', 'n_counts'\n",
       "     obsm: 'X_pca', 'X_umap'\n",
       "     layers: 'norm10k',\n",
       " 'cNMF': AnnData object with n_obs × n_vars = 92284 × 60\n",
       "     obs: 'sample', 'species', 'gene_count', 'tscp_count', 'mread_count', 'leiden', 'n_counts'\n",
       "     uns: 'guide_names', 'guide_targets', 'var_names'\n",
       "     obsm: 'guide_assignment'\n",
       "     varm: 'loadings'}"
      ]
     },
     "execution_count": 142,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mdata.mod"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "metadata": {},
   "outputs": [],
   "source": [
    "# For each non \"rna\" key in the mdata, we want to reconstruct the gene expression matrix\n",
    "# We will multiply the program loadings in varm (program x gene) by the cell loadings in .X (cell x program) to do this\n",
    "# This will give us a cell x gene matrix that we will then take the sum of sum of squares of the element-wise difference with the rna.X\n",
    "\n",
    "# Get the proper rna.X\n",
    "orig_mtx = mdata[\"rna\"].layers[\"norm10k\"].copy().toarray()\n",
    "\n",
    "# Calculate a reconstructed matrix for each non-rna key in mdata\n",
    "errors = {}\n",
    "for key in mdata.mod.keys():\n",
    "    if key == \"rna\":\n",
    "        continue\n",
    "    program_mtx = mdata[key].varm[\"loadings\"]\n",
    "    cell_mtx = mdata[key].X\n",
    "    reconstructed_mtx = np.dot(cell_mtx, program_mtx)\n",
    "    error = np.sum((orig_mtx - reconstructed_mtx) ** 2)\n",
    "    errors[key] = error"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(92284, 5451)"
      ]
     },
     "execution_count": 153,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reconstructed_mtx.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(92284, 5451)"
      ]
     },
     "execution_count": 155,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "orig_mtx.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'cNMF': 436202896.33146477}"
      ]
     },
     "execution_count": 158,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "errors"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# DONE!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def count_unique(categorical_var, count_var, dataframe, cummul=False, unique=False):\n",
    "    counts_df = count(categorical_var, count_var, dataframe)\n",
    "    new_df = []\n",
    "    terms = []\n",
    "    # AK: the order of programs is based on the number of unique terms, not on the most significant terms\n",
    "    for prog in counts_df[categorical_var].unique():\n",
    "        terms_ = dataframe.loc[dataframe[categorical_var] == prog, count_var].unique()\n",
    "        unique_terms = [term for term in terms_ if term not in terms]\n",
    "        terms.extend(unique_terms)\n",
    "        new_df.append([prog, len(unique_terms)])\n",
    "    new_df = pd.DataFrame(new_df, columns=[categorical_var, count_var])\n",
    "    if cummul:\n",
    "        new_df[count_var] = new_df[count_var].cumsum()\n",
    "    if unique:\n",
    "        return new_df\n",
    "    else:\n",
    "        return counts_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>program_name</th>\n",
       "      <th>Term</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4</td>\n",
       "      <td>182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>6</td>\n",
       "      <td>132</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>20</td>\n",
       "      <td>119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>97</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>10</td>\n",
       "      <td>45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>7</td>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>8</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>12</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>16</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>2</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>47</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>34</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>35</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>17</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>26</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>24</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>23</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>14</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>28</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>59</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>13</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>21</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>33</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>19</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>15</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>41</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>49</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>50</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>57</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    program_name  Term\n",
       "0              4   182\n",
       "1              6   132\n",
       "2             20   119\n",
       "3              0    97\n",
       "4             10    45\n",
       "5              7    34\n",
       "6              8    14\n",
       "7             12    14\n",
       "8             16    12\n",
       "9              2     9\n",
       "10            47     8\n",
       "11            34     7\n",
       "12            35     6\n",
       "13            17     6\n",
       "14            26     5\n",
       "15            24     4\n",
       "16            23     3\n",
       "17            14     3\n",
       "18            28     2\n",
       "19            59     2\n",
       "20            13     2\n",
       "21            21     1\n",
       "22            33     1\n",
       "23            19     1\n",
       "24            15     1\n",
       "25            41     1\n",
       "26            49     1\n",
       "27            50     1\n",
       "28            57     1\n",
       "29             5     1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# A term that is shared across multiple programs will be counted once for each program it is associated with\n",
    "counts_df = dataframe.value_counts([categorical_var, count_var])\n",
    "counts_df = counts_df.groupby(categorical_var).sum()\n",
    "counts_df = counts_df.sort_values(ascending=False)\n",
    "counts_df = pd.DataFrame(counts_df.reset_index().values,\n",
    "                             columns=[categorical_var, count_var])\n",
    "counts_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "291"
      ]
     },
     "execution_count": 169,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "count_unique(categorical_var, count_var, filtered_data, cummul=False, unique=False)[\"Term\"].sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>program_name</th>\n",
       "      <th>Term</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4</td>\n",
       "      <td>93</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>6</td>\n",
       "      <td>58</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>20</td>\n",
       "      <td>31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>7</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>47</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>12</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>24</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>14</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>16</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>23</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>8</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>26</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>35</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>13</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>59</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>19</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>28</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>34</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>41</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>49</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>50</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    program_name  Term\n",
       "0              4    93\n",
       "1              6    58\n",
       "2              0    49\n",
       "3             20    31\n",
       "4              7    12\n",
       "5             47     6\n",
       "6             10     6\n",
       "7             12     5\n",
       "8             24     4\n",
       "9             14     3\n",
       "10            16     3\n",
       "11            23     3\n",
       "12             8     3\n",
       "13            26     2\n",
       "14            35     2\n",
       "15            13     2\n",
       "16            59     2\n",
       "17            19     1\n",
       "18             2     1\n",
       "19            28     1\n",
       "20            34     1\n",
       "21            41     1\n",
       "22            49     1\n",
       "23            50     1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# A program that has multiple terms that are shared across multiple programs will only be counted in its most significant program\n",
    "\n",
    "# Start by sorting the dataframe by the FDR q-value and removing duplicates in the Term column\n",
    "\n",
    "\n",
    "# Now count the number of unique terms for each program\n",
    "counts_df = count(categorical_var, count_var, dataframe)\n",
    "counts_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>program_name</th>\n",
       "      <th>Term</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4</td>\n",
       "      <td>182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>6</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>20</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>47</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>8</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>10</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>14</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>7</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>35</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>50</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>41</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>19</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>59</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>23</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>24</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>34</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>26</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>17</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>28</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>21</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>33</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>49</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>57</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    program_name  Term\n",
       "0              4   182\n",
       "3              0    45\n",
       "1              6    28\n",
       "2             20     8\n",
       "10            47     5\n",
       "6              8     4\n",
       "4             10     3\n",
       "17            14     3\n",
       "5              7     2\n",
       "12            35     2\n",
       "27            50     1\n",
       "25            41     1\n",
       "23            19     1\n",
       "19            59     1\n",
       "16            23     1\n",
       "15            24     1\n",
       "11            34     1\n",
       "9              2     1\n",
       "7             12     1\n",
       "14            26     0\n",
       "13            17     0\n",
       "18            28     0\n",
       "20            13     0\n",
       "21            21     0\n",
       "22            33     0\n",
       "24            15     0\n",
       "8             16     0\n",
       "26            49     0\n",
       "28            57     0\n",
       "29             5     0"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "new_df = []\n",
    "terms = []\n",
    "for prog in counts_df[categorical_var].unique():\n",
    "    terms_ = dataframe.loc[dataframe[categorical_var] == prog, count_var].unique()\n",
    "    unique_terms = [term for term in terms_ if term not in terms]\n",
    "    terms.extend(unique_terms)\n",
    "    new_df.append([prog, len(unique_terms)])\n",
    "new_df = pd.DataFrame(new_df, columns=[categorical_var, count_var]).sort_values(count_var, ascending=False)\n",
    "new_df"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.10 test_celloracle",
   "language": "python",
   "name": "test_scenicplus-py310"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
