{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.append('../../../')\n",
    "\n",
    "import mudata\n",
    "\n",
    "from src.evaluation import compute_categorical_association\n",
    "from src.evaluation import compute_explained_variance_ratio\n",
    "from src.evaluation import compute_geneset_enrichment\n",
    "from src.evaluation import compute_trait_enrichment\n",
    "from src.evaluation import compute_motif_enrichment\n",
    "\n",
    "from plotly.subplots import make_subplots\n",
    "\n",
    "import plotly.graph_objects as go\n",
    "import plotly.express as px\n",
    "\n",
    "import ipywidgets as w\n",
    "from IPython.display import display\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from scipy import stats\n",
    "\n",
    "from tqdm.auto import tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/oak/stanford/groups/akundaje/regu/miniforge3/envs/evaluation/lib/python3.9/site-packages/anndata/_core/anndata.py:522: FutureWarning: The dtype argument is deprecated and will be removed in late 2024.\n",
      "  warnings.warn(\n",
      "/oak/stanford/groups/akundaje/regu/miniforge3/envs/evaluation/lib/python3.9/site-packages/anndata/_core/anndata.py:522: FutureWarning: The dtype argument is deprecated and will be removed in late 2024.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<pre>MuData object with n_obs × n_vars = 4000 × 17532\n",
       "  2 modalities\n",
       "    rna:\t4000 x 17472\n",
       "      obs:\t&#x27;barcodes&#x27;, &#x27;n_genes&#x27;, &#x27;n_counts&#x27;, &#x27;sample&#x27;, &#x27;batch&#x27;\n",
       "      var:\t&#x27;n_cells&#x27;\n",
       "    cNMF:\t4000 x 60\n",
       "      obs:\t&#x27;sample&#x27;, &#x27;batch&#x27;\n",
       "      uns:\t&#x27;var_names&#x27;\n",
       "      varm:\t&#x27;loadings&#x27;, &#x27;loadings_zscore&#x27;</pre>"
      ],
      "text/plain": [
       "MuData object with n_obs × n_vars = 4000 × 17532\n",
       "  2 modalities\n",
       "    rna:\t4000 x 17472\n",
       "      obs:\t'barcodes', 'n_genes', 'n_counts', 'sample', 'batch'\n",
       "      var:\t'n_cells'\n",
       "    cNMF:\t4000 x 60\n",
       "      obs:\t'sample', 'batch'\n",
       "      uns:\t'var_names'\n",
       "      varm:\t'loadings', 'loadings_zscore'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Import test data\n",
    "mdata = mudata.read('../../../../../../data/TeloHAEC_Perturb-seq_2kG/2kG.library_K60_kangh.h5mu')\n",
    "\n",
    "sel_idx = []\n",
    "for batch in mdata['cNMF'].obs['batch'].unique():\n",
    "    for samp in mdata['cNMF'].obs['sample'].unique():\n",
    "        mdata_obs_ = mdata['cNMF'].obs.loc[(mdata['cNMF'].obs['batch']==batch) & \\\n",
    "                                           (mdata['cNMF'].obs['sample']==samp)]\n",
    "\n",
    "        sel_idx.extend(mdata_obs_.iloc[:200].index.tolist())\n",
    "\n",
    "mdata = mdata[sel_idx].copy()\n",
    "mdata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "3ccf2abfb8eb4e9abd7630aa0ad5bc3a",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Computing explained variance:   0%|          | 0/60 [00:00<?, ?programs/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Perform testing by averaging over sample\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "bfd92c52f43d44baa5db3db0cbb3c9df",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Testing batch association:   0%|          | 0/60 [00:00<?, ?programs/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/oak/stanford/groups/akundaje/regu/projects/gene_program_evaluation/gene_network_evaluation/src/plotting/dashapp/../../../src/evaluation/association_categorical.py:65: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  prog_data = prog_data.groupby([pseudobulk_key, categorical_key]).mean().dropna().reset_index()\n",
      "/oak/stanford/groups/akundaje/regu/projects/gene_program_evaluation/gene_network_evaluation/src/plotting/dashapp/../../../src/evaluation/association_categorical.py:65: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  prog_data = prog_data.groupby([pseudobulk_key, categorical_key]).mean().dropna().reset_index()\n",
      "/oak/stanford/groups/akundaje/regu/projects/gene_program_evaluation/gene_network_evaluation/src/plotting/dashapp/../../../src/evaluation/association_categorical.py:65: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  prog_data = prog_data.groupby([pseudobulk_key, categorical_key]).mean().dropna().reset_index()\n",
      "/oak/stanford/groups/akundaje/regu/projects/gene_program_evaluation/gene_network_evaluation/src/plotting/dashapp/../../../src/evaluation/association_categorical.py:65: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  prog_data = prog_data.groupby([pseudobulk_key, categorical_key]).mean().dropna().reset_index()\n",
      "/oak/stanford/groups/akundaje/regu/projects/gene_program_evaluation/gene_network_evaluation/src/plotting/dashapp/../../../src/evaluation/association_categorical.py:65: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  prog_data = prog_data.groupby([pseudobulk_key, categorical_key]).mean().dropna().reset_index()\n",
      "/oak/stanford/groups/akundaje/regu/projects/gene_program_evaluation/gene_network_evaluation/src/plotting/dashapp/../../../src/evaluation/association_categorical.py:65: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  prog_data = prog_data.groupby([pseudobulk_key, categorical_key]).mean().dropna().reset_index()\n",
      "/oak/stanford/groups/akundaje/regu/projects/gene_program_evaluation/gene_network_evaluation/src/plotting/dashapp/../../../src/evaluation/association_categorical.py:65: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  prog_data = prog_data.groupby([pseudobulk_key, categorical_key]).mean().dropna().reset_index()\n",
      "/oak/stanford/groups/akundaje/regu/projects/gene_program_evaluation/gene_network_evaluation/src/plotting/dashapp/../../../src/evaluation/association_categorical.py:65: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  prog_data = prog_data.groupby([pseudobulk_key, categorical_key]).mean().dropna().reset_index()\n",
      "/oak/stanford/groups/akundaje/regu/projects/gene_program_evaluation/gene_network_evaluation/src/plotting/dashapp/../../../src/evaluation/association_categorical.py:65: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  prog_data = prog_data.groupby([pseudobulk_key, categorical_key]).mean().dropna().reset_index()\n",
      "/oak/stanford/groups/akundaje/regu/projects/gene_program_evaluation/gene_network_evaluation/src/plotting/dashapp/../../../src/evaluation/association_categorical.py:65: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  prog_data = prog_data.groupby([pseudobulk_key, categorical_key]).mean().dropna().reset_index()\n",
      "/oak/stanford/groups/akundaje/regu/projects/gene_program_evaluation/gene_network_evaluation/src/plotting/dashapp/../../../src/evaluation/association_categorical.py:65: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  prog_data = prog_data.groupby([pseudobulk_key, categorical_key]).mean().dropna().reset_index()\n",
      "/oak/stanford/groups/akundaje/regu/projects/gene_program_evaluation/gene_network_evaluation/src/plotting/dashapp/../../../src/evaluation/association_categorical.py:65: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  prog_data = prog_data.groupby([pseudobulk_key, categorical_key]).mean().dropna().reset_index()\n",
      "/oak/stanford/groups/akundaje/regu/projects/gene_program_evaluation/gene_network_evaluation/src/plotting/dashapp/../../../src/evaluation/association_categorical.py:65: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  prog_data = prog_data.groupby([pseudobulk_key, categorical_key]).mean().dropna().reset_index()\n",
      "/oak/stanford/groups/akundaje/regu/projects/gene_program_evaluation/gene_network_evaluation/src/plotting/dashapp/../../../src/evaluation/association_categorical.py:65: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  prog_data = prog_data.groupby([pseudobulk_key, categorical_key]).mean().dropna().reset_index()\n",
      "/oak/stanford/groups/akundaje/regu/projects/gene_program_evaluation/gene_network_evaluation/src/plotting/dashapp/../../../src/evaluation/association_categorical.py:65: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  prog_data = prog_data.groupby([pseudobulk_key, categorical_key]).mean().dropna().reset_index()\n",
      "/oak/stanford/groups/akundaje/regu/projects/gene_program_evaluation/gene_network_evaluation/src/plotting/dashapp/../../../src/evaluation/association_categorical.py:65: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  prog_data = prog_data.groupby([pseudobulk_key, categorical_key]).mean().dropna().reset_index()\n",
      "/oak/stanford/groups/akundaje/regu/projects/gene_program_evaluation/gene_network_evaluation/src/plotting/dashapp/../../../src/evaluation/association_categorical.py:65: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  prog_data = prog_data.groupby([pseudobulk_key, categorical_key]).mean().dropna().reset_index()\n",
      "/oak/stanford/groups/akundaje/regu/projects/gene_program_evaluation/gene_network_evaluation/src/plotting/dashapp/../../../src/evaluation/association_categorical.py:65: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  prog_data = prog_data.groupby([pseudobulk_key, categorical_key]).mean().dropna().reset_index()\n",
      "/oak/stanford/groups/akundaje/regu/projects/gene_program_evaluation/gene_network_evaluation/src/plotting/dashapp/../../../src/evaluation/association_categorical.py:65: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  prog_data = prog_data.groupby([pseudobulk_key, categorical_key]).mean().dropna().reset_index()\n",
      "/oak/stanford/groups/akundaje/regu/projects/gene_program_evaluation/gene_network_evaluation/src/plotting/dashapp/../../../src/evaluation/association_categorical.py:65: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  prog_data = prog_data.groupby([pseudobulk_key, categorical_key]).mean().dropna().reset_index()\n",
      "/oak/stanford/groups/akundaje/regu/projects/gene_program_evaluation/gene_network_evaluation/src/plotting/dashapp/../../../src/evaluation/association_categorical.py:65: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  prog_data = prog_data.groupby([pseudobulk_key, categorical_key]).mean().dropna().reset_index()\n",
      "/oak/stanford/groups/akundaje/regu/projects/gene_program_evaluation/gene_network_evaluation/src/plotting/dashapp/../../../src/evaluation/association_categorical.py:65: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  prog_data = prog_data.groupby([pseudobulk_key, categorical_key]).mean().dropna().reset_index()\n",
      "/oak/stanford/groups/akundaje/regu/projects/gene_program_evaluation/gene_network_evaluation/src/plotting/dashapp/../../../src/evaluation/association_categorical.py:65: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  prog_data = prog_data.groupby([pseudobulk_key, categorical_key]).mean().dropna().reset_index()\n",
      "/oak/stanford/groups/akundaje/regu/projects/gene_program_evaluation/gene_network_evaluation/src/plotting/dashapp/../../../src/evaluation/association_categorical.py:65: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  prog_data = prog_data.groupby([pseudobulk_key, categorical_key]).mean().dropna().reset_index()\n",
      "/oak/stanford/groups/akundaje/regu/projects/gene_program_evaluation/gene_network_evaluation/src/plotting/dashapp/../../../src/evaluation/association_categorical.py:65: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  prog_data = prog_data.groupby([pseudobulk_key, categorical_key]).mean().dropna().reset_index()\n",
      "/oak/stanford/groups/akundaje/regu/projects/gene_program_evaluation/gene_network_evaluation/src/plotting/dashapp/../../../src/evaluation/association_categorical.py:65: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  prog_data = prog_data.groupby([pseudobulk_key, categorical_key]).mean().dropna().reset_index()\n",
      "/oak/stanford/groups/akundaje/regu/projects/gene_program_evaluation/gene_network_evaluation/src/plotting/dashapp/../../../src/evaluation/association_categorical.py:65: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  prog_data = prog_data.groupby([pseudobulk_key, categorical_key]).mean().dropna().reset_index()\n",
      "/oak/stanford/groups/akundaje/regu/projects/gene_program_evaluation/gene_network_evaluation/src/plotting/dashapp/../../../src/evaluation/association_categorical.py:65: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  prog_data = prog_data.groupby([pseudobulk_key, categorical_key]).mean().dropna().reset_index()\n",
      "/oak/stanford/groups/akundaje/regu/projects/gene_program_evaluation/gene_network_evaluation/src/plotting/dashapp/../../../src/evaluation/association_categorical.py:65: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  prog_data = prog_data.groupby([pseudobulk_key, categorical_key]).mean().dropna().reset_index()\n",
      "/oak/stanford/groups/akundaje/regu/projects/gene_program_evaluation/gene_network_evaluation/src/plotting/dashapp/../../../src/evaluation/association_categorical.py:65: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  prog_data = prog_data.groupby([pseudobulk_key, categorical_key]).mean().dropna().reset_index()\n",
      "/oak/stanford/groups/akundaje/regu/projects/gene_program_evaluation/gene_network_evaluation/src/plotting/dashapp/../../../src/evaluation/association_categorical.py:65: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  prog_data = prog_data.groupby([pseudobulk_key, categorical_key]).mean().dropna().reset_index()\n",
      "/oak/stanford/groups/akundaje/regu/projects/gene_program_evaluation/gene_network_evaluation/src/plotting/dashapp/../../../src/evaluation/association_categorical.py:65: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  prog_data = prog_data.groupby([pseudobulk_key, categorical_key]).mean().dropna().reset_index()\n",
      "/oak/stanford/groups/akundaje/regu/projects/gene_program_evaluation/gene_network_evaluation/src/plotting/dashapp/../../../src/evaluation/association_categorical.py:65: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  prog_data = prog_data.groupby([pseudobulk_key, categorical_key]).mean().dropna().reset_index()\n",
      "/oak/stanford/groups/akundaje/regu/projects/gene_program_evaluation/gene_network_evaluation/src/plotting/dashapp/../../../src/evaluation/association_categorical.py:65: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  prog_data = prog_data.groupby([pseudobulk_key, categorical_key]).mean().dropna().reset_index()\n",
      "/oak/stanford/groups/akundaje/regu/projects/gene_program_evaluation/gene_network_evaluation/src/plotting/dashapp/../../../src/evaluation/association_categorical.py:65: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  prog_data = prog_data.groupby([pseudobulk_key, categorical_key]).mean().dropna().reset_index()\n",
      "/oak/stanford/groups/akundaje/regu/projects/gene_program_evaluation/gene_network_evaluation/src/plotting/dashapp/../../../src/evaluation/association_categorical.py:65: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  prog_data = prog_data.groupby([pseudobulk_key, categorical_key]).mean().dropna().reset_index()\n",
      "/oak/stanford/groups/akundaje/regu/projects/gene_program_evaluation/gene_network_evaluation/src/plotting/dashapp/../../../src/evaluation/association_categorical.py:65: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  prog_data = prog_data.groupby([pseudobulk_key, categorical_key]).mean().dropna().reset_index()\n",
      "/oak/stanford/groups/akundaje/regu/projects/gene_program_evaluation/gene_network_evaluation/src/plotting/dashapp/../../../src/evaluation/association_categorical.py:65: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  prog_data = prog_data.groupby([pseudobulk_key, categorical_key]).mean().dropna().reset_index()\n",
      "/oak/stanford/groups/akundaje/regu/projects/gene_program_evaluation/gene_network_evaluation/src/plotting/dashapp/../../../src/evaluation/association_categorical.py:65: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  prog_data = prog_data.groupby([pseudobulk_key, categorical_key]).mean().dropna().reset_index()\n",
      "/oak/stanford/groups/akundaje/regu/projects/gene_program_evaluation/gene_network_evaluation/src/plotting/dashapp/../../../src/evaluation/association_categorical.py:65: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  prog_data = prog_data.groupby([pseudobulk_key, categorical_key]).mean().dropna().reset_index()\n",
      "/oak/stanford/groups/akundaje/regu/projects/gene_program_evaluation/gene_network_evaluation/src/plotting/dashapp/../../../src/evaluation/association_categorical.py:65: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  prog_data = prog_data.groupby([pseudobulk_key, categorical_key]).mean().dropna().reset_index()\n",
      "/oak/stanford/groups/akundaje/regu/projects/gene_program_evaluation/gene_network_evaluation/src/plotting/dashapp/../../../src/evaluation/association_categorical.py:65: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  prog_data = prog_data.groupby([pseudobulk_key, categorical_key]).mean().dropna().reset_index()\n",
      "/oak/stanford/groups/akundaje/regu/projects/gene_program_evaluation/gene_network_evaluation/src/plotting/dashapp/../../../src/evaluation/association_categorical.py:65: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  prog_data = prog_data.groupby([pseudobulk_key, categorical_key]).mean().dropna().reset_index()\n",
      "/oak/stanford/groups/akundaje/regu/projects/gene_program_evaluation/gene_network_evaluation/src/plotting/dashapp/../../../src/evaluation/association_categorical.py:65: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  prog_data = prog_data.groupby([pseudobulk_key, categorical_key]).mean().dropna().reset_index()\n",
      "/oak/stanford/groups/akundaje/regu/projects/gene_program_evaluation/gene_network_evaluation/src/plotting/dashapp/../../../src/evaluation/association_categorical.py:65: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  prog_data = prog_data.groupby([pseudobulk_key, categorical_key]).mean().dropna().reset_index()\n",
      "/oak/stanford/groups/akundaje/regu/projects/gene_program_evaluation/gene_network_evaluation/src/plotting/dashapp/../../../src/evaluation/association_categorical.py:65: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  prog_data = prog_data.groupby([pseudobulk_key, categorical_key]).mean().dropna().reset_index()\n",
      "/oak/stanford/groups/akundaje/regu/projects/gene_program_evaluation/gene_network_evaluation/src/plotting/dashapp/../../../src/evaluation/association_categorical.py:65: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  prog_data = prog_data.groupby([pseudobulk_key, categorical_key]).mean().dropna().reset_index()\n",
      "/oak/stanford/groups/akundaje/regu/projects/gene_program_evaluation/gene_network_evaluation/src/plotting/dashapp/../../../src/evaluation/association_categorical.py:65: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  prog_data = prog_data.groupby([pseudobulk_key, categorical_key]).mean().dropna().reset_index()\n",
      "/oak/stanford/groups/akundaje/regu/projects/gene_program_evaluation/gene_network_evaluation/src/plotting/dashapp/../../../src/evaluation/association_categorical.py:65: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  prog_data = prog_data.groupby([pseudobulk_key, categorical_key]).mean().dropna().reset_index()\n",
      "/oak/stanford/groups/akundaje/regu/projects/gene_program_evaluation/gene_network_evaluation/src/plotting/dashapp/../../../src/evaluation/association_categorical.py:65: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  prog_data = prog_data.groupby([pseudobulk_key, categorical_key]).mean().dropna().reset_index()\n",
      "/oak/stanford/groups/akundaje/regu/projects/gene_program_evaluation/gene_network_evaluation/src/plotting/dashapp/../../../src/evaluation/association_categorical.py:65: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  prog_data = prog_data.groupby([pseudobulk_key, categorical_key]).mean().dropna().reset_index()\n",
      "/oak/stanford/groups/akundaje/regu/projects/gene_program_evaluation/gene_network_evaluation/src/plotting/dashapp/../../../src/evaluation/association_categorical.py:65: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  prog_data = prog_data.groupby([pseudobulk_key, categorical_key]).mean().dropna().reset_index()\n",
      "/oak/stanford/groups/akundaje/regu/projects/gene_program_evaluation/gene_network_evaluation/src/plotting/dashapp/../../../src/evaluation/association_categorical.py:65: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  prog_data = prog_data.groupby([pseudobulk_key, categorical_key]).mean().dropna().reset_index()\n",
      "/oak/stanford/groups/akundaje/regu/projects/gene_program_evaluation/gene_network_evaluation/src/plotting/dashapp/../../../src/evaluation/association_categorical.py:65: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  prog_data = prog_data.groupby([pseudobulk_key, categorical_key]).mean().dropna().reset_index()\n",
      "/oak/stanford/groups/akundaje/regu/projects/gene_program_evaluation/gene_network_evaluation/src/plotting/dashapp/../../../src/evaluation/association_categorical.py:65: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  prog_data = prog_data.groupby([pseudobulk_key, categorical_key]).mean().dropna().reset_index()\n",
      "/oak/stanford/groups/akundaje/regu/projects/gene_program_evaluation/gene_network_evaluation/src/plotting/dashapp/../../../src/evaluation/association_categorical.py:65: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  prog_data = prog_data.groupby([pseudobulk_key, categorical_key]).mean().dropna().reset_index()\n",
      "/oak/stanford/groups/akundaje/regu/projects/gene_program_evaluation/gene_network_evaluation/src/plotting/dashapp/../../../src/evaluation/association_categorical.py:65: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  prog_data = prog_data.groupby([pseudobulk_key, categorical_key]).mean().dropna().reset_index()\n",
      "/oak/stanford/groups/akundaje/regu/projects/gene_program_evaluation/gene_network_evaluation/src/plotting/dashapp/../../../src/evaluation/association_categorical.py:65: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  prog_data = prog_data.groupby([pseudobulk_key, categorical_key]).mean().dropna().reset_index()\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "97a3c5538df949f7af674ba055268b5a",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Identifying differential batch:   0%|          | 0/60 [00:00<?, ?programs/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Explained variance ratio plot\n",
    "explained_variance = compute_explained_variance_ratio(mdata, prog_key='cNMF', data_key='rna', inplace=False)\n",
    "\n",
    "# Since original data wasn't available we use this as a stand in\n",
    "from sklearn.decomposition import PCA\n",
    "\n",
    "pca = PCA(n_components=60)\n",
    "\n",
    "explained_variance['VarianceExplained'] = pca.fit(mdata['cNMF'].X).explained_variance_ratio_*0.76\n",
    "explained_variance = explained_variance.reset_index()\n",
    "explained_variance['ProgramID'] = explained_variance['index'].apply(lambda x: 'K60_{}'.format(x))\n",
    "explained_variance = explained_variance.drop(['index', 'explained_variance_ratio_X'], axis=1)\n",
    "\n",
    "# Categorical association plot\n",
    "association_df, posthoc_df = compute_categorical_association(mdata, prog_key='cNMF',\n",
    "                                                             pseudobulk_key='sample',\n",
    "                                                             categorical_key='batch',\n",
    "                                                             inplace=False, \n",
    "                                                             n_jobs=-1)\n",
    "\n",
    "# # Geneset enrichment \n",
    "# gsea_df = compute_geneset_enrichment(mdata, prog_key='cNMF', library='Reactome_2013', inplace=False, n_jobs=-1)\n",
    "\n",
    "# # Trait enrichment \n",
    "# gwas_df = compute_trait_enrichment(mdata, '../../smk/resources/OpenTargets_L2G_Filtered.csv.gz', prog_key='cNMF', inplace=False)\n",
    "\n",
    "# # Motif enrichment\n",
    "# motif_match_df, motif_count_df, motif_enrichment_df = \\\n",
    "# compute_motif_enrichment(mdata, prog_key='cNMF', data_key='rna', motif_file='../tests/test_data/motifs.meme',\n",
    "#                          seq_file='../../../../../data/hg38/hg38.fa', coords_file='../tests/test_data/p2g_links.txt',\n",
    "#                          n_jobs=1, inplace=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save trait enrichment data\n",
    "# gwas_df.to_csv('dashapp/example_data/OpenTargets_L2G_cNMF_program_Enrichment_Results.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load example data\n",
    "evaluation_output = pd.ExcelFile('example_data/cNMF_evaluation_output.xlsx')\n",
    "\n",
    "explained_variance = pd.read_excel(evaluation_output, sheet_name='explained_variance')\n",
    "enrichment_motif = pd.read_excel(evaluation_output, sheet_name='enrichment_motif')\n",
    "enrichment_gsea = pd.read_excel(evaluation_output, sheet_name='enrichment_gsea')\n",
    "enrichment_trait = pd.read_excel(evaluation_output, sheet_name='enrichment_trait')\n",
    "\n",
    "program_statistics = pd.read_excel(evaluation_output, sheet_name='program_statistics')\n",
    "regulator_statistitcs = pd.read_excel(evaluation_output, sheet_name='regulator_statistics')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This is the format of your plot grid:\n",
      "[ (1,1) x,y   ]  [ (1,2) x2,y2 ]\n",
      "[ (2,1) x3,y3 ]  [ (2,2) x4,y4 ]\n",
      "[ (3,1) x5,y5 ]  [ (3,2) x6,y6 ]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Assemble dashboard - page 1\n",
    "\n",
    "# Plot unique terms per program\n",
    "def count(categorical_var, count_var, dataframe):\n",
    "\n",
    "    counts_df = dataframe.value_counts([categorical_var, count_var])\n",
    "    counts_df = counts_df.groupby(categorical_var).sum()\n",
    "    counts_df = counts_df.sort_values(ascending=False)\n",
    "\n",
    "    counts_df = pd.DataFrame(counts_df.reset_index().values, \n",
    "                             columns=[categorical_var,\n",
    "                                      count_var])\n",
    "                                      \n",
    "    return counts_df\n",
    "\n",
    "def count_unique(categorical_var, count_var, dataframe, \n",
    "                 cummul=False, unique=False):\n",
    "\n",
    "    counts_df = count(categorical_var, count_var, dataframe)\n",
    "\n",
    "    new_df = []\n",
    "    terms = []\n",
    "    for prog in counts_df[categorical_var].unique():\n",
    "        \n",
    "        terms_ = dataframe.loc[dataframe[categorical_var]==prog, count_var].unique()\n",
    "        unique_terms =  [term for term in terms_ if term not in terms]\n",
    "\n",
    "        terms.extend(unique_terms)\n",
    "        new_df.append([prog, len(unique_terms)])\n",
    "\n",
    "    new_df = pd.DataFrame(new_df, columns=[categorical_var, count_var])\n",
    "\n",
    "    if cummul:\n",
    "        new_df[count_var] = new_df[count_var].cumsum()\n",
    "\n",
    "    if unique:\n",
    "        return counts_df\n",
    "    else:\n",
    "        return new_df\n",
    "\n",
    "fig = make_subplots(\n",
    "    rows=3, cols=2,\n",
    "    specs=[\n",
    "           [{}, {}],\n",
    "           [{}, {}],\n",
    "           [{}, {}]\n",
    "          ],\n",
    "    print_grid=True,\n",
    "    subplot_titles=(' component wise R2 scores', \n",
    "                    'Number of enriched GO terms',\n",
    "                    'Number of enriched promoter Motifs', \n",
    "                    'Number of enriched enhancer Motifs',\n",
    "                    'Number of enriched GWAS traits'),\n",
    "    vertical_spacing = 0.05, horizontal_spacing = 0.1)\n",
    "\n",
    "\n",
    "cummul=False\n",
    "unique=False\n",
    "\n",
    "# Plot unique GSEA Terms per program\n",
    "gsea_unique_df = count_unique('ProgramID', 'ID', enrichment_gsea.loc[enrichment_gsea['qvalue']<=0.05],\n",
    "                              cummul=cummul, unique=unique)\n",
    "\n",
    "# Plot unique GWAS Terms per program\n",
    "gwas_unique_df = count_unique('ProgramID', 'Term', enrichment_trait.loc[enrichment_trait['Adjusted P-value']<=0.05],\n",
    "                              cummul=cummul, unique=unique)\n",
    "\n",
    "# Plot unique Promoter Motif Terms per program\n",
    "motif_unique_promoter_df = count_unique('ProgramID', 'TFMotif', enrichment_motif.loc[(enrichment_motif['FDR']<=0.05) &\\\n",
    "                                                                                     (enrichment_motif['EPType']=='Promoter')],\n",
    "                              cummul=cummul, unique=unique)\n",
    "\n",
    "# Plot unique Enhancer Motif Terms per program\n",
    "motif_unique_enhancer_df = count_unique('ProgramID', 'TFMotif', enrichment_motif.loc[(enrichment_motif['FDR']<=0.05) &\\\n",
    "                                                                                     (enrichment_motif['EPType']=='Enhancer')],\n",
    "                              cummul=cummul, unique=unique)\n",
    "\n",
    "# Generate example data for dashapp\n",
    "\n",
    "a = explained_variance.copy()\n",
    "if cummul:\n",
    "    a[' Component wise R2 scores'] = a['VarianceExplained'].cumsum()\n",
    "else:\n",
    "    a[' Component wise R2 scores'] = a['VarianceExplained']    \n",
    "a = a.drop(['VarianceExplained'], axis=1)\n",
    "\n",
    "gsea_unique_df['Number of enriched GO terms'] = gsea_unique_df['ID']\n",
    "a = a.merge(gsea_unique_df.loc[:, ['ProgramID', 'Number of enriched GO terms']], \n",
    "            left_on='ProgramID', right_on='ProgramID', how='left')\n",
    "\n",
    "motif_unique_promoter_df['Number of enriched promoter Motifs'] = motif_unique_promoter_df['TFMotif']\n",
    "a = a.merge(motif_unique_promoter_df.loc[:, ['ProgramID', 'Number of enriched promoter Motifs']], \n",
    "            left_on='ProgramID', right_on='ProgramID', how='left')\n",
    "\n",
    "motif_unique_enhancer_df['Number of enriched enhancer Motifs'] = motif_unique_enhancer_df['TFMotif']\n",
    "a = a.merge(motif_unique_enhancer_df.loc[:, ['ProgramID', 'Number of enriched enhancer Motifs']], \n",
    "            left_on='ProgramID', right_on='ProgramID', how='left')\n",
    "\n",
    "gwas_unique_df['Number of enriched GWAS traits'] = gwas_unique_df['Term']\n",
    "a = a.merge(gwas_unique_df.loc[:, ['ProgramID', 'Number of enriched GWAS traits']], \n",
    "            left_on='ProgramID', right_on='ProgramID', how='left')\n",
    "\n",
    "\n",
    "for col in ['Number of enriched GWAS traits',\n",
    "            'Number of enriched GO terms',\n",
    "            'Number of enriched promoter Motifs',\n",
    "            'Number of enriched enhancer Motifs']:\n",
    "\n",
    "    if cummul:\n",
    "        a[col] = a[col].fillna(a[col].max())\n",
    "    else:\n",
    "        a[col] = a[col].fillna(0)        \n",
    "\n",
    "plots = {}\n",
    "for i, col in enumerate([col for col in a.columns if 'ProgramID' not in col]):\n",
    "\n",
    "    plots[col] = px.scatter(x=a.sort_values(col, ascending=cummul)['ProgramID'], \n",
    "                            y=a.sort_values(col, ascending=cummul)[col])\n",
    "    plots[col].update_layout(xaxis_title='Components', yaxis_title=col)\n",
    "\n",
    "    row_num = int((i)/2) + 1\n",
    "    col_num = i - (row_num-1)*2 + 1\n",
    "    fig.add_trace(plots[col]['data'][0], row=row_num, col=col_num)\n",
    "    fig.update_xaxes(showticklabels=False, row=row_num, col=col_num)\n",
    "    fig.update_yaxes(ticksuffix = \"  \", row=row_num, col=col_num)\n",
    "\n",
    "# Update plots\n",
    "fig.update_traces(hovertemplate=\"Program Name: %{x} <br> Value: %{y}\", marker_color='black')\n",
    "\n",
    "# Add corrmap\n",
    "corr_map = px.imshow(a.iloc[:,1:].corr(method='spearman'), text_auto=True)\n",
    "fig.add_trace(corr_map['data'][0], row=3, col=2)\n",
    "fig.update_yaxes(showticklabels=False, row=3, col=2)\n",
    "\n",
    "\n",
    "fig.update_layout(height=1050, width=1000, \n",
    "                  coloraxis_showscale=False,\n",
    "                  plot_bgcolor='whitesmoke',\n",
    "                  title_text=\"GEP Dashboard - v0.1 - Goodness of fit measures\")\n",
    "                  \n",
    "fig.write_html('example_output/0_goodness_of_fit.html')\n",
    "# fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This is the format of your plot grid:\n",
      "[ (1,1) x,y   ]  [ (1,2) x2,y2 ]\n",
      "[ (2,1) x3,y3 ]  [ (2,2) x4,y4 ]\n",
      "[ (3,1) x5,y5           -      ]\n",
      "[ (4,1) x6,y6 ]  [ (4,2) x7,y7 ]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Page 2 - program assessment\n",
    "\n",
    "# Batch assocation\n",
    "association_df['batch_sample_kruskall_wallis_fdr'] = stats.false_discovery_control(association_df.batch_sample_kruskall_wallis_pval)\n",
    "association_df['batch_sample_kruskall_wallis_neg_log_fdr'] = association_df['batch_sample_kruskall_wallis_fdr'].apply(lambda x: -np.log(x))\n",
    "\n",
    "association_df['batch_sample_kruskall_wallis_log_stat'] = association_df['batch_sample_kruskall_wallis_stat'].apply(lambda x: np.log(x))\n",
    "\n",
    "# Program scores across batch\n",
    "prog_df = pd.DataFrame(mdata['cNMF'].X, index=mdata['cNMF'].obs.index)\n",
    "prog_df['sample'] = mdata['cNMF'].obs['sample']\n",
    "prog_df['batch'] = mdata['cNMF'].obs['batch']\n",
    "\n",
    "# Loadings\n",
    "loadings = pd.DataFrame(mdata['cNMF'].varm['loadings'], \n",
    "                        index=mdata['cNMF'].var_names,\n",
    "                        columns=mdata['cNMF'].uns['var_names']).T    \n",
    "loadings = loadings/abs(loadings).max(axis=0)\n",
    "\n",
    "# Regulators\n",
    "regulatory = regulator_statistitcs.loc[:, ['Perturbation', 'ProgramsRegulated', 'log2FC', 'AcrossProgramsFDR']]\n",
    "\n",
    "# Make figure\n",
    "fig = make_subplots(rows=4, cols=2,\n",
    "                    specs=[\n",
    "                           [{}, {}],\n",
    "                           [{}, {}],\n",
    "                           [{'colspan':2}, None],\n",
    "                           [{}, {}],\n",
    "                          ],\n",
    "                    print_grid=True,\n",
    "                    subplot_titles=('Enrichment w.r.t. batch', \n",
    "                                    'MAGMA regression',\n",
    "                                    'Program distribution across batch',  \n",
    "                                    'Program-gene KD effect',    \n",
    "                                    'Program-gene loadings (normalised)'\n",
    "                                    ),\n",
    "                    vertical_spacing = 0.1, \n",
    "                    horizontal_spacing = 0.1)\n",
    "\n",
    "for k, r in enumerate([col for col in prog_df.columns if col not in ['sample', 'batch']]):\n",
    "       # Plot volcano\n",
    "       volcano = go.Scatter(x=association_df.batch_sample_kruskall_wallis_log_stat,\n",
    "                            y=association_df.batch_sample_kruskall_wallis_neg_log_fdr,\n",
    "                            customdata=association_df.index.values,\n",
    "                            hovertemplate=\" Program Name: %{customdata}\",\n",
    "                            showlegend=False, mode='markers', \n",
    "                            visible=True if k == 0 else False,\n",
    "                            marker_color='black',\n",
    "                            )\n",
    "       fig.add_trace(volcano, row=1, col=1)\n",
    "       fig.update_xaxes(showticklabels=False, row=1, col=1)\n",
    "       fig.update_yaxes(title='Neg. log adjusted pval', ticksuffix = \"  \", row=1, col=1)\n",
    "\n",
    "# Plot box\n",
    "for k, r in enumerate([col for col in prog_df.columns if col not in ['sample', 'batch']]):\n",
    "\n",
    "    fig.add_trace(\n",
    "                  go.Box(x=prog_df.batch, \n",
    "                         y=prog_df[r],\n",
    "                         visible=True if k == 0 else False,\n",
    "                         customdata=prog_df.index.values,\n",
    "                         hovertemplate=\"Cell barcode: %{customdata}\",\n",
    "                         marker_color='black',\n",
    "                        ), \n",
    "                   row=2, col=1)\n",
    "# fig.update_xaxes(showticklabels=False, row=2, col=1)\n",
    "fig.update_yaxes(ticksuffix = \"  \", row=2, col=1)\n",
    "\n",
    "# Plot loadings\n",
    "for k, r in enumerate(loadings.columns):\n",
    "\n",
    "    dfp = loadings.sort_values(r, ascending=False)[:100]\n",
    "    dfp['regulatory'] = None \n",
    "    fig.add_trace(\n",
    "                  go.Bar(x=dfp.index, \n",
    "                         y=dfp[r],\n",
    "                         name='', \n",
    "                         orientation='v',\n",
    "                         hovertemplate=\"Gene Name: %{x} <br> Loading: %{y}\",\n",
    "                         visible=True if k == 0 else False,\n",
    "                         marker_color='black',\n",
    "                        ), \n",
    "                   row=3, col=1)\n",
    "fig.update_xaxes(showticklabels=False, row=3, col=1)\n",
    "fig.update_yaxes(tickvals=np.arange(0,1.25,0.25), ticksuffix = \"  \", row=3, col=1)\n",
    "\n",
    "# Plot regulators\n",
    "for k, r in enumerate(loadings.columns):\n",
    "\n",
    "       regulatory_ = regulatory.loc[regulatory.ProgramsRegulated=='K60_{}'.format(r)]\n",
    "\n",
    "       # Plot volcano\n",
    "       volcano = go.Scatter(x=regulatory_.log2FC,\n",
    "                            y=regulatory_.AcrossProgramsFDR.apply(lambda x: -np.log10(x)),\n",
    "                            customdata=regulatory_.Perturbation.values,\n",
    "                            hovertemplate=\" Perturbed gene: %{customdata}\",\n",
    "                            showlegend=False, mode='markers', \n",
    "                            visible=True if k == 0 else False,\n",
    "                            marker_color='black',\n",
    "                            )\n",
    "       fig.add_trace(volcano, row=2, col=2)\n",
    "       fig.update_xaxes(showticklabels=False, row=2, col=2)\n",
    "       fig.update_yaxes(title='Neg. log adjusted pval', ticksuffix = \"  \", row=2, col=2)\n",
    "\n",
    "# Plot regulators\n",
    "for k, r in enumerate(loadings.columns):\n",
    "\n",
    "       # Plot volcano\n",
    "       volcano = go.Scatter(x=program_statistics.CAD_MagmaBeta,\n",
    "                            y=program_statistics.CAD_MagmaPvalue.apply(lambda x: -np.log10(x)),\n",
    "                            customdata=program_statistics.ProgramID.values,\n",
    "                            hovertemplate=\" Program: %{customdata}\",\n",
    "                            showlegend=False, mode='markers', \n",
    "                            visible=True if k == 0 else False,\n",
    "                            marker_color='black',\n",
    "                            )\n",
    "       fig.add_trace(volcano, row=1, col=2)\n",
    "       fig.update_xaxes(showticklabels=False, row=1, col=2)\n",
    "       fig.update_yaxes(title='Neg. log adjusted pval', ticksuffix = \"  \", row=1, col=2)\n",
    "       \t\n",
    "        \n",
    "# Define buttons for dropdown\n",
    "col_opts = list(loadings.columns)\n",
    "buttons_opts = []\n",
    "for i, opt in enumerate(col_opts):\n",
    "    args = [False] * len(col_opts)\n",
    "    args[i] = True\n",
    "    buttons_opts.append(\n",
    "        dict(\n",
    "             method='update',\n",
    "             label=opt,\n",
    "             args=[{\n",
    "                    'visible': args, #this is the key line!\n",
    "                    'title': opt,\n",
    "                    'showlegend': False\n",
    "                   }]\n",
    "            ))\n",
    "    \n",
    "fig.update_layout(height=1400, width=1000, \n",
    "                  plot_bgcolor='whitesmoke',\n",
    "\n",
    "                  title_text=\"GEP Dashboard - v0.1 - Investigate GEPs\",\n",
    "                  updatemenus = [go.layout.Updatemenu(\n",
    "                                 active=0,\n",
    "                                 buttons=buttons_opts,\n",
    "                                 x=-0.1,\n",
    "                                 xanchor='left',\n",
    "                                 y=1,\n",
    "                                 yanchor='bottom')]\n",
    "                 )\n",
    "fig.write_html('example_output/1_covarite_association.html')\n",
    "#fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['ProgramID', 'CuratedLabel', 'CategoryEndothelialCellSpecificityZscore',\n",
       "       'CategoryBatchCorrelation', 'ProgramCategoryLabel', 'ProgramGenesTop10',\n",
       "       'ProgramGenesMotifsPromoter', 'ProgramGenesMotifsEnhancer',\n",
       "       'ProgramGeneSetsTop10', 'ProgramGenesMAPK', 'ProgramGenesECM',\n",
       "       'ProgramGenesECCandidate', 'ProgramGenesKnownCAD',\n",
       "       'RegulatorsECCandidate', 'RegulatorsKnownCAD', 'RegulatorsMAPK',\n",
       "       'RegulatorsTF', 'CAD_MagmaBeta', 'CAD_MagmaPvalue',\n",
       "       'CAD_RegulatorECCandidateEnrichment',\n",
       "       'CAD_ProgramGeneECCandidateEnrichment', 'CAD_ECCandidateFDR',\n",
       "       'LDSC_Enrichment', 'LDSC_PValue'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "program_statistics.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PerturbationOriginalName</th>\n",
       "      <th>Perturbation</th>\n",
       "      <th>ProgramsRegulated</th>\n",
       "      <th>log2FC</th>\n",
       "      <th>p.value</th>\n",
       "      <th>AcrossProgramsFDR</th>\n",
       "      <th>ExperimentWideFDR</th>\n",
       "      <th>IsMultiTarget</th>\n",
       "      <th>Gene.target.full.name</th>\n",
       "      <th>Gene2</th>\n",
       "      <th>TeloHAEC_ctrl_TPM</th>\n",
       "      <th>KDEfficiencyLog2FC</th>\n",
       "      <th>KDEfficiencyPValue</th>\n",
       "      <th>KDEfficiencyFDR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ANKRD22</td>\n",
       "      <td>ANKRD22</td>\n",
       "      <td>K60_19</td>\n",
       "      <td>0.1225</td>\n",
       "      <td>0.000171</td>\n",
       "      <td>0.010257</td>\n",
       "      <td>0.041135</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>-6.650000e-17</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>FYCO1</td>\n",
       "      <td>FYCO1</td>\n",
       "      <td>K60_19</td>\n",
       "      <td>-0.24596</td>\n",
       "      <td>0.000118</td>\n",
       "      <td>0.007097</td>\n",
       "      <td>0.032332</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>16.98391</td>\n",
       "      <td>-2.362350e-01</td>\n",
       "      <td>2.149290e-01</td>\n",
       "      <td>3.172970e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>METAP2</td>\n",
       "      <td>METAP2</td>\n",
       "      <td>K60_19</td>\n",
       "      <td>-0.2443</td>\n",
       "      <td>0.000087</td>\n",
       "      <td>0.005203</td>\n",
       "      <td>0.026196</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>122.92320</td>\n",
       "      <td>-9.310180e-01</td>\n",
       "      <td>1.610000e-10</td>\n",
       "      <td>3.330000e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193</th>\n",
       "      <td>PMVK</td>\n",
       "      <td>PMVK</td>\n",
       "      <td>K60_19</td>\n",
       "      <td>0.36288</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000034</td>\n",
       "      <td>0.000455</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>47.97226</td>\n",
       "      <td>-1.032499e+00</td>\n",
       "      <td>3.230000e-06</td>\n",
       "      <td>2.980000e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>222</th>\n",
       "      <td>RNF146</td>\n",
       "      <td>RNF146</td>\n",
       "      <td>K60_19</td>\n",
       "      <td>0.36188</td>\n",
       "      <td>0.000184</td>\n",
       "      <td>0.010216</td>\n",
       "      <td>0.042579</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>27.36403</td>\n",
       "      <td>-5.302900e-01</td>\n",
       "      <td>1.309300e-02</td>\n",
       "      <td>3.633700e-02</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    PerturbationOriginalName Perturbation ProgramsRegulated   log2FC  \\\n",
       "6                    ANKRD22      ANKRD22            K60_19   0.1225   \n",
       "86                     FYCO1        FYCO1            K60_19 -0.24596   \n",
       "141                   METAP2       METAP2            K60_19  -0.2443   \n",
       "193                     PMVK         PMVK            K60_19  0.36288   \n",
       "222                   RNF146       RNF146            K60_19  0.36188   \n",
       "\n",
       "      p.value AcrossProgramsFDR ExperimentWideFDR  IsMultiTarget  \\\n",
       "6    0.000171          0.010257          0.041135          False   \n",
       "86   0.000118          0.007097          0.032332          False   \n",
       "141  0.000087          0.005203          0.026196          False   \n",
       "193  0.000001          0.000034          0.000455          False   \n",
       "222  0.000184          0.010216          0.042579          False   \n",
       "\n",
       "    Gene.target.full.name Gene2  TeloHAEC_ctrl_TPM  KDEfficiencyLog2FC  \\\n",
       "6                     NaN   NaN            0.00000       -6.650000e-17   \n",
       "86                    NaN   NaN           16.98391       -2.362350e-01   \n",
       "141                   NaN   NaN          122.92320       -9.310180e-01   \n",
       "193                   NaN   NaN           47.97226       -1.032499e+00   \n",
       "222                   NaN   NaN           27.36403       -5.302900e-01   \n",
       "\n",
       "     KDEfficiencyPValue  KDEfficiencyFDR  \n",
       "6          1.000000e+00     1.000000e+00  \n",
       "86         2.149290e-01     3.172970e-01  \n",
       "141        1.610000e-10     3.330000e-09  \n",
       "193        3.230000e-06     2.980000e-05  \n",
       "222        1.309300e-02     3.633700e-02  "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "regulator_statistitcs.loc[regulator_statistitcs.ProgramsRegulated=='K60_19']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Program statistics\n",
    "\n",
    "# Magma CAD_MagmaBeta\tCAD_MagmaPvalue\n",
    "\n",
    "# LDSC LDSC_Enrichment\tLDSC_PValue\n",
    "\n",
    "# Similar to loadings - genes whose KD affects the program\n",
    "\n",
    "# Trait heritability\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Regulator statistics\n",
    "\n",
    "# % KD KDEfficiencyLog2FC\tKDEfficiencyPValue\tKDEfficiencyFDR\n",
    "\n",
    "# Log2FC on which program  Perturbation\tProgramsRegulated\tlog2FC\tp.value\tAcrossProgramsFDR\tExperimentWideFDR\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot trait enrichment interpretation\n",
    "def plot_interactive_phewas(data, x_column='trait_reported',\n",
    "                            y_column='-log10(p-value)',\n",
    "                            color_column='trait_category',\n",
    "                            filter_column='program_name',\n",
    "                            significance_threshold=0.05,\n",
    "                            annotation_cols=[\"program_name\", \"trait_reported\",\n",
    "                                             \"trait_category\", \"P-value\",\n",
    "                                             \"Genes\", \"study_id\", \"pmid\"],\n",
    "                           query_string=\"trait_category != 'measurement'\",\n",
    "                           title=\"Cell Program x OpenTargets GWAS L2G Enrichment\"):\n",
    "    \n",
    "    # Get unique values for the filtering column\n",
    "    filter_values = ['All'] + list(data[filter_column].unique())\n",
    "\n",
    "    # Initialize output widget to display the plot\n",
    "    output = Output()\n",
    "    \n",
    "    if query_string:\n",
    "        data=data.query(query_string)\n",
    "\n",
    "    # Function to update plot based on dropdown selection\n",
    "    def update_plot(selected_value):\n",
    "        # Filter data based on selected value\n",
    "        if selected_value == \"All\":\n",
    "            filtered_data = data.copy()  # No selection, show all data\n",
    "        else:\n",
    "            filtered_data = data[data[filter_column] == selected_value]\n",
    "\n",
    "        # Create the plot\n",
    "        fig = px.scatter(filtered_data, x=x_column, y=y_column, color=color_column,\n",
    "                         title=title,\n",
    "                         hover_data=annotation_cols)\n",
    "\n",
    "        # Customize layout\n",
    "        fig.update_layout(\n",
    "            xaxis_title=x_column,\n",
    "            yaxis_title=y_column,\n",
    "            yaxis=dict(tickformat=\".1f\"),\n",
    "            width=1000,  # Adjust width as needed\n",
    "            height=800,  # Adjust height as needed,\n",
    "            xaxis_tickfont=dict(size=4)\n",
    "        )\n",
    "\n",
    "        # Add horizontal dashed line for significance threshold\n",
    "        fig.add_hline(y=-np.log10(significance_threshold), line_dash=\"dash\",\n",
    "                      annotation_text=f'Significance Threshold ({significance_threshold})', annotation_position=\"top right\")\n",
    "\n",
    "        # Clear previous plot and display the new one\n",
    "        with output:\n",
    "            output.clear_output(wait=True)\n",
    "            fig.show()\n",
    "\n",
    "    # Create dropdown widget\n",
    "    dropdown = Dropdown(options=filter_values, description=f\"{filter_column}:\")\n",
    "\n",
    "    # Define function to handle dropdown value change\n",
    "    def on_change(change):\n",
    "        if change['type'] == 'change' and change['name'] == 'value':\n",
    "            update_plot(change['new'])\n",
    "\n",
    "    # Link dropdown change to function\n",
    "    dropdown.observe(on_change)\n",
    "\n",
    "    # Display dropdown widget and initial plot\n",
    "    display(VBox([dropdown, output]))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "evaluation",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
