{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.append('../../../')\n",
    "\n",
    "import mudata\n",
    "\n",
    "from src.evaluation import compute_categorical_association\n",
    "from src.evaluation import compute_explained_variance_ratio\n",
    "from src.evaluation import compute_geneset_enrichment\n",
    "from src.evaluation import compute_trait_enrichment\n",
    "from src.evaluation import compute_motif_enrichment\n",
    "\n",
    "from plotly.subplots import make_subplots\n",
    "\n",
    "import plotly.graph_objects as go\n",
    "import plotly.express as px\n",
    "\n",
    "import ipywidgets as w\n",
    "from IPython.display import display\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from scipy import stats\n",
    "\n",
    "from tqdm.auto import tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import test data\n",
    "mdata = mudata.read('../../../../../../data/TeloHAEC_Perturb-seq_2kG/2kG.library_K60_kangh.h5mu')\n",
    "\n",
    "sel_idx = []\n",
    "for batch in mdata['cNMF'].obs['batch'].unique():\n",
    "    for samp in mdata['cNMF'].obs['sample'].unique():\n",
    "        mdata_obs_ = mdata['cNMF'].obs.loc[(mdata['cNMF'].obs['batch']==batch) & \\\n",
    "                                           (mdata['cNMF'].obs['sample']==samp)]\n",
    "\n",
    "        sel_idx.extend(mdata_obs_.iloc[:200].index.tolist())\n",
    "\n",
    "mdata = mdata[sel_idx].copy()\n",
    "mdata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Explained variance ratio plot\n",
    "# explained_variance = compute_explained_variance_ratio(mdata, prog_key='cNMF', data_key='rna', inplace=False)\n",
    "\n",
    "# # Since original data wasn't available we use this as a stand in\n",
    "# from sklearn.decomposition import PCA\n",
    "\n",
    "# pca = PCA(n_components=60)\n",
    "\n",
    "# explained_variance['explained_variance_ratio_X'] = pca.fit(mdata['cNMF'].X).explained_variance_ratio_*0.93\n",
    "\n",
    "# Categorical association plot\n",
    "association_df, posthoc_df = compute_categorical_association(mdata, prog_key='cNMF',\n",
    "                                                             pseudobulk_key='sample',\n",
    "                                                             categorical_key='batch',\n",
    "                                                             inplace=False, \n",
    "                                                             n_jobs=-1)\n",
    "\n",
    "# # Geneset enrichment \n",
    "# gsea_df = compute_geneset_enrichment(mdata, prog_key='cNMF', library='Reactome_2013', inplace=False, n_jobs=-1)\n",
    "\n",
    "# # Trait enrichment \n",
    "# gwas_df = compute_trait_enrichment(mdata, '../../smk/resources/OpenTargets_L2G_Filtered.csv.gz', prog_key='cNMF', inplace=False)\n",
    "\n",
    "# # Motif enrichment\n",
    "# motif_match_df, motif_count_df, motif_enrichment_df = \\\n",
    "# compute_motif_enrichment(mdata, prog_key='cNMF', data_key='rna', motif_file='../tests/test_data/motifs.meme',\n",
    "#                          seq_file='../../../../../data/hg38/hg38.fa', coords_file='../tests/test_data/p2g_links.txt',\n",
    "#                          n_jobs=1, inplace=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save trait enrichment data\n",
    "# gwas_df.to_csv('dashapp/example_data/OpenTargets_L2G_cNMF_program_Enrichment_Results.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load example data\n",
    "evaluation_output = pd.ExcelFile('example_data/cNMF_evaluation_output.xlsx')\n",
    "\n",
    "explained_variance = pd.read_excel(explained_variance, sheet_name='explained_variance')\n",
    "enrichment_motif = pd.read_excel(evaluation_output, sheet_name='enrichment_motif')\n",
    "enrichment_gsea = pd.read_excel(evaluation_output, sheet_name='enrichment_gsea')\n",
    "enrichment_trait = pd.read_excel(evaluation_output, sheet_name='enrichment_trait')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Assemble dashboard - page 1\n",
    "\n",
    "# Plot unique terms per program\n",
    "def count_unique(categorical_var, count_var, dataframe):\n",
    "\n",
    "    counts_df = dataframe.value_counts([categorical_var, count_var])\n",
    "    counts_df = counts_df.groupby(categorical_var).sum()\n",
    "    counts_df = counts_df.sort_values(ascending=False)#.cumsum()\n",
    "\n",
    "    counts_df = pd.DataFrame(counts_df.reset_index().values, \n",
    "                             columns=[categorical_var,\n",
    "                                      count_var])\n",
    "    return counts_df\n",
    "\n",
    "fig = make_subplots(\n",
    "    rows=3, cols=2,\n",
    "    specs=[\n",
    "           [{}, {}],\n",
    "           [{}, {}],\n",
    "           [{}, None]\n",
    "          ],\n",
    "    print_grid=True,\n",
    "    subplot_titles=('Component wise R2 scores', 'Number of enriched GO terms',\n",
    "                    'Number of enriched promoter Motifs', 'Number of enriched GWAS traits',\n",
    "                    'Number of enriched enhancer Motifs'),\n",
    "    vertical_spacing = 0.05, horizontal_spacing = 0.1)\n",
    "\n",
    "# Explained variance ratios\n",
    "explained_variance = explained_variance.sort_values(explained_variance['VarianceExplained'], ascending=False)\n",
    "\n",
    "# Plot unique GSEA Terms per program\n",
    "gsea_unique_df = count_unique('ProgramID', 'ID', enrichment_gsea.loc[enrichment_gsea['qvalue']<=0.05])\n",
    "\n",
    "# Plot unique GWAS Terms per program\n",
    "gwas_unique_df = count_unique('ProgramID', 'Term', enrichment_trait.loc[enrichment_trait['Adjusted P-value']<=0.05])\n",
    "\n",
    "# Plot unique Promoter Motif Terms per program\n",
    "motif_unique_promoter_df = count_unique('ProgramID', 'TFMotif', enrichment_motif.loc[(enrichment_motif['FDR']<=0.05) &\\\n",
    "                                                                                     (enrichment_motif['EPType']=='Promoter')])\n",
    "\n",
    "# Plot unique Enhancer Motif Terms per program\n",
    "motif_unique_enhancer_df = count_unique('ProgramID', 'TFMotif', enrichment_motif.loc[(enrichment_motif['FDR']<=0.05) &\\\n",
    "                                                                                     (enrichment_motif['EPType']=='Enhancer')])\n",
    "\n",
    "# Generate example data for dashapp\n",
    "\n",
    "a = explained_ratios\n",
    "a['Component wise R2 scores'] = a['VarianceExplained']\n",
    "a = a.drop(['VarianceExplained'], axis=1)\n",
    "\n",
    "gsea_unique_df['Number of enriched GO terms'] = gsea_unique_df['ID']\n",
    "a = a.merge(gsea_unique_df.loc[:, ['ProgramID', 'Number of enriched GO terms']], \n",
    "            left_on='ProgramID', right_on='ProgramID', how='left')\n",
    "\n",
    "motif_unique_promoter_df['Number of enriched promoter Motifs'] = motif_unique_promoter_df['TFMotif']\n",
    "a = a.merge(motif_unique_promoter_df.loc[:, ['ProgramID', 'Number of enriched promoter Motifs']], \n",
    "            left_on='ProgramID', right_on='ProgramID', how='left')\n",
    "\n",
    "motif_unique_enhancer_df['Number of enriched enhancer Motifs'] = motif_unique_enhancer_df['TFMotif']\n",
    "a = a.merge(motif_unique_enhancer_df.loc[:, ['ProgramID', 'Number of enriched enhancer Motifs']], \n",
    "            left_on='ProgramID', right_on='ProgramID', how='left')\n",
    "\n",
    "gwas_unique_df['Number of enriched GWAS traits'] = gwas_unique_df['Term']\n",
    "a = a.merge(gwas_unique_df.loc[:, ['ProgramID', 'Number of enriched GWAS traits']], \n",
    "            left_on='ProgramID', right_on='ProgramID', how='left')\n",
    "\n",
    "\n",
    "a['Number of enriched GWAS traits'] = a['Number of enriched GWAS traits'].fillna(0)\n",
    "a['Number of enriched GO terms'] = a['Number of enriched GO terms'].fillna(0)\n",
    "a['Number of enriched promoter Motifs'] = a['Number of enriched promoter Motifs'].fillna(0)\n",
    "a['Number of enriched enhancer Motifs'] = a['Number of enriched enhancer Motifs'].fillna(0)\n",
    "\n",
    "\n",
    "plots = {}\n",
    "for i, col in enumerate([col for col in a.columns if 'ProgramID' not in col]):\n",
    "\n",
    "    plots[col] = px.scatter(x=a.sort_values(col, ascending=False)['ProgramID'], \n",
    "                            y=a.sort_values(col, ascending=False)[col],)\n",
    "    plots[col].update_layout(xaxis_title='Components', yaxis_title=col)\n",
    "\n",
    "    row_num = int((i)/2) + 1\n",
    "    col_num = i - (row_num-1)*2 + 1\n",
    "    fig.add_trace(plots[col]['data'][0], row=row_num, col=col_num)\n",
    "    fig.update_xaxes(showticklabels=False, row=row_num, col=col_num)\n",
    "    fig.update_yaxes(ticksuffix = \"  \", row=row_num, col=col_num)\n",
    "\n",
    "fig.update_traces(hovertemplate=\"Program Name: %{x} <br> Value: %{y}\", marker_color='black')\n",
    "fig.update_layout(height=1050, width=1000, \n",
    "                  plot_bgcolor='whitesmoke',\n",
    "                  title_text=\"GEP Dashboard - v0.1 - Goodness of fit measures\")\n",
    "                  \n",
    "fig.write_html('example_output/0_goodness_of_fit.html')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Page 2 - program assessment\n",
    "\n",
    "# Batch assocation\n",
    "association_df['batch_sample_kruskall_wallis_fdr'] = stats.false_discovery_control(association_df.batch_sample_kruskall_wallis_pval)\n",
    "association_df['batch_sample_kruskall_wallis_neg_log_fdr'] = association_df['batch_sample_kruskall_wallis_fdr'].apply(lambda x: -np.log(x))\n",
    "\n",
    "association_df['batch_sample_kruskall_wallis_log_stat'] = association_df['batch_sample_kruskall_wallis_stat'].apply(lambda x: np.log(x))\n",
    "\n",
    "# Program scores across batch\n",
    "prog_df = pd.DataFrame(mdata['cNMF'].X, index=mdata['cNMF'].obs.index)\n",
    "prog_df['sample'] = mdata['cNMF'].obs['sample']\n",
    "prog_df['batch'] = mdata['cNMF'].obs['batch']\n",
    "\n",
    "# Loadings\n",
    "loadings = pd.DataFrame(mdata['cNMF'].varm['loadings'], \n",
    "                        index=mdata['cNMF'].var_names,\n",
    "                        columns=mdata['cNMF'].uns['var_names']).T    \n",
    "loadings = loadings/abs(loadings).max(axis=0)\n",
    "\n",
    "# Make figure\n",
    "fig = make_subplots(rows=2, cols=2,\n",
    "                    specs=[\n",
    "                           [{}, {}],\n",
    "                           [{'colspan':2}, None],\n",
    "                          ],\n",
    "                    print_grid=True,\n",
    "                    subplot_titles=('Enrichment w.r.t. batch', \n",
    "                                    'Program distribution across batch', \n",
    "                                    'Program-gene loadings (normalised)'),\n",
    "                    vertical_spacing = 0.1, \n",
    "                    horizontal_spacing = 0.1)\n",
    "\n",
    "for k, r in enumerate([col for col in prog_df.columns if col not in ['sample', 'batch']]):\n",
    "       # Plot volcano\n",
    "       volcano = go.Scatter(x=association_df.batch_sample_kruskall_wallis_log_stat,\n",
    "                            y=association_df.batch_sample_kruskall_wallis_neg_log_fdr,\n",
    "                            customdata=association_df.index.values,\n",
    "                            hovertemplate=\" Program Name: %{customdata}\",\n",
    "                            showlegend=False, mode='markers', \n",
    "                            visible=True if k == 0 else False,\n",
    "                            marker_color='black',\n",
    "                            )\n",
    "       fig.add_trace(volcano, row=1, col=1)\n",
    "       fig.update_xaxes(showticklabels=False, row=1, col=1)\n",
    "       fig.update_yaxes(title='Neg. log adjusted pval', ticksuffix = \"  \", row=1, col=1)\n",
    "\n",
    "# Plot box\n",
    "for k, r in enumerate([col for col in prog_df.columns if col not in ['sample', 'batch']]):\n",
    "\n",
    "    fig.add_trace(\n",
    "                  go.Box(x=prog_df.batch, \n",
    "                         y=prog_df[r],\n",
    "                         visible=True if k == 0 else False,\n",
    "                         customdata=prog_df.index.values,\n",
    "                         hovertemplate=\"Cell barcode: %{customdata}\",\n",
    "                         marker_color='black',\n",
    "                        ), \n",
    "                   row=1, col=2)\n",
    "# fig.update_xaxes(showticklabels=False, row=2, col=1)\n",
    "fig.update_yaxes(ticksuffix = \"  \", row=1, col=2)\n",
    "\n",
    "# Plot loadings\n",
    "for k, r in enumerate(loadings.columns):\n",
    "\n",
    "    dfp = loadings.sort_values(r, ascending=False)[:100]\n",
    "    fig.add_trace(\n",
    "                  go.Bar(x=dfp.index, \n",
    "                         y=dfp[r],\n",
    "                         name='', \n",
    "                         orientation='v',\n",
    "                         hovertemplate=\"Gene Name: %{x} <br> Loading: %{y}\",\n",
    "                         visible=True if k == 0 else False,\n",
    "                         marker_color='black',\n",
    "                        ), \n",
    "                   row=2, col=1)\n",
    "fig.update_xaxes(showticklabels=False, row=2, col=1)\n",
    "fig.update_yaxes(tickvals=np.arange(0,1.25,0.25), ticksuffix = \"  \", row=2, col=1)\n",
    "        \n",
    "# Define buttons for dropdown\n",
    "col_opts = list(loadings.columns)\n",
    "buttons_opts = []\n",
    "for i, opt in enumerate(col_opts):\n",
    "    args = [False] * len(col_opts)\n",
    "    args[i] = True\n",
    "    buttons_opts.append(\n",
    "        dict(\n",
    "             method='update',\n",
    "             label=opt,\n",
    "             args=[{\n",
    "                    'visible': args, #this is the key line!\n",
    "                    'title': opt,\n",
    "                    'showlegend': False\n",
    "                   }]\n",
    "            ))\n",
    "    \n",
    "fig.update_layout(height=700, width=1000, \n",
    "                  plot_bgcolor='whitesmoke',\n",
    "\n",
    "                  title_text=\"GEP Dashboard - v0.1 - Investigate GEPs\",\n",
    "                  updatemenus = [go.layout.Updatemenu(\n",
    "                                 active=0,\n",
    "                                 buttons=buttons_opts,\n",
    "                                 x=-0.1,\n",
    "                                 xanchor='left',\n",
    "                                 y=1,\n",
    "                                 yanchor='bottom')]\n",
    "                 )\n",
    "fig.write_html('example_output/1_covarite_association.html')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot trait enrichment interpretation\n",
    "def plot_interactive_phewas(data, x_column='trait_reported',\n",
    "                            y_column='-log10(p-value)',\n",
    "                            color_column='trait_category',\n",
    "                            filter_column='program_name',\n",
    "                            significance_threshold=0.05,\n",
    "                            annotation_cols=[\"program_name\", \"trait_reported\",\n",
    "                                             \"trait_category\", \"P-value\",\n",
    "                                             \"Genes\", \"study_id\", \"pmid\"],\n",
    "                           query_string=\"trait_category != 'measurement'\",\n",
    "                           title=\"Cell Program x OpenTargets GWAS L2G Enrichment\"):\n",
    "    \n",
    "    # Get unique values for the filtering column\n",
    "    filter_values = ['All'] + list(data[filter_column].unique())\n",
    "\n",
    "    # Initialize output widget to display the plot\n",
    "    output = Output()\n",
    "    \n",
    "    if query_string:\n",
    "        data=data.query(query_string)\n",
    "\n",
    "    # Function to update plot based on dropdown selection\n",
    "    def update_plot(selected_value):\n",
    "        # Filter data based on selected value\n",
    "        if selected_value == \"All\":\n",
    "            filtered_data = data.copy()  # No selection, show all data\n",
    "        else:\n",
    "            filtered_data = data[data[filter_column] == selected_value]\n",
    "\n",
    "        # Create the plot\n",
    "        fig = px.scatter(filtered_data, x=x_column, y=y_column, color=color_column,\n",
    "                         title=title,\n",
    "                         hover_data=annotation_cols)\n",
    "\n",
    "        # Customize layout\n",
    "        fig.update_layout(\n",
    "            xaxis_title=x_column,\n",
    "            yaxis_title=y_column,\n",
    "            yaxis=dict(tickformat=\".1f\"),\n",
    "            width=1000,  # Adjust width as needed\n",
    "            height=800,  # Adjust height as needed,\n",
    "            xaxis_tickfont=dict(size=4)\n",
    "        )\n",
    "\n",
    "        # Add horizontal dashed line for significance threshold\n",
    "        fig.add_hline(y=-np.log10(significance_threshold), line_dash=\"dash\",\n",
    "                      annotation_text=f'Significance Threshold ({significance_threshold})', annotation_position=\"top right\")\n",
    "\n",
    "        # Clear previous plot and display the new one\n",
    "        with output:\n",
    "            output.clear_output(wait=True)\n",
    "            fig.show()\n",
    "\n",
    "    # Create dropdown widget\n",
    "    dropdown = Dropdown(options=filter_values, description=f\"{filter_column}:\")\n",
    "\n",
    "    # Define function to handle dropdown value change\n",
    "    def on_change(change):\n",
    "        if change['type'] == 'change' and change['name'] == 'value':\n",
    "            update_plot(change['new'])\n",
    "\n",
    "    # Link dropdown change to function\n",
    "    dropdown.observe(on_change)\n",
    "\n",
    "    # Display dropdown widget and initial plot\n",
    "    display(VBox([dropdown, output]))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "evaluation",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
